// @baidu/one-ui@4.32.1/src/components/breadcrumb/breadcrumbItem.tsx
import React, {PureComponent} from 'react';
import {noop} from 'lodash';
import {IconChevronRight} from 'dls-icons-react';
import {classnames} from '../../core/commonTools';
import {BreadcrumbItemProps} from './interface';
import {withConfigConsumer} from '../providerConfig/context';

@withConfigConsumer('breadcrumb')
class BreadcrumbItem extends PureComponent<BreadcrumbItemProps> {

    static defaultProps = {
        prefixCls: 'one-breadcrumb',
        separator: <IconChevronRight />,
        onClick: noop,
        disabled: false
    };

    render() {
        const {
            prefixCls, separator, children,
            onClick, disabled, className, ...restProps
        } = this.props;
        let link;
        const linkClassName = classnames({
            [`${prefixCls}-link-disabled`]: disabled,
            [`${prefixCls}-link`]: !disabled
        }, className);
        const linkProps = {
            className: linkClassName,
            ...restProps
        };
        if ('href' in this.props) {
            if (disabled) {
                delete linkProps.href;
            }
            link = <a {...linkProps}>{children}</a>;
        }
        else {
            link = <span {...linkProps}>{children}</span>;
        }
        if (children) {
            return (
                <span onClick={onClick}>
                    {link}
                    <span className={`${prefixCls}-separator`}>{separator}</span>
                </span>
            );
        }
        return null;
    }
}

export default BreadcrumbItem;
