// @baidu/one-ui@4.41.1/es/components/drawer/common/drawerChild.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import getScrollBarSize from 'rc-util/lib/getScrollBarSize';
import KeyCode from 'rc-util/lib/KeyCode';
import React from 'react';
import omit from 'omit.js';
import { addEventListener, dataToArray, getTouchParentScroll, removeEventListener, transitionEnd, isNumeric } from '../../../core/drawerTools';
import { classnames, windowIsUndefined, passiveSupported } from '../../../core/commonTools';
var currentDrawer = {};
var openCount = 0;
;

var DrawerChild = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(DrawerChild, _React$Component);

  function DrawerChild(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.passive = void 0;
    _this.drawerId = void 0;
    _this.dom = void 0;
    _this.handlerDom = void 0;
    _this.maskDom = void 0;
    _this.contentDom = void 0;
    _this.levelDom = void 0;
    _this.contentWrapper = void 0;
    _this.startPos = void 0;

    _this.handleClickBody = function (e) {
      var onClose = _this.props.onClose;

      if (onClose) {
        onClose(e);
      }
    };

    _this.domFocus = function () {
      if (_this.dom) {
        _this.dom.focus();
      }
    };

    _this.removeStartHandler = function (e) {
      if (e.touches.length > 1) {
        return;
      }

      _this.startPos = {
        x: e.touches[0].clientX,
        y: e.touches[0].clientY
      };
    };

    _this.removeMoveHandler = function (e) {
      if (e.changedTouches.length > 1) {
        return;
      }

      var currentTarget = e.currentTarget;
      var differX = e.changedTouches[0].clientX - _this.startPos.x;
      var differY = e.changedTouches[0].clientY - _this.startPos.y;

      if (currentTarget === _this.maskDom || currentTarget === _this.handlerDom || currentTarget === _this.contentDom && getTouchParentScroll(currentTarget, e.target, differX, differY)) {
        e.preventDefault();
      }
    };

    _this.transitionEnd = function (e) {
      var dom = e.target;
      removeEventListener(dom, transitionEnd, _this.transitionEnd);
      dom.style.transition = '';
    };

    _this.onKeyDown = function (e) {
      if (e.keyCode === KeyCode.ESC) {
        var onClose = _this.props.onClose;
        e.stopPropagation();

        if (onClose) {
          onClose(e);
        }
      }
    };

    _this.onWrapperTransitionEnd = function (e) {
      var _this$props = _this.props,
          open = _this$props.open,
          afterVisibleChange = _this$props.afterVisibleChange;

      if (e.target === _this.contentWrapper && e.propertyName.match(/transform$/)) {
        _this.dom.style.transition = '';

        if (!open && _this.getCurrentDrawerSome()) {
          document.body.style.overflowX = '';

          if (_this.maskDom) {
            _this.maskDom.style.left = '';
            _this.maskDom.style.width = '';
          }
        }

        if (afterVisibleChange) {
          afterVisibleChange(!!open);
        }
      }
    };

    _this.openLevelTransition = function () {
      var _this$props2 = _this.props,
          open = _this$props2.open,
          width = _this$props2.width,
          height = _this$props2.height;

      var _this$getHorizontalBo = _this.getHorizontalBoolAndPlacementName(),
          isHorizontal = _this$getHorizontalBo.isHorizontal,
          placementName = _this$getHorizontalBo.placementName;

      var contentValue = _this.contentDom ? _this.contentDom.getBoundingClientRect()[isHorizontal ? 'width' : 'height'] : 0;
      var value = (isHorizontal ? width : height) || contentValue;

      _this.setLevelAndScrolling(open, placementName, value);
    };

    _this.setLevelTransform = function (open, placementName, value, right) {
      var _this$props3 = _this.props,
          placement = _this$props3.placement,
          duration = _this$props3.duration,
          ease = _this$props3.ease,
          showMask = _this$props3.showMask; // router 切换时可能会导至页面失去滚动条，所以需要时时获取。

      _this.levelDom.forEach(function (dom) {
        dom.style.transition = "transform " + duration + " " + ease;
        addEventListener(dom, transitionEnd, _this.transitionEnd);
        var levelValue = open ? value : 0;
        var $value = typeof levelValue === 'number' ? levelValue + "px" : levelValue;
        var placementPos = placement === 'left' || placement === 'top' ? $value : "-" + $value;
        placementPos = showMask && placement === 'right' && right ? "calc(" + placementPos + " + " + right + "px)" : placementPos;
        dom.style.transform = levelValue ? placementName + "(" + placementPos + ")" : '';
      });
    };

    _this.setLevelAndScrolling = function (open, placementName, value) {
      var onChange = _this.props.onChange;

      if (!windowIsUndefined) {
        var right = document.body.scrollHeight > (window.innerHeight || document.documentElement.clientHeight) && window.innerWidth > document.body.offsetWidth ? getScrollBarSize(true) : 0;

        _this.setLevelTransform(open, placementName, value, right);

        _this.toggleScrollingToDrawerAndBody(right);
      }

      if (onChange) {
        onChange(open);
      }
    };

    _this.toggleScrollingToDrawerAndBody = function (right) {
      var _this$props4 = _this.props,
          getOpenCount = _this$props4.getOpenCount,
          getContainer = _this$props4.getContainer,
          showMask = _this$props4.showMask,
          open = _this$props4.open;
      var container = getContainer && getContainer(); // 处理 body 滚动

      if (container && container.parentNode === document.body && showMask) {
        var eventArray = ['touchstart'];
        var domArray = [document.body, _this.maskDom, _this.handlerDom, _this.contentDom];

        if (open && document.body.style.overflow !== 'hidden') {
          if (openCount === 1) {
            document.body.style.overflow = 'hidden';
            document.body.style.width = "calc(100% - " + right + "px)";
          }

          document.body.style.touchAction = 'none'; // 手机禁滚

          domArray.forEach(function (item, i) {
            if (!item) {
              return;
            }

            addEventListener(item, eventArray[i] || 'touchmove', i ? _this.removeMoveHandler : _this.removeStartHandler, _this.passive);
          });
        } else if (_this.getCurrentDrawerSome()) {
          // 没有弹框的状态下清除 overflow;
          if (!openCount) {
            document.body.style.overflow = '';
            document.body.style.width = '';
          }

          document.body.style.touchAction = ''; // 恢复事件

          domArray.forEach(function (item, i) {
            if (!item) {
              return;
            }

            removeEventListener(item, eventArray[i] || 'touchmove', i ? _this.removeMoveHandler : _this.removeStartHandler, _this.passive);
          });
        }
      }
    };

    _this.getCurrentDrawerSome = function () {
      return !Object.keys(currentDrawer).some(function (key) {
        return currentDrawer[key];
      });
    };

    _this.getLevelDom = function (_ref) {
      var level = _ref.level,
          getContainer = _ref.getContainer;

      if (windowIsUndefined) {
        return;
      }

      var container = getContainer && getContainer();
      var parent = container ? container.parentNode : null;
      _this.levelDom = [];

      if (level === 'all') {
        var children = parent ? Array.prototype.slice.call(parent.children) : [];
        children.forEach(function (child) {
          if (child.nodeName !== 'SCRIPT' && child.nodeName !== 'STYLE' && child.nodeName !== 'LINK' && child !== container) {
            _this.levelDom.push(child);
          }
        });
      } else if (level) {
        dataToArray(level).forEach(function (key) {
          document.querySelectorAll(key).forEach(function (item) {
            _this.levelDom.push(item);
          });
        });
      }
    };

    _this.getHorizontalBoolAndPlacementName = function () {
      var placement = _this.props.placement;
      var isHorizontal = placement === 'left' || placement === 'right';
      var placementName = "translate" + (isHorizontal ? 'X' : 'Y');
      return {
        isHorizontal: isHorizontal,
        placementName: placementName
      };
    };

    _this.onClick = function (e) {
      e.stopPropagation();
    };

    _this.state = {
      _self: _assertThisInitialized(_this)
    };
    return _this;
  }

  var _proto = DrawerChild.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var _this2 = this;

    if (!windowIsUndefined) {
      this.passive = passiveSupported ? {
        passive: false
      } : false;

      if (this.props.closeDrawerByClickBody && this.props.open) {
        window.addEventListener('click', this.handleClickBody);
      }
    }

    var _this$props5 = this.props,
        open = _this$props5.open,
        showMask = _this$props5.showMask;

    if (open && showMask) {
      openCount += 1;
    }

    this.drawerId = "drawer_id_" + Number((Date.now() + Math.random()).toString().replace('.', Math.round(Math.random() * 9).toString())).toString(16);
    this.getLevelDom(this.props);

    if (open) {
      currentDrawer[this.drawerId] = open; // 默认打开状态时推出 level;

      this.openLevelTransition();
      this.forceUpdate(function () {
        _this2.domFocus();
      });
    }
  };

  _proto.componentDidUpdate = function componentDidUpdate(prevProps) {
    var _this$props6 = this.props,
        open = _this$props6.open,
        showMask = _this$props6.showMask;

    if (!prevProps.open && open && showMask) {
      openCount += 1;
    }

    if (prevProps.open && !open && showMask) {
      openCount -= 1;
    }

    if (open !== prevProps.open) {
      if (open) {
        this.domFocus();
      }

      currentDrawer[this.drawerId] = !!open;
      this.openLevelTransition();
    }
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    var _this$props7 = this.props,
        getOpenCount = _this$props7.getOpenCount,
        open = _this$props7.open,
        showMask = _this$props7.showMask;

    if (open && showMask) {
      openCount -= 1;
    }

    delete currentDrawer[this.drawerId];

    if (!windowIsUndefined) {
      window.removeEventListener('click', this.handleClickBody);
    }

    if (open) {
      this.setLevelTransform(false);
      document.body.style.touchAction = '';
    }

    if (typeof getOpenCount === 'function' && !getOpenCount()) {
      document.body.style.overflow = '';
      document.body.style.width = '';
    }
  };

  _proto.render = function render() {
    var _classnames,
        _this3 = this;

    var _this$props8 = this.props,
        className = _this$props8.className,
        children = _this$props8.children,
        style = _this$props8.style,
        width = _this$props8.width,
        height = _this$props8.height,
        $open = _this$props8.open,
        prefixCls = _this$props8.prefixCls,
        placement = _this$props8.placement,
        level = _this$props8.level,
        ease = _this$props8.ease,
        duration = _this$props8.duration,
        getContainer = _this$props8.getContainer,
        handler = _this$props8.handler,
        onChange = _this$props8.onChange,
        afterVisibleChange = _this$props8.afterVisibleChange,
        showMask = _this$props8.showMask,
        maskClosable = _this$props8.maskClosable,
        maskStyle = _this$props8.maskStyle,
        drawerStyle = _this$props8.drawerStyle,
        onClose = _this$props8.onClose,
        onHandleClick = _this$props8.onHandleClick,
        keyboard = _this$props8.keyboard,
        getOpenCount = _this$props8.getOpenCount,
        lightPrefix = _this$props8.lightPrefix,
        props = _objectWithoutPropertiesLoose(_this$props8, ["className", "children", "style", "width", "height", "open", "prefixCls", "placement", "level", "ease", "duration", "getContainer", "handler", "onChange", "afterVisibleChange", "showMask", "maskClosable", "maskStyle", "drawerStyle", "onClose", "onHandleClick", "keyboard", "getOpenCount", "lightPrefix"]); // 首次渲染都将是关闭状态。


    var open = this.dom ? $open : false;
    var wrapperClassName = classnames(prefixCls, (_classnames = {}, _classnames[prefixCls + "-" + placement] = true, _classnames[prefixCls + "-open"] = open, _classnames[className || ''] = !!className, _classnames['no-mask'] = !showMask, _classnames));
    var placementName = this.getHorizontalBoolAndPlacementName().placementName; // 百分比与像素动画不同步，第一次打用后全用像素动画。
    // const defaultValue = !this.contentDom || !level ? '100%' : `${value}px`;

    var placementPos = placement === 'left' || placement === 'top' ? '-100%' : '100%';
    var transform = open ? '' : placementName + "(" + placementPos + ")";
    var handlerChildren = handler && /*#__PURE__*/React.cloneElement(handler, {
      onClick: function onClick(e) {
        if (handler.props.onClick) {
          handler.props.onClick();
        }

        if (onHandleClick) {
          onHandleClick(e);
        }
      },
      ref: function ref(c) {
        _this3.handlerDom = c;
      }
    });
    return /*#__PURE__*/React.createElement("div", _extends({}, omit(props, ['switchScrollingEffect', 'closeDrawerByClickBody', 'scrollLocker']), {
      tabIndex: -1,
      className: wrapperClassName,
      style: style,
      ref: function ref(c) {
        _this3.dom = c;
      },
      onKeyDown: open && keyboard ? this.onKeyDown : undefined,
      onTransitionEnd: this.onWrapperTransitionEnd,
      onClick: this.onClick
    }), showMask && /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-mask",
      onClick: maskClosable ? onClose : undefined,
      style: maskStyle,
      ref: function ref(c) {
        _this3.maskDom = c;
      }
    }), /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-content-wrapper",
      style: _extends({
        transform: transform,
        msTransform: transform,
        width: isNumeric(width) ? width + "px" : width,
        height: isNumeric(height) ? height + "px" : height
      }, drawerStyle),
      ref: function ref(c) {
        _this3.contentWrapper = c;
      }
    }, /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-content",
      ref: function ref(c) {
        _this3.contentDom = c;
      },
      onTouchStart: open && showMask ? this.removeStartHandler : undefined // 跑用例用
      ,
      onTouchMove: open && showMask ? this.removeMoveHandler : undefined // 跑用例用

    }, children), handlerChildren));
  };

  return DrawerChild;
}(React.Component);

DrawerChild.getDerivedStateFromProps = function (props, _ref2) {
  var prevProps = _ref2.prevProps,
      _self = _ref2._self;
  var nextState = {
    prevProps: props
  };

  if (prevProps !== undefined) {
    var placement = props.placement,
        level = props.level;

    if (placement !== prevProps.placement) {
      // test 的 bug, 有动画过场，删除 dom
      _self.contentDom = null;
    }

    if (level !== prevProps.level) {
      _self.getLevelDom(props);
    }
  }

  return nextState;
};

export default DrawerChild;
