// @baidu/one-ui@4.41.1/es/components/tag/editableGroup.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _partial from "lodash/partial";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import { IconPlus } from 'dls-icons-react';
import Tag from './tag';
import Tooltip from '../tooltip';
import Input from '../input';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
var inputWidth = 58;
;
var TagEditableGroup = (_dec = withConfigConsumer('tag'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(TagEditableGroup, _PureComponent);

  function TagEditableGroup(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.handleClose = function (value) {
      var dataSource = _this.state.dataSource;
      var onClose = _this.props.onClose;
      onClose(dataSource, value);

      if (!('dataSource' in _this.props)) {
        dataSource = dataSource.filter(function (tag) {
          return tag.value !== value;
        });

        _this.setState({
          dataSource: dataSource
        });
      }
    };

    _this.input = void 0;

    _this.showInput = function () {
      _this.setState({
        inputVisible: true
      }, function () {
        return _this.input.focus();
      });
    };

    _this.handleInputChange = function (e) {
      _this.setState({
        inputValue: e.value
      });
    };

    _this.handleInputConfirm = function () {
      var _this$state = _this.state,
          dataSource = _this$state.dataSource,
          inputValue = _this$state.inputValue;
      var onInputConfirm = _this.props.onInputConfirm;
      onInputConfirm(dataSource, inputValue);
      var newState = {
        inputVisible: false,
        inputValue: ''
      };

      if (!('dataSource' in _this.props)) {
        var currentTags = dataSource;

        if (inputValue) {
          currentTags = [].concat(currentTags, [{
            label: inputValue,
            value: inputValue
          }]);
        }

        newState.dataSource = currentTags;
      }

      _this.setState(newState);
    };

    _this.saveInputRef = function (input) {
      _this.input = input;
    };

    _this.state = {
      dataSource: props.defaultDataSource || props.dataSource || [],
      inputVisible: false,
      inputValue: ''
    };
    return _this;
  }

  var _proto = TagEditableGroup.prototype;

  _proto.render = function render() {
    var _classnames,
        _this2 = this;

    var _this$props = this.props,
        className = _this$props.className,
        prefixCls = _this$props.prefixCls,
        size = _this$props.size,
        normalized = _this$props.normalized;
    var _this$state2 = this.state,
        dataSource = _this$state2.dataSource,
        inputVisible = _this$state2.inputVisible,
        inputValue = _this$state2.inputValue;
    var groupTagClassName = classnames(className, prefixCls + "-group-wrapper", (_classnames = {}, _classnames[prefixCls + "-group-wrapper-normalized"] = normalized, _classnames));
    return /*#__PURE__*/React.createElement("div", {
      className: groupTagClassName
    }, dataSource.map(function (tag, index) {
      var label = tag.label,
          value = tag.value,
          tagProps = tag.tagProps;
      var isLongTag = typeof label === 'string' && label.length > 20;
      var tagElem = /*#__PURE__*/React.createElement(Tag, _extends({
        key: index,
        closable: true,
        onClose: _partial(_this2.handleClose, value),
        noClosed: true
      }, tagProps), isLongTag ? label.slice(0, 20) + "..." : label);
      return isLongTag ? /*#__PURE__*/React.createElement(Tooltip, {
        title: label,
        key: index
      }, tagElem) : tagElem;
    }), inputVisible && /*#__PURE__*/React.createElement(Input, {
      inputRef: this.saveInputRef,
      size: size,
      style: {
        width: inputWidth
      },
      value: inputValue,
      onChange: this.handleInputChange,
      onBlur: this.handleInputConfirm,
      onPressEnter: this.handleInputConfirm
    }), !inputVisible && /*#__PURE__*/React.createElement(Tag, {
      onClick: this.showInput,
      className: prefixCls + "-add-tag"
    }, /*#__PURE__*/React.createElement(IconPlus, {
      className: prefixCls + "-add-icon"
    }), "\u6DFB\u52A0"));
  };

  return TagEditableGroup;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-tag',
  className: '',
  size: 'medium',
  onClose: function onClose() {},
  onInputConfirm: function onInputConfirm() {}
}, _class2.getDerivedStateFromProps = function (nextProps, prevState) {
  if ('dataSource' in nextProps && nextProps.dataSource !== prevState.dataSource) {
    return {
      dataSource: nextProps.dataSource
    };
  }

  return null;
}, _temp)) || _class);
export default TagEditableGroup;
