// @baidu/one-ui@4.41.1/lib/components/tag/group.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _pull2 = _interopRequireDefault(require("lodash/pull"));

var _partial2 = _interopRequireDefault(require("lodash/partial"));

var _react = _interopRequireWildcard(require("react"));

var _tag = _interopRequireDefault(require("./tag"));

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

;
var TagGroup = (_dec = (0, _context.withConfigConsumer)('tag'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(TagGroup, _PureComponent);

  function TagGroup(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.isUniqueMode = function () {
      return _this.props.mode === 'unique';
    };

    _this.onChangeGroup = function (value, checked) {
      var selectedValue = [].concat(_this.state.value);

      if (checked) {
        // 选中
        if (_this.isUniqueMode()) {
          // 单选模式
          selectedValue = [value];
        } else {
          selectedValue.push(value);
        }
      } else if (_this.isUniqueMode()) {
        // 没选中，单选模式
        selectedValue = [];
      } else {
        (0, _pull2["default"])(selectedValue, value);
      }

      if (!('value' in _this.props)) {
        _this.setState({
          value: selectedValue
        });
      }

      var onChange = _this.props.onChange;
      onChange(selectedValue);
    };

    _this.state = {
      value: props.defaultValue || props.value || [],
      prevProps: props
    };
    return _this;
  }

  var _proto = TagGroup.prototype;

  _proto.render = function render() {
    var _classnames,
        _this2 = this;

    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        className = _this$props.className,
        dataSource = _this$props.dataSource,
        size = _this$props.size,
        normalized = _this$props.normalized;
    var value = this.state.value;
    var selectedValue = this.isUniqueMode() ? value && value.length && [value[0]] || [] : value;
    var groupTagClassName = (0, _commonTools.classnames)(className, prefixCls + "-group-wrapper", (_classnames = {}, _classnames[prefixCls + "-group-wrapper-normalized"] = normalized, _classnames));
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: groupTagClassName
    }, dataSource.map(function (tag, index) {
      var value = tag.value,
          label = tag.label,
          _tag$tagProps = tag.tagProps,
          tagProps = _tag$tagProps === void 0 ? {} : _tag$tagProps;
      var checked = selectedValue.indexOf(value) > -1;
      var onClose = tagProps.onClose;

      if (onClose) {
        tagProps.onClose = (0, _partial2["default"])(onClose, value);
      }

      return /*#__PURE__*/_react["default"].createElement(_tag["default"], (0, _extends2["default"])({
        key: index,
        size: size,
        checkable: true,
        checked: checked,
        onChange: (0, _partial2["default"])(_this2.onChangeGroup, value)
      }, tagProps), label);
    }));
  };

  return TagGroup;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-tag',
  mode: 'unique',
  dataSource: [],
  className: '',
  onChange: function onChange() {},
  size: 'medium'
}, _class2.getDerivedStateFromProps = function (nextProps, prevState) {
  var newState = {
    prevProps: nextProps
  };
  var prevProps = prevState.prevProps;

  if ('value' in nextProps && nextProps.value !== prevState.value) {
    newState.value = nextProps.value;
  }

  if ('dataSource' in nextProps && nextProps.dataSource !== prevProps.dataSource) {
    newState.dataSource = nextProps.dataSource;
  }

  return newState;
}, _temp)) || _class);
var _default = TagGroup;
exports["default"] = _default;
module.exports = exports.default;
