// @baidu/one-ui@4.32.1/lib/components/uploader/common/fileItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _dlsIconsReact = require("dls-icons-react");

var _progress = _interopRequireDefault(require("../../progress"));

var _button = _interopRequireDefault(require("../../button"));

var _uploaderTools = require("../../../core/uploaderTools");

var _commonTools = require("../../../core/commonTools");

var FileItem = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(FileItem, _PureComponent);

  function FileItem() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = FileItem.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        status = _this$props.status,
        name = _this$props.name,
        progressStep = _this$props.progressStep,
        errorMessage = _this$props.errorMessage,
        onRemove = _this$props.onRemove,
        index = _this$props.index,
        className = _this$props.className,
        controls = _this$props.controls,
        file = _this$props.file;
    var itemClassNames = (0, _commonTools.classnames)(prefixCls + "-file-item", prefixCls + "-file-item-" + status, className);
    var actions = [{
      key: 'delete',
      icon: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconTrash, null),
      onClick: function onClick(file, event) {
        return onRemove(index, event);
      }
    }];

    if (typeof controls === 'function') {
      actions = controls(file, actions) || [];
    }

    return /*#__PURE__*/_react["default"].createElement("div", {
      className: itemClassNames
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-file-item-container"
    }, /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconFile, {
      className: prefixCls + "-file-item-file-icon"
    }), /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-file-item-label"
    }, name), actions.map(function (_ref) {
      var visible = _ref.visible,
          disabled = _ref.disabled,
          key = _ref.key,
          icon = _ref.icon,
          _onClick = _ref.onClick;

      if (visible === false) {
        return null;
      }

      return /*#__PURE__*/_react["default"].createElement(_button["default"], {
        key: key,
        type: "text-aux",
        disabled: disabled,
        className: prefixCls + "-file-item-control",
        onClick: function onClick(event) {
          return _onClick(file, event);
        }
      }, icon);
    })), status === _uploaderTools.originStatus.UPLOADING ? /*#__PURE__*/_react["default"].createElement(_progress["default"], {
      size: "small",
      percent: progressStep,
      showInfo: false,
      className: prefixCls + "-file-item-progress"
    }) : null, status === _uploaderTools.originStatus.ERROR && errorMessage && errorMessage.length ? /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-file-item-message"
    }, errorMessage.join('，')) : null);
  };

  return FileItem;
}(_react.PureComponent);

exports["default"] = FileItem;
module.exports = exports.default;
