// @baidu/one-ui@4.41.1/es/components/checkbox/common/commonCheckbox.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _noop from "lodash/noop";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import { classnames } from '../../../core/commonTools';
import { IconCheck, IconMinus } from 'dls-icons-react';
import { withConfigConsumer } from '../../providerConfig/context';
var Checkbox = (_dec = withConfigConsumer('checkbox'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Checkbox, _PureComponent);

  function Checkbox(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.input = void 0;

    _this.saveInput = function (node) {
      _this.input = node;
    };

    _this.onHandleChange = function (e) {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          onChange = _this$props.onChange;

      if (disabled) {
        return;
      }

      var checked = e.target.checked;

      if (!('checked' in _this.props)) {
        _this.setState({
          checked: checked
        });
      }

      if (onChange) {
        onChange({
          target: _extends({}, _this.props, {
            checked: checked
          }),
          stopPropagation: function stopPropagation() {
            e.stopPropagation();
          },
          preventDefault: function preventDefault() {
            e.preventDefault();
          },
          nativeEvent: e.nativeEvent
        });
      }
    };

    var _checked = 'checked' in props ? props.checked : props.defaultChecked;

    _this.state = {
      checked: _checked
    };
    return _this;
  }

  var _proto = Checkbox.prototype;

  _proto.focus = function focus() {
    this.input.focus();
  };

  _proto.blur = function blur() {
    this.input.blur();
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props2 = this.props,
        prefixCls = _this$props2.prefixCls,
        className = _this$props2.className,
        style = _this$props2.style,
        type = _this$props2.type,
        mode = _this$props2.mode,
        disabled = _this$props2.disabled,
        onClick = _this$props2.onClick,
        onFocus = _this$props2.onFocus,
        onBlur = _this$props2.onBlur,
        autoFocus = _this$props2.autoFocus,
        indeterminate = _this$props2.indeterminate,
        inputOriginProps = _this$props2.inputOriginProps,
        name = _this$props2.name;
    var checked = this.state.checked;
    var classString = classnames(prefixCls, className, (_classnames = {}, _classnames[prefixCls + "-checked"] = checked, _classnames[prefixCls + "-disabled"] = disabled, _classnames));
    return /*#__PURE__*/React.createElement("span", {
      className: classString,
      style: style
    }, /*#__PURE__*/React.createElement("input", _extends({
      type: type,
      disabled: disabled,
      className: prefixCls + "-input",
      checked: !!checked,
      onClick: onClick,
      onFocus: onFocus,
      onBlur: onBlur,
      onChange: this.onHandleChange,
      autoFocus: autoFocus,
      ref: this.saveInput,
      name: name
    }, inputOriginProps)), /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-inner"
    }, type === 'checkbox' && indeterminate && mode !== 'strong' ? /*#__PURE__*/React.createElement(IconMinus, null) : null, type === 'checkbox' && checked && mode !== 'strong' ? /*#__PURE__*/React.createElement(IconCheck, null) : null));
  };

  return Checkbox;
}(PureComponent), _class2.defaultProps = {
  type: 'checkbox',
  defaultChecked: false,
  onFocus: _noop,
  onBlur: _noop,
  onChange: _noop
}, _class2.getDerivedStateFromProps = function (nextProps) {
  if ('checked' in nextProps) {
    return {
      checked: nextProps.checked
    };
  }

  return null;
}, _temp)) || _class);
export default Checkbox;
