// @baidu/one-ui@4.41.1/es/components/alert/alert.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _noop from "lodash/noop";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import classes from 'component-classes';
import { IconTimes, IconCheckCircleSolid, IconExclamationCircleSolid, IconTimesCircleSolid, IconInfoCircleSolid } from 'dls-icons-react';
import { classnames } from '../../core/commonTools';
import IconSvg from '../iconSvg';
import { withConfigConsumer } from '../providerConfig/context';
import omit from 'omit.js';
var defaultHeight = 32;
var Alert = (_dec = withConfigConsumer('alert'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Alert, _PureComponent);

  function Alert(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.alertContentRef = void 0;
    _this.alertRef = void 0;

    _this.componentDidMount = function () {
      _this.setDomClassByHeight();
    };

    _this.componentDidUpdate = function () {
      _this.setDomClassByHeight();
    };

    _this.setDomClassByHeight = function () {
      var alertContentRef = _this.alertContentRef;
      var alertRef = _this.alertRef;

      if (!alertContentRef || !alertRef) {
        return null;
      }

      var height = alertContentRef.offsetHeight;
      var multipleClassName = _this.props.prefixCls + "-multiple";
      var singleClassName = _this.props.prefixCls + "-single";
      var dom = classes(alertRef);

      if (height > defaultHeight) {
        dom.remove(singleClassName);
        dom.add(multipleClassName);
      } else if (dom.has(multipleClassName)) {
        dom.remove(multipleClassName);
        dom.add(singleClassName);
      }
    };

    _this.onHandleClose = function (e) {
      e.preventDefault();

      if (!('visble' in _this.props)) {
        _this.setState({
          visible: false
        });
      }

      var onClose = _this.props.onClose;

      if (onClose) {
        onClose(e);
      }
    };

    _this.saveRef = function (type) {
      return function (ref) {
        _this[type] = ref;
      };
    };

    _this.state = {
      visible: props.visible || true
    };
    return _this;
  }

  var _proto = Alert.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        className = _this$props.className,
        prefixCls = _this$props.prefixCls,
        showIcon = _this$props.showIcon,
        icon = _this$props.icon,
        type = _this$props.type,
        content = _this$props.content,
        closable = _this$props.closable,
        title = _this$props.title,
        size = _this$props.size,
        rest = _objectWithoutPropertiesLoose(_this$props, ["className", "prefixCls", "showIcon", "icon", "type", "content", "closable", "title", "size"]);

    var visible = this.state.visible;

    if (!visible) {
      return null;
    }

    var iconNode = null;

    if (!icon) {
      switch (type) {
        case 'success':
          iconNode = /*#__PURE__*/React.createElement(IconCheckCircleSolid, {
            className: prefixCls + "-title-icon"
          });
          break;

        case 'info':
          iconNode = /*#__PURE__*/React.createElement(IconInfoCircleSolid, {
            className: prefixCls + "-title-icon"
          });
          break;

        case 'error':
          iconNode = /*#__PURE__*/React.createElement(IconTimesCircleSolid, {
            className: prefixCls + "-title-icon"
          });
          break;

        case 'warning':
          iconNode = /*#__PURE__*/React.createElement(IconExclamationCircleSolid, {
            className: prefixCls + "-title-icon"
          });
          break;

        default:
          iconNode = null;
      }
    } else if (typeof icon === 'string') {
      iconNode = /*#__PURE__*/React.createElement(IconSvg, {
        className: prefixCls + "-title-icon",
        type: icon
      });
    } else if ( /*#__PURE__*/React.isValidElement(icon)) {
      iconNode = icon;
    }

    var alertCls = classnames(prefixCls, prefixCls + "-" + type, prefixCls + "-" + size, (_classnames = {}, _classnames[prefixCls + "-no-title"] = !title, _classnames[prefixCls + "-with-title"] = !!title, _classnames[prefixCls + "-show-icon"] = showIcon, _classnames[prefixCls + "-has-close-icon"] = closable, _classnames), className);
    var closeIcon = closable ? /*#__PURE__*/React.createElement("span", {
      onClick: this.onHandleClose,
      className: prefixCls + "-close-icon"
    }, /*#__PURE__*/React.createElement(IconTimes, null)) : null;
    return /*#__PURE__*/React.createElement("div", _extends({
      className: alertCls,
      ref: this.saveRef('alertRef')
    }, omit(rest, ['visible', 'onClose', 'lightPrefix'])), showIcon ? /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-icon"
    }, iconNode) : null, /*#__PURE__*/React.createElement("div", null, title ? /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-title"
    }, title) : null, /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-content",
      ref: this.saveRef('alertContentRef')
    }, content)), closeIcon);
  };

  return Alert;
}(PureComponent), _class2.defaultProps = {
  onClose: _noop,
  prefixCls: 'one-alert',
  type: 'info',
  size: 'medium'
}, _class2.Page = void 0, _class2.getDerivedStateFromProps = function (nextProps, prevState) {
  if ('visible' in nextProps && nextProps.visible !== prevState.visible) {
    return {
      visible: nextProps.visible
    };
  }

  return null;
}, _temp)) || _class);
export default Alert;
