// @baidu/one-ui@4.41.1/es/components/sidenav/index.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

/**
 * @file 侧导航组件
 * @author tangjinhua
 * @date 2021-03-12
 */
import React, { PureComponent } from 'react';
import Menu from '../menu';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
var Sidenav = (_dec = withConfigConsumer('side-nav'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Sidenav, _PureComponent);

  function Sidenav() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Sidenav.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        collapsed = _this$props.collapsed,
        prefixCls = _this$props.prefixCls,
        className = _this$props.className,
        lightPrefix = _this$props.lightPrefix,
        restProps = _objectWithoutPropertiesLoose(_this$props, ["collapsed", "prefixCls", "className", "lightPrefix"]);

    return /*#__PURE__*/React.createElement(Menu, _extends({}, restProps, {
      mode: "inline",
      inlineCollapsed: collapsed,
      className: classnames(prefixCls, className)
    }));
  };

  return Sidenav;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-side-nav',
  collapsed: false,
  paddingLeft: 24
}, _temp)) || _class);
export default Sidenav;
export * from './interface';
