// @baidu/one-ui@4.32.1/lib/components/grid/row.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _col = _interopRequireDefault(require("./col"));

var _dec, _class, _class2, _temp;

var GridRow = (_dec = (0, _context.withConfigConsumer)('row'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(GridRow, _PureComponent);

  function GridRow() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = GridRow.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        type = _this$props.type,
        justify = _this$props.justify,
        align = _this$props.align,
        className = _this$props.className,
        gutter = _this$props.gutter,
        style = _this$props.style,
        children = _this$props.children,
        _this$props$prefixCls = _this$props.prefixCls,
        prefixCls = _this$props$prefixCls === void 0 ? 'one-row' : _this$props$prefixCls,
        others = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, ["type", "justify", "align", "className", "gutter", "style", "children", "prefixCls"]);
    var classes = (0, _commonTools.classnames)((_classnames = {}, _classnames[prefixCls] = !type, _classnames[prefixCls + "-" + type] = type, _classnames[prefixCls + "-" + type + "-" + justify] = type && justify, _classnames[prefixCls + "-" + type + "-" + align] = type && align, _classnames), className);
    var rowStyle = gutter > 0 ? (0, _extends2["default"])({
      marginLeft: gutter / -2,
      marginRight: gutter / -2
    }, style) : style;

    var cols = _react.Children.map(children, function (col) {
      if ( /*#__PURE__*/(0, _react.isValidElement)(col) && col.type === _col["default"] && gutter > 0) {
        return /*#__PURE__*/(0, _react.cloneElement)(col, {
          style: (0, _extends2["default"])({
            paddingLeft: gutter / 2,
            paddingRight: gutter / 2
          }, col.props.style)
        });
      }

      return col;
    });

    return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({}, others, {
      className: classes,
      style: rowStyle
    }), cols);
  };

  return GridRow;
}(_react.PureComponent), _class2.defaultProps = {
  gutter: 0
}, _temp)) || _class);
var _default = GridRow;
exports["default"] = _default;
module.exports = exports.default;
