// @baidu/one-ui@4.32.1/src/core/cascaderPaneTools.js
/**
 * @file 级联面板tools
 * @author huangshiming
 * @date 2020-05-04
 */
import {uniq, pullAll} from 'lodash';

// 将options转成对象
export const transOptionsToObject = (options, fieldName = 'value', optionsObject = {}, parent) => {
    options.forEach(option => {
        const value = option[fieldName];
        optionsObject[value] = {
            value,
            parent: parent || '',
            label: option.label
        };
        if (option.disabled) {
            optionsObject[value].disabled = true;
        }
        if (option.children && option.children.length) {
            optionsObject[value].children = option.children.map(child => child[fieldName]);
            transOptionsToObject(option.children, 'value', optionsObject, value);
        }
    });
    return optionsObject;
};

// 获取选中状态
export const formatCheckedKeys = (
    optionsObject,
    checkedKeys,
    newCheckedKeys = []
) => {
    checkedKeys.forEach(checkedKey => {
        newCheckedKeys.push(checkedKey);
        if (optionsObject[checkedKey] && optionsObject[checkedKey].children) {
            formatCheckedKeys(optionsObject, optionsObject[checkedKey].children, newCheckedKeys);
        }
    });
    return newCheckedKeys;
};

// 获取选中后的半选状态的值
export const getIndeterminateKeys = (
    optionsObject,
    targetKeys,
    indeterminates = [],
    checkedKeys = []
) => {
    targetKeys.forEach(checkedKey => {
        if (optionsObject[checkedKey] && optionsObject[checkedKey].parent) {
            // 存在父节点，遍历父节点
            let allChecked = true;
            const parentKey = optionsObject[checkedKey].parent;
            const children = optionsObject[parentKey].children;
            if (children.length === 1 && indeterminates.indexOf(children[0]) > -1) {
                // 只有一个子节点
                allChecked = false;
            }
            else {
                optionsObject[parentKey].children.forEach(child => {
                    if (checkedKeys.indexOf(child) === -1) {
                        allChecked = false;
                    }
                });
            }
            if (!allChecked) {
                indeterminates.push(parentKey);
                getIndeterminateKeys(optionsObject, [parentKey], indeterminates, checkedKeys);
            }
            else if (allChecked) {
                // 移出半选
                if (indeterminates.indexOf(parentKey) > -1) {
                    indeterminates.splice(indeterminates.indexOf(parentKey), 1);
                }
                getIndeterminateKeys(optionsObject, [parentKey], indeterminates, checkedKeys);
            }
        }
    });
    return indeterminates;
};

// 获取取消选中的半选
export const getCancelIndeterminateKeys = (
    optionsObject,
    targetKeys,
    indeterminates = [],
    checkedKeys = []
) => {
    targetKeys.forEach(checkedKey => {
        if (optionsObject[checkedKey] && optionsObject[checkedKey].parent) {
            let hasChecked = false;
            let allChecked = true;
            const parentKey = optionsObject[checkedKey].parent;
            optionsObject[parentKey].children.forEach(child => {
                if (checkedKeys.indexOf(child) > -1 || indeterminates.indexOf(child) > -1) {
                    hasChecked = true;
                }
                if (checkedKeys.indexOf(child) === -1) {
                    allChecked = false;
                }
            });
            if (hasChecked) {
                if (indeterminates.indexOf(parentKey) === -1 && allChecked === false) {
                    indeterminates.push(parentKey);
                }
                getCancelIndeterminateKeys(optionsObject, [parentKey], indeterminates, checkedKeys);
            }
            else {
                if (indeterminates.indexOf(parentKey) > -1) {
                    indeterminates.splice(indeterminates.indexOf(parentKey), 1);
                }
                getCancelIndeterminateKeys(optionsObject, [parentKey], indeterminates, checkedKeys);
            }
        }
    });
    return indeterminates;
};

// 获取父亲全选的节点
export const formatAllCheckedKeys = (
    optionsObject, checkedKey, alCheckedKeys, checkedKeys = []
) => {
    checkedKeys.push(checkedKey);
    const newCheckedKeys = [...alCheckedKeys, checkedKey];
    if (optionsObject[checkedKey] && optionsObject[checkedKey].parent) {
        let allChecked = true;
        const parentKey = optionsObject[checkedKey].parent;
        optionsObject[parentKey].children.forEach(child => {
            if (newCheckedKeys.indexOf(child) === -1) {
                allChecked = false;
            }
        });
        if (allChecked) {
            formatAllCheckedKeys(optionsObject, parentKey, newCheckedKeys, checkedKeys);
        }
    }
    return checkedKeys;
};

// 遍历parent节点
export const formatParentKeys = (optionsObject, checkedKey, parentKeys = []) => {
    if (optionsObject[checkedKey] && optionsObject[checkedKey].parent) {
        parentKeys.push(optionsObject[checkedKey].parent);
        formatParentKeys(optionsObject, optionsObject[checkedKey].parent, parentKeys);
    }
    return parentKeys;
};

// 删除选中的checked keys
export const formatDeleteCheckedKeys = (optionsObject, deleteKey, interCheckedKeys) => {
    const deleteCheckedKeys = [
        ...formatCheckedKeys(optionsObject, [deleteKey]),
        ...formatParentKeys(optionsObject, deleteKey)
    ];
    return [...pullAll(interCheckedKeys, deleteCheckedKeys)];
};

// 获取筛选的key
export const filterSearchValue = (flattenTrees, inputValue) => {
    const newFlattenTrees = [];
    flattenTrees.forEach(flattenTree => {
        flattenTree.forEach(tree => {
            if (tree.label.indexOf(inputValue) > -1) {
                newFlattenTrees.push(flattenTree);
            }
        });
    });
    return newFlattenTrees;
};

// 展平树 - 单选的情况
export const flattenTree = (options, ancestor = []) => {
    let flattenOptions = [];
    options.forEach(option => {
        const path = ancestor.concat(option);
        if (!option.children || !option.children.length) {
            flattenOptions.push(path);
        }
        if (option.children) {
            flattenOptions = flattenOptions.concat(flattenTree(option.children, path));
        }
    });
    return flattenOptions;
};

// 寻找祖先
export const findAncestor = (optionsObject, curKey, ancestor = []) => {
    if (optionsObject[curKey] && optionsObject[curKey].children) {
        ancestor.unshift(optionsObject[curKey]);
    }
    if (optionsObject[curKey] && optionsObject[curKey].parent) {
        const parentKey = optionsObject[curKey].parent;
        ancestor.unshift(optionsObject[curKey]);
        ancestor.unshift(optionsObject[parentKey]);
        if (optionsObject[parentKey] && optionsObject[parentKey].parent) {
            findAncestor(optionsObject, parentKey, ancestor);
        }
    }
    return uniq(ancestor);
};

// 展平多选树 - 多选的情况
export const flattenMultipleTree = (optionsObject, options, ancestor = []) => {
    options.forEach(option => {
        if (option.children && option.children.length) {
            ancestor.push(findAncestor(optionsObject, option.value));
            flattenMultipleTree(optionsObject, option.children, ancestor);
        }
    });
    return ancestor;
};

export const getDisabledKeys = (disabledKeys, optionsObject) => {
    let newDisabledKeys = [...disabledKeys];
    disabledKeys.forEach(key => {
        if (optionsObject[key] && optionsObject[key].children) {
            newDisabledKeys = [
                ...newDisabledKeys,
                ...getDisabledKeys(optionsObject[key].children, optionsObject)
            ];
        }
    });
    return newDisabledKeys;
};

// 获取所有disabled的key，如果该节点有child则，child也应该disabled
export const formatDisabledKeys = optionsObject => {
    const disabledKeys = [];
    Object.keys(optionsObject).forEach(key => {
        if (optionsObject[key] && optionsObject[key].disabled) {
            disabledKeys.push(key);
        }
    });
    return uniq(getDisabledKeys(disabledKeys, optionsObject));
};

export const filterDisabledParentCheckedKeys = (optionsObject, checkedKeys) => {
    const removeFilters = [];
    checkedKeys.forEach(key => {
        if (optionsObject[key]
            && optionsObject[key].children
            && optionsObject[key].children.length
        ) {
            optionsObject[key].children.forEach(child => {
                if (checkedKeys.indexOf(child) === -1) {
                    removeFilters.push(key);
                }
            });
        }
    });
    return [...pullAll(checkedKeys, removeFilters)];
};
