// @baidu/one-ui@4.32.1/es/components/stack/index.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { forwardRef } from 'react';
import classNames from 'classnames';

/**
 * @visibleName Stack
 */
export var Stack = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var _classNames, _classNames2, _ref2;

  var prefixCls = _ref.prefixCls,
      children = _ref.children,
      className = _ref.className,
      direction = _ref.direction,
      wrap = _ref.wrap,
      inline = _ref.inline,
      align = _ref.align,
      justify = _ref.justify,
      gap = _ref.gap,
      restProps = _objectWithoutPropertiesLoose(_ref, ["prefixCls", "children", "className", "direction", "wrap", "inline", "align", "justify", "gap"]);

  var rootClass = classNames(prefixCls, className, (_classNames = {}, _classNames[prefixCls + "-wrap"] = wrap, _classNames[prefixCls + "-inline"] = inline, _classNames[prefixCls + "-" + direction] = direction, _classNames[prefixCls + "-gap"] = gap, _classNames));
  var realAlign = align || (direction === 'row' ? 'center' : 'stretch');
  var innerClass = classNames(prefixCls + "-inner", (_classNames2 = {}, _classNames2[prefixCls + "-inner-gap-" + gap] = typeof gap === 'string', _classNames2[prefixCls + "-inner-align-" + realAlign] = realAlign, _classNames2[prefixCls + "-inner-justify-" + justify] = justify, _classNames2));
  var innerStyle = typeof gap === 'number' && gap > 0 ? (_ref2 = {}, _ref2["--" + (prefixCls + "-gap")] = gap + "px", _ref2) : null;
  return /*#__PURE__*/React.createElement("div", _extends({
    className: rootClass
  }, restProps, {
    ref: ref
  }), /*#__PURE__*/React.createElement("div", {
    className: innerClass,
    style: innerStyle
  }, children));
});
Stack.defaultProps = {
  prefixCls: 'one-stack',
  direction: 'row',
  wrap: false,
  inline: false
};
Stack.displayName = 'Stack';
export default Stack;
export * from './interface';
