// @baidu/one-ui@4.41.1/lib/components/providerConfig/context.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.withConfigConsumer = exports["default"] = exports.Consumer = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _config = require("../config");

/**
 * @file 顶层context
 * @author huangshiming
 * @date 2020-04-28
 */
var context = /*#__PURE__*/_react["default"].createContext(undefined);

var lightThemeComponents = [// components
'select', 'dropdown', 'cascader', 'uploader', 'pagination', 'radio-group', 'checkbox-group', 'input', 'input-group', 'number-input', 'textarea', 'search-box', 'time-picker', 'date-picker', 'button', // containers
'popover', 'loading', 'tooltip', 'overlay', 'drawer', 'dialog', 'row', 'col', 'card', 'tabs', 'form', 'stack', 'layout', 'layout-header', 'layout-footer', 'layout-sidebar', 'layout-content'];
var Consumer = context.Consumer;
exports.Consumer = Consumer;
var _default = context;
/**
 * 组件decorator，用于处理全局config
 *
 * @param suffixCls 组件className后缀
 */

exports["default"] = _default;

var withConfigConsumer = function withConfigConsumer(suffixCls) {
  return function (Component) {
    var ComponentWithConsumer = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
      return /*#__PURE__*/_react["default"].createElement(Consumer, null, function (config) {
        var _ref = config || {},
            size = _ref.size,
            _ref$prefixCls = _ref.prefixCls,
            prefixCls = _ref$prefixCls === void 0 ? _config.PREFIX : _ref$prefixCls,
            theme = _ref.theme,
            normalized = _ref.normalized,
            table = _ref.table;

        var normalizedConfig = suffixCls === 'table' && table && table.loadingOption ? {
          loadingOption: table.loadingOption
        } : {};

        if (normalized) {
          normalizedConfig.normalized = normalized;
        }

        if (size) {
          normalizedConfig.size = size;
        }

        var className = props.className,
            compnentTheme = props.theme,
            restProps = (0, _objectWithoutPropertiesLoose2["default"])(props, ["className", "theme"]);
        var normalizedTheme = compnentTheme || theme;
        var normalizedPrefixCls = prefixCls;

        if (normalizedTheme === _config.THEME_LIGHT_AI) {
          normalizedPrefixCls = _config.THEME_LIGHT_AI_PREFIX;
          normalizedTheme = _config.THEME_LIGHT_D22;
        }

        if (prefixCls) {
          normalizedConfig.lightPrefix = normalizedPrefixCls;
          normalizedConfig.prefixCls = normalizedPrefixCls + "-" + suffixCls;
        }

        if (normalizedTheme) {
          normalizedConfig.theme = normalizedTheme;

          if (normalizedTheme !== _config.THEME_LIGHT_D22 || lightThemeComponents.includes(suffixCls)) {
            var themeClassName = normalizedPrefixCls + "-theme-" + normalizedTheme;

            if (!className || className.indexOf(themeClassName) === -1) {
              className = (0, _classnames["default"])(className, themeClassName);
            }
          }
        }

        var themeName = normalizedConfig.lightPrefix === _config.THEME_LIGHT_AI_PREFIX ? _config.THEME_LIGHT_AI : compnentTheme || theme;
        normalizedConfig.className = !className ? themeName : RegExp("(^|s)" + themeName).test(className) ? className : (0, _classnames["default"])(className, themeName);

        var component = /*#__PURE__*/_react["default"].createElement(Component, (0, _extends2["default"])({
          ref: ref
        }, normalizedConfig, restProps));

        if (normalizedTheme === _config.THEME_LIGHT_D22 && !lightThemeComponents.includes(suffixCls)) {
          return /*#__PURE__*/_react["default"].createElement(context.Provider, {
            value: (0, _extends2["default"])({}, config, {
              prefixCls: normalizedPrefixCls,
              theme: ''
            })
          }, component);
        }

        return /*#__PURE__*/_react["default"].createElement(context.Provider, {
          value: (0, _extends2["default"])({}, config, {
            theme: normalizedTheme,
            prefixCls: normalizedPrefixCls
          })
        }, component);
      });
    });
    ComponentWithConsumer.displayName = Component.displayName || Component.name;
    return ComponentWithConsumer;
  };
};

exports.withConfigConsumer = withConfigConsumer;
