// less-plugin-dls@11.12.0/variables.global.ai.js
export const dlsColorBrand = '#4d79ff'
export const dlsColorBrand0 = '#fff'
export const dlsColorBrand1 = '#f2f7ff'
export const dlsColorBrand2 = '#e3edff'
export const dlsColorBrand3 = '#c4daff'
export const dlsColorBrand4 = '#99beff'
export const dlsColorBrand5 = '#729cfe'
export const dlsColorBrand6 = '#4d79ff'
export const dlsColorBrand7 = '#3a5bfd'
export const dlsColorBrand8 = '#333fe6'
export const dlsColorBrand9 = '#0f2dbd'
export const dlsColorBrand10 = '#1c244a'
export const dlsColorBrand11 = '#000'
export const dlsColorInfo = '#4d79ff'
export const dlsColorInfo0 = '#fff'
export const dlsColorInfo1 = '#f2f7ff'
export const dlsColorInfo2 = '#e3edff'
export const dlsColorInfo3 = '#c4daff'
export const dlsColorInfo4 = '#99beff'
export const dlsColorInfo5 = '#729cfe'
export const dlsColorInfo6 = '#4d79ff'
export const dlsColorInfo7 = '#3a5bfd'
export const dlsColorInfo8 = '#333fe6'
export const dlsColorInfo9 = '#0f2dbd'
export const dlsColorInfo10 = '#1c244a'
export const dlsColorInfo11 = '#000'
export const dlsColorSuccess = '#3ac222'
export const dlsColorSuccess0 = '#fff'
export const dlsColorSuccess1 = '#eeffeb'
export const dlsColorSuccess2 = '#e3ffde'
export const dlsColorSuccess3 = '#cfffc7'
export const dlsColorSuccess4 = '#b2ffa5'
export const dlsColorSuccess5 = '#8cff78'
export const dlsColorSuccess6 = '#59f53d'
export const dlsColorSuccess7 = '#3ac222'
export const dlsColorSuccess8 = '#19a500'
export const dlsColorSuccess9 = '#137e00'
export const dlsColorSuccess10 = '#0c4e00'
export const dlsColorSuccess11 = '#000'
export const dlsColorWarning = '#fea800'
export const dlsColorWarning0 = '#fff'
export const dlsColorWarning1 = '#fff8eb'
export const dlsColorWarning2 = '#fff4dd'
export const dlsColorWarning3 = '#ffebc4'
export const dlsColorWarning4 = '#ffdf9f'
export const dlsColorWarning5 = '#ffce6f'
export const dlsColorWarning6 = '#ffba32'
export const dlsColorWarning7 = '#fea800'
export const dlsColorWarning8 = '#d88f00'
export const dlsColorWarning9 = '#a56d00'
export const dlsColorWarning10 = '#664300'
export const dlsColorWarning11 = '#000'
export const dlsColorError = '#f53f3f'
export const dlsColorError0 = '#fff'
export const dlsColorError1 = '#ffebeb'
export const dlsColorError2 = '#ffe2e2'
export const dlsColorError3 = '#ffd3d3'
export const dlsColorError4 = '#ffbcbc'
export const dlsColorError5 = '#ff9e9e'
export const dlsColorError6 = '#ff7878'
export const dlsColorError7 = '#f53f3f'
export const dlsColorError8 = '#d01b1b'
export const dlsColorError9 = '#9f0000'
export const dlsColorError10 = '#620000'
export const dlsColorError11 = '#000'
export const dlsColorGray0 = '#fff'
export const dlsColorGray1 = '#f6f7fa'
export const dlsColorGray2 = '#ebedf5'
export const dlsColorGray3 = '#e2e6f0'
export const dlsColorGray4 = '#d3d9e6'
export const dlsColorGray5 = '#c5ccdb'
export const dlsColorGray6 = '#a8b0bf'
export const dlsColorGray7 = '#848b99'
export const dlsColorGray8 = '#545b66'
export const dlsColorGray9 = '#191b1e'
export const dlsColorGray10 = '#0a0b0d'
export const dlsColorGray11 = '#000'
export const dlsColorTranslucent0 = 'transparent'
export const dlsColorTranslucent1 = 'rgba(109, 159, 247, 0.07)'
export const dlsColorTranslucent2 = 'rgba(108, 156, 240, 0.1)'
export const dlsColorTranslucent3 = 'rgba(102, 146, 222, 0.15)'
export const dlsColorTranslucent4 = 'rgba(96, 134, 199, 0.2)'
export const dlsColorTranslucent5 = 'rgba(90, 119, 168, 0.3)'
export const dlsColorTranslucent6 = 'rgba(83, 101, 138, 0.5)'
export const dlsColorTranslucent7 = 'rgba(73, 86, 107, 0.8)'
export const dlsColorTranslucent8 = 'rgba(46, 52, 64, 0.9)'
export const dlsColorTranslucent9 = 'rgba(13, 15, 18, 0.95)'
export const dlsColorTranslucent10 = 'rgba(4, 4, 5, 0.95)'
export const dlsColorTranslucent11 = '#000'
export const dlsFontFamily = '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "PingFang SC", "Hiragino Sans GB", "Microsoft YaHei", sans-serif'
export const dlsFontFamilyNumber = '"Baidu Number", sans-serif'
export const dlsFontSize0 = '12px'
export const dlsFontSize1 = '14px'
export const dlsFontSize2 = '16px'
export const dlsFontSize3 = '18px'
export const dlsFontSize4 = '20px'
export const dlsFontSize5 = '24px'
export const dlsFontSize6 = '32px'
export const dlsFontWeight1 = '400'
export const dlsFontWeight2 = '500'
export const dlsFontWeight3 = '600'
export const dlsLineHeight1 = '1.4'
export const dlsLineHeight2 = '1.6'
export const dlsLineHeight3 = '1.8'
export const dlsSpacingUnit = '4px'
export const dlsPaddingUnit = '4px'
export const dlsHeightUnit = '4px'
export const dlsSpacing1 = '4px'
export const dlsSpacing2 = '8px'
export const dlsSpacing3 = '12px'
export const dlsSpacing4 = '16px'
export const dlsSpacing5 = '24px'
export const dlsSpacing6 = '32px'
export const dlsSpacing7 = '40px'
export const dlsHeightXs = '28px'
export const dlsHeightS = '32px'
export const dlsHeightM = '36px'
export const dlsHeightL = '40px'
export const dlsHeightXl = '48px'
export const dlsForegroundColorReverse = '#fff'
export const dlsForegroundColorPrimaryReverse = '#fff'
export const dlsForegroundColorPrimary = '#3a5bfd'
export const dlsForegroundColorPrimaryHover = '#4d79ff'
export const dlsForegroundColorPrimaryActive = '#333fe6'
export const dlsForegroundColorPrimaryDisabled = '#99beff'
export const dlsForegroundColorNeutral = 'rgba(13, 15, 18, 0.95)'
export const dlsForegroundColorNeutralHover = 'rgba(46, 52, 64, 0.9)'
export const dlsForegroundColorNeutralActive = '#000'
export const dlsForegroundColorNeutralDisabled = 'rgba(83, 101, 138, 0.5)'
export const dlsForegroundColorNeutralStrong = '#000'
export const dlsForegroundColorNeutralStrongHover = 'rgba(4, 4, 5, 0.95)'
export const dlsForegroundColorNeutralStrongActive = '#000'
export const dlsForegroundColorNeutralStrongDisabled = 'rgba(83, 101, 138, 0.5)'
export const dlsForegroundColorNeutralLight = 'rgba(46, 52, 64, 0.9)'
export const dlsForegroundColorNeutralLightHover = 'rgba(73, 86, 107, 0.8)'
export const dlsForegroundColorNeutralLightActive = 'rgba(13, 15, 18, 0.95)'
export const dlsForegroundColorNeutralLightDisabled = 'rgba(83, 101, 138, 0.5)'
export const dlsForegroundColorNeutralWeak = 'rgba(73, 86, 107, 0.8)'
export const dlsForegroundColorNeutralWeakHover = 'rgba(46, 52, 64, 0.9)'
export const dlsForegroundColorNeutralWeakActive = 'rgba(13, 15, 18, 0.95)'
export const dlsForegroundColorNeutralWeakDisabled = 'rgba(90, 119, 168, 0.3)'
export const dlsForegroundColorNeutralDim = 'rgba(83, 101, 138, 0.5)'
export const dlsForegroundColorNeutralDimHover = 'rgba(73, 86, 107, 0.8)'
export const dlsForegroundColorNeutralDimActive = 'rgba(46, 52, 64, 0.9)'
export const dlsForegroundColorNeutralDimDisabled = 'rgba(102, 146, 222, 0.15)'
export const dlsForegroundColorNeutralSubtle = 'rgba(90, 119, 168, 0.3)'
export const dlsForegroundColorNeutralSubtleHover = 'rgba(83, 101, 138, 0.5)'
export const dlsForegroundColorNeutralSubtleActive = 'rgba(73, 86, 107, 0.8)'
export const dlsForegroundColorNeutralSubtleDisabled = 'rgba(102, 146, 222, 0.15)'
export const dlsForegroundColorNeutralHighlightable = 'rgba(13, 15, 18, 0.95)'
export const dlsForegroundColorNeutralHighlightableHover = '#4d79ff'
export const dlsForegroundColorNeutralHighlightableActive = '#333fe6'
export const dlsForegroundColorNeutralHighlightableDisabled = 'rgba(83, 101, 138, 0.5)'
export const dlsForegroundColorInfoPrimary = '#3a5bfd'
export const dlsForegroundColorInfoPrimaryDisabled = '#99beff'
export const dlsForegroundColorSuccessPrimary = '#3ac222'
export const dlsForegroundColorSuccessPrimaryDisabled = '#b2ffa5'
export const dlsForegroundColorWarningPrimary = '#ffba32'
export const dlsForegroundColorWarningPrimaryDisabled = '#ffdf9f'
export const dlsForegroundColorErrorPrimary = '#f53f3f'
export const dlsForegroundColorErrorPrimaryDisabled = '#ffbcbc'
export const dlsForegroundColorHighlighted = '#fea800'
export const dlsBackgroundColorBase1 = '#fff'
export const dlsBackgroundColorBase1Hover = '#f6f7fa'
export const dlsBackgroundColorBase1Active = '#e2e6f0'
export const dlsBackgroundColorBase1Disabled = '#fff'
export const dlsBackgroundColorBase2 = '#f6f7fa'
export const dlsBackgroundColorBase2Hover = '#ebedf5'
export const dlsBackgroundColorBase2Active = '#e2e6f0'
export const dlsBackgroundColorBase2Disabled = '#f6f7fa'
export const dlsBackgroundColorBase3 = '#ebedf5'
export const dlsBackgroundColorBase3Hover = '#e2e6f0'
export const dlsBackgroundColorBase3Active = '#d3d9e6'
export const dlsBackgroundColorBase3Disabled = '#f6f7fa'
export const dlsBackgroundColorPrimary = '#3a5bfd'
export const dlsBackgroundColorPrimaryHover = '#4d79ff'
export const dlsBackgroundColorPrimaryActive = '#333fe6'
export const dlsBackgroundColorPrimaryDisabled = '#e3edff'
export const dlsBorderColorPrimary = 'transparent'
export const dlsBorderColorPrimaryDisabled = 'transparent'
export const dlsBackgroundColorInfoPrimary = '#3a5bfd'
export const dlsBackgroundColorInfoPrimaryHover = '#4d79ff'
export const dlsBackgroundColorInfoPrimaryActive = '#333fe6'
export const dlsBackgroundColorInfoPrimaryDisabled = '#e3edff'
export const dlsBorderColorInfoPrimary = 'transparent'
export const dlsBorderColorInfoPrimaryDisabled = 'transparent'
export const dlsBackgroundColorSuccessPrimary = '#3ac222'
export const dlsBackgroundColorSuccessPrimaryHover = '#59f53d'
export const dlsBackgroundColorSuccessPrimaryActive = '#19a500'
export const dlsBackgroundColorSuccessPrimaryDisabled = '#e3ffde'
export const dlsBorderColorSuccessPrimary = 'transparent'
export const dlsBorderColorSuccessPrimaryDisabled = 'transparent'
export const dlsBackgroundColorWarningPrimary = '#ffba32'
export const dlsBackgroundColorWarningPrimaryHover = '#ffce6f'
export const dlsBackgroundColorWarningPrimaryActive = '#fea800'
export const dlsBackgroundColorWarningPrimaryDisabled = '#fff4dd'
export const dlsBorderColorWarningPrimary = 'transparent'
export const dlsBorderColorWarningPrimaryDisabled = 'transparent'
export const dlsBackgroundColorErrorPrimary = '#f53f3f'
export const dlsBackgroundColorErrorPrimaryHover = '#ff7878'
export const dlsBackgroundColorErrorPrimaryActive = '#d01b1b'
export const dlsBackgroundColorErrorPrimaryDisabled = '#ffe2e2'
export const dlsBorderColorErrorPrimary = 'transparent'
export const dlsBorderColorErrorPrimaryDisabled = 'transparent'
export const dlsBackgroundColorNeutralPrimary = '#a8b0bf'
export const dlsBackgroundColorNeutralPrimaryHover = '#848b99'
export const dlsBackgroundColorNeutralPrimaryActive = '#545b66'
export const dlsBackgroundColorNeutralPrimaryDisabled = '#e2e6f0'
export const dlsBorderColorNeutralPrimary = 'transparent'
export const dlsBorderColorNeutralPrimaryDisabled = 'transparent'
export const dlsBackgroundColorInfo = '#f2f7ff'
export const dlsBackgroundColorInfoHover = '#e3edff'
export const dlsBackgroundColorInfoActive = '#c4daff'
export const dlsBackgroundColorInfoDisabled = '#f2f7ff'
export const dlsBorderColorInfo = '#99beff'
export const dlsBorderColorInfoDisabled = '#c4daff'
export const dlsBackgroundColorSuccess = '#eeffeb'
export const dlsBackgroundColorSuccessHover = '#e3ffde'
export const dlsBackgroundColorSuccessActive = '#cfffc7'
export const dlsBackgroundColorSuccessDisabled = '#eeffeb'
export const dlsBorderColorSuccess = '#b2ffa5'
export const dlsBorderColorSuccessDisabled = '#cfffc7'
export const dlsBackgroundColorWarning = '#fff8eb'
export const dlsBackgroundColorWarningHover = '#fff4dd'
export const dlsBackgroundColorWarningActive = '#ffebc4'
export const dlsBackgroundColorWarningDisabled = '#fff8eb'
export const dlsBorderColorWarning = '#ffdf9f'
export const dlsBorderColorWarningDisabled = '#ffebc4'
export const dlsBackgroundColorError = '#ffebeb'
export const dlsBackgroundColorErrorHover = '#ffe2e2'
export const dlsBackgroundColorErrorActive = '#ffd3d3'
export const dlsBackgroundColorErrorDisabled = '#ffebeb'
export const dlsBorderColorError = '#ffbcbc'
export const dlsBorderColorErrorDisabled = '#ffd3d3'
export const dlsBackgroundColorNeutral = '#ebedf5'
export const dlsBackgroundColorNeutralHover = '#e2e6f0'
export const dlsBackgroundColorNeutralActive = '#d3d9e6'
export const dlsBackgroundColorNeutralDisabled = '#f6f7fa'
export const dlsBorderColorNeutral = '#c5ccdb'
export const dlsBorderColorNeutralDisabled = '#e2e6f0'
export const dlsBackgroundColorCurrent = '#f2f7ff'
export const dlsBackgroundColorCurrentHover = '#e3edff'
export const dlsBackgroundColorCurrentActive = '#c4daff'
export const dlsBackgroundColorCurrentDisabled = '#f2f7ff'
export const dlsBorderColorCurrent = '#3a5bfd'
export const dlsBorderColorCurrentDisabled = '#c4daff'
export const dlsBackgroundColorStrong = '#fff'
export const dlsBackgroundColorStrongHover = '#f2f7ff'
export const dlsBackgroundColorStrongActive = '#e3edff'
export const dlsBackgroundColorStrongDisabled = '#f2f7ff'
export const dlsBorderColorStrong = '#3a5bfd'
export const dlsBorderColorStrongDisabled = '#c4daff'
export const dlsBackgroundColorErrorStrong = '#fff'
export const dlsBackgroundColorErrorStrongHover = '#ffebeb'
export const dlsBackgroundColorErrorStrongActive = '#ffe2e2'
export const dlsBackgroundColorErrorStrongDisabled = '#ffebeb'
export const dlsBorderColorErrorStrong = '#f53f3f'
export const dlsBorderColorErrorStrongDisabled = '#ffd3d3'
export const dlsBackgroundColorFillable = '#fff'
export const dlsBackgroundColorFillableHover = '#ebedf5'
export const dlsBackgroundColorFillableActive = '#d3d9e6'
export const dlsBackgroundColorFillableDisabled = '#f6f7fa'
export const dlsBorderColorFillable = '#d3d9e6'
export const dlsBorderColorFillableHover = '#a8b0bf'
export const dlsBorderColorFillableActive = '#a8b0bf'
export const dlsBorderColorFillableDisabled = '#e2e6f0'
export const dlsBackgroundColorPressable = '#fff'
export const dlsBackgroundColorPressableHover = '#f6f7fa'
export const dlsBackgroundColorPressableActive = '#e2e6f0'
export const dlsBackgroundColorPressableDisabled = '#f6f7fa'
export const dlsBorderColorPressable = '#e2e6f0'
export const dlsBorderColorPressableHover = '#d3d9e6'
export const dlsBorderColorPressableActive = '#d3d9e6'
export const dlsBorderColorPressableDisabled = '#e2e6f0'
export const dlsBackgroundColorTranslucent = 'rgba(0, 0, 0, 0.6)'
export const dlsBackgroundColorTranslucentHover = 'rgba(0, 0, 0, 0.5)'
export const dlsBackgroundColorTranslucentActive = 'rgba(0, 0, 0, 0.7)'
export const dlsBackgroundColorTranslucentDisabled = 'rgba(0, 0, 0, 0.2)'
export const dlsBorderColorTranslucent = 'transparent'
export const dlsBorderColorTranslucentDisabled = 'transparent'
export const dlsBackgroundColorTranslucentReverse = 'transparent'
export const dlsBackgroundColorTranslucentReverseHover = 'rgba(255, 255, 255, 0.15)'
export const dlsBackgroundColorTranslucentReverseActive = 'rgba(255, 255, 255, 0.25)'
export const dlsBackgroundColorTranslucentReverseDisabled = 'transparent'
export const dlsBorderColorTranslucentReverse = 'transparent'
export const dlsBorderColorTranslucentReverseDisabled = 'transparent'
export const dlsBackgroundColorModal = 'rgba(0, 0, 0, 0.6)'
export const dlsLineColor0 = '#ebedf5'
export const dlsLineColor1 = '#e2e6f0'
export const dlsLineColor2 = '#d3d9e6'
export const dlsBorderRadius0 = '4px'
export const dlsBorderRadius1 = '6px'
export const dlsBorderRadius2 = '10px'
export const dlsBorderRadius3 = '16px'
export const dlsBorderRadius4 = '26px'
export const dlsShadowColor = '#0047c2'
export const dlsShadow1 = '0 4px 10px rgba(0, 71, 194, 0.03), 0 3px 9px rgba(0, 71, 194, 0.02), 0 2px 8px rgba(0, 71, 194, 0.01)'
export const dlsShadow2 = '0 6px 28px 2px rgba(0, 71, 194, 0.04), 0 4px 26px 2px rgba(0, 71, 194, 0.03), 0 2px 24px 1px rgba(0, 71, 194, 0.02)'
export const dlsShadow3 = '0 6px 32px 2px rgba(0, 71, 194, 0.06), 0 5px 30px 1px rgba(0, 71, 194, 0.05), 0 4px 28px 1px rgba(0, 71, 194, 0.04)'
export const dlsShadow4 = '0 1px 8px rgba(0, 71, 194, 0.06), 0 7px 14px 2px rgba(0, 71, 194, 0.05), 0 8px 16px 4px rgba(0, 71, 194, 0.04)'
export const dlsShadowColorLight = '#000'
export const dlsShadowLight1 = '0 4px 10px rgba(0, 0, 0, 0.03), 0 3px 9px rgba(0, 0, 0, 0.02), 0 2px 8px rgba(0, 0, 0, 0.01)'
export const dlsShadowLight2 = '0 6px 28px 2px rgba(0, 0, 0, 0.04), 0 4px 26px 2px rgba(0, 0, 0, 0.03), 0 2px 24px 1px rgba(0, 0, 0, 0.02)'
export const dlsShadowLight3 = '0 6px 32px 2px rgba(0, 0, 0, 0.06), 0 5px 30px 1px rgba(0, 0, 0, 0.05), 0 4px 28px 1px rgba(0, 0, 0, 0.04)'
export const dlsShadowOpacityFocus = '0.2'
export const dlsShadowSpreadFocus = '2px'
export const dlsShadowColorFocus = '#4d79ff'
export const dlsShadowColorErrorFocus = '#ff7878'
export const dlsShadowFocus = '0 0 0 2px rgba(77, 121, 255, 0.2)'
export const dlsShadowErrorFocus = '0 0 0 2px rgba(255, 120, 120, 0.2)'
export const dlsBackgroundColorFocus = '#e3edff'
export const dlsBorderColorFocus = '#3a5bfd'
export const dlsBorderColorErrorFocus = '#f53f3f'
export const dlsIconSizeNormal = 'calc(1em + 2px)'
export const dlsIconSizeMinor = 'calc(1em + 2px)'
export const dlsIconSizeAux = 'calc(1em + 2px)'
export const dlsIconColorAux = 'rgba(73, 86, 107, 0.8)'
export const dlsIconColorAuxDisabled = 'rgba(83, 101, 138, 0.5)'
export const dlsTransitionDurationInstant = '100ms'
export const dlsTransitionDurationFast = '200ms'
export const dlsTransitionDurationNormal = '300ms'
export const dlsTransitionDurationSlow = '400ms'
export const dlsTransitionTimingFunctionEase = 'ease'
export const dlsTransitionTimingFunctionEaseIn = 'ease-in'
export const dlsTransitionTimingFunctionEaseOut = 'ease-out'
export const dlsTransitionTimingFunctionLinear = 'linear'
