// less-plugin-dls@11.12.0/variables.palette.js
export const dlsColorBrand = '#0054e6'
export const dlsColorBrand0 = '#fff'
export const dlsColorBrand1 = '#ebf2ff'
export const dlsColorBrand2 = '#dce9ff'
export const dlsColorBrand3 = '#bfd7ff'
export const dlsColorBrand4 = '#95bcff'
export const dlsColorBrand5 = '#5e99ff'
export const dlsColorBrand6 = '#1a6eff'
export const dlsColorBrand7 = '#0054e6'
export const dlsColorBrand8 = '#0047c4'
export const dlsColorBrand9 = '#003796'
export const dlsColorBrand10 = '#00225c'
export const dlsColorBrand11 = '#000'
export const dlsColorInfo = '#0054e6'
export const dlsColorInfo0 = '#fff'
export const dlsColorInfo1 = '#ebf2ff'
export const dlsColorInfo2 = '#dce9ff'
export const dlsColorInfo3 = '#bfd7ff'
export const dlsColorInfo4 = '#95bcff'
export const dlsColorInfo5 = '#5e99ff'
export const dlsColorInfo6 = '#1a6eff'
export const dlsColorInfo7 = '#0054e6'
export const dlsColorInfo8 = '#0047c4'
export const dlsColorInfo9 = '#003796'
export const dlsColorInfo10 = '#00225c'
export const dlsColorInfo11 = '#000'
export const dlsColorSuccess = '#009940'
export const dlsColorSuccess0 = '#fff'
export const dlsColorSuccess1 = '#ebfff3'
export const dlsColorSuccess2 = '#daffe9'
export const dlsColorSuccess3 = '#baffd7'
export const dlsColorSuccess4 = '#8bffbb'
export const dlsColorSuccess5 = '#4dff97'
export const dlsColorSuccess6 = '#00cc55'
export const dlsColorSuccess7 = '#009940'
export const dlsColorSuccess8 = '#008236'
export const dlsColorSuccess9 = '#00632a'
export const dlsColorSuccess10 = '#003d1a'
export const dlsColorSuccess11 = '#000'
export const dlsColorWarning = '#f27318'
export const dlsColorWarning0 = '#fff'
export const dlsColorWarning1 = '#fff3eb'
export const dlsColorWarning2 = '#ffecdf'
export const dlsColorWarning3 = '#ffdfc9'
export const dlsColorWarning4 = '#ffcca8'
export const dlsColorWarning5 = '#ffb37d'
export const dlsColorWarning6 = '#ff9448'
export const dlsColorWarning7 = '#f27318'
export const dlsColorWarning8 = '#ce5600'
export const dlsColorWarning9 = '#9d4200'
export const dlsColorWarning10 = '#612800'
export const dlsColorWarning11 = '#000'
export const dlsColorError = '#d9150b'
export const dlsColorError0 = '#fff'
export const dlsColorError1 = '#ffeceb'
export const dlsColorError2 = '#ffdedc'
export const dlsColorError3 = '#ffc4c0'
export const dlsColorError4 = '#ff9d98'
export const dlsColorError5 = '#ff6a63'
export const dlsColorError6 = '#ff2b20'
export const dlsColorError7 = '#d9150b'
export const dlsColorError8 = '#b80900'
export const dlsColorError9 = '#8d0700'
export const dlsColorError10 = '#570400'
export const dlsColorError11 = '#000'
export const dlsColorGray0 = '#fff'
export const dlsColorGray1 = '#f6f7fa'
export const dlsColorGray2 = '#ebedf5'
export const dlsColorGray3 = '#e2e6f0'
export const dlsColorGray4 = '#d3d9e6'
export const dlsColorGray5 = '#c5ccdb'
export const dlsColorGray6 = '#a8b0bf'
export const dlsColorGray7 = '#848b99'
export const dlsColorGray8 = '#545b66'
export const dlsColorGray9 = '#191b1e'
export const dlsColorGray10 = '#0a0b0d'
export const dlsColorGray11 = '#000'
export const dlsColorTranslucent0 = 'transparent'
export const dlsColorTranslucent1 = 'rgba(109, 159, 247, 0.07)'
export const dlsColorTranslucent2 = 'rgba(108, 156, 240, 0.1)'
export const dlsColorTranslucent3 = 'rgba(102, 146, 222, 0.15)'
export const dlsColorTranslucent4 = 'rgba(96, 134, 199, 0.2)'
export const dlsColorTranslucent5 = 'rgba(90, 119, 168, 0.3)'
export const dlsColorTranslucent6 = 'rgba(83, 101, 138, 0.5)'
export const dlsColorTranslucent7 = 'rgba(73, 86, 107, 0.8)'
export const dlsColorTranslucent8 = 'rgba(46, 52, 64, 0.9)'
export const dlsColorTranslucent9 = 'rgba(13, 15, 18, 0.95)'
export const dlsColorTranslucent10 = 'rgba(4, 4, 5, 0.95)'
export const dlsColorTranslucent11 = '#000'
