// @baidu/one-ui@4.41.1/es/components/link/navLink.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

/**
 * @file 支持react-router的link
 * @author huangshiming@baidu.com
 * @date 2020-11-26
 */
import React, { forwardRef } from 'react';
import Link from './textLink';

function isModifiedEvent(event) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
}

var NavLink = /*#__PURE__*/forwardRef(function (props, ref) {
  var target = props.target,
      onClick = props.onClick,
      navigate = props.navigate,
      lightPrefix = props.lightPrefix,
      restProps = _objectWithoutPropertiesLoose(props, ["target", "onClick", "navigate", "lightPrefix"]);

  var handleOnClick = function handleOnClick(event) {
    try {
      if (onClick) {
        onClick(event);
      }
    } catch (ex) {
      event.preventDefault();
      throw ex;
    }

    if (!event.defaultPrevented && event.button === 0 && (!target || target === '_self') && !isModifiedEvent(event)) {
      event.preventDefault();
      navigate();
    }
  };

  return /*#__PURE__*/React.createElement(Link, _extends({}, restProps, {
    target: target,
    isAtag: true,
    ref: ref,
    onClick: handleOnClick
  }));
});
export default NavLink;
