// @baidu/one-ui@4.32.1/src/components/popover/popover.tsx
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import Tooltip from '../tooltip';
import {classnames} from '../../core/commonTools';
import {PopoverProps} from './interface';
import {withConfigConsumer} from '../providerConfig/context';

@withConfigConsumer('popover')
class Popover extends PureComponent<PopoverProps> {

    static defaultProps = {
        prefixCls: 'one-popover',
        placement: 'bottom',
        // transitionName: 'zoom-big',
        size: 'medium',
        trigger: 'hover',
        mouseEnterDelay: 0.1,
        mouseLeaveDelay: 0.1,
        overlayStyle: {}
    };

    getPopupDomNode() {
        return this.tooltip.getPopupDomNode();
    }

    getOverlay() {
        const {title, prefixCls, content} = this.props;

        if (!title && !content) {
            return null;
        }

        return (
            <>
                {title && <div className={`${prefixCls}-title`}>{title}</div>}
                {
                    content && (
                        <div className={`${prefixCls}-inner-content`}>
                            {content}
                        </div>
                    )
                }
            </>
        );
    }

    tooltip;

    saveTooltip = node => {
        this.tooltip = node;
    }

    render() {
        const props = {...this.props};
        delete props.title;
        const {
            prefixCls,
            size,
            overlayClassName
        } = props;
        return (
            <Tooltip
                {...props}
                overlayClassName={classnames(`${prefixCls}-${size}`, overlayClassName)}
                ref={this.saveTooltip}
                overlay={this.getOverlay()}
            />
        );
    }
}

export default Popover;
