// @baidu/one-ui@4.32.1/es/components/radio/button.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
import Radio from './radio';
import Context from './context';

var RadioButton = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(RadioButton, _PureComponent);

  function RadioButton() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = RadioButton.prototype;

  _proto.render = function render() {
    var _this = this;

    return /*#__PURE__*/React.createElement(Context.Consumer, null, function (context) {
      return _this.renderRadio(context);
    });
  };

  _proto.renderRadio = function renderRadio(context) {
    var _this$props = this.props,
        type = _this$props.type,
        props = _objectWithoutPropertiesLoose(_this$props, ["type"]);

    return /*#__PURE__*/React.createElement(Radio, _extends({}, props, {
      type: type || (context == null ? void 0 : context.type) || 'strong'
    }));
  };

  return RadioButton;
}(PureComponent);

RadioButton.defaultProps = {
  prefixCls: 'one-radio-button',
  size: 'medium'
};
export default RadioButton;
