// @baidu/one-ui@4.41.1/es/components/radio/button.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import Radio from './radio';
import Context from './context';
import { withConfigConsumer } from '../providerConfig/context';
var RadioButton = (_dec = withConfigConsumer('radio-button'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(RadioButton, _PureComponent);

  function RadioButton() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = RadioButton.prototype;

  _proto.render = function render() {
    var _this = this;

    return /*#__PURE__*/React.createElement(Context.Consumer, null, function (context) {
      return _this.renderRadio(context);
    });
  };

  _proto.renderRadio = function renderRadio(context) {
    var _this$props = this.props,
        type = _this$props.type,
        props = _objectWithoutPropertiesLoose(_this$props, ["type"]);

    return /*#__PURE__*/React.createElement(Radio, _extends({}, props, {
      type: type || (context == null ? void 0 : context.type) || 'strong'
    }));
  };

  return RadioButton;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-radio-button',
  size: 'medium'
}, _temp)) || _class);
export default RadioButton;
