// @baidu/one-ui@4.32.1/src/components/transition/interface.ts
import {BasePropsWithChildren} from '../interface';

export interface TransitionProps extends BasePropsWithChildren {

    /**
     * 进入或退出
     */
    in?: boolean;

    /**
     * 渲染时动效
     */
    appear?: boolean;

    /**
     * 超时
     */
    timeout?: number;

    /**
     * 动效名称
     */
    name?:
        'fade'
        | 'scale-fade'
        | 'slide-left'
        | 'slide-right'
        | 'slide-up'
        | 'slide-down'
        | 'translate'
        | 'translate-scale-fade'
        | 'translate-gravity'
        | 'move'
        | string;

    /**
     * 退出时卸载
     */
    unmountOnExit?: boolean;

    /**
     * 进入开始回调
     */
    onEnter?: (node: HTMLElement, isAppearing: boolean) => void;
    /**
     * 进入中回调
     */
    onEntering?: (node: HTMLElement, isAppearing: boolean) => void;
    /**
     * 进入完毕回调
     */
    onEntered?: (node: HTMLElement, isAppearing: boolean) => void;

    /**
     * 退出开始回调
     */
    onExit?: (node: HTMLElement) => void;

    /**
     * 退出中回调
     */
    onExiting?: (node: HTMLElement) => void;

    /**
     * 退出完毕回调
     */
    onExited?: (node: HTMLElement) => void;
};
