// @baidu/one-ui@4.32.1/es/components/embedded/embedded.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { IconTimes } from 'dls-icons-react';
import Button from '../button';
import { classnames } from '../../core/commonTools';
/**
 * @deprecated
 */

var Embedded = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Embedded, _PureComponent);

  function Embedded(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.onClose = function () {
      _this.setState({
        visible: false
      });

      _this.props.onClose();
    };

    _this.renderFooter = function () {
      var _this$props = _this.props,
          footer = _this$props.footer,
          onOk = _this$props.onOk,
          onCancel = _this$props.onCancel,
          okProps = _this$props.okProps,
          cancelProps = _this$props.cancelProps,
          okOrder = _this$props.okOrder,
          cancelOrder = _this$props.cancelOrder,
          buttonSize = _this$props.buttonSize,
          size = _this$props.size,
          okText = _this$props.okText,
          cancelText = _this$props.cancelText,
          hideDefaultFooter = _this$props.hideDefaultFooter;

      if (footer) {
        return footer;
      }

      var okStyle = okProps.style ? _extends({}, okProps.style, {
        order: okOrder
      }) : {
        order: okOrder
      };
      var cancelStyle = cancelProps.style ? _extends({}, cancelProps.style, {
        order: cancelOrder
      }) : {
        order: cancelOrder
      };
      var defaultFooter = [/*#__PURE__*/React.createElement(Button, _extends({
        size: buttonSize || size,
        key: "confirm",
        onClick: onOk,
        type: "primary"
      }, okProps, {
        style: okStyle
      }), okText), /*#__PURE__*/React.createElement(Button, _extends({
        size: buttonSize || size,
        onClick: onCancel,
        key: "cancel",
        type: "normal"
      }, cancelProps, {
        style: cancelStyle
      }), cancelText)];
      return !hideDefaultFooter ? defaultFooter : null;
    };

    _this.state = {
      visible: true
    };
    return _this;
  }

  var _proto = Embedded.prototype;

  _proto.render = function render() {
    var _this$props2 = this.props,
        title = _this$props2.title,
        prefixCls = _this$props2.prefixCls,
        className = _this$props2.className,
        closable = _this$props2.closable,
        children = _this$props2.children,
        size = _this$props2.size,
        position = _this$props2.position;
    var visible = this.state.visible;

    if ('visible' in this.props) {
      visible = this.props.visible;
    }

    if (!visible) {
      return null;
    }

    var footer = this.renderFooter();
    var embeddedClassNames = classnames(prefixCls, className, prefixCls + "-" + size);
    return /*#__PURE__*/React.createElement("div", {
      className: embeddedClassNames
    }, closable ? /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-close",
      onClick: this.onClose
    }, /*#__PURE__*/React.createElement(IconTimes, {
      className: prefixCls + "-close-icon"
    })) : null, /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-header"
    }, title), /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-body"
    }, children), footer ? /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-footer " + prefixCls + "-" + position
    }, footer) : null);
  };

  return Embedded;
}(PureComponent);

Embedded.propTypes = {
  /** 面板关闭时触发 */
  onClose: PropTypes.func,

  /** 传入底部的button数组 */
  footer: PropTypes.array,

  /** 面板的标题 */
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),

  /** 面板是否可以关闭 */
  closable: PropTypes.bool,

  /** 类名前缀 */
  prefixCls: PropTypes.string,

  /** 自定义类名 */
  className: PropTypes.string,
  children: PropTypes.node,

  /** 面板是否可见 */
  visible: PropTypes.bool,
  onOk: PropTypes.func,
  onCancel: PropTypes.func,
  okProps: PropTypes.object,
  cancelProps: PropTypes.object,
  okOrder: PropTypes.number,
  cancelOrder: PropTypes.number,

  /** 按钮的类型 */
  buttonSize: PropTypes.oneOf(['small', 'medium', 'large', 'xsmall', 'xlarge']),
  size: PropTypes.oneOf(['small', 'medium']),
  okText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  cancelText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  hideDefaultFooter: PropTypes.bool,
  position: PropTypes.oneOf(['left', 'center', 'right'])
};
Embedded.defaultProps = {
  footer: null,
  closable: true,
  prefixCls: 'one-embedded',
  className: '',
  onOk: function onOk() {},
  onCancel: function onCancel() {},
  okProps: {},
  cancelProps: {},
  okOrder: 1,
  cancelOrder: 2,
  size: 'medium',
  okText: '确定',
  cancelText: '取消',
  hideDefaultFooter: true,
  position: 'left'
};
export { Embedded as default };
