// @baidu/one-ui@4.32.1/src/components/searchBox/interface.ts
import {ReactNode, ChangeEventHandler} from "react";
import {BaseProps} from "../interface";
import {MenuProps} from "../menu";

export type SearchBoxOption = {

    /**
     * 文本
     */
    label: ReactNode;

    /**
     * 值
     */
    value: string;

    /**
     * 禁用
     */
    disabled?: boolean;
}

export interface SearchBoxProps extends BaseProps,
    Omit<JSX.IntrinsicElements['input'], 'size' | 'type' | 'ref'>  {

        /**
         * 尺寸
         */
        size?: 'xsmall' | 'small' | 'medium' | 'large';

        /**
         * 搜索框按钮宽度
         */
        buttonWidth?: number;

        /**
         * 搜索框宽度
         *
         * @deprecated use `className` or `style`
         */
        width?: number | string;

        /**
         * 搜索框高度
         *
         * @deprecated use `className` or `style`
         */
        height?: number;

        /**
         * 搜索框是否可用
         */
        disabled?: boolean;

        /**
         * 搜索框默认值
         */
        defaultValue?: string;

        /**
         * 搜索框值
         */
        value?: string;

        /**
         * 回车按键回调函数
         */
        onPressEnter?(e: {
            target: HTMLInputElement
        }): void;

        /**
         * 值改变时回调函数
         */
        onChange?: ChangeEventHandler<HTMLInputElement>;

        /**
         * 点击搜索时回调函数
         */
        onSearch?(e: {
            target: HTMLInputElement
        }): void;

        /**
         * 点击清空按钮时回调函数
         */
        onClearClick?(e: {
            target: HTMLInputElement
        });

        /**
         * 面板展开/收起时回调函数
         */
        onVisibleChange?(visible: boolean);

        /**
         * 值为空时占位字符串
         */
        placeholder?: string;

        /**
         * @deprecated
         * @internal
         * 值为空，点击搜索时代入输入框的值
         */
        defaultQuery?: {
            prefix: string;
            value: string;
        };

        /**
         * 下拉框中数据源
         */
        options?: SearchBoxOption[];

        /**
         * 下拉框高度
         */
        dropdownHeight?: string | number;

        /**
         * 是否展现下拉框 为true时，options或overlay不为空时，才出现下拉框
         * @default false
         */
        isShowDropDown?: boolean;

        /**
         * 是否展示搜索icon
         */
        showSearchIcon?: boolean;

        /**
         * @deprecated
         */
        isShowSearchIcon?: boolean;

        /**
         * 是否展示关闭icon
         */
        showCloseIcon?: boolean;

        /**
         * 下拉面板等宽
         */
        dropdownMatchSelectWidth?: boolean;

        /**
         * 下拉框中item点击回调函数
         */
        handleMenuClick?: MenuProps['onClick'];

        /**
         * 下拉框内容自定义render，需要返回dom结构
         */
        overlay?: ReactNode | (() => ReactNode);

        /**
         * 搜索功能的按钮样式
         */
        searchIconType?: 'icon' | 'button' | 'custom';

        /**
         * @deprecated
         * @internal
         * 自定义图标
         */
        customRender?: ReactNode;

        /**
         * 搜索框是否是可读状态
         */
        readOnly?: boolean;
    };
