// @baidu/one-ui@4.41.1/lib/components/dialog/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {};
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _dialog = _interopRequireDefault(require("./dialog"));

var _confirm = _interopRequireDefault(require("./confirm"));

var _toast = _interopRequireDefault(require("../toast"));

var _interface = require("./interface");

Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  exports[key] = _interface[key];
});

/**
 * @file Dialog
 * @author huangshiming@baidu.com
 * @date 2020-05-19
 */
_dialog["default"].confirm = function (props) {
  var config = (0, _extends2["default"])({
    type: 'confirm',
    okCancel: true,
    okType: 'primary',
    okText: '确定',
    cancelText: '取消',
    cancelType: 'normal',
    size: 'medium'
  }, props);
  return (0, _confirm["default"])(config);
};

_dialog["default"].alert = function (props) {
  var config = (0, _extends2["default"])({
    type: 'alert',
    okCancel: false,
    okType: 'primary',
    okText: '确定',
    buttonPosition: 'center',
    size: 'medium'
  }, props);
  return (0, _confirm["default"])(config);
};

_dialog["default"].toast = _toast["default"];
var _default = _dialog["default"];
exports["default"] = _default;
module.exports = exports.default;
