// @baidu/one-ui@4.32.1/src/components/textArea/interface.ts
import {ComponentPropsWithoutRef, TextareaHTMLAttributes} from "react";
import {BaseProps} from "../interface";

export interface TextAreaProps extends BaseProps, Omit<ComponentPropsWithoutRef<'textarea'>, 'onChange' | 'onFocus' | 'onBlur'> {

    /**
     * 尺寸
     */
    size?: 'small' | 'medium';

    /**
     * 输入框值-非受控
     */
    defaultValue?: string;

    /**
     * 输入框值-受控
     */
    value?: string;

    /**
     * placeholder
     */
    placeholder?: string;

    /**
     * 输入变化回调
     */
    onChange?: (e: {
        value: string;
        errorMessage: string;
        target: HTMLTextAreaElement;
    }) => void;

    /**
     * 禁用
     */
    disabled?: boolean;

    /**
     * 只读
     */
    readOnly?: boolean;

    /**
     * 宽度
     */
    width?: number | string;

    /**
     * 最小可调节的高度
     */
    minRows?: number;

    /**
     * 最大可调节的高度
     */
    maxRows?: number;

    /**
     * 最短字符长度
     */
    minLen?: number;

    /**
     * 当比最短字符短时的错误信息
     */
    minLenErrorMessage?: string;

    /**
     * 最长字符长度
     */
    maxLen?: number;

    /**
     * 当比最长字符长时的错误信息
     */
    maxLenErrorMessage?: string;

    /**
     * 是否必填
     */
    isRequired?: boolean;

    /**
     * 当为空时的错误信息
     */
    requiredErrorMessage?: string;

    /**
     * 错误信息
     */
    errorMessage?: string;

    /**
     * 错误信息；当不传此值时，会内部校验是否非空、最短字符、最长字符；当传此值时，错误信息以此值为准
     */
    showErrorMessage?: boolean;

    /**
     * 是否报错，在没有errorMessage的情况下
     */
    showErrorWithoutErrorMessage?: boolean;

    /**
     * 失去焦点
     */
    onBlur?: (e: {
        value: string;
        errorMessage: string;
        event: FocusEvent
    }) => void;

    /**
     * 获得焦点
     */
    onFocus?: (e: {
        value: string;
        errorMessage: string;
        event: FocusEvent
    }) => void;

    /**
     * 提示信息
     */
    tipText?: string;

    /**
     * 信息位置
     * @deprecated
     */
    location?: 'right' | 'layer' | 'bottom';

    /**
     * 提示信息位置
     * @deprecated
     */
    tipLocation?: 'right' | 'layer' | 'bottom';

    /**
     * 错误信息位置
     * @deprecated
     */
    errorLocation?: 'right' | 'layer' | 'bottom';

    /**
     * 计数时不算的字符
     */
    filterArray?: string[];

    /**
     * 计数方式，当cn时，汉字记为2个字符，当en时，汉字记为1个字符。默认为cn
     */
    countMode?: 'cn' | 'en';

    /**
     * 自定义计数方式
     */
    getLength?: (value: string) => number;

    /**
     * @deprecated
     * 挂载在原生的textarea上的属性，比较粗暴，不建议使用
     */
    originTextAreaProps?: TextareaHTMLAttributes<HTMLTextAreaElement>;

    /**
     * @deprecated
     */
    containerRef?;
};
