// @baidu/one-ui@4.41.1/lib/components/popover/popover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _tooltip = _interopRequireDefault(require("../tooltip"));

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

var Popover = (_dec = (0, _context.withConfigConsumer)('popover'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Popover, _PureComponent);

  function Popover() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.tooltip = void 0;

    _this.saveTooltip = function (node) {
      _this.tooltip = node;
    };

    return _this;
  }

  var _proto = Popover.prototype;

  _proto.getPopupDomNode = function getPopupDomNode() {
    return this.tooltip.getPopupDomNode();
  };

  _proto.getOverlay = function getOverlay() {
    var _this$props = this.props,
        title = _this$props.title,
        prefixCls = _this$props.prefixCls,
        content = _this$props.content;

    if (!title && !content) {
      return null;
    }

    return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, title && /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-title"
    }, title), content && /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-inner-content"
    }, content));
  };

  _proto.render = function render() {
    var props = (0, _extends2["default"])({}, this.props);
    delete props.title;
    var prefixCls = props.prefixCls,
        size = props.size,
        overlayClassName = props.overlayClassName;
    return /*#__PURE__*/_react["default"].createElement(_tooltip["default"], (0, _extends2["default"])({}, props, {
      overlayClassName: (0, _commonTools.classnames)(prefixCls + "-" + size, overlayClassName),
      ref: this.saveTooltip,
      overlay: this.getOverlay()
    }));
  };

  return Popover;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-popover',
  placement: 'bottom',
  // transitionName: 'zoom-big',
  size: 'medium',
  trigger: 'hover',
  mouseEnterDelay: 0.1,
  mouseLeaveDelay: 0.1,
  overlayStyle: {}
}, _temp)) || _class);
var _default = Popover;
exports["default"] = _default;
module.exports = exports.default;
