// @baidu/one-ui@4.41.1/lib/components/sidenav/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
var _exportNames = {};
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _menu = _interopRequireDefault(require("../menu"));

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _interface = require("./interface");

Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  exports[key] = _interface[key];
});

var _dec, _class, _class2, _temp;

var Sidenav = (_dec = (0, _context.withConfigConsumer)('side-nav'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Sidenav, _PureComponent);

  function Sidenav() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Sidenav.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        collapsed = _this$props.collapsed,
        prefixCls = _this$props.prefixCls,
        className = _this$props.className,
        lightPrefix = _this$props.lightPrefix,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, ["collapsed", "prefixCls", "className", "lightPrefix"]);
    return /*#__PURE__*/_react["default"].createElement(_menu["default"], (0, _extends2["default"])({}, restProps, {
      mode: "inline",
      inlineCollapsed: collapsed,
      className: (0, _commonTools.classnames)(prefixCls, className)
    }));
  };

  return Sidenav;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-side-nav',
  collapsed: false,
  paddingLeft: 24
}, _temp)) || _class);
var _default = Sidenav;
exports["default"] = _default;
module.exports = exports.default;
