// @baidu/one-ui@4.41.1/lib/components/datePicker/common/calendar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _partial2 = _interopRequireDefault(require("lodash/partial"));

var _react = _interopRequireWildcard(require("react"));

var _dayjs = _interopRequireDefault(require("dayjs"));

var _datePickerTools = require("../../../core/datePickerTools");

var _commonTools = require("../../../core/commonTools");

var betweenTime = function betweenTime(startTime, endTime, time) {
  return startTime && endTime && time >= startTime && time <= endTime;
};

var normalizeWeekDays = function normalizeWeekDays(day, _ref, selectedValue) {
  var minTime = _ref.minTime,
      maxTime = _ref.maxTime,
      validator = _ref.validator;
  var dayTime = (0, _datePickerTools.getTimeTramp)(day.value);
  var disabled = false;

  if (dayTime > maxTime || dayTime < minTime) {
    disabled = true;
  } else if (validator && typeof validator === 'function') {
    disabled = validator({
      selectedValue: selectedValue,
      dayItem: day.value,
      timeStamp: dayTime,
      getTimeStamp: _datePickerTools.getTimeTramp
    });
  }

  day.disabled = disabled;
  return day;
};

var RangeDayItemRender = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(RangeDayItemRender, _PureComponent);

  function RangeDayItemRender() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.cache = {};

    _this.memoize = function (param) {
      if (!this.cache[param]) {
        var date = param.split('/');
        var result = (0, _datePickerTools.formatPerMonthInDay)({
          year: date[0],
          month: date[1]
        });
        this.cache[param] = result;
      }

      return this.cache[param];
    };

    _this.onClickDay = function (value, disabled, readOnly) {
      if (disabled) {
        return;
      }

      var _this$props = _this.props,
          onChange = _this$props.onChange,
          dateFormat = _this$props.dateFormat,
          multiple = _this$props.multiple;
      onChange((0, _dayjs["default"])(value).format(dateFormat), multiple && readOnly);
    };

    _this.onMouseEnter = function (value, disabled) {
      if (disabled) {
        return;
      }

      _this.props.onMouseEnter(value);
    };

    _this.onMouseLeave = function (value, disabled) {
      if (disabled) {
        return;
      }

      _this.props.onMouseLeave(value);
    };

    _this.renderMonthItem = function () {
      var _this$props2 = _this.props,
          showYear = _this$props2.showYear,
          showMonth = _this$props2.showMonth,
          prefixCls = _this$props2.prefixCls,
          beginDate = _this$props2.beginDate,
          validateMinDate = _this$props2.validateMinDate,
          validateMaxDate = _this$props2.validateMaxDate,
          endDate = _this$props2.endDate,
          hoverDate = _this$props2.hoverDate,
          mode = _this$props2.mode,
          multiple = _this$props2.multiple,
          validateDisabled = _this$props2.validateDisabled;

      var perMonthInDay = _this.memoize(showYear + "/" + showMonth);

      var selectedValue = [beginDate, endDate].filter(Boolean);
      var minTime = (0, _datePickerTools.getTimeTramp)(validateMinDate);
      var maxTime = (0, _datePickerTools.getTimeTramp)(validateMaxDate);
      var validation = {
        minTime: minTime,
        maxTime: maxTime,
        validator: validateDisabled
      };
      var weeks = (0, _datePickerTools.formatWeek)(perMonthInDay).map(function (days) {
        return days.map(function (day) {
          return normalizeWeekDays(day, validation, selectedValue);
        });
      });
      var itemDaysClx = prefixCls + "-body-month";
      var todayTime = (0, _datePickerTools.getTimeTramp)((0, _dayjs["default"])().startOf('day'));
      var startTime = beginDate ? (0, _datePickerTools.getTimeTramp)(beginDate) : null;
      var endTime = endDate ? (0, _datePickerTools.getTimeTramp)(endDate) : startTime;
      var singleSelected = multiple && beginDate && !endDate;
      var hoverTime = hoverDate ? (0, _datePickerTools.getTimeTramp)(hoverDate) : null;

      if (singleSelected && hoverTime && hoverTime < startTime && mode === 'week') {
        startTime = Math.min(maxTime, (0, _datePickerTools.getTimeTramp)((0, _dayjs["default"])(endTime).endOf('week').startOf('day')));
        endTime = startTime;
      }

      var hoverStartTime = hoverTime;
      var hoverEndTime = hoverTime;

      if (mode === 'week' && hoverTime) {
        hoverStartTime = Math.max(minTime, (0, _datePickerTools.getTimeTramp)((0, _dayjs["default"])(hoverTime).startOf('week')));
        hoverEndTime = Math.min(maxTime, (0, _datePickerTools.getTimeTramp)((0, _dayjs["default"])(hoverTime).endOf('week').startOf('day')));
      }

      var rangeStartTime = singleSelected && hoverStartTime ? Math.min(startTime, hoverStartTime) : startTime;
      var rangeEndTime = singleSelected && hoverEndTime ? Math.max(endTime, hoverEndTime) : endTime;
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: itemDaysClx
      }, weeks.map(function (days, index) {
        return /*#__PURE__*/_react["default"].createElement("div", {
          key: "week-" + index,
          className: itemDaysClx + "-week"
        }, days.map(function (_ref2) {
          var _classnames;

          var value = _ref2.value,
              label = _ref2.label,
              disabled = _ref2.disabled,
              isCurrentMonth = _ref2.isCurrentMonth;
          var dayTime = (0, _datePickerTools.getTimeTramp)(value);
          var readOnly = !isCurrentMonth;
          var range = betweenTime(rangeStartTime, rangeEndTime, dayTime) && !readOnly && !disabled;
          var startSelected = dayTime === startTime && !readOnly;
          var endSelected = dayTime === endTime && !readOnly;
          var selected = startSelected || endSelected;
          var className = (0, _commonTools.classnames)(itemDaysClx + "-item", (_classnames = {}, _classnames[itemDaysClx + "-item-read-only"] = readOnly, _classnames[itemDaysClx + "-item-today"] = dayTime === todayTime && !readOnly, _classnames[itemDaysClx + "-item-disabled"] = disabled, _classnames[itemDaysClx + "-item-selected"] = selected, _classnames[itemDaysClx + "-item-selected-start"] = startSelected, _classnames[itemDaysClx + "-item-selected-end"] = endSelected, _classnames[itemDaysClx + "-item-range"] = range, _classnames[itemDaysClx + "-item-range-start"] = rangeStartTime === dayTime && !readOnly && !disabled, _classnames[itemDaysClx + "-item-range-end"] = rangeEndTime === dayTime && !readOnly && !disabled, _classnames[itemDaysClx + "-item-hover"] = !singleSelected && !readOnly && betweenTime(hoverStartTime, hoverEndTime, dayTime), _classnames[itemDaysClx + "-item-hover-end"] = !singleSelected && !readOnly && hoverEndTime === dayTime, _classnames));
          return /*#__PURE__*/_react["default"].createElement("span", {
            key: value,
            className: className,
            onClick: (0, _partial2["default"])(_this.onClickDay, value, disabled, readOnly),
            onMouseEnter: (0, _partial2["default"])(_this.onMouseEnter, value, disabled),
            onMouseLeave: (0, _partial2["default"])(_this.onMouseLeave, value, disabled)
          }, /*#__PURE__*/_react["default"].createElement("span", null, label));
        }));
      }));
    };

    return _this;
  }

  var _proto = RangeDayItemRender.prototype;

  _proto.render = function render() {
    return this.renderMonthItem();
  };

  return RangeDayItemRender;
}(_react.PureComponent);

exports["default"] = RangeDayItemRender;
RangeDayItemRender.defaultProps = {
  onChange: function onChange() {},
  onMouseEnter: function onMouseEnter() {},
  onMouseLeave: function onMouseLeave() {}
};
module.exports = exports.default;
