// @baidu/one-ui@4.41.1/lib/components/datePicker/common/monthPanel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _partial2 = _interopRequireDefault(require("lodash/partial"));

var _react = _interopRequireWildcard(require("react"));

var _miniStore = require("mini-store");

var _dayjs = _interopRequireDefault(require("dayjs"));

var _datePickerTools = require("../../../core/datePickerTools");

var _commonTools = require("../../../core/commonTools");

var monthItems = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];
;
;

var MonthRender = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(MonthRender, _PureComponent);

  function MonthRender(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.store = void 0;

    _this.onClickMonth = function (month, disabled) {
      if (disabled) {
        return;
      }

      var newState = {};
      var _this$props = _this.props,
          type = _this$props.type,
          isMonthRender = _this$props.isMonthRender,
          showYear = _this$props.showYear,
          dateFormat = _this$props.dateFormat;

      if (type === 'nextMultiple') {
        newState.endDateShowMonth = month;
        newState.endDatePanelType = 'date';
      } else {
        newState.showMonth = month;
        newState.panelType = isMonthRender ? 'month' : 'date';
      }

      _this.store.setState(newState);

      if (_this.props.onClickMonth) {
        var _value = (0, _dayjs["default"])(showYear + "/" + month).format(dateFormat);

        _this.props.onClickMonth(_value);
      }
    };

    _this.formatItemClassName = function (showYear, showMonth, month) {
      var _classnames;

      var _this$props2 = _this.props,
          prefixCls = _this$props2.prefixCls,
          type = _this$props2.type,
          endDateShowYear = _this$props2.endDateShowYear,
          endDateShowMonth = _this$props2.endDateShowMonth;
      var _this$props3 = _this.props,
          validateMinDate = _this$props3.validateMinDate,
          validateMaxDate = _this$props3.validateMaxDate;
      var range = (0, _datePickerTools.monthDayRange)(showYear);

      if (type === 'prevMultiple') {
        var endDate = endDateShowYear + "/" + (endDateShowMonth - 1) + "/01";

        if (endDateShowMonth === 1) {
          endDate = endDateShowYear - 1 + "/12/01";
        }

        if ((0, _datePickerTools.getTimeTramp)(endDate) < (0, _datePickerTools.getTimeTramp)(validateMaxDate)) {
          validateMaxDate = endDate;
        }
      } else if (type === 'nextMultiple') {
        var beginDate = _this.props.showYear + "/" + (_this.props.showMonth + 1) + "/01";

        if (_this.props.showMonth === 12) {
          beginDate = _this.props.showYear + 1 + "/01/01";
        }

        if ((0, _datePickerTools.getTimeTramp)(beginDate) > (0, _datePickerTools.getTimeTramp)(validateMinDate)) {
          validateMinDate = beginDate;
        }
      }

      var monthClassName = prefixCls + "-month-container-item";
      var currentMonthLastDay = (0, _datePickerTools.getTimeTramp)(showYear + "/" + month + "/" + range[month - 1]);
      var currentMonthFirstDay = (0, _datePickerTools.getTimeTramp)(showYear + "/" + month + "/01");
      var disabled = currentMonthLastDay < (0, _datePickerTools.getTimeTramp)(validateMinDate) || currentMonthFirstDay > (0, _datePickerTools.getTimeTramp)(validateMaxDate);
      var monthClassNames = (0, _commonTools.classnames)(monthClassName, (_classnames = {}, _classnames[monthClassName + "-selected"] = month === showMonth, _classnames[monthClassName + "-disabled"] = disabled, _classnames[monthClassName + "-last-row"] = month > 9, _classnames));
      return {
        disabled: disabled,
        monthClassNames: monthClassNames
      };
    };

    _this.renderMonth = function () {
      var monthContainers = [];
      var _this$props4 = _this.props,
          showMonth = _this$props4.showMonth,
          showYear = _this$props4.showYear,
          endDateShowYear = _this$props4.endDateShowYear,
          endDateShowMonth = _this$props4.endDateShowMonth,
          type = _this$props4.type;
      var currentMonth = type === 'nextMultiple' ? endDateShowMonth : showMonth;
      var currentYear = type === 'nextMultiple' ? endDateShowYear : showYear;
      monthItems.forEach(function (month) {
        var _this$formatItemClass = _this.formatItemClassName(currentYear, currentMonth, month),
            monthClassNames = _this$formatItemClass.monthClassNames,
            disabled = _this$formatItemClass.disabled;

        monthContainers.push( /*#__PURE__*/_react["default"].createElement("span", {
          className: monthClassNames,
          key: month,
          onClick: (0, _partial2["default"])(_this.onClickMonth, month, disabled)
        }, month + "\u6708"));
      });
      return monthContainers;
    };

    _this.store = _this.props.store;
    return _this;
  }

  var _proto = MonthRender.prototype;

  _proto.render = function render() {
    var _this$props5 = this.props,
        prefixCls = _this$props5.prefixCls,
        panelType = _this$props5.panelType,
        endDatePanelType = _this$props5.endDatePanelType,
        type = _this$props5.type;
    var currentPanelType = type === 'nextMultiple' ? endDatePanelType : panelType;

    if (currentPanelType !== 'month') {
      return null;
    }

    return /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-month-container"
    }, this.renderMonth());
  };

  return MonthRender;
}(_react.PureComponent);

var _default = (0, _miniStore.connect)(function (state) {
  return {
    showYear: state.showYear,
    showMonth: state.showMonth,
    endDateShowYear: state.endDateShowYear,
    endDateShowMonth: state.endDateShowMonth,
    validateMinDate: state.validateMinDate,
    validateMaxDate: state.validateMaxDate,
    endDatePanelType: state.endDatePanelType,
    panelType: state.panelType
  };
})(MonthRender);

exports["default"] = _default;
module.exports = exports.default;
