// @baidu/one-ui@4.41.1/lib/components/datePicker/common/yearPanel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _partial2 = _interopRequireDefault(require("lodash/partial"));

var _react = _interopRequireWildcard(require("react"));

var _miniStore = require("mini-store");

var _datePickerTools = require("../../../core/datePickerTools");

var _commonTools = require("../../../core/commonTools");

var yearCellPadding = 4;
var yearCellHeight = 28;
var mediumYearCellHeight = 32;

var YearRender = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(YearRender, _PureComponent);

  function YearRender(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.store = void 0;
    _this.yearRenderElement = void 0;

    _this.componentDidMount = function () {
      _this.scrollToCurrentYear();
    };

    _this.componentDidUpdate = function () {
      _this.scrollToCurrentYear();
    };

    _this.scrollToCurrentYear = function () {
      var yearRenderElement = _this.yearRenderElement;

      if (!yearRenderElement) {
        return;
      }

      var _this$props = _this.props,
          showYear = _this$props.showYear,
          endDateShowYear = _this$props.endDateShowYear,
          type = _this$props.type,
          isMonthRender = _this$props.isMonthRender,
          panelType = _this$props.panelType,
          size = _this$props.size;
      var currentYear = type === 'nextMultiple' ? endDateShowYear : showYear;

      var minYear = _this.getMaxAndMinYear().minYear;

      var gapYear = currentYear - minYear;
      var yearCellCurrentHeigt = size === 'medium' ? mediumYearCellHeight : yearCellHeight;
      var scrollTop = (yearCellCurrentHeigt + yearCellPadding) * gapYear;

      if (isMonthRender && panelType === 'year') {
        gapYear = Math.ceil(gapYear / 3) - 3;
        scrollTop = (yearCellCurrentHeigt + yearCellPadding * 5) * gapYear;
      }

      yearRenderElement.scrollTop = scrollTop;
    };

    _this.getMaxAndMinYear = function () {
      var _this$props2 = _this.props,
          validateMinDate = _this$props2.validateMinDate,
          validateMaxDate = _this$props2.validateMaxDate,
          type = _this$props2.type,
          endDateShowYear = _this$props2.endDateShowYear,
          showYear = _this$props2.showYear;
      var maxYear = +(0, _datePickerTools.getDetailDate)(validateMaxDate).fullYear;
      var minYear = +(0, _datePickerTools.getDetailDate)(validateMinDate).fullYear;

      if (type === 'prevMultiple') {
        maxYear = endDateShowYear;
      } else if (type === 'nextMultiple') {
        minYear = showYear;
      }

      return {
        maxYear: maxYear,
        minYear: minYear
      };
    };

    _this.onClickYear = function (currentYear) {
      var type = _this.props.type;
      var newState = {};

      if (type === 'nextMultiple') {
        newState.endDateShowYear = currentYear;
      } else {
        newState.showYear = currentYear;
      }

      _this.store.setState(newState);
    };

    _this.getBodyElement = function (ref) {
      _this.yearRenderElement = ref;
    };

    _this.renderYear = function () {
      var _classnames2;

      var _this$props3 = _this.props,
          showYear = _this$props3.showYear,
          prefixCls = _this$props3.prefixCls,
          endDateShowYear = _this$props3.endDateShowYear,
          type = _this$props3.type,
          isMonthRender = _this$props3.isMonthRender,
          panelType = _this$props3.panelType;

      if (isMonthRender && panelType !== 'year') {
        return null;
      }

      var _this$getMaxAndMinYea = _this.getMaxAndMinYear(),
          minYear = _this$getMaxAndMinYea.minYear,
          maxYear = _this$getMaxAndMinYea.maxYear;

      var yearContainer = [];
      var currentYear = type === 'nextMultiple' ? endDateShowYear : showYear;

      for (var i = minYear; i <= maxYear; i++) {
        var _classnames;

        var className = (0, _commonTools.classnames)(prefixCls + "-year-container-item", (_classnames = {}, _classnames[prefixCls + "-year-container-item-selected"] = currentYear === i, _classnames));
        yearContainer.push( /*#__PURE__*/_react["default"].createElement("div", {
          className: className,
          key: i,
          onClick: (0, _partial2["default"])(_this.onClickYear, i)
        }, /*#__PURE__*/_react["default"].createElement("span", null, i, isMonthRender ? '年' : null)));
      }

      var yearClassNames = (0, _commonTools.classnames)(prefixCls + "-year-container", (_classnames2 = {}, _classnames2[prefixCls + "-year-container-is-month-render"] = isMonthRender && panelType === 'year', _classnames2));
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: yearClassNames,
        ref: _this.getBodyElement
      }, yearContainer);
    };

    _this.store = _this.props.store;
    return _this;
  }

  var _proto = YearRender.prototype;

  _proto.render = function render() {
    return this.renderYear();
  };

  return YearRender;
}(_react.PureComponent);

var _default = (0, _miniStore.connect)(function (state) {
  return {
    endDateShowYear: state.endDateShowYear,
    showYear: state.showYear,
    validateMinDate: state.validateMinDate,
    validateMaxDate: state.validateMaxDate,
    panelType: state.panelType
  };
})(YearRender);

exports["default"] = _default;
module.exports = exports.default;
