// @baidu/one-ui@4.32.1/lib/components/empty/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {
  Empty: true
};
exports["default"] = exports.Empty = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../core/commonTools");

var _dlsIllustrationsReact = require("dls-illustrations-react");

var _interface = require("../interface");

Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  exports[key] = _interface[key];
});

function renderPart(part, className) {
  if (part) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: className
    }, part);
  }

  return null;
}
/**
 * @visibleName Empty
 */


var Empty = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var className = _ref.className,
      prefixCls = _ref.prefixCls,
      size = _ref.size,
      title = _ref.title,
      description = _ref.description,
      actions = _ref.actions,
      children = _ref.children,
      restProps = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["className", "prefixCls", "size", "title", "description", "actions", "children"]);
  return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
    ref: ref,
    className: (0, _commonTools.classnames)("" + prefixCls, prefixCls + "-" + size, className)
  }, restProps), renderPart(children, prefixCls + "-image"), renderPart(title, prefixCls + "-title"), renderPart(description, prefixCls + "-description"), renderPart(actions, prefixCls + "-actions"));
});
exports.Empty = Empty;
Empty.defaultProps = {
  prefixCls: 'one-empty',
  description: '暂无数据',
  size: 'medium',
  children: /*#__PURE__*/_react["default"].createElement(_dlsIllustrationsReact.IllustrationSpotNoContent, null)
};
Empty.displayName = 'Empty';
var _default = Empty;
exports["default"] = _default;
