// @baidu/one-ui@4.41.1/lib/components/select/common/popSelectTrigger.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _rcTrigger = _interopRequireDefault(require("rc-trigger"));

var _popContent = _interopRequireDefault(require("./popContent"));

_rcTrigger["default"].displayName = 'Trigger';
var BUILT_IN_PLACEMENTS = {
  bottomLeft: {
    points: ['tl', 'bl'],
    offset: [0, 4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  topLeft: {
    points: ['bl', 'tl'],
    offset: [0, -4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  bottomRight: {
    points: ['tr', 'br'],
    offset: [0, 4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  topRight: {
    points: ['br', 'tr'],
    offset: [0, -4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  }
};
;
;

var PopSelectTrigger = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(PopSelectTrigger, _PureComponent);

  function PopSelectTrigger(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.trigger = void 0;
    _this.root = void 0;

    _this.setDropdownWidth = function () {
      var width = _this.root.offsetWidth;

      if (width !== _this.state.dropdownWidth) {
        _this.setState({
          dropdownWidth: width
        });
      }
    };

    _this.getPopupElement = function () {
      var _this$props = _this.props,
          overlay = _this$props.overlay,
          prefixCls = _this$props.prefixCls,
          id = _this$props.id;
      return /*#__PURE__*/_react["default"].createElement(_popContent["default"], {
        key: "content",
        trigger: _this.trigger,
        prefixCls: prefixCls,
        id: id,
        overlay: overlay
      });
    };

    _this.saveTrigger = function (node) {
      _this.trigger = node;
    };

    _this.saveRoot = function (node) {
      _this.root = node;
    };

    _this.state = {
      dropdownWidth: null
    };
    return _this;
  }

  var _proto = PopSelectTrigger.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.setDropdownWidth();
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    this.setDropdownWidth();
  };

  _proto.getPopupDomNode = function getPopupDomNode() {
    return this.trigger.getPopupDomNode();
  };

  _proto.render = function render() {
    var _this$props2 = this.props,
        overlayClassName = _this$props2.overlayClassName,
        trigger = _this$props2.trigger,
        mouseEnterDelay = _this$props2.mouseEnterDelay,
        mouseLeaveDelay = _this$props2.mouseLeaveDelay,
        overlayStyle = _this$props2.overlayStyle,
        prefixCls = _this$props2.prefixCls,
        children = _this$props2.children,
        onVisibleChange = _this$props2.onVisibleChange,
        afterVisibleChange = _this$props2.afterVisibleChange,
        transitionName = _this$props2.transitionName,
        animation = _this$props2.animation,
        align = _this$props2.align,
        destroyPopUpOnHide = _this$props2.destroyPopUpOnHide,
        defaultVisible = _this$props2.defaultVisible,
        getPopupContainer = _this$props2.getPopupContainer,
        dropdownMatchSelectWidth = _this$props2.dropdownMatchSelectWidth,
        popupPlacement = _this$props2.popupPlacement,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props2, ["overlayClassName", "trigger", "mouseEnterDelay", "mouseLeaveDelay", "overlayStyle", "prefixCls", "children", "onVisibleChange", "afterVisibleChange", "transitionName", "animation", "align", "destroyPopUpOnHide", "defaultVisible", "getPopupContainer", "dropdownMatchSelectWidth", "popupPlacement"]);
    var extraProps = (0, _extends2["default"])({}, restProps);

    if ('visible' in this.props) {
      // @ts-ignore
      extraProps.popupVisible = this.props.visible;
    }

    var widthProp = dropdownMatchSelectWidth ? 'width' : 'minWidth';
    var popupStyle = (0, _extends2["default"])({}, overlayStyle);

    if (this.state.dropdownWidth) {
      popupStyle[widthProp] = this.state.dropdownWidth + "px";
    }

    return /*#__PURE__*/_react["default"].createElement(_rcTrigger["default"], (0, _extends2["default"])({
      popupClassName: overlayClassName,
      ref: this.saveTrigger,
      prefixCls: prefixCls,
      popup: this.getPopupElement,
      action: trigger,
      builtinPlacements: BUILT_IN_PLACEMENTS,
      popupPlacement: popupPlacement,
      popupAlign: align,
      getPopupContainer: getPopupContainer,
      onPopupVisibleChange: onVisibleChange,
      afterPopupVisibleChange: afterVisibleChange,
      popupTransitionName: transitionName,
      popupAnimation: animation,
      defaultPopupVisible: defaultVisible,
      destroyPopupOnHide: destroyPopUpOnHide,
      mouseLeaveDelay: mouseLeaveDelay,
      popupStyle: popupStyle,
      mouseEnterDelay: mouseEnterDelay
    }, extraProps), /*#__PURE__*/_react["default"].cloneElement(children, {
      ref: this.saveRoot
    }));
  };

  return PopSelectTrigger;
}(_react.PureComponent);

PopSelectTrigger.defaultProps = {
  prefixCls: 'one-select-pop',
  mouseEnterDelay: 0,
  destroyPopUpOnHide: false,
  mouseLeaveDelay: 0.1,
  align: {},
  trigger: ['hover'],
  dropdownMatchSelectWidth: true,
  popupPlacement: 'bottomLeft'
};
var _default = PopSelectTrigger;
exports["default"] = _default;
module.exports = exports.default;
