// @baidu/one-ui@4.32.1/lib/components/datePicker/common/dayItemRender.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _miniStore = require("mini-store");

var _weekRender = _interopRequireDefault(require("./weekRender"));

var _calendar = _interopRequireDefault(require("./calendar"));

var _classnames = _interopRequireDefault(require("classnames"));

var DayItemRender = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(DayItemRender, _PureComponent);

  function DayItemRender() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = DayItemRender.prototype;

  _proto.render = function render() {
    var _classNames;

    var _this$props = this.props,
        panelType = _this$props.panelType,
        prefixCls = _this$props.prefixCls,
        mode = _this$props.mode;

    if (panelType === 'month') {
      return null;
    }

    var bodyContainerClassName = (0, _classnames["default"])(prefixCls + "-day-container", (_classNames = {}, _classNames[prefixCls + "-week-mode"] = mode === 'week', _classNames));
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: bodyContainerClassName
    }, /*#__PURE__*/_react["default"].createElement(_weekRender["default"], this.props), /*#__PURE__*/_react["default"].createElement(_calendar["default"], this.props));
  };

  return DayItemRender;
}(_react.PureComponent);

var _default = (0, _miniStore.connect)(function (state) {
  var _ref = state._value,
      beginDate = _ref[0],
      endDate = _ref[1];
  return {
    showYear: state.showYear,
    showMonth: state.showMonth,
    panelType: state.panelType,
    beginDate: beginDate,
    endDate: endDate,
    validateMinDate: state.validateMinDate,
    validateMaxDate: state.validateMaxDate
  };
})(DayItemRender);

exports["default"] = _default;
module.exports = exports.default;
