// @baidu/one-ui@4.32.1/lib/components/datePicker/common/rangeDayRender.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _miniStore = require("mini-store");

var _weekRender = _interopRequireDefault(require("./weekRender"));

var _calendar = _interopRequireDefault(require("./calendar"));

var DayItemRender = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(DayItemRender, _PureComponent);

  function DayItemRender() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = DayItemRender.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        panelType = _this$props.panelType,
        prefixCls = _this$props.prefixCls,
        type = _this$props.type,
        endDatePanelType = _this$props.endDatePanelType,
        endDateShowYear = _this$props.endDateShowYear,
        endDateShowMonth = _this$props.endDateShowMonth,
        currentDate = _this$props.currentDate;
    var currentPanelType = type === 'nextMultiple' ? endDatePanelType : panelType;

    if (currentPanelType === 'month') {
      return null;
    }

    var bodyContainerClassName = prefixCls + "-day-container";
    var otherProps = {
      beginDate: currentDate[0] || '',
      endDate: currentDate[1] || ''
    };

    if (type === 'nextMultiple') {
      otherProps = (0, _extends2["default"])({}, otherProps, {
        showYear: endDateShowYear,
        showMonth: endDateShowMonth
      });
    }

    return /*#__PURE__*/_react["default"].createElement("div", {
      className: bodyContainerClassName
    }, /*#__PURE__*/_react["default"].createElement(_weekRender["default"], this.props), /*#__PURE__*/_react["default"].createElement(_calendar["default"], (0, _extends2["default"])({}, this.props, otherProps, {
      multiple: true
    })));
  };

  return DayItemRender;
}(_react.PureComponent);

var _default = (0, _miniStore.connect)(function (state) {
  return {
    showYear: state.showYear,
    showMonth: state.showMonth,
    panelType: state.panelType,
    validateMinDate: state.validateMinDate,
    validateMaxDate: state.validateMaxDate,
    endDateShowYear: state.endDateShowYear,
    endDateShowMonth: state.endDateShowMonth,
    endDatePanelType: state.endDatePanelType
  };
})(DayItemRender);

exports["default"] = _default;
module.exports = exports.default;
