// less-plugin-dls@1.5.0/variables.js
export const dlsColorBrand = '#0052cc'
export const dlsColorBrand0 = '#fff'
export const dlsColorBrand1 = '#f2f7ff'
export const dlsColorBrand2 = '#e1edff'
export const dlsColorBrand3 = '#c0d9ff'
export const dlsColorBrand4 = '#8fbcff'
export const dlsColorBrand5 = '#4f96ff'
export const dlsColorBrand6 = '#0066ff'
export const dlsColorBrand7 = '#0052cc'
export const dlsColorBrand8 = '#0046ad'
export const dlsColorBrand9 = '#003585'
export const dlsColorBrand10 = '#002152'
export const dlsColorBrand11 = '#000'
export const dlsColorInfo = '#0052cc'
export const dlsColorInfo0 = '#fff'
export const dlsColorInfo1 = '#f2f7ff'
export const dlsColorInfo2 = '#e1edff'
export const dlsColorInfo3 = '#c0d9ff'
export const dlsColorInfo4 = '#8fbcff'
export const dlsColorInfo5 = '#4f96ff'
export const dlsColorInfo6 = '#0066ff'
export const dlsColorInfo7 = '#0052cc'
export const dlsColorInfo8 = '#0046ad'
export const dlsColorInfo9 = '#003585'
export const dlsColorInfo10 = '#002152'
export const dlsColorInfo11 = '#000'
export const dlsColorSuccess = '#00bf5c'
export const dlsColorSuccess0 = '#fff'
export const dlsColorSuccess1 = '#f2fff8'
export const dlsColorSuccess2 = '#e1ffef'
export const dlsColorSuccess3 = '#c0ffde'
export const dlsColorSuccess4 = '#8fffc5'
export const dlsColorSuccess5 = '#4fffa4'
export const dlsColorSuccess6 = '#00f275'
export const dlsColorSuccess7 = '#00bf5c'
export const dlsColorSuccess8 = '#00a24e'
export const dlsColorSuccess9 = '#007c3c'
export const dlsColorSuccess10 = '#004c25'
export const dlsColorSuccess11 = '#000'
export const dlsColorWarning = '#ff9100'
export const dlsColorWarning0 = '#fff'
export const dlsColorWarning1 = '#fffaf2'
export const dlsColorWarning2 = '#fff4e4'
export const dlsColorWarning3 = '#ffe8ca'
export const dlsColorWarning4 = '#ffd8a4'
export const dlsColorWarning5 = '#ffc272'
export const dlsColorWarning6 = '#ffa733'
export const dlsColorWarning7 = '#ff9100'
export const dlsColorWarning8 = '#d97b00'
export const dlsColorWarning9 = '#a65e00'
export const dlsColorWarning10 = '#663a00'
export const dlsColorWarning11 = '#000'
export const dlsColorError = '#cc1800'
export const dlsColorError0 = '#fff'
export const dlsColorError1 = '#fff4f2'
export const dlsColorError2 = '#ffe4e1'
export const dlsColorError3 = '#ffc7c0'
export const dlsColorError4 = '#ff9d8f'
export const dlsColorError5 = '#ff644f'
export const dlsColorError6 = '#ff1e00'
export const dlsColorError7 = '#cc1800'
export const dlsColorError8 = '#ad1400'
export const dlsColorError9 = '#851000'
export const dlsColorError10 = '#520a00'
export const dlsColorError11 = '#000'
export const dlsColorGray0 = '#fff'
export const dlsColorGray1 = '#f6f7fa'
export const dlsColorGray2 = '#ebedf5'
export const dlsColorGray3 = '#e2e6f0'
export const dlsColorGray4 = '#d3d9e6'
export const dlsColorGray5 = '#c5ccdb'
export const dlsColorGray6 = '#a8b0bf'
export const dlsColorGray7 = '#848b99'
export const dlsColorGray8 = '#545b66'
export const dlsColorGray9 = '#282c33'
export const dlsColorGray10 = '#0a0b0d'
export const dlsColorGray11 = '#000'
export const dlsFontFamily = '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "PingFang SC", "Hiragino Sans GB", "Microsoft YaHei", sans-serif'
export const dlsFontSize0 = '12px'
export const dlsFontSize1 = '14px'
export const dlsFontSize2 = '16px'
export const dlsFontSize3 = '20px'
export const dlsFontSize4 = '24px'
export const dlsFontSize5 = '32px'
export const dlsFontWeight1 = '400'
export const dlsFontWeight2 = '500'
export const dlsFontWeight3 = '600'
export const dlsLineHeight1 = '1.4'
export const dlsLineHeight2 = '1.6'
export const dlsLineHeight3 = '1.8'
export const dlsSpacingUnit = '4px'
export const dlsPaddingUnit = '4px'
export const dlsHeightUnit = '4px'
export const dlsHeightXs = '24px'
export const dlsHeightS = '28px'
export const dlsHeightM = '32px'
export const dlsHeightL = '36px'
export const dlsHeightXl = '40px'
export const dlsForegroundColorReverse = '#fff'
export const dlsForegroundColorPrimaryReverse = '#fff'
export const dlsForegroundColorPrimary = '#0052cc'
export const dlsForegroundColorPrimaryHover = '#0066ff'
export const dlsForegroundColorPrimaryActive = '#0046ad'
export const dlsForegroundColorPrimaryDisabled = '#8fbcff'
export const dlsForegroundColorNeutral = '#282c33'
export const dlsForegroundColorNeutralHover = '#545b66'
export const dlsForegroundColorNeutralActive = '#000'
export const dlsForegroundColorNeutralDisabled = '#a8b0bf'
export const dlsForegroundColorNeutralStrong = '#000'
export const dlsForegroundColorNeutralStrongHover = '#0a0b0d'
export const dlsForegroundColorNeutralStrongActive = '#000'
export const dlsForegroundColorNeutralStrongDisabled = '#a8b0bf'
export const dlsForegroundColorNeutralLight = '#545b66'
export const dlsForegroundColorNeutralLightHover = '#848b99'
export const dlsForegroundColorNeutralLightActive = '#282c33'
export const dlsForegroundColorNeutralLightDisabled = '#a8b0bf'
export const dlsForegroundColorNeutralWeak = '#848b99'
export const dlsForegroundColorNeutralWeakHover = '#545b66'
export const dlsForegroundColorNeutralWeakActive = '#282c33'
export const dlsForegroundColorNeutralWeakDisabled = '#c5ccdb'
export const dlsForegroundColorNeutralDim = '#a8b0bf'
export const dlsForegroundColorNeutralDimHover = '#848b99'
export const dlsForegroundColorNeutralDimActive = '#545b66'
export const dlsForegroundColorNeutralDimDisabled = '#e2e6f0'
export const dlsForegroundColorNeutralHighlightable = '#282c33'
export const dlsForegroundColorNeutralHighlightableHover = '#0066ff'
export const dlsForegroundColorNeutralHighlightableActive = '#0046ad'
export const dlsForegroundColorNeutralHighlightableDisabled = '#a8b0bf'
export const dlsForegroundColorInfoPrimary = '#0052cc'
export const dlsForegroundColorInfoPrimaryDisabled = '#8fbcff'
export const dlsForegroundColorSuccessPrimary = '#00bf5c'
export const dlsForegroundColorSuccessPrimaryDisabled = '#8fffc5'
export const dlsForegroundColorWarningPrimary = '#ffa733'
export const dlsForegroundColorWarningPrimaryDisabled = '#ffd8a4'
export const dlsForegroundColorErrorPrimary = '#cc1800'
export const dlsForegroundColorErrorPrimaryDisabled = '#ff9d8f'
export const dlsForegroundColorHighlighted = '#ff9100'
export const dlsBackgroundColorBase1 = '#fff'
export const dlsBackgroundColorBase1Hover = '#f6f7fa'
export const dlsBackgroundColorBase1Active = '#e2e6f0'
export const dlsBackgroundColorBase1Disabled = '#fff'
export const dlsBackgroundColorBase2 = '#f6f7fa'
export const dlsBackgroundColorBase2Hover = '#ebedf5'
export const dlsBackgroundColorBase2Active = '#e2e6f0'
export const dlsBackgroundColorBase2Disabled = '#f6f7fa'
export const dlsBackgroundColorBase3 = '#ebedf5'
export const dlsBackgroundColorBase3Hover = '#e2e6f0'
export const dlsBackgroundColorBase3Active = '#d3d9e6'
export const dlsBackgroundColorBase3Disabled = '#f6f7fa'
export const dlsBackgroundColorPrimary = '#0052cc'
export const dlsBackgroundColorPrimaryHover = '#0066ff'
export const dlsBackgroundColorPrimaryActive = '#0046ad'
export const dlsBackgroundColorPrimaryDisabled = '#e1edff'
export const dlsBorderColorPrimary = 'transparent'
export const dlsBorderColorPrimaryDisabled = 'transparent'
export const dlsBackgroundColorInfoPrimary = '#0052cc'
export const dlsBackgroundColorInfoPrimaryHover = '#0066ff'
export const dlsBackgroundColorInfoPrimaryActive = '#0046ad'
export const dlsBackgroundColorInfoPrimaryDisabled = '#e1edff'
export const dlsBorderColorInfoPrimary = 'transparent'
export const dlsBorderColorInfoPrimaryDisabled = 'transparent'
export const dlsBackgroundColorSuccessPrimary = '#00bf5c'
export const dlsBackgroundColorSuccessPrimaryHover = '#00f275'
export const dlsBackgroundColorSuccessPrimaryActive = '#00a24e'
export const dlsBackgroundColorSuccessPrimaryDisabled = '#e1ffef'
export const dlsBorderColorSuccessPrimary = 'transparent'
export const dlsBorderColorSuccessPrimaryDisabled = 'transparent'
export const dlsBackgroundColorWarningPrimary = '#ffa733'
export const dlsBackgroundColorWarningPrimaryHover = '#ffc272'
export const dlsBackgroundColorWarningPrimaryActive = '#ff9100'
export const dlsBackgroundColorWarningPrimaryDisabled = '#fff4e4'
export const dlsBorderColorWarningPrimary = 'transparent'
export const dlsBorderColorWarningPrimaryDisabled = 'transparent'
export const dlsBackgroundColorErrorPrimary = '#cc1800'
export const dlsBackgroundColorErrorPrimaryHover = '#ff1e00'
export const dlsBackgroundColorErrorPrimaryActive = '#ad1400'
export const dlsBackgroundColorErrorPrimaryDisabled = '#ffe4e1'
export const dlsBorderColorErrorPrimary = 'transparent'
export const dlsBorderColorErrorPrimaryDisabled = 'transparent'
export const dlsBackgroundColorNeutralPrimary = '#a8b0bf'
export const dlsBackgroundColorNeutralPrimaryHover = '#848b99'
export const dlsBackgroundColorNeutralPrimaryActive = '#545b66'
export const dlsBackgroundColorNeutralPrimaryDisabled = '#e2e6f0'
export const dlsBorderColorNeutralPrimary = 'transparent'
export const dlsBorderColorNeutralPrimaryDisabled = 'transparent'
export const dlsBackgroundColorInfo = '#f2f7ff'
export const dlsBackgroundColorInfoHover = '#e1edff'
export const dlsBackgroundColorInfoActive = '#c0d9ff'
export const dlsBackgroundColorInfoDisabled = '#f2f7ff'
export const dlsBorderColorInfo = '#8fbcff'
export const dlsBorderColorInfoDisabled = '#c0d9ff'
export const dlsBackgroundColorSuccess = '#f2fff8'
export const dlsBackgroundColorSuccessHover = '#e1ffef'
export const dlsBackgroundColorSuccessActive = '#c0ffde'
export const dlsBackgroundColorSuccessDisabled = '#f2fff8'
export const dlsBorderColorSuccess = '#8fffc5'
export const dlsBorderColorSuccessDisabled = '#c0ffde'
export const dlsBackgroundColorWarning = '#fffaf2'
export const dlsBackgroundColorWarningHover = '#fff4e4'
export const dlsBackgroundColorWarningActive = '#ffe8ca'
export const dlsBackgroundColorWarningDisabled = '#fffaf2'
export const dlsBorderColorWarning = '#ffd8a4'
export const dlsBorderColorWarningDisabled = '#ffe8ca'
export const dlsBackgroundColorError = '#fff4f2'
export const dlsBackgroundColorErrorHover = '#ffe4e1'
export const dlsBackgroundColorErrorActive = '#ffc7c0'
export const dlsBackgroundColorErrorDisabled = '#fff4f2'
export const dlsBorderColorError = '#ff9d8f'
export const dlsBorderColorErrorDisabled = '#ffc7c0'
export const dlsBackgroundColorNeutral = '#ebedf5'
export const dlsBackgroundColorNeutralHover = '#e2e6f0'
export const dlsBackgroundColorNeutralActive = '#d3d9e6'
export const dlsBackgroundColorNeutralDisabled = '#f6f7fa'
export const dlsBorderColorNeutral = '#c5ccdb'
export const dlsBorderColorNeutralDisabled = '#e2e6f0'
export const dlsBackgroundColorCurrent = '#f2f7ff'
export const dlsBackgroundColorCurrentHover = '#e1edff'
export const dlsBackgroundColorCurrentActive = '#c0d9ff'
export const dlsBackgroundColorCurrentDisabled = '#f2f7ff'
export const dlsBorderColorCurrent = '#0052cc'
export const dlsBorderColorCurrentDisabled = '#c0d9ff'
export const dlsBackgroundColorStrong = '#fff'
export const dlsBackgroundColorStrongHover = '#f2f7ff'
export const dlsBackgroundColorStrongActive = '#e1edff'
export const dlsBackgroundColorStrongDisabled = '#f2f7ff'
export const dlsBorderColorStrong = '#0052cc'
export const dlsBorderColorStrongDisabled = '#c0d9ff'
export const dlsBackgroundColorErrorStrong = '#fff'
export const dlsBackgroundColorErrorStrongHover = '#fff4f2'
export const dlsBackgroundColorErrorStrongActive = '#ffe4e1'
export const dlsBackgroundColorErrorStrongDisabled = '#fff4f2'
export const dlsBorderColorErrorStrong = '#cc1800'
export const dlsBorderColorErrorStrongDisabled = '#ffc7c0'
export const dlsBackgroundColorFillable = '#fff'
export const dlsBackgroundColorFillableHover = '#ebedf5'
export const dlsBackgroundColorFillableActive = '#d3d9e6'
export const dlsBackgroundColorFillableDisabled = '#f6f7fa'
export const dlsBorderColorFillable = '#d3d9e6'
export const dlsBorderColorFillableHover = '#a8b0bf'
export const dlsBorderColorFillableActive = '#a8b0bf'
export const dlsBorderColorFillableDisabled = '#e2e6f0'
export const dlsBackgroundColorPressable = '#fff'
export const dlsBackgroundColorPressableHover = '#f6f7fa'
export const dlsBackgroundColorPressableActive = '#e2e6f0'
export const dlsBackgroundColorPressableDisabled = '#f6f7fa'
export const dlsBorderColorPressable = '#e2e6f0'
export const dlsBorderColorPressableHover = '#d3d9e6'
export const dlsBorderColorPressableActive = '#d3d9e6'
export const dlsBorderColorPressableDisabled = '#e2e6f0'
export const dlsBackgroundColorTranslucent = 'rgba(0, 0, 0, 0.6)'
export const dlsBackgroundColorTranslucentHover = 'rgba(0, 0, 0, 0.5)'
export const dlsBackgroundColorTranslucentActive = 'rgba(0, 0, 0, 0.7)'
export const dlsBackgroundColorTranslucentDisabled = 'rgba(0, 0, 0, 0.2)'
export const dlsBorderColorTranslucent = 'transparent'
export const dlsBorderColorTranslucentDisabled = 'transparent'
export const dlsBackgroundColorModal = 'rgba(0, 0, 0, 0.6)'
export const dlsLineColor0 = '#ebedf5'
export const dlsLineColor1 = '#e2e6f0'
export const dlsLineColor2 = '#d3d9e6'
export const dlsBorderRadius0 = '2px'
export const dlsBorderRadius1 = '4px'
export const dlsBorderRadius2 = '6px'
export const dlsBorderRadius3 = '10px'
export const dlsShadowColor = '#000'
export const dlsShadow1 = '0 4px 6px rgba(0, 0, 0, 0.06), 0 1px 10px rgba(0, 0, 0, 0.05), 0 2px 4px -1px rgba(0, 0, 0, 0.01)'
export const dlsShadow2 = '0 8px 10px 1px rgba(0, 0, 0, 0.06), 0 3px 14px 2px rgba(0, 0, 0, 0.05), 0 5px 5px -3px rgba(0, 0, 0, 0.01)'
export const dlsShadow3 = '0 16px 24px 2px rgba(0, 0, 0, 0.06), 0 6px 30px 5px rgba(0, 0, 0, 0.05), 0 8px 10px -5px rgba(0, 0, 0, 0.01)'
export const dlsShadowColorLight = '#000'
export const dlsShadowLight1 = '0 4px 10px rgba(0, 0, 0, 0.03), 0 3px 9px rgba(0, 0, 0, 0.02), 0 2px 8px rgba(0, 0, 0, 0.01)'
export const dlsShadowLight2 = '0 6px 28px 2px rgba(0, 0, 0, 0.04), 0 4px 26px 2px rgba(0, 0, 0, 0.03), 0 2px 24px 1px rgba(0, 0, 0, 0.02)'
export const dlsShadowLight3 = '0 6px 32px 2px rgba(0, 0, 0, 0.06), 0 5px 30px 1px rgba(0, 0, 0, 0.05), 0 4px 28px 1px rgba(0, 0, 0, 0.04)'
export const dlsShadowOpacityFocus = '0.2'
export const dlsShadowSpreadFocus = '2px'
export const dlsShadowColorFocus = '#0066ff'
export const dlsShadowColorErrorFocus = '#ff1e00'
export const dlsShadowFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsShadowErrorFocus = '0 0 0 2px rgba(255, 30, 0, 0.2)'
export const dlsBackgroundColorFocus = '#e1edff'
export const dlsBorderColorFocus = '#0052cc'
export const dlsBorderColorErrorFocus = '#cc1800'
export const dlsIconSizeNormal = 'calc(1em + 2px)'
export const dlsIconSizeMinor = 'calc(1em + 2px)'
export const dlsIconSizeAux = 'calc(1em + 2px)'
export const dlsIconColorAux = '#848b99'
export const dlsIconColorAuxDisabled = '#a8b0bf'
export const dlsButtonFontSizeXs = '12px'
export const dlsButtonFontSizeS = '12px'
export const dlsButtonFontSizeM = '14px'
export const dlsButtonFontSizeL = '16px'
export const dlsButtonFontSizeXl = '16px'
export const dlsButtonHeightXs = '24px'
export const dlsButtonHeightS = '28px'
export const dlsButtonHeightM = '32px'
export const dlsButtonHeightL = '36px'
export const dlsButtonHeightXl = '40px'
export const dlsButtonPaddingXs = '8px'
export const dlsButtonPaddingS = '12px'
export const dlsButtonPaddingM = '16px'
export const dlsButtonPaddingL = '20px'
export const dlsButtonPaddingXl = '20px'
export const dlsButtonBorderRadiusXs = '2px'
export const dlsButtonBorderRadiusS = '2px'
export const dlsButtonBorderRadiusM = '4px'
export const dlsButtonBorderRadiusL = '6px'
export const dlsButtonBorderRadiusXl = '6px'
export const dlsButtonContentSpacing = '4px'
export const dlsButtonFontColorNormal = '#282c33'
export const dlsButtonFontColorNormalDisabled = '#a8b0bf'
export const dlsButtonBorderColorNormal = 'transparent'
export const dlsButtonBorderColorNormalHover = 'transparent'
export const dlsButtonBorderColorNormalActive = 'transparent'
export const dlsButtonBorderColorNormalDisabled = 'transparent'
export const dlsButtonBackgroundColorNormal = '#ebedf5'
export const dlsButtonBackgroundColorNormalHover = '#e2e6f0'
export const dlsButtonBackgroundColorNormalFocus = '#e2e6f0'
export const dlsButtonBackgroundColorNormalActive = '#d3d9e6'
export const dlsButtonBackgroundColorNormalDisabled = '#f6f7fa'
export const dlsButtonFontColorBasic = '#282c33'
export const dlsButtonFontColorBasicDisabled = '#a8b0bf'
export const dlsButtonBorderColorBasic = '#e2e6f0'
export const dlsButtonBorderColorBasicHover = '#d3d9e6'
export const dlsButtonBorderColorBasicActive = '#d3d9e6'
export const dlsButtonBorderColorBasicDisabled = '#e2e6f0'
export const dlsButtonBackgroundColorBasic = '#fff'
export const dlsButtonBackgroundColorBasicHover = '#f6f7fa'
export const dlsButtonBackgroundColorBasicFocus = '#f6f7fa'
export const dlsButtonBackgroundColorBasicActive = '#e2e6f0'
export const dlsButtonBackgroundColorBasicDisabled = '#f6f7fa'
export const dlsButtonFontColorStrong = '#0052cc'
export const dlsButtonFontColorStrongDisabled = '#8fbcff'
export const dlsButtonBorderColorStrong = '#0052cc'
export const dlsButtonBorderColorStrongHover = '#0052cc'
export const dlsButtonBorderColorStrongActive = '#0052cc'
export const dlsButtonBorderColorStrongDisabled = '#c0d9ff'
export const dlsButtonBackgroundColorStrong = '#fff'
export const dlsButtonBackgroundColorStrongHover = '#f2f7ff'
export const dlsButtonBackgroundColorStrongFocus = '#f2f7ff'
export const dlsButtonBackgroundColorStrongActive = '#e1edff'
export const dlsButtonBackgroundColorStrongDisabled = '#f2f7ff'
export const dlsButtonFontColorPrimary = '#fff'
export const dlsButtonFontColorPrimaryDisabled = '#fff'
export const dlsButtonBorderColorPrimary = 'transparent'
export const dlsButtonBorderColorPrimaryHover = 'transparent'
export const dlsButtonBorderColorPrimaryActive = 'transparent'
export const dlsButtonBorderColorPrimaryDisabled = 'transparent'
export const dlsButtonBackgroundColorPrimary = '#0052cc'
export const dlsButtonBackgroundColorPrimaryHover = '#0066ff'
export const dlsButtonBackgroundColorPrimaryFocus = '#0066ff'
export const dlsButtonBackgroundColorPrimaryActive = '#0046ad'
export const dlsButtonBackgroundColorPrimaryDisabled = '#e1edff'
export const dlsButtonFontColorTranslucent = '#fff'
export const dlsButtonFontColorTranslucentDisabled = 'rgba(255, 255, 255, 0.6)'
export const dlsButtonBorderColorTranslucent = 'transparent'
export const dlsButtonBorderColorTranslucentHover = 'transparent'
export const dlsButtonBorderColorTranslucentActive = 'transparent'
export const dlsButtonBorderColorTranslucentDisabled = 'transparent'
export const dlsButtonBackgroundColorTranslucent = 'rgba(0, 0, 0, 0.6)'
export const dlsButtonBackgroundColorTranslucentHover = 'rgba(0, 0, 0, 0.5)'
export const dlsButtonBackgroundColorTranslucentFocus = 'rgba(0, 0, 0, 0.5)'
export const dlsButtonBackgroundColorTranslucentActive = 'rgba(0, 0, 0, 0.7)'
export const dlsButtonBackgroundColorTranslucentDisabled = 'rgba(0, 0, 0, 0.2)'
export const dlsButtonFontColorText = '#282c33'
export const dlsButtonFontColorTextHover = '#545b66'
export const dlsButtonFontColorTextFocus = '#282c33'
export const dlsButtonFontColorTextActive = '#000'
export const dlsButtonFontColorTextDisabled = '#a8b0bf'
export const dlsButtonFontColorTextStrong = '#0052cc'
export const dlsButtonFontColorTextStrongHover = '#0066ff'
export const dlsButtonFontColorTextStrongFocus = '#0052cc'
export const dlsButtonFontColorTextStrongActive = '#0046ad'
export const dlsButtonFontColorTextStrongDisabled = '#8fbcff'
export const dlsButtonFontColorTextAux = '#848b99'
export const dlsButtonFontColorTextAuxHover = '#545b66'
export const dlsButtonFontColorTextAuxFocus = '#848b99'
export const dlsButtonFontColorTextAuxActive = '#282c33'
export const dlsButtonFontColorTextAuxDisabled = '#c5ccdb'
export const dlsButtonBackgroundColorText = 'transparent'
export const dlsButtonBackgroundColorTextHover = 'transparent'
export const dlsButtonBackgroundColorTextFocus = 'transparent'
export const dlsButtonBackgroundColorTextActive = 'transparent'
export const dlsButtonBackgroundColorTextDisabled = 'transparent'
export const dlsButtonPaddingText = '2px'
export const dlsButtonBackgroundColorChecked = '#f2f7ff'
export const dlsButtonBackgroundColorCheckedHover = '#e1edff'
export const dlsButtonBackgroundColorCheckedFocus = '#e1edff'
export const dlsButtonBackgroundColorCheckedActive = '#c0d9ff'
export const dlsButtonBackgroundColorCheckedDisabled = '#f2f7ff'
export const dlsButtonBorderColorChecked = '#0052cc'
export const dlsButtonBorderColorCheckedDisabled = '#c0d9ff'
export const dlsButtonForegroundColorChecked = '#0052cc'
export const dlsButtonBorderColorFocus = '#0052cc'
export const dlsButtonShadowFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsButtonGroupSeparatorOpacity = '0.15'
export const dlsButtonFontColorSelected = '#0052cc'
export const dlsButtonFontColorSelectedDisabled = '#8fbcff'
export const dlsButtonBorderColorSelected = '#0052cc'
export const dlsButtonBorderColorSelectedDisabled = '#c0d9ff'
export const dlsButtonBackgroundColorSelected = '#f2f7ff'
export const dlsButtonBackgroundColorSelectedHover = '#e1edff'
export const dlsButtonBackgroundColorSelectedFocus = '#e1edff'
export const dlsButtonBackgroundColorSelectedActive = '#c0d9ff'
export const dlsButtonBackgroundColorSelectedDisabled = '#f2f7ff'
export const dlsButtonMinWidthStableS = 'calc(24px + 4em)'
export const dlsButtonMinWidthStableM = 'calc(32px + 4em)'
export const dlsLinkFontSizeS = '12px'
export const dlsLinkFontSizeM = '14px'
export const dlsLinkTextDecoration = 'none'
export const dlsLinkTextDecorationHover = 'underline'
export const dlsLinkTextDecorationFocus = 'underline'
export const dlsLinkTextDecorationActive = 'underline'
export const dlsLinkFontColorNormal = '#282c33'
export const dlsLinkFontColorNormalHover = '#545b66'
export const dlsLinkFontColorNormalFocus = '#545b66'
export const dlsLinkFontColorNormalActive = '#000'
export const dlsLinkFontColorNormalDisabled = '#a8b0bf'
export const dlsLinkFontColorStrong = '#0052cc'
export const dlsLinkFontColorStrongHover = '#0066ff'
export const dlsLinkFontColorStrongFocus = '#0066ff'
export const dlsLinkFontColorStrongActive = '#0046ad'
export const dlsLinkFontColorStrongDisabled = '#8fbcff'
export const dlsRadioFontSizeS = '12px'
export const dlsRadioFontSizeM = '14px'
export const dlsRadioFontColor = '#282c33'
export const dlsRadioFontColorDisabled = '#a8b0bf'
export const dlsRadioSizeS = '14px'
export const dlsRadioSizeM = '16px'
export const dlsRadioDotSizeS = '6px'
export const dlsRadioDotSizeM = '6px'
export const dlsRadioSpacing = '8px'
export const dlsRadioBorderColor = '#d3d9e6'
export const dlsRadioBorderColorHover = '#a8b0bf'
export const dlsRadioBorderColorFocus = '#0052cc'
export const dlsRadioBorderColorActive = '#a8b0bf'
export const dlsRadioBorderColorDisabled = '#e2e6f0'
export const dlsRadioBorderColorChecked = 'transparent'
export const dlsRadioBorderColorCheckedHover = 'transparent'
export const dlsRadioBorderColorCheckedFocus = '#0052cc'
export const dlsRadioBorderColorCheckedActive = 'transparent'
export const dlsRadioBorderColorCheckedDisabled = 'transparent'
export const dlsRadioBackgroundColor = '#fff'
export const dlsRadioBackgroundColorHover = '#ebedf5'
export const dlsRadioBackgroundColorFocus = '#ebedf5'
export const dlsRadioBackgroundColorActive = '#d3d9e6'
export const dlsRadioBackgroundColorDisabled = '#f6f7fa'
export const dlsRadioBackgroundColorChecked = '#0052cc'
export const dlsRadioBackgroundColorCheckedHover = '#0066ff'
export const dlsRadioBackgroundColorCheckedFocus = '#0066ff'
export const dlsRadioBackgroundColorCheckedActive = '#0046ad'
export const dlsRadioBackgroundColorCheckedDisabled = '#e1edff'
export const dlsRadioDotColor = '#fff'
export const dlsRadioDotColorHover = '#fff'
export const dlsRadioDotColorFocus = '#fff'
export const dlsRadioDotColorActive = '#fff'
export const dlsRadioDotColorDisabled = '#fff'
export const dlsRadioShadowFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsRadioGroupSpacing = '48px'
export const dlsRadioGroupSpacingSimple = '8px'
export const dlsRadioGroupLineSpacing = '8px'
export const dlsCheckboxFontSizeS = '12px'
export const dlsCheckboxFontSizeM = '14px'
export const dlsCheckboxFontColor = '#282c33'
export const dlsCheckboxFontColorDisabled = '#a8b0bf'
export const dlsCheckboxSizeS = '14px'
export const dlsCheckboxSizeM = '16px'
export const dlsCheckboxIconMinSize = '8px'
export const dlsCheckboxIconSize = '1em'
export const dlsCheckboxBorderRadius = '2px'
export const dlsCheckboxSpacing = '8px'
export const dlsCheckboxBorderColor = '#d3d9e6'
export const dlsCheckboxBorderColorHover = '#a8b0bf'
export const dlsCheckboxBorderColorFocus = '#0052cc'
export const dlsCheckboxBorderColorActive = '#a8b0bf'
export const dlsCheckboxBorderColorDisabled = '#e2e6f0'
export const dlsCheckboxBorderColorChecked = 'transparent'
export const dlsCheckboxBorderColorCheckedHover = 'transparent'
export const dlsCheckboxBorderColorCheckedFocus = '#0052cc'
export const dlsCheckboxBorderColorCheckedActive = 'transparent'
export const dlsCheckboxBorderColorCheckedDisabled = 'transparent'
export const dlsCheckboxBackgroundColor = '#fff'
export const dlsCheckboxBackgroundColorHover = '#ebedf5'
export const dlsCheckboxBackgroundColorFocus = '#ebedf5'
export const dlsCheckboxBackgroundColorActive = '#d3d9e6'
export const dlsCheckboxBackgroundColorDisabled = '#f6f7fa'
export const dlsCheckboxBackgroundColorChecked = '#0052cc'
export const dlsCheckboxBackgroundColorCheckedHover = '#0066ff'
export const dlsCheckboxBackgroundColorCheckedFocus = '#0066ff'
export const dlsCheckboxBackgroundColorCheckedActive = '#0046ad'
export const dlsCheckboxBackgroundColorCheckedDisabled = '#e1edff'
export const dlsCheckboxForegroundColor = '#fff'
export const dlsCheckboxForegroundColorHover = '#fff'
export const dlsCheckboxForegroundColorFocus = '#fff'
export const dlsCheckboxForegroundColorActive = '#fff'
export const dlsCheckboxForegroundColorDisabled = '#fff'
export const dlsCheckboxShadowFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsCheckboxGroupSpacing = '48px'
export const dlsCheckboxGroupSpacingSimple = '8px'
export const dlsCheckboxGroupLineSpacing = '8px'
export const dlsCheckboxIconSizeStrong = '6px'
export const dlsCheckboxIconSpacingStrong = '3px'
export const dlsCheckboxIconColorStrong = '#d3d9e6'
export const dlsCheckboxIconColorStrongDisabled = '#e2e6f0'
export const dlsCheckboxIconColorStrongChecked = '#0052cc'
export const dlsCheckboxIconColorStrongCheckedDisabled = '#c0d9ff'
export const dlsSwitchWidthXs = '20px'
export const dlsSwitchWidthS = '26px'
export const dlsSwitchWidthM = '32px'
export const dlsSwitchHeightXs = '12px'
export const dlsSwitchHeightS = '16px'
export const dlsSwitchHeightM = '20px'
export const dlsSwitchThumbSizeXs = '8px'
export const dlsSwitchThumbSizeS = '12px'
export const dlsSwitchThumbSizeM = '14px'
export const dlsSwitchPaddingXs = '2px'
export const dlsSwitchPaddingS = '2px'
export const dlsSwitchPaddingM = '3px'
export const dlsSwitchTrackColorOn = '#0052cc'
export const dlsSwitchTrackColorOnHover = '#0066ff'
export const dlsSwitchTrackColorOnFocus = '#0066ff'
export const dlsSwitchTrackColorOnActive = '#0046ad'
export const dlsSwitchTrackColorOnDisabled = '#e1edff'
export const dlsSwitchTrackColorOff = '#a8b0bf'
export const dlsSwitchTrackColorOffHover = '#848b99'
export const dlsSwitchTrackColorOffFocus = '#848b99'
export const dlsSwitchTrackColorOffActive = '#545b66'
export const dlsSwitchTrackColorOffDisabled = '#e2e6f0'
export const dlsSwitchThumbColorOn = '#fff'
export const dlsSwitchThumbColorOnHover = '#fff'
export const dlsSwitchThumbColorOnFocus = '#fff'
export const dlsSwitchThumbColorOnActive = '#fff'
export const dlsSwitchThumbColorOnDisabled = '#fff'
export const dlsSwitchThumbColorOff = '#fff'
export const dlsSwitchThumbColorOffHover = '#fff'
export const dlsSwitchThumbColorOffFocus = '#fff'
export const dlsSwitchThumbColorOffActive = '#fff'
export const dlsSwitchThumbColorOffDisabled = '#fff'
export const dlsSwitchShadowOnFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsSwitchShadowOffFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsInputFontSizeXs = '12px'
export const dlsInputFontSizeS = '12px'
export const dlsInputFontSizeM = '14px'
export const dlsInputFontSizeL = '16px'
export const dlsInputHeightXs = '24px'
export const dlsInputHeightS = '28px'
export const dlsInputHeightM = '32px'
export const dlsInputHeightL = '36px'
export const dlsInputWidth = '300px'
export const dlsInputBorderRadiusXs = '2px'
export const dlsInputBorderRadiusS = '2px'
export const dlsInputBorderRadiusM = '4px'
export const dlsInputBorderRadiusL = '6px'
export const dlsInputPadding = '12px'
export const dlsInputContentSpacing = '8px'
export const dlsInputBorderColor = '#d3d9e6'
export const dlsInputBorderColorHover = '#a8b0bf'
export const dlsInputBorderColorFocus = '#0052cc'
export const dlsInputBorderColorError = '#cc1800'
export const dlsInputBorderColorErrorHover = '#cc1800'
export const dlsInputBorderColorErrorFocus = '#cc1800'
export const dlsInputBorderColorReadOnly = '#d3d9e6'
export const dlsInputBorderColorReadOnlyHover = '#a8b0bf'
export const dlsInputBorderColorReadOnlyFocus = '#0052cc'
export const dlsInputBorderColorDisabled = '#e2e6f0'
export const dlsInputBackgroundColor = '#fff'
export const dlsInputBackgroundColorReadOnly = '#f6f7fa'
export const dlsInputBackgroundColorDisabled = '#f6f7fa'
export const dlsInputFontColor = '#282c33'
export const dlsInputFontColorError = '#282c33'
export const dlsInputFontColorReadOnly = '#848b99'
export const dlsInputFontColorDisabled = '#a8b0bf'
export const dlsInputPlaceholderColor = '#848b99'
export const dlsInputCountColor = '#848b99'
export const dlsInputCountColorError = '#cc1800'
export const dlsInputShadowFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsInputShadowInlineFocus = '0 1px 0 0 #0066ff'
export const dlsInputShadowErrorFocus = '0 0 0 2px rgba(255, 30, 0, 0.2)'
export const dlsInputShadowInlineErrorFocus = '0 1px 0 0 #ff1e00'
export const dlsInputIconSize = 'calc(1em + 2px)'
export const dlsInputIconColor = '#848b99'
export const dlsInputIconColorDisabled = '#a8b0bf'
export const dlsTextareaFontSizeS = '12px'
export const dlsTextareaFontSizeM = '14px'
export const dlsTextareaLineHeight = '1.4'
export const dlsTextareaPaddingX = '12px'
export const dlsTextareaPaddingY = '8px'
export const dlsTextareaWidth = '300px'
export const dlsTextareaRows = '3'
export const dlsTextareaHeightS = '68.4px'
export const dlsTextareaHeightM = '76.8px'
export const dlsTextareaCountHeight = '16px'
export const dlsTextareaCountPadding = '4px'
export const dlsTextareaCountFontSize = '12px'
export const dlsTextareaCountFontColor = '#fff'
export const dlsTextareaCountFontColorError = '#fff'
export const dlsTextareaCountBackgroundColor = 'rgba(0, 0, 0, 0.6)'
export const dlsTextareaCountBackgroundColorError = 'rgba(204, 24, 0, 0.6)'
export const dlsTextareaCountBorderRadius = '2px'
export const dlsNumberInputHeightXs = '24px'
export const dlsNumberInputHeightS = '28px'
export const dlsNumberInputHeightM = '32px'
export const dlsNumberInputSpinIconSizeNormal = '0.85em'
export const dlsNumberInputSpinButtonMinWidthNormal = '18px'
export const dlsNumberInputSpinButtonWidthNormalXs = '18px'
export const dlsNumberInputSpinButtonWidthNormalS = '18px'
export const dlsNumberInputSpinButtonWidthNormalM = '18px'
export const dlsNumberInputDigitsNormal = '6'
export const dlsNumberInputWidthNormalXs = 'calc(6ch + 43px)'
export const dlsNumberInputWidthNormalS = 'calc(6ch + 43px)'
export const dlsNumberInputWidthNormalM = 'calc(6ch + 43px)'
export const dlsNumberInputSpinIconSizeStrong = '1em'
export const dlsNumberInputDigitsStrong = '6'
export const dlsNumberInputWidthStrongXs = 'calc(6ch + 72px)'
export const dlsNumberInputWidthStrongS = 'calc(6ch + 80px)'
export const dlsNumberInputWidthStrongM = 'calc(6ch + 88px)'
export const dlsNumberInputSpinIconColor = '#848b99'
export const dlsNumberInputSpinIconColorDisabled = '#c5ccdb'
export const dlsOptionFontSizeXs = '12px'
export const dlsOptionFontSizeS = '12px'
export const dlsOptionFontSizeM = '14px'
export const dlsOptionFontSizeL = '16px'
export const dlsOptionHeightXs = '24px'
export const dlsOptionHeightS = '28px'
export const dlsOptionHeightM = '32px'
export const dlsOptionHeightL = '36px'
export const dlsOptionPaddingXs = '16px'
export const dlsOptionPaddingS = '16px'
export const dlsOptionPaddingM = '16px'
export const dlsOptionPaddingL = '16px'
export const dlsOptionMaxWidthXs = 'calc(40em + 32px)'
export const dlsOptionMaxWidthS = 'calc(40em + 32px)'
export const dlsOptionMaxWidthM = 'calc(40em + 32px)'
export const dlsOptionMaxWidthL = 'calc(40em + 32px)'
export const dlsOptionIconSizeAux = 'calc(1em + 2px)'
export const dlsOptionLabelSpacingBefore = '8px'
export const dlsOptionLabelSpacingAfter = '12px'
export const dlsOptionFontColor = '#282c33'
export const dlsOptionFontColorHover = '#282c33'
export const dlsOptionFontColorFocus = '#282c33'
export const dlsOptionFontColorActive = '#282c33'
export const dlsOptionFontColorDisabled = '#a8b0bf'
export const dlsOptionFontColorHighlighted = '#ff9100'
export const dlsOptionIconColorAux = '#848b99'
export const dlsOptionIconColorAuxDisabled = '#a8b0bf'
export const dlsOptionFontColorSelected = '#0052cc'
export const dlsOptionFontColorSelectedHover = '#0066ff'
export const dlsOptionFontColorSelectedFocus = '#0066ff'
export const dlsOptionFontColorSelectedActive = '#0046ad'
export const dlsOptionBackgroundColor = '#fff'
export const dlsOptionBackgroundColorHover = '#f6f7fa'
export const dlsOptionBackgroundColorFocus = '#e1edff'
export const dlsOptionBackgroundColorActive = '#e2e6f0'
export const dlsOptionBackgroundColorSelectedParent = '#f6f7fa'
export const dlsOptionFontWeightSelectedParent = '400'
export const dlsOptionFontWeightSelected = '400'
export const dlsOptionGroupLabelLineHeight = '1.4'
export const dlsOptionGroupLabelFontColor = '#848b99'
export const dlsOptionGroupLabelPaddingY = '4px'
export const dlsOptionGroupPaddingY = '4px'
export const dlsOptionGroupSeparatorColor = '#e2e6f0'
export const dlsOptionDropdownMaxHeightXs = '200px'
export const dlsOptionDropdownMaxHeightS = '232px'
export const dlsOptionDropdownMaxHeightM = '264px'
export const dlsOptionDropdownMaxHeightL = '296px'
export const dlsOptionDropdownBorderRadiusXs = '2px'
export const dlsOptionDropdownBorderRadiusS = '2px'
export const dlsOptionDropdownBorderRadiusM = '4px'
export const dlsOptionDropdownBorderRadiusL = '6px'
export const dlsOptionDropdownTargetSpacing = '4px'
export const dlsOptionDropdownShadow = '0 4px 6px rgba(0, 0, 0, 0.06), 0 1px 10px rgba(0, 0, 0, 0.05), 0 2px 4px -1px rgba(0, 0, 0, 0.01)'
export const dlsOptionDropdownPlaceholderPaddingY = '8px'
export const dlsOptionDropdownPlaceholderFontColor = '#848b99'
export const dlsTimePickerWidth = '150px'
export const dlsTimePickerDropdownMinWidth = '150px'
export const dlsTimePickerOptionContentMinWidth = '2ch'
export const dlsTimePickerIconColor = '#848b99'
export const dlsTimePickerIconColorDisabled = '#a8b0bf'
export const dlsPaginationFontSizeXs = '12px'
export const dlsPaginationFontSizeS = '12px'
export const dlsPaginationFontSizeM = '14px'
export const dlsPaginationPageItemMinWidthXs = '24px'
export const dlsPaginationPageItemMinWidthS = '28px'
export const dlsPaginationPageItemMinWidthM = '32px'
export const dlsPaginationPageItemPaddingX = '4px'
export const dlsPaginationPageItemSpacing = '4px'
export const dlsPaginationPageItemsMarginBefore = '48px'
export const dlsPaginationPageItemsMarginAfter = '24px'
export const dlsPaginationSpacing1 = '8px'
export const dlsPaginationSpacing2 = '8px'
export const dlsPaginationGotoInputPaddingX = '8px'
export const dlsPaginationGotoInputDigits = '3'
export const dlsPaginationGotoInputWidth = 'calc(3ch + 18px)'
export const dlsPaginationPageItemFontColor = '#282c33'
export const dlsPaginationPageItemFontColorDisabled = '#a8b0bf'
export const dlsPaginationPageItemBackgroundColor = '#fff'
export const dlsPaginationPageItemBackgroundColorHover = '#f6f7fa'
export const dlsPaginationPageItemBackgroundColorFocus = '#f6f7fa'
export const dlsPaginationPageItemBackgroundColorActive = '#e2e6f0'
export const dlsPaginationPageItemBackgroundColorDisabled = '#fff'
export const dlsBadgeFontSize = '12px'
export const dlsBadgeDotSize = '8px'
export const dlsBadgeLabelHeight = '16px'
export const dlsBadgeLabelMinWidth = '16px'
export const dlsBadgeLabelPaddingX = '4px'
export const dlsBadgeLabelSpacingStatus = '8px'
export const dlsBadgeBackgroundColorInfo = '#0052cc'
export const dlsBadgeBackgroundColorSuccess = '#00bf5c'
export const dlsBadgeBackgroundColorWarning = '#ffa733'
export const dlsBadgeBackgroundColorError = '#cc1800'
export const dlsBadgeBackgroundColorAux = '#a8b0bf'
export const dlsBadgeFontColor = '#fff'
export const dlsBreadcrumbFontSizeS = '12px'
export const dlsBreadcrumbFontSizeM = '14px'
export const dlsBreadcrumbSeparatorIconSize = 'calc(1em + 2px)'
export const dlsBreadcrumbSeparatorColor = '#848b99'
export const dlsBreadcrumbSeparatorSpacing = '8px'
export const dlsBreadcrumbItemCurrentFontColor = '#282c33'
export const dlsBreadcrumbItemCurrentFontWeight = '500'
export const dlsStepsTitleFontSizeS = '12px'
export const dlsStepsTitleFontSizeM = '14px'
export const dlsStepsDescFontSizeS = '12px'
export const dlsStepsDescFontSizeM = '12px'
export const dlsStepsMarkerFontSizeS = '14px'
export const dlsStepsMarkerFontSizeM = '16px'
export const dlsStepsTitleFontWeight = '500'
export const dlsStepsTitleFontColor = '#282c33'
export const dlsStepsTitleFontColorHover = '#282c33'
export const dlsStepsTitleFontColorFocus = '#282c33'
export const dlsStepsTitleFontColorActive = '#282c33'
export const dlsStepsDescFontColor = '#848b99'
export const dlsStepsTitleFontColorIncomplete = '#848b99'
export const dlsStepsDescFontColorIncomplete = '#848b99'
export const dlsStepsMarkerSizeS = '28px'
export const dlsStepsMarkerSizeM = '32px'
export const dlsStepsMarkerIconSize = 'calc(1em + 2px)'
export const dlsStepsTitleMaxWidthHorizontal = '10em'
export const dlsStepsDescMaxWidthHorizontal = '12em'
export const dlsStepsMarkerSpacing = '8px'
export const dlsStepsTitleSpacing = '4px'
export const dlsStepsConnectorSpacingHorizontal = '12px'
export const dlsStepsConnectorSpacingVertical = '8px'
export const dlsStepsConnectorMinLength = '24px'
export const dlsStepsMarkerBackgroundColorIncomplete = '#fff'
export const dlsStepsMarkerBorderColorIncomplete = '#e2e6f0'
export const dlsStepsMarkerForegroundColorIncomplete = '#848b99'
export const dlsStepsConnectorColorIncomplete = '#e2e6f0'
export const dlsStepsMarkerBackgroundColorCompleted = '#fff'
export const dlsStepsMarkerBackgroundColorCompletedHover = '#f2f7ff'
export const dlsStepsMarkerBackgroundColorCompletedFocus = '#f2f7ff'
export const dlsStepsMarkerBackgroundColorCompletedActive = '#e1edff'
export const dlsStepsMarkerBorderColorCompleted = '#0052cc'
export const dlsStepsMarkerForegroundColorCompleted = '#0052cc'
export const dlsStepsMarkerShadowCompletedFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsStepsConnectorColorCompleted = '#0052cc'
export const dlsStepsMarkerBackgroundColorError = '#fff'
export const dlsStepsMarkerBackgroundColorErrorHover = '#fff4f2'
export const dlsStepsMarkerBackgroundColorErrorFocus = '#fff4f2'
export const dlsStepsMarkerBackgroundColorErrorActive = '#ffe4e1'
export const dlsStepsMarkerBorderColorError = '#cc1800'
export const dlsStepsMarkerForegroundColorError = '#cc1800'
export const dlsStepsMarkerShadowErrorFocus = '0 0 0 2px rgba(255, 30, 0, 0.2)'
export const dlsStepsConnectorColorError = '#cc1800'
export const dlsStepsMarkerBackgroundColorCurrent = '#0052cc'
export const dlsStepsMarkerBackgroundColorCurrentHover = '#0066ff'
export const dlsStepsMarkerBackgroundColorCurrentFocus = '#0066ff'
export const dlsStepsMarkerBackgroundColorCurrentActive = '#0046ad'
export const dlsStepsMarkerBorderColorCurrent = 'transparent'
export const dlsStepsMarkerBorderColorCurrentFocus = '#0052cc'
export const dlsStepsMarkerForegroundColorCurrent = '#fff'
export const dlsStepsMarkerShadowCurrentFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsStepsConnectorColorCurrent = '#0052cc'
export const dlsStepsMarkerBackgroundColorErrorCurrent = '#cc1800'
export const dlsStepsMarkerBackgroundColorErrorCurrentHover = '#ff1e00'
export const dlsStepsMarkerBackgroundColorErrorCurrentFocus = '#ff1e00'
export const dlsStepsMarkerBackgroundColorErrorCurrentActive = '#ad1400'
export const dlsStepsMarkerBorderColorErrorCurrent = 'transparent'
export const dlsStepsMarkerBorderColorErrorCurrentFocus = '#cc1800'
export const dlsStepsMarkerForegroundColorErrorCurrent = '#fff'
export const dlsStepsMarkerShadowErrorCurrentFocus = '0 0 0 2px rgba(255, 30, 0, 0.2)'
export const dlsStepsConnectorColorErrorCurrent = '#cc1800'
export const dlsTagHeightS = '24px'
export const dlsTagHeightM = '28px'
export const dlsTagBorderRadiusS = '2px'
export const dlsTagBorderRadiusM = '4px'
export const dlsTagPadding = '8px'
export const dlsTagContentSpacing = '8px'
export const dlsTagListSpacing = '4px'
export const dlsTagFontSizeS = '12px'
export const dlsTagFontSizeM = '14px'
export const dlsTagIconSizeAux = 'calc(1em + 2px)'
export const dlsTagColor = '#545b66'
export const dlsTagColorSelected = '#0052cc'
export const dlsTagColorInfo = '#0052cc'
export const dlsTagColorSuccess = '#00bf5c'
export const dlsTagColorWarning = '#ffa733'
export const dlsTagColorError = '#cc1800'
export const dlsTagFontColorSelected = '#fff'
export const dlsTagFontColorReverse = '#fff'
export const dlsTagFontOpacityDisabled = '0.3'
export const dlsTagFontOpacitySelectedDisabled = '1'
export const dlsTagFontOpacityReverseDisabled = '1'
export const dlsTagBackgroundOpacity = '0.1'
export const dlsTagBackgroundOpacityHover = '0.2'
export const dlsTagBackgroundOpacityActive = '0.3'
export const dlsTagBackgroundOpacityDisabled = '0.1'
export const dlsTagBackgroundOpacitySelected = '1'
export const dlsTagBackgroundOpacitySelectedHover = '0.6'
export const dlsTagBackgroundOpacitySelectedActive = '0.8'
export const dlsTagBackgroundOpacitySelectedDisabled = '0.1'
export const dlsTagBackgroundOpacityReverse = '1'
export const dlsTagBackgroundOpacityReverseHover = '0.6'
export const dlsTagBackgroundOpacityReverseActive = '0.8'
export const dlsTagBackgroundOpacityReverseDisabled = '0.1'
export const dlsTagBorderOpacity = '0.2'
export const dlsTagBorderOpacityDisabled = '0.1'
export const dlsAnchorPaddingX = '16px'
export const dlsAnchorItemSpacing = '12px'
export const dlsAnchorLevelIndent = '16px'
export const dlsAnchorTrackColor = '#e2e6f0'
export const dlsAnchorTrackWidth = '1px'
export const dlsAnchorIndicatorWidth = '2px'
export const dlsAnchorIndicatorColor = '#0052cc'
export const dlsAnchorIndicatorColorHover = '#0066ff'
export const dlsAnchorIndicatorColorFocus = '#0066ff'
export const dlsAnchorIndicatorColorActive = '#0046ad'
export const dlsLoadingFontSizeS = '12px'
export const dlsLoadingFontSizeM = '14px'
export const dlsLoadingFontSizeL = '16px'
export const dlsLoadingSizeS = 'calc(1em + 2px)'
export const dlsLoadingSizeM = 'calc(1em + 2px)'
export const dlsLoadingSizeL = '24px'
export const dlsLoadingPadding = '8px'
export const dlsLoadingContentSpacing = '8px'
export const dlsLoadingFontColorNormal = '#282c33'
export const dlsLoadingFontColorStrong = '#282c33'
export const dlsLoadingFontColorReverse = '#fff'
export const dlsLoadingColorNormal = '#848b99'
export const dlsLoadingColorStrong = '#0052cc'
export const dlsLoadingColorReverse = '#fff'
export const dlsLoadingBackgroundColor = 'rgba(255, 255, 255, 0.8)'
export const dlsLoadingBackgroundColorReverse = 'rgba(0, 0, 0, 0.6)'
export const dlsProgressWidthBarS = '200px'
export const dlsProgressWidthBarM = '300px'
export const dlsProgressSizeCircular = '108px'
export const dlsProgressFontSizeS = '12px'
export const dlsProgressFontSizeM = '14px'
export const dlsProgressFontSizeCircular = '16px'
export const dlsProgressFontColor = '#545b66'
export const dlsProgressIconSizeBar = 'calc(1em + 2px)'
export const dlsProgressIconSizeCircular = '32px'
export const dlsProgressTrackWidthS = '4px'
export const dlsProgressTrackWidthM = '6px'
export const dlsProgressBorderRadiusS = '2px'
export const dlsProgressBorderRadiusM = '3px'
export const dlsProgressTrackSpacingBar = '8px'
export const dlsProgressIconSpacingBar = '8px'
export const dlsProgressTrackColor = '#d3d9e6'
export const dlsProgressIndicatorColor = '#0052cc'
export const dlsProgressIndicatorColorSuccess = '#00bf5c'
export const dlsProgressIndicatorColorError = '#cc1800'
export const dlsProgressIconColorSuccess = '#00bf5c'
export const dlsProgressIconColorError = '#cc1800'
export const dlsTreeNodeHeightS = '28px'
export const dlsTreeNodeHeightM = '32px'
export const dlsTreeFontSizeS = '12px'
export const dlsTreeFontSizeM = '14px'
export const dlsTreeIconSize = 'calc(1em + 2px)'
export const dlsTreeIconSizeAux = 'calc(1em + 2px)'
export const dlsTreePaddingX = '16px'
export const dlsTreeNodeContentSpacing = '8px'
export const dlsTreeIconColorAux = '#848b99'
export const dlsTreeIconColorAuxDisabled = '#a8b0bf'
export const dlsTreeNodeFontColor = '#282c33'
export const dlsTreeNodeFontColorDisabled = '#a8b0bf'
export const dlsTreeNodeFontColorSelected = '#0052cc'
export const dlsTreeNodeFontColorSelectedHover = '#0066ff'
export const dlsTreeNodeFontColorSelectedFocus = '#0066ff'
export const dlsTreeNodeFontColorSelectedActive = '#0046ad'
export const dlsTreeNodeFontColorHighlighted = '#ff9100'
export const dlsTreeNodeBackgroundColor = '#fff'
export const dlsTreeNodeBackgroundColorHover = '#f6f7fa'
export const dlsTreeNodeBackgroundColorFocus = '#f6f7fa'
export const dlsTreeNodeBackgroundColorActive = '#e2e6f0'
export const dlsTreeNodeBackgroundColorSelected = '#f2f7ff'
export const dlsTreeNodeBackgroundColorSelectedHover = '#e1edff'
export const dlsTreeNodeBackgroundColorSelectedFocus = '#e1edff'
export const dlsTreeNodeBackgroundColorSelectedActive = '#c0d9ff'
export const dlsTreeNodeFontWeightSelected = '400'
export const dlsTransferFontSizeS = '12px'
export const dlsTransferFontSizeM = '14px'
export const dlsTransferBorderRadiusS = '2px'
export const dlsTransferBorderRadiusM = '4px'
export const dlsTransferPaddingX = '12px'
export const dlsTransferHeaderContentSpacing1 = '4px'
export const dlsTransferHeaderContentSpacing2 = '8px'
export const dlsTransferContentPaddingY = '4px'
export const dlsTransferItemHeightS = '28px'
export const dlsTransferItemHeightM = '32px'
export const dlsTransferItemContentSpacing = '8px'
export const dlsTransferContentMinHeightS = '288px'
export const dlsTransferContentMinHeightM = '328px'
export const dlsTransferPanelWidthS = '228px'
export const dlsTransferPanelWidthM = '260px'
export const dlsTransferPanelSpacing = '12px'
export const dlsTransferHeaderHeightS = '32px'
export const dlsTransferHeaderHeightM = '36px'
export const dlsTransferHeaderBackgroundColor = '#f6f7fa'
export const dlsTransferHeaderFontColor = '#282c33'
export const dlsTransferHeaderFontWeight = '400'
export const dlsTransferHeaderButtonFontWeight = '500'
export const dlsTransferBorderColor = '#d3d9e6'
export const dlsTransferItemBackgroundColor = '#fff'
export const dlsTransferItemBackgroundColorHover = '#f6f7fa'
export const dlsTransferItemBackgroundColorFocus = '#f6f7fa'
export const dlsTransferItemBackgroundColorActive = '#e2e6f0'
export const dlsTransferItemBackgroundColorDisabled = 'transparent'
export const dlsTransferItemFontColor = '#282c33'
export const dlsTransferItemFontColorHover = '#282c33'
export const dlsTransferItemFontColorFocus = '#282c33'
export const dlsTransferItemFontColorActive = '#282c33'
export const dlsTransferItemFontColorDisabled = '#a8b0bf'
export const dlsTabHeightS = '40px'
export const dlsTabHeightM = '44px'
export const dlsTabHeightL = '48px'
export const dlsTabPaddingXS = '16px'
export const dlsTabPaddingXM = '16px'
export const dlsTabPaddingXL = '16px'
export const dlsTabSpacingXS = '20px'
export const dlsTabSpacingXM = '24px'
export const dlsTabSpacingXL = '28px'
export const dlsTabDropdownSpacingX = '8px'
export const dlsTabDropdownSpacingY = '8px'
export const dlsTabContentSpacing = '4px'
export const dlsTabNavSpacingS = '8px'
export const dlsTabNavSpacingM = '12px'
export const dlsTabNavSpacingL = '16px'
export const dlsTabAddSpacingS = '8px'
export const dlsTabAddSpacingM = '12px'
export const dlsTabAddSpacingL = '16px'
export const dlsTabFontSizeS = '12px'
export const dlsTabFontSizeM = '14px'
export const dlsTabFontSizeL = '16px'
export const dlsTabFontColor = '#282c33'
export const dlsTabFontColorHover = '#545b66'
export const dlsTabFontColorFocus = '#545b66'
export const dlsTabFontColorActive = '#000'
export const dlsTabFontColorDisabled = '#a8b0bf'
export const dlsTabFontColorCurrent = '#0052cc'
export const dlsTabFontColorCurrentHover = '#0066ff'
export const dlsTabFontColorCurrentFocus = '#0066ff'
export const dlsTabFontColorCurrentActive = '#0046ad'
export const dlsTabBackgroundColorFocus = '#e1edff'
export const dlsTabFontWeightCurrent = '400'
export const dlsTabContainerSeparatorColor = '#d3d9e6'
export const dlsTabContainerBorderWidth = '1px'
export const dlsTabPanelPaddingTop = '16px'
export const dlsTabOverflowShadowRight = '0 4px 6px rgba(0, 0, 0, 0.06), 0 1px 10px rgba(0, 0, 0, 0.05), 0 2px 4px -1px rgba(0, 0, 0, 0.01)'
export const dlsTabOverflowShadowLeft = '0 4px 6px rgba(0, 0, 0, 0.06), 0 1px 10px rgba(0, 0, 0, 0.05), 0 2px 4px -1px rgba(0, 0, 0, 0.01)'
export const dlsTabIndicatorColor = '#0052cc'
export const dlsTabIndicatorWidth = '2px'
export const dlsTabIconSize = 'calc(1em + 2px)'
export const dlsTabIconSizeAux = 'calc(1em + 2px)'
export const dlsTabSeparatorHeightSimple = '1em'
export const dlsTabSeparatorColor = '#d3d9e6'
export const dlsTabMenuHeightStrongM = '48px'
export const dlsTabPaddingXStrongM = '24px'
export const dlsTabSpacingXStrongM = '0'
export const dlsTabMenuPaddingXStrongM = '16px'
export const dlsTabBorderRadiusStrongM = '10px'
export const dlsTabContainerBorderRadiusStrongM = '6px'
export const dlsTabContainerBorderColorStrong = '#ebedf5'
export const dlsTabMenuBackgroundColorStrong = '#ebedf5'
export const dlsTabBackgroundColorStrong = 'transparent'
export const dlsTabBackgroundColorStrongHover = '#f6f7fa'
export const dlsTabBackgroundColorStrongFocus = '#e1edff'
export const dlsTabBackgroundColorStrongActive = '#e2e6f0'
export const dlsTabBackgroundColorStrongDisabled = 'transparent'
export const dlsTabBackgroundColorStrongCurrent = '#fff'
export const dlsTabBackgroundColorStrongCurrentHover = '#fff'
export const dlsTabBackgroundColorStrongCurrentFocus = '#fff'
export const dlsTabBackgroundColorStrongCurrentActive = '#fff'
export const dlsTabFontColorStrong = '#282c33'
export const dlsTabFontColorStrongHover = '#545b66'
export const dlsTabFontColorStrongFocus = '#545b66'
export const dlsTabFontColorStrongActive = '#000'
export const dlsTabFontColorStrongDisabled = '#a8b0bf'
export const dlsTabFontColorStrongCurrent = '#0052cc'
export const dlsTabFontColorStrongCurrentHover = '#0066ff'
export const dlsTabFontColorStrongCurrentFocus = '#0066ff'
export const dlsTabFontColorStrongCurrentActive = '#0046ad'
export const dlsMenuItemFontWeightCurrent = '400'
export const dlsMenuItemIconColor = '#848b99'
export const dlsMenuItemIconColorDisabled = '#a8b0bf'
export const dlsMenuItemIconColorCurrent = 'currentColor'
export const dlsMenuItemIconColorCurrentDisabled = 'currentColor'
export const dlsMenuIndicatorWidth = '2px'
export const dlsMenuIndicatorColor = '#0052cc'
export const dlsMenuIndicatorColorHover = '#0052cc'
export const dlsMenuIndicatorColorFocus = '#0052cc'
export const dlsMenuIndicatorColorActive = '#0052cc'
export const dlsMenuWidth = '200px'
export const dlsMenuWidthCollapsed = '60px'
export const dlsMenuFontSizeS = '12px'
export const dlsMenuFontSizeM = '14px'
export const dlsMenuFontSizeL = '16px'
export const dlsMenuItemHeightS = '28px'
export const dlsMenuItemHeightM = '32px'
export const dlsMenuItemHeightL = '36px'
export const dlsMenuPaddingX = '20px'
export const dlsMenuPaddingY = '8px'
export const dlsMenuItemSpacing = '8px'
export const dlsMenuItemContentSpacing = '12px'
export const dlsMenuLevelIndent = '16px'
export const dlsMenuFooterHeightS = '28px'
export const dlsMenuFooterHeightM = '28px'
export const dlsMenuFooterHeightL = '28px'
export const dlsMenuItemIconSize = 'calc(1em + 2px)'
export const dlsMenuItemIconSizeAux = 'calc(1em + 2px)'
export const dlsMenuBorderColor = '#e2e6f0'
export const dlsMenuSeparatorColor = '#e2e6f0'
export const dlsMenuItemFontColor = '#282c33'
export const dlsMenuItemFontColorHover = '#282c33'
export const dlsMenuItemFontColorFocus = '#282c33'
export const dlsMenuItemFontColorActive = '#282c33'
export const dlsMenuItemFontColorDisabled = '#a8b0bf'
export const dlsMenuItemBackgroundColor = '#fff'
export const dlsMenuItemBackgroundColorHover = '#f6f7fa'
export const dlsMenuItemBackgroundColorFocus = '#e1edff'
export const dlsMenuItemBackgroundColorActive = '#e2e6f0'
export const dlsMenuItemBackgroundColorDisabled = '#fff'
export const dlsMenuItemFontColorCurrent = '#0052cc'
export const dlsMenuItemFontColorCurrentHover = '#0066ff'
export const dlsMenuItemFontColorCurrentFocus = '#0066ff'
export const dlsMenuItemFontColorCurrentActive = '#0046ad'
export const dlsMenuItemBackgroundColorCurrent = '#f2f7ff'
export const dlsMenuItemBackgroundColorCurrentHover = '#e1edff'
export const dlsMenuItemBackgroundColorCurrentFocus = '#e1edff'
export const dlsMenuItemBackgroundColorCurrentActive = '#c0d9ff'
export const dlsMenuItemGroupLabelFontColor = '#282c33'
export const dlsMenuItemGroupLabelFontWeight = '500'
export const dlsMenuFontSizeHorizontalM = '14px'
export const dlsMenuItemHeightHorizontalM = '64px'
export const dlsMenuItemPaddingXHorizontal = '16px'
export const dlsMenuBorderColorHorizontal = '#d3d9e6'
export const dlsMenuItemFontColorHorizontal = '#282c33'
export const dlsMenuItemFontColorHorizontalHover = '#545b66'
export const dlsMenuItemFontColorHorizontalFocus = '#545b66'
export const dlsMenuItemFontColorHorizontalActive = '#000'
export const dlsMenuItemFontColorHorizontalDisabled = '#a8b0bf'
export const dlsMenuItemBackgroundColorHorizontal = '#fff'
export const dlsMenuItemBackgroundColorHorizontalHover = '#fff'
export const dlsMenuItemBackgroundColorHorizontalFocus = '#e1edff'
export const dlsMenuItemBackgroundColorHorizontalActive = '#fff'
export const dlsMenuItemBackgroundColorHorizontalDisabled = '#fff'
export const dlsMenuItemFontColorHorizontalCurrent = '#0052cc'
export const dlsMenuItemFontColorHorizontalCurrentHover = '#0066ff'
export const dlsMenuItemFontColorHorizontalCurrentFocus = '#0066ff'
export const dlsMenuItemFontColorHorizontalCurrentActive = '#0046ad'
export const dlsMenuItemBackgroundColorHorizontalCurrent = '#fff'
export const dlsMenuItemBackgroundColorHorizontalCurrentHover = '#fff'
export const dlsMenuItemBackgroundColorHorizontalCurrentFocus = '#e1edff'
export const dlsMenuItemBackgroundColorHorizontalCurrentActive = '#fff'
export const dlsMenuItemShadowHorizontalCurrent = '0 1px 2px rgba(0, 82, 204, 0.4)'
export const dlsMenuCardBackgroundColor = '#fff'
export const dlsMenuCardBorderRadiusM = '4px'
export const dlsMenuCardPaddingX = '24px'
export const dlsMenuCardPaddingY = '24px'
export const dlsMenuCardColumnSpacing = '48px'
export const dlsMenuCardTargetSpacing = '4px'
export const dlsMenuCardTitleFontColor = '#282c33'
export const dlsMenuCardTitleFontSizeM = '14px'
export const dlsMenuCardTitleFontWeight = '500'
export const dlsMenuCardItemSpacing = '16px'
export const dlsMenuCardItemFontSizeM = '14px'
export const dlsMenuCardItemIconSize = 'calc(1em + 2px)'
export const dlsSelectWidth = '300px'
export const dlsSelectTagHeightXs = '16px'
export const dlsSelectTagHeightS = '20px'
export const dlsSelectTagHeightM = '24px'
export const dlsSelectTagHeightL = '28px'
export const dlsSelectPaddingX = '12px'
export const dlsSelectPaddingY = '4px'
export const dlsSelectContentSpacing = '8px'
export const dlsSelectTagSpacing = '4px'
export const dlsSelectIconSizeAux = 'calc(1em + 2px)'
export const dlsSelectIconColorAux = '#848b99'
export const dlsSelectIconColorAuxDisabled = '#a8b0bf'
export const dlsDropdownIconSizeAux = 'calc(1em + 2px)'
export const dlsDropdownIconSpacingAux = '8px'
export const dlsDropdownSplitButtonPaddingX = '12px'
export const dlsDropdownIconColorAuxNormal = '#848b99'
export const dlsDropdownIconColorAuxNormalDisabled = '#a8b0bf'
export const dlsDropdownIconColorAuxPrimary = 'currentColor'
export const dlsDropdownIconColorAuxPrimaryDisabled = 'currentColor'
export const dlsSliderThumbSizeS = '8px'
export const dlsSliderThumbSizeM = '12px'
export const dlsSliderTrackMinWidth = '2px'
export const dlsSliderTrackWidthS = '2px'
export const dlsSliderTrackWidthM = '4px'
export const dlsSliderTrackBorderRadiusS = '1px'
export const dlsSliderTrackBorderRadiusM = '2px'
export const dlsSliderThumbDotMinSize = '2px'
export const dlsSliderThumbDotSizeS = '2px'
export const dlsSliderThumbDotSizeM = '4px'
export const dlsSliderMarkMinSize = '4px'
export const dlsSliderMarkSizeS = '4px'
export const dlsSliderMarkSizeM = '8px'
export const dlsSliderFontSizeS = '12px'
export const dlsSliderFontSizeM = '14px'
export const dlsSliderFontColor = '#282c33'
export const dlsSliderFontColorDisabled = '#a8b0bf'
export const dlsSliderThumbBackgroundColor = '#0052cc'
export const dlsSliderThumbBackgroundColorHover = '#0066ff'
export const dlsSliderThumbBackgroundColorFocus = '#0066ff'
export const dlsSliderThumbBackgroundColorActive = '#0046ad'
export const dlsSliderThumbBackgroundColorReadOnly = '#a8b0bf'
export const dlsSliderThumbBackgroundColorReadOnlyHover = '#848b99'
export const dlsSliderThumbBackgroundColorReadOnlyFocus = '#848b99'
export const dlsSliderThumbBackgroundColorDisabled = '#e1edff'
export const dlsSliderThumbBorderColorFocus = '#0052cc'
export const dlsSliderThumbBorderColorReadOnlyFocus = '#0052cc'
export const dlsSliderThumbShadowFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsSliderThumbShadowReadOnlyFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsSliderThumbDotColor = '#fff'
export const dlsSliderTrackColor = '#e2e6f0'
export const dlsSliderTrackColorHover = '#d3d9e6'
export const dlsSliderTrackColorReadOnly = '#e2e6f0'
export const dlsSliderTrackColorDisabled = '#e2e6f0'
export const dlsSliderTrackProgressColor = '#0052cc'
export const dlsSliderTrackProgressColorHover = '#0066ff'
export const dlsSliderTrackProgressColorReadOnly = '#a8b0bf'
export const dlsSliderTrackProgressColorDisabled = '#e1edff'
export const dlsCollapseHeaderPaddingX = '16px'
export const dlsCollapseHeaderPaddingY = '16px'
export const dlsCollapseHeaderContentSpacing = '8px'
export const dlsCollapseHeaderFontSizeS = '12px'
export const dlsCollapseHeaderFontSizeM = '14px'
export const dlsCollapseHeaderLineHeight = '1.4'
export const dlsCollapseBodyPaddingX = '16px'
export const dlsCollapseBodyPaddingY = '16px'
export const dlsCollapseBodyFontSizeS = '12px'
export const dlsCollapseBodyFontSizeM = '14px'
export const dlsCollapseBodyLineHeight = '1.8'
export const dlsCollapseBorderRadiusS = '2px'
export const dlsCollapseBorderRadiusM = '4px'
export const dlsCollapseIconSizeAux = 'calc(1em + 2px)'
export const dlsCollapseBorderColor = '#e2e6f0'
export const dlsCollapseBorderColorFocus = '#e2e6f0'
export const dlsCollapseHeaderFontColor = '#282c33'
export const dlsCollapseHeaderFontColorDisabled = '#a8b0bf'
export const dlsCollapseHeaderBackgroundColor = '#f6f7fa'
export const dlsCollapseHeaderBackgroundColorHover = '#ebedf5'
export const dlsCollapseHeaderBackgroundColorFocus = '#e1edff'
export const dlsCollapseHeaderBackgroundColorActive = '#e2e6f0'
export const dlsCollapseHeaderBackgroundColorDisabled = '#f6f7fa'
export const dlsCollapseBodyFontColor = '#545b66'
export const dlsCollapseBodyBackgroundColor = '#fff'
export const dlsCollapseIconColorAux = '#848b99'
export const dlsCollapseIconColorAuxDisabled = '#a8b0bf'
export const dlsTooltipMaxWidth = 'calc(16em + 16px)'
export const dlsTooltipFontSize = '12px'
export const dlsTooltipLineHeight = '1.8'
export const dlsTooltipPaddingX = '8px'
export const dlsTooltipPaddingY = '4px'
export const dlsTooltipTargetSpacing = '4px'
export const dlsTooltipArrowWidth = '8px'
export const dlsTooltipArrowSpacing = '8px'
export const dlsTooltipBorderRadius = '2px'
export const dlsTooltipFontColorNormal = '#282c33'
export const dlsTooltipBackgroundColorNormal = '#fff'
export const dlsTooltipShadowNormal = '0 4px 6px rgba(0, 0, 0, 0.06), 0 1px 10px rgba(0, 0, 0, 0.05), 0 2px 4px -1px rgba(0, 0, 0, 0.01)'
export const dlsTooltipFontColorReverse = '#fff'
export const dlsTooltipBackgroundColorReverse = 'rgba(0, 0, 0, 0.6)'
export const dlsTooltipShadowReverse = 'none'
export const dlsPopoverMaxWidth = '400px'
export const dlsPopoverLineHeight = '1.8'
export const dlsPopoverPaddingX = '16px'
export const dlsPopoverPaddingY = '16px'
export const dlsPopoverArrowSpacing = '12px'
export const dlsPopoverBorderRadiusS = '2px'
export const dlsPopoverBorderRadiusM = '4px'
export const dlsPopoverFontSizeS = '12px'
export const dlsPopoverFontSizeM = '14px'
export const dlsDialogWidthWide = '800px'
export const dlsDialogWidthMedium = '600px'
export const dlsDialogWidthNarrow = '400px'
export const dlsDialogWidthContextual = '400px'
export const dlsDialogBorderRadius = '6px'
export const dlsDialogSpacingY = '20px'
export const dlsDialogMaxHeight = 'calc(100vh - 40px)'
export const dlsDialogPaddingX = '24px'
export const dlsDialogPaddingY = '24px'
export const dlsDialogTitleContentSpacing = '16px'
export const dlsDialogTitleSpacingAfter = '12px'
export const dlsDialogActionSpacingBefore = '20px'
export const dlsDialogActionContentSpacing = '12px'
export const dlsDialogTitleSpacingAfterContextual = '4px'
export const dlsDialogIconSpacingContextual = '12px'
export const dlsDialogIconSizeContextual = 'calc(1em + 2px)'
export const dlsDialogTitleLineHeight = '1.4'
export const dlsDialogBodyLineHeight = '1.8'
export const dlsDialogTitleFontSizeS = '16px'
export const dlsDialogTitleFontSizeM = '20px'
export const dlsDialogBodyFontSizeS = '12px'
export const dlsDialogBodyFontSizeM = '14px'
export const dlsDialogTitleFontWeight = '500'
export const dlsDialogTitleFontColor = '#282c33'
export const dlsDialogBodyFontColor = '#282c33'
export const dlsDialogCloseIconSize = 'calc(1em + 2px)'
export const dlsDialogCloseIconSizeFullscreen = 'calc(1em + 2px)'
export const dlsDialogCloseIconMinSize = '10px'
export const dlsDialogIconColorInfo = '#0052cc'
export const dlsDialogIconColorSuccess = '#00bf5c'
export const dlsDialogIconColorWarning = '#ffa733'
export const dlsDialogIconColorError = '#cc1800'
export const dlsDialogBackgroundColor = '#fff'
export const dlsDialogBackdropColor = 'rgba(0, 0, 0, 0.6)'
export const dlsDialogShadow = '0 8px 10px 1px rgba(0, 0, 0, 0.06), 0 3px 14px 2px rgba(0, 0, 0, 0.05), 0 5px 5px -3px rgba(0, 0, 0, 0.01)'
export const dlsDrawerWidth = '800px'
export const dlsDrawerLevelIndent = '60px'
export const dlsDrawerShadowTop = '0 8px 10px 1px rgba(0, 0, 0, 0.06), 0 3px 14px 2px rgba(0, 0, 0, 0.05), 0 5px 5px -3px rgba(0, 0, 0, 0.01)'
export const dlsDrawerShadowRight = '0 8px 10px 1px rgba(0, 0, 0, 0.06), 0 3px 14px 2px rgba(0, 0, 0, 0.05), 0 5px 5px -3px rgba(0, 0, 0, 0.01)'
export const dlsDrawerShadowBottom = '0 8px 10px 1px rgba(0, 0, 0, 0.06), 0 3px 14px 2px rgba(0, 0, 0, 0.05), 0 5px 5px -3px rgba(0, 0, 0, 0.01)'
export const dlsDrawerShadowLeft = '0 8px 10px 1px rgba(0, 0, 0, 0.06), 0 3px 14px 2px rgba(0, 0, 0, 0.05), 0 5px 5px -3px rgba(0, 0, 0, 0.01)'
export const dlsDrawerTitleFontSizeS = '16px'
export const dlsDrawerTitleFontSizeM = '20px'
export const dlsDrawerCloseIconSize = 'calc(1em + 2px)'
export const dlsEmbeddedTitleFontSizeS = '16px'
export const dlsEmbeddedTitleFontSizeM = '20px'
export const dlsEmbeddedCloseIconSize = 'calc(1em + 2px)'
export const dlsEmbeddedShadow = '0 8px 10px 1px rgba(0, 0, 0, 0.06), 0 3px 14px 2px rgba(0, 0, 0, 0.05), 0 5px 5px -3px rgba(0, 0, 0, 0.01)'
export const dlsEmbeddedBorderColor = '#e2e6f0'
export const dlsToastFontSizeS = '12px'
export const dlsToastFontSizeM = '14px'
export const dlsToastBorderRadiusS = '2px'
export const dlsToastBorderRadiusM = '4px'
export const dlsToastLineHeight = '1.8'
export const dlsToastIconSize = 'calc(1em + 2px)'
export const dlsToastIconSizeTitled = 'calc(1em + 2px)'
export const dlsToastIconSizeAux = 'calc(1em + 2px)'
export const dlsToastPaddingX = '16px'
export const dlsToastPaddingY = '8px'
export const dlsToastPaddingYTitled = '16px'
export const dlsToastContentSpacingBefore = '8px'
export const dlsToastContentSpacingAfter = '16px'
export const dlsToastSpacing = '8px'
export const dlsToastMaxWidth = '400px'
export const dlsToastListSpacingBefore = '8px'
export const dlsToastTitleFontSizeS = '14px'
export const dlsToastTitleFontSizeM = '16px'
export const dlsToastTitleFontWeight = '500'
export const dlsToastTitleLineHeight = '1.4'
export const dlsToastTitleSpacingAfter = '4px'
export const dlsToastBackgroundColor = '#fff'
export const dlsToastFontColor = '#282c33'
export const dlsToastTitleFontColor = '#282c33'
export const dlsToastShadow = '0 4px 6px rgba(0, 0, 0, 0.06), 0 1px 10px rgba(0, 0, 0, 0.05), 0 2px 4px -1px rgba(0, 0, 0, 0.01)'
export const dlsToastIconColorInfo = '#0052cc'
export const dlsToastIconColorSuccess = '#00bf5c'
export const dlsToastIconColorWarning = '#ffa733'
export const dlsToastIconColorError = '#cc1800'
export const dlsAlertFontSizeS = '12px'
export const dlsAlertFontSizeM = '14px'
export const dlsAlertBorderRadiusS = '2px'
export const dlsAlertBorderRadiusM = '4px'
export const dlsAlertLineHeight = '1.8'
export const dlsAlertIconSize = 'calc(1em + 2px)'
export const dlsAlertIconSizeTitled = 'calc(1em + 2px)'
export const dlsAlertIconSizeAux = 'calc(1em + 2px)'
export const dlsAlertPaddingX = '16px'
export const dlsAlertPaddingY = '4px'
export const dlsAlertPaddingYMultiline = '8px'
export const dlsAlertPaddingYTitled = '16px'
export const dlsAlertContentSpacingBefore = '8px'
export const dlsAlertContentSpacingAfter = '16px'
export const dlsAlertTitleFontSizeS = '12px'
export const dlsAlertTitleFontSizeM = '14px'
export const dlsAlertTitleFontWeight = '500'
export const dlsAlertTitleLineHeight = '1.4'
export const dlsAlertTitleSpacingAfter = '4px'
export const dlsAlertFontColor = '#282c33'
export const dlsAlertTitleFontColor = '#282c33'
export const dlsAlertBackgroundColorInfo = '#f2f7ff'
export const dlsAlertBackgroundColorSuccess = '#f2fff8'
export const dlsAlertBackgroundColorWarning = '#fffaf2'
export const dlsAlertBackgroundColorError = '#fff4f2'
export const dlsAlertBorderColorInfo = 'transparent'
export const dlsAlertBorderColorSuccess = 'transparent'
export const dlsAlertBorderColorWarning = 'transparent'
export const dlsAlertBorderColorError = 'transparent'
export const dlsAlertIconColorInfo = '#0052cc'
export const dlsAlertIconColorSuccess = '#00bf5c'
export const dlsAlertIconColorWarning = '#ffa733'
export const dlsAlertIconColorError = '#cc1800'
export const dlsAlertPagesSpacing = '4px'
export const dlsAlertPagesFontColor = '#848b99'
export const dlsCarouselIndicatorWidth = '12px'
export const dlsCarouselIndicatorWidthCurrent = '36px'
export const dlsCarouselIndicatorHeight = '3px'
export const dlsCarouselIndicatorDotSize = '6px'
export const dlsCarouselIndicatorBorderRadius = '1.5px'
export const dlsCarouselIndicatorSpacingOuter = '12px'
export const dlsCarouselIndicatorSpacingInner = '4px'
export const dlsCarouselIndicatorBackgroundColor = 'rgba(255, 255, 255, 0.3)'
export const dlsCarouselIndicatorBackgroundColorHover = 'rgba(255, 255, 255, 0.4)'
export const dlsCarouselIndicatorBackgroundColorActive = '#fff'
export const dlsCarouselIndicatorBackgroundColorCurrent = '#fff'
export const dlsCarouselIndicatorBackgroundColorOutside = '#ebedf5'
export const dlsCarouselIndicatorBackgroundColorOutsideHover = '#e2e6f0'
export const dlsCarouselIndicatorBackgroundColorOutsideActive = '#d3d9e6'
export const dlsCarouselIndicatorBackgroundColorOutsideCurrent = '#0052cc'
export const dlsCarouselControlSize = '32px'
export const dlsCarouselControlBorderRadius = '50%'
export const dlsCarouselControlShadow = '0 4px 6px rgba(0, 0, 0, 0.06), 0 1px 10px rgba(0, 0, 0, 0.05), 0 2px 4px -1px rgba(0, 0, 0, 0.01)'
export const dlsCarouselControlIconSize = '16px'
export const dlsCarouselControlSpacingInside = '24px'
export const dlsCarouselControlSpacingOutside = '16px'
export const dlsCarouselPagesHeight = '20px'
export const dlsCarouselPagesFontSize = '12px'
export const dlsCarouselPagesFontColor = '#fff'
export const dlsCarouselPagesBackgroundColor = 'rgba(0, 0, 0, 0.6)'
export const dlsCarouselPagesPaddingX = '4px'
export const dlsCarouselPagesBorderRadius = '2px'
export const dlsCarouselSlideGutter = '12px'
export const dlsDatePickerWidth = '160px'
export const dlsDatePickerWidthRange = '248px'
export const dlsDatePickerFontSizeS = '12px'
export const dlsDatePickerFontSizeM = '14px'
export const dlsDatePickerDropdownBackgroundColor = '#fff'
export const dlsDatePickerDropdownBorderRadiusS = '2px'
export const dlsDatePickerDropdownBorderRadiusM = '4px'
export const dlsDatePickerDropdownShadow = '0 4px 6px rgba(0, 0, 0, 0.06), 0 1px 10px rgba(0, 0, 0, 0.05), 0 2px 4px -1px rgba(0, 0, 0, 0.01)'
export const dlsDatePickerIconColor = '#848b99'
export const dlsDatePickerIconColorDisabled = '#a8b0bf'
export const dlsDatePickerRangeConnectorPaddingX = '12px'
export const dlsDatePickerRangePanelSpacing = '8px'
export const dlsDatePickerPanelSeparatorColor = '#e2e6f0'
export const dlsDatePickerPanelPaddingX = '16px'
export const dlsDatePickerHeaderPaginationIconSize = 'calc(1em + 2px)'
export const dlsDatePickerHeaderIconSizeAux = 'calc(1em + 2px)'
export const dlsDatePickerHeaderPaddingY = '8px'
export const dlsDatePickerHeaderPaginationSpacing = '12px'
export const dlsDatePickerHeaderIconAuxSpacing = '8px'
export const dlsDatePickerDayBodyPaddingTop = '4px'
export const dlsDatePickerDayBodyPaddingBottom = '12px'
export const dlsDatePickerWeekdaysSpacingAfter = '12px'
export const dlsDatePickerMonthBodyPaddingTop = '24px'
export const dlsDatePickerYearBodyPaddingTop = '24px'
export const dlsDatePickerCellHeightS = '28px'
export const dlsDatePickerCellHeightM = '32px'
export const dlsDatePickerCellWidthS = '28px'
export const dlsDatePickerCellWidthM = '32px'
export const dlsDatePickerCellBorderRadiusS = '2px'
export const dlsDatePickerCellBorderRadiusM = '4px'
export const dlsDatePickerCellShadowFocus = '0 0 0 2px rgba(0, 102, 255, 0.2)'
export const dlsDatePickerCellFontColor = '#282c33'
export const dlsDatePickerCellFontColorCurrent = '#0052cc'
export const dlsDatePickerCellFontColorAux = '#848b99'
export const dlsDatePickerCellFontColorDisabled = '#a8b0bf'
export const dlsDatePickerCellFontColorSelected = '#fff'
export const dlsDatePickerCellFontWeight = '400'
export const dlsDatePickerCellFontWeightSelected = '500'
export const dlsDatePickerDayCellSpacingY = '4px'
export const dlsDatePickerDayCellSpacingX = '8px'
export const dlsDatePickerMonthCellSpacingY = '20px'
export const dlsDatePickerMonthCellPaddingX = '8px'
export const dlsDatePickerYearCellSpacingY = '20px'
export const dlsDatePickerYearCellPaddingX = '8px'
export const dlsDatePickerCellBackgroundColor = 'transparent'
export const dlsDatePickerCellBackgroundColorHover = '#f6f7fa'
export const dlsDatePickerCellBackgroundColorFocus = '#f6f7fa'
export const dlsDatePickerCellBackgroundColorActive = '#e2e6f0'
export const dlsDatePickerCellBackgroundColorDisabled = '#f6f7fa'
export const dlsDatePickerCellBackgroundColorSelected = '#0052cc'
export const dlsDatePickerCellBackgroundColorSelectedHover = '#0066ff'
export const dlsDatePickerCellBackgroundColorSelectedFocus = '#0066ff'
export const dlsDatePickerCellBackgroundColorSelectedActive = '#0046ad'
export const dlsDatePickerCellBackgroundColorAux = 'transparent'
export const dlsDatePickerCellBackgroundColorAuxHover = '#f6f7fa'
export const dlsDatePickerCellBackgroundColorAuxFocus = '#f6f7fa'
export const dlsDatePickerCellBackgroundColorAuxActive = '#e2e6f0'
export const dlsDatePickerCellBackgroundColorRangeEndHover = '#e1edff'
export const dlsDatePickerCellBackgroundColorRangeEndActive = '#c0d9ff'
export const dlsDatePickerCellBackgroundColorRangeEndFocus = '#e1edff'
export const dlsDatePickerCellBorderColor = 'transparent'
export const dlsDatePickerCellBorderColorFocus = '#0052cc'
export const dlsDatePickerCellBorderColorRangeEndHover = 'transparent'
export const dlsDatePickerCellBorderColorRangeEndActive = 'transparent'
export const dlsDatePickerListItemHeightM = '32px'
export const dlsDatePickerListItemHeightS = '28px'
export const dlsDatePickerListPaddingY = '8px'
export const dlsDatePickerListItemSpacing = '4px'
export const dlsDatePickerListItemPaddingX = '16px'
export const dlsDatePickerListBackgroundColor = 'transparent'
export const dlsDatePickerListBackgroundColorHover = '#f6f7fa'
export const dlsDatePickerListBackgroundColorFocus = '#f6f7fa'
export const dlsDatePickerListBackgroundColorActive = '#e2e6f0'
export const dlsDatePickerListBackgroundColorSelected = '#f2f7ff'
export const dlsDatePickerListBackgroundColorSelectedHover = '#e1edff'
export const dlsDatePickerListBackgroundColorSelectedFocus = '#e1edff'
export const dlsDatePickerListBackgroundColorSelectedActive = '#c0d9ff'
export const dlsDatePickerListFontColor = '#282c33'
export const dlsDatePickerListFontColorHover = '#282c33'
export const dlsDatePickerListFontColorFocus = '#282c33'
export const dlsDatePickerListFontColorActive = '#282c33'
export const dlsDatePickerListFontColorSelected = '#0052cc'
export const dlsDatePickerListFontColorSelectedHover = '#0052cc'
export const dlsDatePickerListFontColorSelectedFocus = '#0052cc'
export const dlsDatePickerListFontColorSelectedActive = '#0052cc'
export const dlsDatePickerListFontWeight = '400'
export const dlsDatePickerListFontWeightSelected = '500'
export const dlsDatePickerYearScrollerWidth = '80px'
export const dlsUploaderFontSizeS = '12px'
export const dlsUploaderFontSizeM = '14px'
export const dlsUploaderHelperTextSpacingFile = '8px'
export const dlsUploaderHelperTextSpacingMedia = '12px'
export const dlsUploaderHelperTextFontColor = '#848b99'
export const dlsUploaderHelperTextMaxWidthMedia = '10em'
export const dlsUploaderFileItemSpacingY = '12px'
export const dlsUploaderFileItemPaddingX = '4px'
export const dlsUploaderFileItemLabelSpacingBefore = '4px'
export const dlsUploaderFileItemLabelSpacingAfter = '8px'
export const dlsUploaderFileItemWidth = '256px'
export const dlsUploaderFileItemHeight = '20px'
export const dlsUploaderFileItemBackgroundColor = 'transparent'
export const dlsUploaderFileItemBackgroundColorHover = '#f6f7fa'
export const dlsUploaderFileItemFontColor = '#282c33'
export const dlsUploaderFileItemIconColor = '#848b99'
export const dlsUploaderForegroundColorError = '#cc1800'
export const dlsUploaderForegroundColorSuccess = '#00bf5c'
export const dlsUploaderMediaBorderColorError = '#cc1800'
export const dlsUploaderMediaHeightM = '160px'
export const dlsUploaderMediaHeightS = '80px'
export const dlsUploaderMediaWidthM = '160px'
export const dlsUploaderMediaWidthS = '80px'
export const dlsUploaderMediaIconSizeM = '40px'
export const dlsUploaderMediaIconSizeS = '20px'
export const dlsUploaderMediaIconColor = '#848b99'
export const dlsUploaderMediaPaddingXM = '12px'
export const dlsUploaderMediaPaddingXS = '8px'
export const dlsUploaderMediaContentSpacingM = '12px'
export const dlsUploaderMediaContentSpacingS = '12px'
export const dlsUploaderMaskIconColor = '#fff'
export const dlsUploaderMaskIconSpacing = '12px'
export const dlsUploaderMaskPaddingY = '4px'
export const dlsUploaderMaskPaddingX = '12px'
export const dlsUploaderMaskBackgroundColor = 'rgba(0, 0, 0, 0.6)'
export const dlsUploaderMediaEntrySpacing = '24px'
export const dlsUploaderMediaEntryContentSpacing = '8px'
export const dlsUploaderMediaEntryIconSize = 'calc(1em + 2px)'
export const dlsUploaderMediaSpacingX = '16px'
export const dlsUploaderMediaSpacingY = '24px'
export const dlsUploaderMediaErrorSpacing = '4px'
export const dlsUploaderMediaPreviewMinWidth = '240px'
export const dlsUploaderMediaPreviewMaxWidth = '480px'
export const dlsTableColumnMinWidth = 'calc(3em + 16px)'
export const dlsTableCellBorderColor = '#ebedf5'
export const dlsTableHeaderBorderColor = '#d3d9e6'
export const dlsTableBorderColorDragging = '#d3d9e6'
export const dlsTablePaddingX = '16px'
export const dlsTableCellPaddingX = '8px'
export const dlsTableCellPaddingY = '12px'
export const dlsTableCellPaddingYCompact = '8px'
export const dlsTableCellPaddingYLoose = '16px'
export const dlsTableCellContentSpacing = '4px'
export const dlsTableCellBackgroundColor = '#fff'
export const dlsTableCellBackgroundColorHover = '#f6f7fa'
export const dlsTableCellFontColor = '#282c33'
export const dlsTableHeaderPaddingX = '8px'
export const dlsTableHeaderPaddingY = '12px'
export const dlsTableHeaderPaddingYCompact = '8px'
export const dlsTableHeaderPaddingYLoose = '16px'
export const dlsTableHeaderBackgroundColor = '#ebedf5'
export const dlsTableHeaderBackgroundColorHover = '#e2e6f0'
export const dlsTableHeaderFontWeight = '500'
export const dlsTableHeaderFontColor = '#282c33'
export const dlsTableSubRowBackgroundColor = '#f6f7fa'
export const dlsTableSubRowBackgroundColorHover = '#ebedf5'
export const dlsTableFontSizeS = '12px'
export const dlsTableFontSizeM = '14px'
export const dlsTableBorderRadiusS = '2px'
export const dlsTableBorderRadiusM = '4px'
export const dlsTableIconSize = 'calc(1em + 2px)'
export const dlsTableIconSizeAux = 'calc(1em + 2px)'
export const dlsTableHeaderIconColor = '#a8b0bf'
export const dlsTableHeaderIconColorHover = '#848b99'
export const dlsTableHeaderIconColorActive = '#545b66'
export const dlsTableHeaderIconColorDisabled = '#e2e6f0'
export const dlsTableOverflowShadowTop = 'none'
export const dlsTableOverflowShadowRight = '0 8px 10px 1px rgba(0, 0, 0, 0.06), 0 3px 14px 2px rgba(0, 0, 0, 0.05), 0 5px 5px -3px rgba(0, 0, 0, 0.01)'
export const dlsTableOverflowShadowBottom = 'none'
export const dlsTableOverflowShadowLeft = '0 8px 10px 1px rgba(0, 0, 0, 0.06), 0 3px 14px 2px rgba(0, 0, 0, 0.05), 0 5px 5px -3px rgba(0, 0, 0, 0.01)'
export const dlsTableBodyHeightNoData = '108px'
export const dlsTableActionSpacing = '16px'
export const dlsTableFilterActionPaddingY = '16px'
export const dlsTableFilterActionSpacing = '16px'
export const dlsFormLabelSpacingAfterHorizontal = '32px'
export const dlsFormLabelSpacingAfterVertical = '8px'
export const dlsFormLabelContentSpacing = '4px'
export const dlsFormLabelWidth = '7em'
export const dlsFormLabelFontColor = '#282c33'
export const dlsFormLabelIconColor = '#545b66'
export const dlsFormLabelRequiredColor = '#cc1800'
export const dlsFormFieldSpacingVertical = '32px'
export const dlsFormFieldErrorFontColor = '#cc1800'
export const dlsFormFieldErrorSpacingBeforeVertical = '4px'
export const dlsFormActionsSpacingBefore = '40px'
export const dlsFormActionsContentSpacing = '20px'
export const dlsLightboxPaddingX = '24px'
export const dlsLightboxPaddingY = '40px'
export const dlsLightboxContentSpacingY = '24px'
export const dlsLightboxContentSpacingX = '24px'
export const dlsLightboxBackdropColor = 'rgba(0, 0, 0, 0.6)'
export const dlsLightboxFontColor = '#fff'
export const dlsLightboxFontSize = '14px'
export const dlsLightboxCloseIconSize = '24px'
export const dlsLightboxControlSize = '48px'
export const dlsLightboxControlBorderRadius = '50%'
export const dlsLightboxControlIconSize = '24px'
export const dlsCascaderColumnMinWidth = '120px'
export const dlsCascaderOptionPaddingXs = '12px'
export const dlsCascaderOptionPaddingS = '12px'
export const dlsCascaderOptionPaddingM = '12px'
export const dlsCascaderOptionPaddingL = '12px'
export const dlsCascaderHeaderPaddingY = '8px'
export const dlsCascaderFooterPaddingY = '8px'
