// @baidu/one-ui@4.41.1/es/components/popover/popover.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import Tooltip from '../tooltip';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
var Popover = (_dec = withConfigConsumer('popover'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Popover, _PureComponent);

  function Popover() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.tooltip = void 0;

    _this.saveTooltip = function (node) {
      _this.tooltip = node;
    };

    return _this;
  }

  var _proto = Popover.prototype;

  _proto.getPopupDomNode = function getPopupDomNode() {
    return this.tooltip.getPopupDomNode();
  };

  _proto.getOverlay = function getOverlay() {
    var _this$props = this.props,
        title = _this$props.title,
        prefixCls = _this$props.prefixCls,
        content = _this$props.content;

    if (!title && !content) {
      return null;
    }

    return /*#__PURE__*/React.createElement(React.Fragment, null, title && /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-title"
    }, title), content && /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-inner-content"
    }, content));
  };

  _proto.render = function render() {
    var props = _extends({}, this.props);

    delete props.title;
    var prefixCls = props.prefixCls,
        size = props.size,
        overlayClassName = props.overlayClassName;
    return /*#__PURE__*/React.createElement(Tooltip, _extends({}, props, {
      overlayClassName: classnames(prefixCls + "-" + size, overlayClassName),
      ref: this.saveTooltip,
      overlay: this.getOverlay()
    }));
  };

  return Popover;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-popover',
  placement: 'bottom',
  // transitionName: 'zoom-big',
  size: 'medium',
  trigger: 'hover',
  mouseEnterDelay: 0.1,
  mouseLeaveDelay: 0.1,
  overlayStyle: {}
}, _temp)) || _class);
export default Popover;
