// @baidu/one-ui@4.32.1/es/components/form/formFieldGroup.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { useContext } from 'react';
import { classnames } from '../../core/commonTools';
import { FormContext } from './context';
/**
 * @visibleName FormFieldGroup
 */

export var FormFieldGroup = function FormFieldGroup(_ref) {
  var _classnames;

  var children = _ref.children,
      className = _ref.className,
      prefixCls = _ref.prefixCls,
      layout = _ref.layout,
      labelAlign = _ref.labelAlign,
      labelPosition = _ref.labelPosition,
      density = _ref.density,
      size = _ref.size,
      props = _objectWithoutPropertiesLoose(_ref, ["children", "className", "prefixCls", "layout", "labelAlign", "labelPosition", "density", "size"]);

  var formContext = useContext(FormContext);
  return /*#__PURE__*/React.createElement(FormContext.Provider, {
    value: _extends({}, formContext, {
      labelAlign: labelAlign,
      labelPosition: labelPosition
    })
  }, /*#__PURE__*/React.createElement("div", _extends({
    className: classnames(prefixCls + "-field-group", className, prefixCls + "-" + layout, prefixCls + "-" + size, prefixCls + "-type-" + density, (_classnames = {}, _classnames[prefixCls + "-label-position-" + labelPosition] = labelPosition, _classnames))
  }, props), children));
};
FormFieldGroup.defaultProps = {
  prefixCls: 'one-form',
  layout: 'default',
  labelAlign: 'left',
  labelPosition: 'side',
  density: 'compact'
};
export default FormFieldGroup;
