// @baidu/one-ui@4.32.1/es/components/form/formField.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _isEmpty from "lodash/isEmpty";
import _pickBy from "lodash/pickBy";
import _flattenDeep from "lodash/flattenDeep";
import React, { isValidElement, PureComponent } from 'react';
import { classnames } from '../../core/commonTools';
import Message from '../message';
import { IconQuestionCircle } from 'dls-icons-react';
import { FormContext, FormFieldContext } from './context';
import Loading from '../loading';
import { COMPONENT_INVALID } from '../config';

var FormField = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(FormField, _PureComponent);

  function FormField() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.formContext = void 0;
    _this.fieldNode = void 0;
    _this.state = {
      abstractFields: []
    };

    _this.registerAbstractField = function (name) {
      _this.setState(function (state) {
        var fields = state.abstractFields.slice();
        fields.push(name);
        return {
          abstractFields: fields
        };
      });
    };

    _this.unregisterAbstractField = function (name) {
      _this.setState(function (state) {
        var fields = state.abstractFields.filter(function (o) {
          return o !== name;
        });
        return {
          abstractFields: fields
        };
      });
    };

    return _this;
  }

  var _proto = FormField.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var _this$props = this.props,
        name = _this$props.name,
        _abstract = _this$props["abstract"];

    if (this.context && _abstract && name) {
      this.context.register(name);
    }
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    var _this$props2 = this.props,
        name = _this$props2.name,
        _abstract2 = _this$props2["abstract"];

    if (this.context && _abstract2 && name) {
      this.context.unregister(name);
    }
  };

  _proto.className = function className(name) {
    var fieldName = this.props.prefixCls + "-field";

    if (name) {
      return fieldName + '-' + name;
    }

    return fieldName;
  };

  _proto.renderFieldNode = function renderFieldNode() {
    var _this$props3 = this.props,
        name = _this$props3.name,
        rules = _this$props3.rules,
        children = _this$props3.children,
        initialValue = _this$props3.initialValue,
        trigger = _this$props3.trigger,
        validateTrigger = _this$props3.validateTrigger,
        valuePropName = _this$props3.valuePropName,
        validate = _this$props3.validate,
        normalize = _this$props3.normalize,
        getValueFromEvent = _this$props3.getValueFromEvent,
        validateFirst = _this$props3.validateFirst,
        validateStatus = _this$props3.validateStatus,
        hidden = _this$props3.hidden,
        preserve = _this$props3.preserve;
    var form = this.formContext.form;
    var fieldNode = children;

    if (name && form && /*#__PURE__*/React.isValidElement(children)) {
      var _classnames;

      fieldNode = /*#__PURE__*/React.cloneElement(children, {
        className: classnames(children.props.className, (_classnames = {}, _classnames[COMPONENT_INVALID] = validateStatus === 'error' || !_isEmpty(form.getFieldError(name)), _classnames))
      });
      var options = {
        rules: rules,
        initialValue: initialValue,
        trigger: trigger,
        validateTrigger: validateTrigger,
        valuePropName: valuePropName,
        validate: validate,
        normalize: normalize,
        getValueFromEvent: getValueFromEvent,
        validateFirst: validateFirst,
        hidden: hidden,
        preserve: preserve
      };
      fieldNode = form.getFieldDecorator(name, _pickBy(options, function (val) {
        return val !== undefined;
      }))(fieldNode);
    }

    return fieldNode;
  };

  _proto.required = function required() {
    var _this$props4 = this.props,
        required = _this$props4.required,
        rules = _this$props4.rules;
    return required == null ? rules && rules.some(function (o) {
      return o.required;
    }) : required;
  };

  _proto.renderTip = function renderTip() {
    var tip = this.props.tip;

    if (tip) {
      if (typeof tip === 'object' && ! /*#__PURE__*/isValidElement(tip)) {
        // message属性自定义
        return /*#__PURE__*/React.createElement(Message, _extends({
          className: this.className('tip'),
          icon: /*#__PURE__*/React.createElement(IconQuestionCircle, null),
          type: "aux",
          size: "small",
          display: "popup"
        }, tip));
      }

      return /*#__PURE__*/React.createElement(Message, {
        className: this.className('tip'),
        icon: /*#__PURE__*/React.createElement(IconQuestionCircle, null),
        type: "aux",
        size: "small",
        display: "popup"
      }, tip);
    }

    return null;
  };

  _proto.renderMain = function renderMain() {
    var _this2 = this;

    var _this$props5 = this.props,
        help = _this$props5.help,
        helpPosition = _this$props5.helpPosition;
    var labelPosition = this.formContext.labelPosition;
    var children = this.renderFieldNode();

    var childrenNode = function childrenNode(helpNode) {
      return /*#__PURE__*/React.createElement("div", {
        key: "Content",
        className: _this2.className('content')
      }, children, helpNode);
    };

    if (!help) {
      return [childrenNode(null), this.renderMessages()];
    }

    var helpNode = this.renderHelp();

    if (helpPosition === 'top') {
      return [labelPosition !== 'top' ? helpNode : null, childrenNode(null), this.renderMessages()];
    } else if (helpPosition === 'side') {
      return [childrenNode(helpNode), this.renderMessages()];
    }

    return [childrenNode(null), this.renderMessages(helpNode)];
  };

  _proto.renderHelp = function renderHelp() {
    var _this$props6 = this.props,
        help = _this$props6.help,
        helpPosition = _this$props6.helpPosition,
        messageDisplay = _this$props6.messageDisplay;

    if (!help) {
      return null;
    }

    return /*#__PURE__*/React.createElement(Message, {
      type: "aux",
      size: "small",
      display: messageDisplay,
      key: "Help",
      className: classnames(this.className('help'), this.className("help-position-" + helpPosition))
    }, help);
  };

  _proto.getMessage = function getMessage(fn, name, type) {
    return (fn(name) || []).map(function (content) {
      return {
        type: type,
        content: content
      };
    });
  };

  _proto.renderMessages = function renderMessages(helpNode) {
    var _this3 = this;

    if (helpNode === void 0) {
      helpNode = null;
    }

    var _this$props7 = this.props,
        messageDisplay = _this$props7.messageDisplay,
        _abstract3 = _this$props7["abstract"],
        name = _this$props7.name,
        validateStatus = _this$props7.validateStatus,
        validateMessage = _this$props7.validateMessage;

    if (!helpNode && _abstract3) {
      return null;
    }

    var form = this.formContext.form;
    var messages = [];

    if (!_abstract3 && form) {
      if (validateMessage && validateStatus) {
        messages.push({
          type: validateStatus,
          content: validateMessage
        });
      } else {
        var fields = this.state.abstractFields.concat(name ? name : []);
        messages = fields.map(function (name) {
          var items = [];
          var validating = form.isFieldValidating(name);

          if (validating && !form.isFormValidating()) {
            items.push({
              type: 'validating',
              content: '校验中...'
            });
          }

          items.push.apply(items, _this3.getMessage(form.getFieldError, name, 'error'));
          items.push.apply(items, _this3.getMessage(form.getFieldWarning, name, 'warning'));
          items.push.apply(items, _this3.getMessage(form.getFieldSuccess, name, 'success'));
          return items;
        });
      }
    }

    var flattenMessages = _flattenDeep(messages);

    return flattenMessages.length ? /*#__PURE__*/React.createElement("div", {
      key: "Messages",
      className: this.className('messages')
    }, flattenMessages.map(function (message, index) {
      if (message.type === 'validating') {
        return /*#__PURE__*/React.createElement(Loading, {
          key: index,
          size: "small",
          tip: "\u6821\u9A8C\u4E2D..."
        });
      }

      return /*#__PURE__*/React.createElement(Message, {
        size: "small",
        type: message.type,
        key: index,
        display: messageDisplay
      }, message.content);
    }), helpNode) : helpNode ? /*#__PURE__*/React.createElement("div", {
      key: "Messages",
      className: this.className('message')
    }, helpNode) : null;
  };

  _proto.renderField = function renderField(formContext) {
    var _classnames2;

    if (formContext === void 0) {
      formContext = {};
    }

    var _this$props8 = this.props,
        style = _this$props8.style,
        className = _this$props8.className,
        _abstract4 = _this$props8["abstract"],
        actions = _this$props8.actions,
        display = _this$props8.display;
    this.formContext = formContext;
    return /*#__PURE__*/React.createElement("div", {
      className: classnames(className, this.className(), (_classnames2 = {}, _classnames2[this.className('abstract')] = _abstract4, _classnames2[this.className('actions')] = actions, _classnames2[this.className('standalone')] = display === 'standalone', _classnames2)),
      style: style
    }, this.renderLabel(), /*#__PURE__*/React.createElement("div", {
      className: this.className('main')
    }, _abstract4 ? this.renderMain() : /*#__PURE__*/React.createElement(FormFieldContext.Provider, {
      value: {
        register: this.registerAbstractField,
        unregister: this.unregisterAbstractField
      }
    }, this.renderMain())));
  };

  _proto.renderLabel = function renderLabel(helpNode) {
    var _this$props9 = this.props,
        label = _this$props9.label,
        _abstract5 = _this$props9["abstract"],
        htmlFor = _this$props9.htmlFor,
        helpPosition = _this$props9.helpPosition,
        help = _this$props9.help;
    var _this$formContext = this.formContext,
        labelPosition = _this$formContext.labelPosition,
        labelAlign = _this$formContext.labelAlign;

    if (label != null && !_abstract5) {
      var _classnames3;

      return /*#__PURE__*/React.createElement("div", {
        className: classnames(this.className('label'), (_classnames3 = {}, _classnames3[this.className('label-left')] = labelAlign === 'left', _classnames3[this.className('label-required')] = this.required(), _classnames3[this.className('label-align-help')] = help && labelPosition === 'side' && helpPosition === 'top', _classnames3))
      }, /*#__PURE__*/React.createElement("label", {
        title: typeof label === 'string' ? label : '',
        htmlFor: htmlFor
      }, label), this.renderTip(), labelPosition === 'top' && helpPosition === 'top' ? this.renderHelp() : helpNode);
    }

    return labelPosition === 'top' && helpPosition === 'top' ? this.renderHelp() : null;
  };

  _proto.render = function render() {
    var _this4 = this;

    return /*#__PURE__*/React.createElement(FormContext.Consumer, null, function (context) {
      return _this4.renderField(context);
    });
  };

  return FormField;
}(PureComponent);

FormField.defaultProps = {
  prefixCls: 'one-form',
  helpPosition: 'bottom',
  messageDisplay: 'simple',
  validateFirst: true
};
FormField.contextType = FormFieldContext;
export default FormField;
