// @baidu/one-ui@4.32.1/lib/components/select/common/oneSelect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _KeyCode = _interopRequireDefault(require("rc-util/lib/KeyCode"));

var _toArray = _interopRequireDefault(require("rc-util/lib/Children/toArray"));

var _dlsIconsReact = require("dls-icons-react");

var _commonTools = require("../../../core/commonTools");

var _componentClasses = _interopRequireDefault(require("component-classes"));

var _rcMenu = require("rc-menu");

var _warning = _interopRequireDefault(require("warning"));

var _option = _interopRequireDefault(require("../option"));

var _selectTrigger = _interopRequireDefault(require("./selectTrigger"));

var _selectTools = require("../../../core/selectTools");

var _checkbox = _interopRequireDefault(require("../../checkbox"));

var _searchText = _interopRequireDefault(require("../searchText"));

var _iconSvg = _interopRequireDefault(require("../../iconSvg"));

var _button = _interopRequireDefault(require("../../button"));

var _loading = _interopRequireDefault(require("../../loading"));

var _config = require("../../config");

var _checkboxText = _interopRequireDefault(require("../checkboxText"));

var heightToSizeMap = {
  xsmall: 24,
  small: 28,
  medium: 32,
  large: 36
};
var SELECT_EMPTY_VALUE_KEY = 'ONE_UI_SELECT_EMPTY_VALUE_KEY';
var CHECK_ALL_VALUE = 'ONE_UI_SELECT_CHECK_ALL_VALUE';

var noop = function noop() {};

function chaining() {
  for (var _len = arguments.length, fns = new Array(_len), _key = 0; _key < _len; _key++) {
    fns[_key] = arguments[_key];
  }

  return function () {
    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    // eslint-disable-line
    // eslint-disable-line
    for (var i = 0; i < fns.length; i++) {
      if (fns[i] && typeof fns[i] === 'function') {
        fns[i].apply(this, args);
      }
    }
  };
}

;
;
;

var OneSelect = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(OneSelect, _PureComponent);

  function OneSelect(_props) {
    var _this;

    _this = _PureComponent.call(this, _props) || this;
    _this.inputRef = void 0;
    _this.inputMirrorRef = void 0;
    _this.topCtrlRef = void 0;
    _this.selectTriggerRef = void 0;
    _this.rootRef = void 0;
    _this.selectionRef = void 0;
    _this.saveInputRef = void 0;
    _this.saveInputMirrorRef = void 0;
    _this.saveTopCtrlRef = void 0;
    _this.saveSelectTriggerRef = void 0;
    _this.saveRootRef = void 0;
    _this.saveSelectionRef = void 0;
    _this.dropdownContainer = void 0;
    _this._focused = void 0;
    _this.blurTimer = void 0;
    _this.focusTimer = void 0;
    _this._mouseDown = void 0;
    _this._options = void 0;

    _this.componentDidUpdate = function (prevProps, prevState) {
      if ((0, _selectTools.isMultipleOrTags)(_this.props)) {
        var inputNode = _this.getInputDOMNode();

        var mirrorNode = _this.getInputMirrorDOMNode();

        if (inputNode) {
          if (inputNode.value) {
            inputNode.style.width = '';
            inputNode.style.width = mirrorNode.clientWidth + "px";
          } else {
            inputNode.style.width = '';
          }
        }

        if (!(0, _isEqual2["default"])(_this.state.value, prevState.value)) {
          var refHeight = _this.selectionRef.offsetHeight;

          _this.setState({
            ctrNodeHeight: refHeight
          });
        }
      }

      _this.forcePopupAlign();
    };

    _this.onInputChange = function (event) {
      var tokenSeparators = _this.props.tokenSeparators;
      var val = event.target.value;

      if ((0, _selectTools.isMultipleOrTags)(_this.props) && tokenSeparators.length && (0, _selectTools.includesSeparators)(val, tokenSeparators)) {
        var nextValue = _this.getValueByInput(val);

        if (nextValue !== undefined) {
          _this.fireChange(nextValue);
        }

        _this.setOpenState(false, true);

        _this.setInputValue('', false);

        return;
      }

      _this.setInputValue(val);

      _this.setState({
        open: true
      });

      if ((0, _selectTools.isCombobox)(_this.props)) {
        _this.fireChange([val]);
      }
    };

    _this.onDropdownVisibleChange = function (open) {
      if (open && !_this._focused) {
        _this.clearBlurTime();

        _this.timeoutFocus();

        _this._focused = true;

        _this.updateFocusClassName();
      }

      _this.setOpenState(open);
    };

    _this.onKeyDown = function (event) {
      var open = _this.state.open;
      var disabled = _this.props.disabled;

      if (disabled) {
        return;
      }

      var keyCode = event.keyCode;

      if (open && !_this.getInputDOMNode()) {
        _this.onInputKeyDown(event);
      } else if (keyCode === _KeyCode["default"].ENTER || keyCode === _KeyCode["default"].DOWN) {
        if (!open) {
          _this.setOpenState(true);
        }

        event.preventDefault();
      } else if (keyCode === _KeyCode["default"].SPACE) {
        // Not block space if popup is shown
        if (!open) {
          _this.setOpenState(true);

          event.preventDefault();
        }
      }
    };

    _this.onInputKeyDown = function (event) {
      var props = _this.props;

      if (props.disabled) {
        return;
      }

      var state = _this.state;
      var keyCode = event.keyCode;

      if ((0, _selectTools.isMultipleOrTags)(props) && !event.target.value && keyCode === _KeyCode["default"].BACKSPACE) {
        event.preventDefault();
        var _value = state.value;

        if (_value.length) {
          _this.removeSelected(_value[_value.length - 1]);
        }

        return;
      }

      if (keyCode === _KeyCode["default"].DOWN) {
        if (!state.open) {
          _this.openIfHasChildren();

          event.preventDefault();
          event.stopPropagation();
          return;
        }
      } else if (keyCode === _KeyCode["default"].ENTER && state.open) {
        event.preventDefault();
      } else if (keyCode === _KeyCode["default"].ESC) {
        if (state.open) {
          _this.setOpenState(false);

          event.preventDefault();
          event.stopPropagation();
        }

        return;
      }

      if (_this.getRealOpenState(state)) {
        var menu = _this.selectTriggerRef.getInnerManualRef();

        if (menu && menu.onKeyDown(event, _this.handleBackfill)) {
          event.preventDefault();
          event.stopPropagation();
        }
      }
    };

    _this.onMenuSelect = function (_ref) {
      var item = _ref.item;

      if (!item) {
        return;
      }

      var itemProps = item.props;
      var type = itemProps && itemProps.type; // 自定义属性的区域

      if (type === 'custom') {
        return;
      }

      var value = _this.state.value;
      var props = _this.props;
      var selectedValue = (0, _selectTools.getValuePropValue)(item);
      var exclusive = (0, _selectTools.getPropValue)(item, 'exclusive');
      var lastValue = value[value.length - 1];

      _this.fireSelect(selectedValue);

      if (exclusive) {
        value = [selectedValue];

        _this.setOpenState(false, true);
      } else if (selectedValue === CHECK_ALL_VALUE) {
        value = _this.getToggleAllValue();
      } else {
        if ((0, _selectTools.isMultipleOrTags)(props)) {
          if ((0, _selectTools.findIndexInValueBySingleValue)(value, selectedValue) !== -1) {
            return;
          }

          value = value.filter(function (val) {
            var _this$getOptionInfoBy;

            return !((_this$getOptionInfoBy = _this.getOptionInfoBySingleValue(val)) != null && _this$getOptionInfoBy.exclusive);
          }).concat([selectedValue]);
        } else {
          if (lastValue !== undefined && lastValue === selectedValue && selectedValue !== _this.state.backfillValue) {
            _this.setOpenState(false, true);

            return;
          }

          value = [selectedValue];

          _this.setOpenState(false, true);
        }
      }

      _this.fireChange(value);

      var inputValue;

      if ((0, _selectTools.isCombobox)(props)) {
        inputValue = (0, _selectTools.getPropValue)(item, props.optionLabelProp);
      } else {
        inputValue = '';
      }

      if (props.autoClearSearchValue) {
        _this.setInputValue(inputValue, false);
      }
    };

    _this.onMenuDeselect = function (_ref2) {
      var item = _ref2.item,
          domEvent = _ref2.domEvent;

      if (domEvent.type === 'click') {
        _this.removeSelected((0, _selectTools.getValuePropValue)(item));
      }

      var _assertThisInitialize = (0, _assertThisInitialized2["default"])(_this),
          props = _assertThisInitialize.props;

      if (props.autoClearSearchValue) {
        _this.setInputValue('', false);
      }
    };

    _this.onArrowClick = function (e) {
      e.stopPropagation();
      e.preventDefault();

      if (!_this.props.disabled) {
        _this.setOpenState(!_this.state.open, !_this.state.open);
      }
    };

    _this.onPlaceholderClick = function () {
      if (_this.getInputDOMNode()) {
        _this.getInputDOMNode().focus();
      }
    };

    _this.onOuterFocus = function (e) {
      if (_this.props.disabled) {
        e.preventDefault();
        return;
      }

      _this.clearBlurTime();

      if (!(0, _selectTools.isMultipleOrTagsOrCombobox)(_this.props) && e.target === _this.getInputDOMNode()) {
        return;
      }

      if (_this._focused) {
        return;
      }

      _this._focused = true;

      _this.updateFocusClassName(); // only effect multiple or tag mode


      if (!(0, _selectTools.isMultipleOrTags)(_this.props) || !_this._mouseDown) {
        _this.timeoutFocus();
      }
    };

    _this.onPopupFocus = function () {
      // fix ie scrollbar, focus element again
      _this.maybeFocus(true, true);
    };

    _this.onOuterBlur = function (e) {
      if (_this.props.disabled) {
        e.preventDefault();
        return;
      }

      _this.blurTimer = setTimeout(function () {
        _this._focused = false;

        _this.updateFocusClassName();

        var props = _this.props;
        var value = _this.state.value;
        var inputValue = _this.state.inputValue;

        if ((0, _selectTools.isSingleMode)(props) && props.showSearch && inputValue && props.defaultActiveFirstOption) {
          var options = _this._options || [];

          if (options.length) {
            var firstOption = (0, _selectTools.findFirstMenuItem)(options);

            if (firstOption) {
              value = [(0, _selectTools.getValuePropValue)(firstOption)];

              _this.fireChange(value);
            }
          }
        } else if ((0, _selectTools.isMultipleOrTags)(props) && inputValue) {
          if (_this._mouseDown) {
            // need update dropmenu when not blur
            _this.setInputValue('');
          } else {
            // this.state.inputValue = (this.getInputDOMNode().value || '');
            // this.state.inputValue = '';
            _this.setState({
              inputValue: ''
            });
          }

          if (props.tags) {
            value = _this.getValueByInput(inputValue);

            if (value !== undefined) {
              _this.fireChange(value);
            }
          }
        } // if click the rest space of Select in multiple mode


        if ((0, _selectTools.isMultipleOrTags)(props) && _this._mouseDown) {
          _this.maybeFocus(true, true);

          _this._mouseDown = false;
          return;
        }

        _this.setOpenState(false);

        props.onBlur(_this.getVLForOnChange(value));
      }, 10);
    };

    _this.onMouseEnter = function () {
      var _this$props = _this.props,
          onMouseEnter = _this$props.onMouseEnter,
          disabled = _this$props.disabled,
          disabledReason = _this$props.disabledReason;

      if (disabled && disabledReason) {
        _this.setState({
          showDisabledReason: true
        });
      }

      if (onMouseEnter) {
        onMouseEnter();
      }
    };

    _this.onMouseLeave = function () {
      var onMouseLeave = _this.props.onMouseLeave;

      _this.setState({
        showDisabledReason: false
      });

      if (onMouseLeave) {
        onMouseLeave();
      }
    };

    _this.onClearSelection = function (event) {
      var props = _this.props;
      var state = _this.state;

      if (props.disabled) {
        return;
      }

      var inputValue = state.inputValue,
          value = state.value;
      event.stopPropagation();

      if (inputValue || value.length) {
        if (value.length) {
          _this.fireChange([]);
        }

        _this.setOpenState(false, true);

        if (inputValue) {
          _this.setInputValue('');
        }
      }
    };

    _this.onChoiceAnimationLeave = function () {
      _this.forcePopupAlign();
    };

    _this.getOptionInfoBySingleValue = function (value, optionsInfo) {
      var info;
      optionsInfo = optionsInfo || _this.state.optionsInfo;

      if (optionsInfo[(0, _selectTools.getMapKey)(value)]) {
        info = optionsInfo[(0, _selectTools.getMapKey)(value)];
      }

      if (info) {
        return info;
      }

      var defaultLabel = value;

      if (_this.props.labelInValue) {
        var label = (0, _selectTools.getLabelFromPropsValue)(_this.props.value, value);

        if (label !== undefined) {
          defaultLabel = label;
        }
      }

      var defaultInfo = {
        option: /*#__PURE__*/_react["default"].createElement(_option["default"], {
          value: value,
          key: value
        }, value),
        value: value,
        label: defaultLabel
      };
      return defaultInfo;
    };

    _this.getOptionBySingleValue = function (value) {
      var _this$getOptionInfoBy2 = _this.getOptionInfoBySingleValue(value),
          option = _this$getOptionInfoBy2.option;

      return option;
    };

    _this.getOptionsBySingleValue = function (values) {
      return values.map(function (value) {
        return _this.getOptionBySingleValue(value);
      });
    };

    _this.getValueByLabel = function (label) {
      if (label === undefined) {
        return null;
      }

      var value = null;
      Object.keys(_this.state.optionsInfo).forEach(function (key) {
        var info = _this.state.optionsInfo[key];

        if ((0, _selectTools.toArray)(info.label).join('') === label) {
          value = info.value;
        }
      });
      return value;
    };

    _this.getVLBySingleValue = function (value) {
      if (_this.props.labelInValue) {
        return {
          key: value,
          label: _this.getLabelBySingleValue(value)
        };
      }

      return value;
    };

    _this.getVLForOnChange = function (vls_) {
      var vls = vls_;

      if (vls !== undefined) {
        if (!_this.props.labelInValue) {
          vls = vls.map(function (v) {
            return v;
          });
        } else {
          vls = vls.map(function (vl) {
            return {
              key: vl,
              label: _this.getLabelBySingleValue(vl)
            };
          });
        }

        return (0, _selectTools.isMultipleOrTags)(_this.props) ? vls : vls[0];
      }

      return vls;
    };

    _this.getLabelBySingleValue = function (value, optionsInfo) {
      var _this$getOptionInfoBy3 = _this.getOptionInfoBySingleValue(value, optionsInfo),
          label = _this$getOptionInfoBy3.label;

      return label;
    };

    _this.getDropdownContainer = function () {
      if (!_this.dropdownContainer) {
        _this.dropdownContainer = document.createElement('div');
        document.body.appendChild(_this.dropdownContainer);
      }

      return _this.dropdownContainer;
    };

    _this.getPlaceholderElement = function () {
      var _this$props2 = _this.props,
          placeholder = _this$props2.placeholder,
          prefixCls = _this$props2.prefixCls,
          selectorName = _this$props2.selectorName;
      var _this$state = _this.state,
          inputValue = _this$state.inputValue,
          value = _this$state.value;
      var hidden = false;

      if (value.length) {
        hidden = true;
      }

      if (value.length === 1 && value[0] === null) {
        hidden = !!_this.getOptionInfoBySingleValue(null).label;
      }

      if (inputValue) {
        hidden = true;
      }

      if ((0, _selectTools.isCombobox)(_this.props) && value.length === 1 && !value[0]) {
        hidden = false;
      }

      if (selectorName || placeholder) {
        return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
          onMouseDown: _selectTools.preventDefaultEvent,
          style: (0, _extends2["default"])({
            display: hidden ? 'none' : 'block'
          }, _selectTools.UNSELECTABLE_STYLE)
        }, _selectTools.UNSELECTABLE_ATTRIBUTE, {
          onClick: _this.onPlaceholderClick,
          className: prefixCls + "-selection__placeholder"
        }), selectorName || placeholder);
      }

      return null;
    };

    _this.getInputElement = function () {
      var _classnames;

      var props = _this.props;
      var inputElement = props.getInputElement ? props.getInputElement() : /*#__PURE__*/_react["default"].createElement("input", {
        id: props.id,
        autoComplete: "off"
      });
      var inputCls = (0, _commonTools.classnames)(inputElement.props.className, (_classnames = {}, _classnames[props.prefixCls + "-search__field"] = true, _classnames)); // Add space to the end of the inputValue as the width measurement tolerance

      return /*#__PURE__*/_react["default"].createElement("div", {
        className: props.prefixCls + "-search__field__wrap"
      }, /*#__PURE__*/_react["default"].cloneElement(inputElement, {
        ref: _this.saveInputRef,
        onChange: _this.onInputChange,
        onKeyDown: chaining(_this.onInputKeyDown, inputElement.props.onKeyDown, _this.props.onInputKeyDown),
        value: _this.state.inputValue,
        disabled: props.disabled,
        className: inputCls
      }), /*#__PURE__*/_react["default"].createElement("span", {
        ref: _this.saveInputMirrorRef,
        className: props.prefixCls + "-search__field__mirror"
      }, _this.state.inputValue, "\xA0"));
    };

    _this.getInputDOMNode = function () {
      return _this.topCtrlRef ? _this.topCtrlRef.querySelector('input,textarea,div[contentEditable]') : _this.inputRef;
    };

    _this.getInputMirrorDOMNode = function () {
      return _this.inputMirrorRef;
    };

    _this.getPopupDOMNode = function () {
      return _this.selectTriggerRef.getPopupDOMNode();
    };

    _this.getPopupMenuComponent = function () {
      return _this.selectTriggerRef.getInnerMenu();
    };

    _this.setOpenState = function (open, needFocus) {
      var _assertThisInitialize2 = (0, _assertThisInitialized2["default"])(_this),
          props = _assertThisInitialize2.props,
          state = _assertThisInitialize2.state;

      if (state.open === open) {
        _this.maybeFocus(open, needFocus);

        return;
      }

      if (_this.props.onDropdownVisibleChange) {
        _this.props.onDropdownVisibleChange(open);
      }

      var nextState = {
        open: open,
        backfillValue: undefined
      }; // clear search input value when open is false in singleMode.

      if (!open && (0, _selectTools.isSingleMode)(props) && props.showSearch) {
        _this.setInputValue('', false);
      }

      if (!open) {
        _this.maybeFocus(open, needFocus);
      }

      _this.setState(nextState, function () {
        if (open) {
          _this.maybeFocus(open, needFocus);
        }
      });
    };

    _this.setInputValue = function (inputValue, fireSearch) {
      if (fireSearch === void 0) {
        fireSearch = true;
      }

      if (inputValue !== _this.state.inputValue) {
        _this.setState({
          inputValue: inputValue
        }, _this.forcePopupAlign);

        if (fireSearch) {
          _this.props.onSearch(inputValue);
        }
      }
    };

    _this.getValueByInput = function (string) {
      var _this$props3 = _this.props,
          multiple = _this$props3.multiple,
          tokenSeparators = _this$props3.tokenSeparators;
      var nextValue = _this.state.value;
      var hasNewValue = false;
      (0, _selectTools.splitBySeparators)(string, tokenSeparators).forEach(function (label) {
        var selectedValue = [label];

        if (multiple) {
          var _value2 = _this.getValueByLabel(label);

          if (_value2 && (0, _selectTools.findIndexInValueBySingleValue)(nextValue, _value2) === -1) {
            nextValue = nextValue.concat(_value2);
            hasNewValue = true;

            _this.fireSelect(_value2);
          }
        } else if ((0, _selectTools.findIndexInValueBySingleValue)(nextValue, label) === -1) {
          nextValue = nextValue.concat(selectedValue);
          hasNewValue = true;

          _this.fireSelect(label);
        }
      });
      return hasNewValue ? nextValue : undefined;
    };

    _this.getRealOpenState = function (state) {
      var props = _this.props;
      var _open = props.open,
          showSearch = props.showSearch;

      if (typeof _open === 'boolean') {
        return _open;
      }

      var open = (state || _this.state).open; // const inputValue = this.state.inputValue;

      var options = _this._options || [];
      var optionsLength = options.length;

      if ((0, _selectTools.isMultipleOrTagsOrCombobox)(props) || !showSearch) {
        if (open && !optionsLength) {
          open = false;
        }
      }

      open = _this.hidePopElementInSearchMode() ? open : _this.hidePopElementInSearchMode();
      return open;
    };

    _this.hidePopElementInSearchMode = function () {
      var showSearch = _this.props.showSearch;
      var inputValue = _this.state.inputValue;
      var options = _this._options || [];
      var optionsLength = options.length;
      var open = true; // 如果搜索模式, 没有输入不展示下拉

      if (showSearch && !inputValue && !optionsLength) {
        open = false;
      }

      return open;
    };

    _this.focus = function () {
      if ((0, _selectTools.isSingleMode)(_this.props) && _this.selectionRef) {
        _this.selectionRef.focus();
      } else if (_this.getInputDOMNode()) {
        _this.getInputDOMNode().focus();
      }
    };

    _this.blur = function () {
      if ((0, _selectTools.isSingleMode)(_this.props) && _this.selectionRef) {
        _this.selectionRef.blur();
      } else if (_this.getInputDOMNode()) {
        _this.getInputDOMNode().blur();
      }
    };

    _this.markMouseDown = function () {
      _this._mouseDown = true;
    };

    _this.markMouseLeave = function () {
      _this._mouseDown = false;
    };

    _this.handleBackfill = function (item) {
      if (!_this.props.backfill || !((0, _selectTools.isSingleMode)(_this.props) || (0, _selectTools.isCombobox)(_this.props))) {
        return;
      }

      var key = (0, _selectTools.getValuePropValue)(item);

      if ((0, _selectTools.isCombobox)(_this.props)) {
        _this.setInputValue(key, false);
      }

      _this.setState({
        value: [key],
        backfillValue: key
      });
    };

    _this.filterOption = function (input, child, defaultFilter) {
      if (defaultFilter === void 0) {
        defaultFilter = _selectTools.defaultFilterFn;
      }

      var value = _this.state.value;
      var lastValue = value[value.length - 1];

      if (!input || lastValue && lastValue === _this.state.backfillValue) {
        return true;
      }

      var filterFn = _this.props.filterOption;

      if ('filterOption' in _this.props) {
        if (_this.props.filterOption === true) {
          filterFn = defaultFilter;
        }
      } else {
        filterFn = defaultFilter;
      }

      if (!filterFn) {
        return true;
      }

      if (typeof filterFn === 'function') {
        return filterFn.call((0, _assertThisInitialized2["default"])(_this), input, child);
      }

      if (child.props.disabled) {
        return false;
      }

      return true;
    };

    _this.timeoutFocus = function () {
      if (_this.focusTimer) {
        _this.clearFocusTime();
      }

      _this.focusTimer = setTimeout(function () {
        _this.props.onFocus();
      }, 10);
    };

    _this.clearFocusTime = function () {
      if (_this.focusTimer) {
        clearTimeout(_this.focusTimer);
        _this.focusTimer = null;
      }
    };

    _this.clearBlurTime = function () {
      if (_this.blurTimer) {
        clearTimeout(_this.blurTimer);
        _this.blurTimer = null;
      }
    };

    _this.updateFocusClassName = function () {
      var _assertThisInitialize3 = (0, _assertThisInitialized2["default"])(_this),
          rootRef = _assertThisInitialize3.rootRef,
          props = _assertThisInitialize3.props;

      if (!rootRef) {
        return;
      } // avoid setState and its side effect


      if (_this._focused) {
        (0, _componentClasses["default"])(rootRef).add(props.prefixCls + "-focus").add(_config.COMPONENT_FOCUS);
      } else {
        (0, _componentClasses["default"])(rootRef).remove(props.prefixCls + "-focus").remove(_config.COMPONENT_FOCUS);
      }
    };

    _this.maybeFocus = function (open, needFocus) {
      if (needFocus || open) {
        var input = _this.getInputDOMNode();

        var _document = document,
            activeElement = _document.activeElement;

        if (input && (open || (0, _selectTools.isMultipleOrTagsOrCombobox)(_this.props))) {
          if (activeElement !== input) {
            input.focus();
            _this._focused = true;
          }
        } else if (activeElement !== _this.selectionRef) {
          _this.selectionRef.focus();

          _this._focused = true;
        }
      }
    };

    _this.removeSelected = function (selectedKey, e) {
      var props = _this.props;

      if (props.disabled || _this.isChildDisabled(selectedKey)) {
        return;
      } // Do not trigger Trigger popup


      if (e && e.stopPropagation) {
        e.stopPropagation();
      }

      var value = selectedKey === CHECK_ALL_VALUE ? _this.getToggleAllValue() : _this.state.value.filter(function (singleValue) {
        return singleValue !== selectedKey;
      });
      var canMultiple = (0, _selectTools.isMultipleOrTags)(props);

      if (canMultiple) {
        var _event;

        if (props.labelInValue) {
          _event = {
            key: selectedKey,
            label: _this.getLabelBySingleValue(selectedKey)
          };
        } else {
          _event = selectedKey;
        }

        props.onDeselect(_event, _this.getOptionBySingleValue(selectedKey));
      }

      _this.fireChange(value);
    };

    _this.openIfHasChildren = function () {
      var props = _this.props;

      if (_react["default"].Children.count(props.children) || (0, _selectTools.isSingleMode)(props)) {
        _this.setOpenState(true);
      }
    };

    _this.fireSelect = function (value) {
      _this.props.onSelect(_this.getVLBySingleValue(value), _this.getOptionBySingleValue(value));
    };

    _this.fireChange = function (value) {
      var props = _this.props;

      if (!('value' in props)) {
        _this.setState({
          value: value
        }, _this.forcePopupAlign);
      }

      var vls = _this.getVLForOnChange(value);

      var options = _this.getOptionsBySingleValue(value);

      props.onChange(vls, (0, _selectTools.isMultipleOrTags)(_this.props) ? options : options[0]);
    };

    _this.isChildDisabled = function (key) {
      return (0, _toArray["default"])(_this.props.children).some(function (child) {
        var childValue = (0, _selectTools.getValuePropValue)(child);
        return childValue === key && child.props && child.props.disabled;
      });
    };

    _this.forcePopupAlign = function () {
      if (!_this.state.open) {
        return;
      }

      _this.selectTriggerRef.triggerRef.forcePopupAlign();
    };

    _this.renderFilterOptions = function () {
      var _this$state2 = _this.state,
          inputValue = _this$state2.inputValue,
          optionsInfo = _this$state2.optionsInfo;
      var _this$props4 = _this.props,
          children = _this$props4.children,
          tags = _this$props4.tags,
          filterOption = _this$props4.filterOption,
          notFoundContent = _this$props4.notFoundContent,
          showSearch = _this$props4.showSearch,
          checkboxPrefixCls = _this$props4.checkboxPrefixCls,
          prefixCls = _this$props4.prefixCls,
          loading = _this$props4.loading,
          loadingText = _this$props4.loadingText,
          size = _this$props4.size,
          multiple = _this$props4.multiple,
          showCheckAll = _this$props4.showCheckAll;
      var menuItems = [];
      var childrenKeys = [];

      var options = _this.renderFilterOptionsFromChildren(children, childrenKeys, menuItems);

      if (tags) {
        // tags value must be string
        var _value3 = _this.state.value;
        _value3 = _value3.filter(function (singleValue) {
          return childrenKeys.indexOf(singleValue) === -1 && (!inputValue || String(singleValue).indexOf(String(inputValue)) > -1);
        });

        _value3.forEach(function (singleValue) {
          var key = singleValue;

          var children = /*#__PURE__*/_react["default"].createElement("span", null, /*#__PURE__*/_react["default"].createElement(_checkbox["default"], {
            checked: true,
            prefixCls: checkboxPrefixCls
          }), /*#__PURE__*/_react["default"].createElement(_searchText["default"], {
            showSearch: true,
            text: key,
            searchValue: inputValue,
            prefixCls: prefixCls
          }));

          var menuItem = /*#__PURE__*/_react["default"].createElement(_rcMenu.Item, {
            style: _selectTools.UNSELECTABLE_STYLE,
            role: "option",
            attribute: _selectTools.UNSELECTABLE_ATTRIBUTE,
            value: key,
            key: key,
            type: "normal"
          }, children);

          options.push(menuItem);
          menuItems.push(menuItem);
        });

        if (inputValue) {
          var notFindInputItem = menuItems.every(function (option) {
            // this.filterOption return true has two meaning,
            // 1, some one exists after filtering
            // 2, filterOption is set to false
            // condition 2 does not mean the option has same value with inputValue
            var filterFn = function filterFn() {
              return (0, _selectTools.getValuePropValue)(option) === inputValue;
            };

            if (filterOption !== false) {
              return !_this.filterOption.call((0, _assertThisInitialized2["default"])(_this), inputValue, option, filterFn);
            }

            return !filterFn();
          });

          if (notFindInputItem) {
            options.unshift( /*#__PURE__*/_react["default"].createElement(_rcMenu.Item, {
              style: _selectTools.UNSELECTABLE_STYLE,
              role: "option",
              attribute: _selectTools.UNSELECTABLE_ATTRIBUTE,
              value: inputValue,
              key: inputValue,
              type: "normal"
            }, inputValue));
          }
        }
      }

      if (!options.length && notFoundContent) {
        options = [/*#__PURE__*/_react["default"].createElement(_rcMenu.Item, {
          style: _selectTools.UNSELECTABLE_STYLE,
          attribute: _selectTools.UNSELECTABLE_ATTRIBUTE,
          disabled: true,
          role: "option",
          value: "NOT_FOUND",
          key: "NOT_FOUND",
          type: "normal",
          className: prefixCls + "-dropdown-menu-empty"
        }, notFoundContent)];
      }

      if (loading) {
        var loadingSize = size === 'xsmall' ? 'small' : size;
        options = [/*#__PURE__*/_react["default"].createElement(_rcMenu.Item, {
          style: _selectTools.UNSELECTABLE_STYLE,
          attribute: _selectTools.UNSELECTABLE_ATTRIBUTE,
          disabled: true,
          role: "option",
          value: "NOT_FOUND",
          key: "NOT_FOUND",
          type: "normal"
        }, /*#__PURE__*/_react["default"].createElement("div", {
          className: prefixCls + "-loading-holder-container"
        }, /*#__PURE__*/_react["default"].createElement(_loading["default"], {
          className: prefixCls + "-loading-icon",
          tip: loadingText,
          size: loadingSize
        })))];
      }

      if (multiple && showCheckAll && Object.keys(optionsInfo).length > 0 && !inputValue && !loading) {
        options.unshift( /*#__PURE__*/_react["default"].createElement(_rcMenu.Item, {
          key: CHECK_ALL_VALUE,
          role: "option",
          type: "normal",
          value: CHECK_ALL_VALUE
        }, /*#__PURE__*/_react["default"].createElement(_checkbox["default"], (0, _extends2["default"])({
          prefixCls: checkboxPrefixCls
        }, _this.getCheckAllState())), /*#__PURE__*/_react["default"].createElement("span", null, "\u5168\u9009")));
      }

      return options;
    };

    _this.renderFilterOptionsFromChildren = function (children, childrenKeys, menuItems) {
      var sel = [];
      var props = _this.props;
      var inputValue = _this.state.inputValue;
      var tags = props.tags,
          multiple = props.multiple,
          checkboxPrefixCls = props.checkboxPrefixCls,
          prefixCls = props.prefixCls;

      _react["default"].Children.forEach(children, function (child) {
        if (!child) {
          return;
        }

        if (child.type && child.type.isSelectOptGroup) {
          var innerItems = _this.renderFilterOptionsFromChildren(child.props.children, childrenKeys, menuItems);

          if (innerItems.length) {
            var label = child.props.label;
            var _key3 = child.key;

            if (!_key3 && typeof label === 'string') {
              _key3 = label;
            } else if (!label && _key3) {
              label = _key3;
            }

            sel.push(
            /*#__PURE__*/
            // @ts-ignore
            _react["default"].createElement(_rcMenu.ItemGroup, {
              key: _key3,
              title: label
            }, innerItems));
          }

          return;
        }

        (0, _warning["default"])(child.type && child.type.isSelectOption, 'the children of `Select` should be `Select.Option` or `Select.OptGroup`, ' + ("instead of `" + (child.type && child.type.name || child.type && child.type.displayName || child.type) + "`."));
        var childValue = (0, _selectTools.getValuePropValue)(child);
        (0, _selectTools.validateOptionValue)(childValue, _this.props);

        if (_this.filterOption(inputValue, child)) {
          var _child$props = child.props,
              childType = _child$props.type,
              _children = _child$props.children,
              disabled = _child$props.disabled,
              exclusive = _child$props.exclusive;
          var type = childType !== 'custom' ? 'normal' : 'custom';
          var childNode = typeof _children === 'string' ? /*#__PURE__*/_react["default"].createElement(_searchText["default"], {
            showSearch: true,
            text: _children,
            searchValue: inputValue,
            prefixCls: prefixCls
          }) : /*#__PURE__*/_react["default"].createElement("span", null, _children);

          if ((tags || multiple // 兼容原Option包含 CheckboxText 逻辑（包含时，则原样输出）
          ) && !_react["default"].Children.toArray(_children).some(function (node) {
            return /*#__PURE__*/(0, _react.isValidElement)(node) && node.type === _checkboxText["default"];
          })) {
            var checked = false;

            if (_this.state.value.indexOf(childValue) > -1) {
              checked = true;
            }

            childNode = /*#__PURE__*/_react["default"].createElement("span", null, exclusive ? null : /*#__PURE__*/_react["default"].createElement(_checkbox["default"], {
              checked: checked,
              prefixCls: checkboxPrefixCls,
              disabled: disabled
            }), childNode);
          }

          var menuItem = /*#__PURE__*/_react["default"].createElement(_rcMenu.Item, (0, _extends2["default"])({
            style: _selectTools.UNSELECTABLE_STYLE,
            attribute: _selectTools.UNSELECTABLE_ATTRIBUTE,
            value: childValue,
            key: childValue,
            role: "option"
          }, child.props, {
            type: type
          }), childNode);

          sel.push(menuItem);
          menuItems.push(menuItem);
        }

        if (tags) {
          childrenKeys.push(childValue);
        }
      });

      return sel;
    };

    _this.renderTopControlNode = function () {
      var _this$state3 = _this.state,
          value = _this$state3.value,
          open = _this$state3.open,
          inputValue = _this$state3.inputValue;
      var props = _this.props;
      var prefixCls = props.prefixCls,
          removeIcon = props.removeIcon,
          maxTagCount = props.maxTagCount,
          titleCallback = props.titleCallback,
          showSearch = props.showSearch,
          selectorName = props.selectorName,
          customRenderTarget = props.customRenderTarget;
      var className = prefixCls + "-selection__rendered";
      var innerNode = null;

      if ((0, _selectTools.isSingleMode)(props)) {
        var selectedValue = null;

        if (value.length === 1 && value[0] !== undefined) {
          var _classnames2;

          var showSelectedValue = false;
          var opacity = 1;

          if (!showSearch) {
            showSelectedValue = true;
          } else if (open) {
            showSelectedValue = !inputValue;

            if (showSelectedValue) {
              opacity = 0.4;
            }
          } else {
            showSelectedValue = true;
          }

          var _singleValue = value[0];

          var _optionInfo = _this.getOptionInfoBySingleValue(_singleValue);

          var label = _optionInfo.label,
              title = _optionInfo.title;
          var showLabel = label || selectorName;

          if (customRenderTarget && typeof customRenderTarget === 'function') {
            showLabel = customRenderTarget(_singleValue, _optionInfo);
          }

          var itemClassName = (0, _commonTools.classnames)(prefixCls + "-selection-selected-value", (_classnames2 = {}, _classnames2[prefixCls + "-selection-selected-value-selector-name"] = !label && selectorName, _classnames2));
          selectedValue = /*#__PURE__*/_react["default"].createElement("div", {
            key: "value",
            className: itemClassName,
            title: (0, _selectTools.toTitle)(title || label),
            style: {
              display: showSelectedValue ? 'block' : 'none',
              opacity: opacity
            }
          }, showLabel);
        }

        if (!showSearch) {
          innerNode = [selectedValue];
        } else {
          innerNode = [selectedValue, /*#__PURE__*/_react["default"].createElement("div", {
            className: prefixCls + "-search " + prefixCls + "-search--inline",
            key: "input",
            style: {
              display: open ? 'block' : 'none'
            }
          }, _this.getInputElement())];
        }

        return /*#__PURE__*/_react["default"].createElement("div", {
          className: className,
          ref: _this.saveTopCtrlRef
        }, _this.getPlaceholderElement(), innerNode);
      }

      var selectedValueNodes = [];
      var limitedCountValue = value;

      if ((0, _selectTools.isMultipleOrTags)(props)) {
        var _classnames3;

        var type = titleCallback.type,
            _selectorName = titleCallback.selectorName;

        if (type === 'enum' || type === 'count' || type === 'custom') {
          // 枚举类型
          var firstLabel = limitedCountValue[0] != null ? (0, _selectTools.getMultipleCheckboxLabel)(_this.getOptionInfoBySingleValue(limitedCountValue[0]).label || '') : '';
          var secondLabel = limitedCountValue[1] != null ? (0, _selectTools.getMultipleCheckboxLabel)(_this.getOptionInfoBySingleValue(limitedCountValue[1]).label || '') : '';
          var limitedCountValueLength = limitedCountValue.length;
          var text = limitedCountValueLength === 1 ? firstLabel : firstLabel + "\u3001" + secondLabel + "\u7B49" + limitedCountValueLength + "\u4E2A";

          if (type === 'count') {
            text = limitedCountValueLength === 1 ? firstLabel : _selectorName + "(" + limitedCountValueLength + "\u4E2A)";
          } else if (type === 'custom') {
            if (typeof titleCallback.customRenderTarget === 'function') {
              text = titleCallback.customRenderTarget(value);
            } else {
              text = _selectorName;
            }
          }

          var _showSelectedValue = false;
          var _opacity = 1;

          if (!showSearch) {
            _showSelectedValue = true;
          } else if (open) {
            _showSelectedValue = !inputValue;

            if (_showSelectedValue) {
              _opacity = 0.4;
            }
          } else {
            _showSelectedValue = true;
          }

          var _selectedValue = limitedCountValueLength ? /*#__PURE__*/_react["default"].createElement("div", {
            key: "value",
            className: prefixCls + "-selection-selected-value",
            title: (0, _selectTools.toTitle)(text),
            style: {
              display: _showSelectedValue ? 'block' : 'none',
              opacity: _opacity
            }
          }, text) : null;

          return /*#__PURE__*/_react["default"].createElement("div", {
            className: className,
            ref: _this.saveTopCtrlRef
          }, _this.getPlaceholderElement(), _selectedValue, showSearch && typeof text === 'string' // 字符串回填支持搜索
          && /*#__PURE__*/_react["default"].createElement("div", {
            className: prefixCls + "-search " + prefixCls + "-search--inline",
            key: "input",
            style: {
              display: open ? 'block' : 'none'
            }
          }, _this.getInputElement()));
        }

        selectedValueNodes = limitedCountValue.map(function (singleValue) {
          var info = _this.getOptionInfoBySingleValue(singleValue);

          var infoLabel = /*#__PURE__*/_react["default"].isValidElement(info.label) && info.label.props.label || info.label;
          var content = typeof infoLabel === 'string' ? infoLabel : singleValue;
          var title = info.title || content;

          var disabled = _this.isChildDisabled(singleValue);

          var choiceClassName = disabled ? prefixCls + "-selection__choice " + prefixCls + "-selection__choice__disabled" : prefixCls + "-selection__choice";
          return /*#__PURE__*/_react["default"].createElement("li", (0, _extends2["default"])({
            style: _selectTools.UNSELECTABLE_STYLE
          }, _selectTools.UNSELECTABLE_ATTRIBUTE, {
            onMouseDown: _selectTools.preventDefaultEvent,
            className: choiceClassName,
            key: singleValue || SELECT_EMPTY_VALUE_KEY,
            title: (0, _selectTools.toTitle)(title)
          }), /*#__PURE__*/_react["default"].createElement("div", {
            className: prefixCls + "-selection__choice__content"
          }, content), /*#__PURE__*/_react["default"].createElement(_button["default"], {
            type: "text-aux",
            onClick: function onClick(event) {
              _this.removeSelected(singleValue, event);
            },
            className: prefixCls + "-selection__choice__remove",
            disabled: disabled
          }, removeIcon || /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconTimes, null)));
        });
        var inputClasses = (0, _commonTools.classnames)(prefixCls + "-search", prefixCls + "-search--inline", (_classnames3 = {}, _classnames3[prefixCls + "-search-hidden"] = !open && value.length > maxTagCount, _classnames3));

        if (showSearch) {
          selectedValueNodes.push( /*#__PURE__*/_react["default"].createElement("li", {
            className: inputClasses,
            key: "input"
          }, _this.getInputElement()));
        }

        innerNode = selectedValueNodes.length ? /*#__PURE__*/_react["default"].createElement("ul", {
          className: prefixCls + "-search-ul"
        }, selectedValueNodes) : null;
        return /*#__PURE__*/_react["default"].createElement("div", {
          className: className,
          ref: _this.saveTopCtrlRef
        }, _this.getPlaceholderElement(), innerNode);
      }

      if (showSearch) {
        selectedValueNodes.push( /*#__PURE__*/_react["default"].createElement("li", {
          className: prefixCls + "-search " + prefixCls + "-search--inline",
          key: "__input"
        }, _this.getInputElement()));
      }

      innerNode = /*#__PURE__*/_react["default"].createElement("ul", null, selectedValueNodes);
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: className,
        ref: _this.saveTopCtrlRef
      }, _this.getPlaceholderElement(), innerNode);
    };

    var _optionsInfo = OneSelect.getOptionsInfoFromProps(_props);

    _this.state = {
      value: OneSelect.getValueFromProps(_props, true),
      // true: use default value
      inputValue: _props.combobox ? OneSelect.getInputValueForCombobox(_props, _optionsInfo, true // use default value
      ) : '',
      open: _props.defaultOpen,
      optionsInfo: _optionsInfo,
      // a flag for aviod redundant getOptionsInfoFromProps call
      skipBuildOptionsInfo: true,
      showDisabledReason: false,
      // 获取dom高度
      ctrNodeHeight: 28
    };
    _this.saveInputRef = (0, _selectTools.saveRef)((0, _assertThisInitialized2["default"])(_this), 'inputRef');
    _this.saveInputMirrorRef = (0, _selectTools.saveRef)((0, _assertThisInitialized2["default"])(_this), 'inputMirrorRef');
    _this.saveTopCtrlRef = (0, _selectTools.saveRef)((0, _assertThisInitialized2["default"])(_this), 'topCtrlRef');
    _this.saveSelectTriggerRef = (0, _selectTools.saveRef)((0, _assertThisInitialized2["default"])(_this), 'selectTriggerRef');
    _this.saveRootRef = (0, _selectTools.saveRef)((0, _assertThisInitialized2["default"])(_this), 'rootRef');
    _this.saveSelectionRef = (0, _selectTools.saveRef)((0, _assertThisInitialized2["default"])(_this), 'selectionRef');
    return _this;
  }

  var _proto = OneSelect.prototype;

  _proto.componentDidMount = function componentDidMount() {
    if (this.props.autoFocus) {
      this.focus();
    }
  };

  OneSelect.getDerivedStateFromProps = function getDerivedStateFromProps(props, state) {
    var optionsInfo = state.skipBuildOptionsInfo ? state.optionsInfo : OneSelect.getOptionsInfoFromProps(props, state);

    if ('children' in props && props.children !== state.children) {
      optionsInfo = OneSelect.getOptionsInfoFromProps(props, state);
    }

    var newState = {
      optionsInfo: optionsInfo,
      skipBuildOptionsInfo: false,
      children: props.children
    };

    if ('open' in props && typeof props.open === 'boolean') {
      newState.open = props.open;
    }

    if ('value' in props) {
      var _value4 = OneSelect.getValueFromProps(props);

      newState.value = _value4;

      if (props.combobox) {
        newState.inputValue = OneSelect.getInputValueForCombobox(props, optionsInfo);
      }
    }

    return newState;
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    this.clearFocusTime();
    this.clearBlurTime();

    if (this.dropdownContainer) {
      _reactDom["default"].unmountComponentAtNode(this.dropdownContainer);

      document.body.removeChild(this.dropdownContainer);
      this.dropdownContainer = null;
    }
  };

  _proto.getToggleAllValue = function getToggleAllValue() {
    var _this$state4 = this.state,
        value = _this$state4.value,
        optionsInfo = _this$state4.optionsInfo;

    if (value.length > 0 && !value.every(function (val) {
      return optionsInfo[(0, _selectTools.getMapKey)(val)].disabled;
    })) {
      // 取消全选
      return value.filter(function (val) {
        return optionsInfo[(0, _selectTools.getMapKey)(val)].disabled;
      });
    } // 全选


    var restValue = Object.keys(optionsInfo).filter(function (key) {
      var item = optionsInfo[key];

      if (value.includes(item.value)) {
        return false;
      }

      return !item.disabled;
    }).map(function (key) {
      return optionsInfo[key].value;
    });
    return value.concat(restValue);
  };

  _proto.getCheckAllState = function getCheckAllState() {
    var _this$state5 = this.state,
        value = _this$state5.value,
        optionsInfo = _this$state5.optionsInfo;
    var allValue = Object.keys(optionsInfo).map(function (key) {
      return optionsInfo[key].value;
    });
    var allChecked = allValue.every(function (val) {
      return value.includes(val);
    });
    var existChecked = allValue.some(function (val) {
      return value.includes(val);
    });
    return {
      checked: allChecked,
      indeterminate: existChecked && !allChecked
    };
  };

  _proto.renderClear = function renderClear() {
    var _this$props5 = this.props,
        prefixCls = _this$props5.prefixCls,
        allowClear = _this$props5.allowClear,
        clearIcon = _this$props5.clearIcon;
    var _this$state6 = this.state,
        value = _this$state6.value,
        inputValue = _this$state6.inputValue;

    var clear = /*#__PURE__*/_react["default"].createElement("span", (0, _extends2["default"])({
      key: "clear",
      className: prefixCls + "-selection__clear",
      onMouseDown: _selectTools.preventDefaultEvent,
      style: _selectTools.UNSELECTABLE_STYLE
    }, _selectTools.UNSELECTABLE_ATTRIBUTE, {
      onClick: this.onClearSelection
    }), clearIcon || /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconTimesCircle, null));

    if (!allowClear) {
      return null;
    }

    if ((0, _selectTools.isCombobox)(this.props)) {
      if (inputValue) {
        return clear;
      }

      return null;
    }

    if (inputValue || value.length) {
      return clear;
    }

    return null;
  };

  _proto.renderTotalDom = function renderTotalDom() {
    var _classnames4;

    var _this$props6 = this.props,
        maxTagCount = _this$props6.maxTagCount,
        prefixCls = _this$props6.prefixCls,
        titleCallback = _this$props6.titleCallback,
        size = _this$props6.size;

    if (!maxTagCount) {
      return null;
    }

    var value = this.state.value;
    var valueLength = value.length;

    if (!valueLength || titleCallback.type !== 'list') {
      return null;
    }

    var ctrNodeHeight = this.state.ctrNodeHeight;
    var totalCountCls = (0, _commonTools.classnames)(prefixCls + "-selection__total_count", (_classnames4 = {}, _classnames4[prefixCls + "-selection__total_count-error"] = valueLength > maxTagCount, _classnames4[prefixCls + "-selection__total_count-min"] = ctrNodeHeight <= (heightToSizeMap[size] || heightToSizeMap.small), _classnames4));
    return /*#__PURE__*/_react["default"].createElement("span", {
      className: totalCountCls
    }, valueLength, "/", maxTagCount);
  };

  _proto.render = function render() {
    var _rootCls;

    var props = this.props;
    var multiple = (0, _selectTools.isMultipleOrTags)(props);
    var state = this.state;
    var className = props.className,
        disabled = props.disabled,
        prefixCls = props.prefixCls,
        inputIcon = props.inputIcon,
        titleCallback = props.titleCallback,
        suffixIcon = props.suffixIcon,
        footer = props.footer;
    var ctrlNode = this.renderTopControlNode();
    var open = this.state.open;

    if (open) {
      this._options = this.renderFilterOptions();
    }

    var realOpen = this.getRealOpenState();
    var options = this._options || [];
    var dataOrAriaAttributeProps = {}; // eslint-disable-next-line no-restricted-syntax

    for (var _key4 in props) {
      if (Object.prototype.hasOwnProperty.call(props, _key4) && (_key4.substr(0, 5) === 'data-' || _key4 === 'role')) {
        dataOrAriaAttributeProps[_key4] = props[_key4];
      }
    }

    var extraSelectionProps = (0, _extends2["default"])({}, dataOrAriaAttributeProps);

    if (!(0, _selectTools.isMultipleOrTagsOrCombobox)(props)) {
      extraSelectionProps = (0, _extends2["default"])({}, extraSelectionProps, {
        onKeyDown: this.onKeyDown,
        tabIndex: props.disabled ? -1 : 0
      });
    }

    var hidePopElementInSearchMode = this.hidePopElementInSearchMode();
    var rootCls = (_rootCls = {}, _rootCls[className] = !!className, _rootCls[prefixCls] = 1, _rootCls[prefixCls + "-open"] = open && hidePopElementInSearchMode, _rootCls[prefixCls + "-focus"] = open || !!this._focused, _rootCls[_config.COMPONENT_FOCUS] = open || !!this._focused, _rootCls[prefixCls + "-combobox"] = (0, _selectTools.isCombobox)(props), _rootCls[prefixCls + "-disabled"] = disabled, _rootCls[prefixCls + "-enabled"] = !disabled, _rootCls[prefixCls + "-allow-clear"] = !!props.allowClear, _rootCls[prefixCls + "-no-arrow"] = !props.showArrow, _rootCls);
    var sectionCls = (0, _commonTools.classnames)(prefixCls + "-selection", prefixCls + "-selection--" + (multiple ? 'multiple' : 'single'));
    var type = multiple ? titleCallback.type : 'single';
    var trigger = 'click';

    if ('trigger' in props && !multiple) {
      trigger = props.trigger;
    }

    var action = disabled ? [] : [trigger];
    var suffixIconRender = null;

    if (suffixIcon) {
      if (typeof suffixIcon === 'string') {
        suffixIconRender = /*#__PURE__*/_react["default"].createElement(_iconSvg["default"], {
          type: suffixIcon
        });
      } else {
        suffixIconRender = typeof suffixIcon === 'function' ? suffixIcon() : suffixIcon;
      }
    }

    var defaultTitleRender = /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
      ref: this.saveSelectionRef,
      key: "selection",
      className: sectionCls // @ts-ignore
      ,
      type: type
    }, extraSelectionProps), ctrlNode, this.renderClear(), suffixIconRender ? /*#__PURE__*/_react["default"].createElement("span", (0, _extends2["default"])({
      key: "custom-key",
      className: prefixCls + "-custom-key",
      style: _selectTools.UNSELECTABLE_STYLE
    }, _selectTools.UNSELECTABLE_ATTRIBUTE), suffixIconRender) : /*#__PURE__*/_react["default"].createElement("span", (0, _extends2["default"])({
      key: "arrow",
      className: prefixCls + "-arrow",
      style: _selectTools.UNSELECTABLE_STYLE
    }, _selectTools.UNSELECTABLE_ATTRIBUTE, {
      onClick: this.onArrowClick
    }), inputIcon || /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconChevronDown, null)), multiple ? this.renderTotalDom() : null);

    return /*#__PURE__*/_react["default"].createElement(_selectTrigger["default"], {
      onPopupFocus: this.onPopupFocus,
      onMouseEnter: this.onMouseEnter,
      onMouseLeave: this.onMouseLeave,
      dropdownAlign: props.dropdownAlign,
      dropdownClassName: props.dropdownClassName,
      dropdownMatchSelectWidth: props.dropdownMatchSelectWidth,
      defaultActiveFirstOption: props.defaultActiveFirstOption,
      dropdownMenuStyle: props.dropdownMenuStyle,
      transitionName: props.transitionName,
      animation: props.animation,
      prefixCls: props.prefixCls,
      dropdownStyle: props.dropdownStyle,
      combobox: props.combobox,
      showSearch: props.showSearch,
      options: options,
      multiple: multiple,
      disabled: disabled,
      visible: realOpen,
      inputValue: state.inputValue,
      value: state.value,
      backfillValue: state.backfillValue,
      firstActiveValue: props.firstActiveValue,
      onDropdownVisibleChange: this.onDropdownVisibleChange,
      getPopupContainer: props.getPopupContainer,
      onMenuSelect: this.onMenuSelect,
      onMenuDeselect: this.onMenuDeselect,
      onPopupScroll: props.onPopupScroll,
      showAction: props.showAction,
      ref: this.saveSelectTriggerRef,
      menuItemSelectedIcon: props.menuItemSelectedIcon,
      action: action,
      footer: footer
    }, /*#__PURE__*/_react["default"].createElement("div", {
      id: props.id,
      style: props.style,
      ref: this.saveRootRef,
      onBlur: this.onOuterBlur,
      onFocus: this.onOuterFocus,
      className: (0, _commonTools.classnames)(rootCls),
      onMouseDown: this.markMouseDown,
      onMouseUp: this.markMouseLeave,
      onMouseOut: this.markMouseLeave
    }, defaultTitleRender));
  };

  return OneSelect;
}(_react.PureComponent);

exports["default"] = OneSelect;
OneSelect.defaultProps = {
  prefixCls: 'one-select',
  defaultOpen: false,
  labelInValue: false,
  defaultActiveFirstOption: true,
  allowClear: false,
  placeholder: '',
  onChange: noop,
  onFocus: noop,
  onBlur: noop,
  onSelect: noop,
  onSearch: noop,
  onDeselect: noop,
  onInputKeyDown: noop,
  showArrow: true,
  dropdownMatchSelectWidth: true,
  dropdownStyle: {},
  dropdownMenuStyle: {},
  optionFilterProp: 'value',
  optionLabelProp: 'value',
  notFoundContent: 'Not Found',
  backfill: false,
  showAction: ['click'],
  tokenSeparators: [],
  autoClearSearchValue: true,
  titleCallback: {
    type: 'list',
    // 三类回填 enum, list, count
    selectorName: '' // 选择已选计数需要填写

  },
  disabledReason: '',
  selectorName: '',
  size: 'medium',
  suffixIcon: null,
  checkboxPrefixCls: 'one-checkbox',
  loading: false,
  loadingText: '加载中...'
};

OneSelect.getOptionsFromChildren = function (children, options) {
  if (options === void 0) {
    options = [];
  }

  _react["default"].Children.forEach(children, function (child) {
    if (!child) {
      return;
    }

    if (child.type && child.type.isSelectOptGroup) {
      OneSelect.getOptionsFromChildren(child.props.children, options);
    } else {
      options.push(child);
    }
  });

  return options;
};

OneSelect.getInputValueForCombobox = function (props, optionsInfo, useDefaultValue) {
  var values = [];

  if ('value' in props && !useDefaultValue) {
    values = (0, _selectTools.toArray)(props.value);
  }

  if ('defaultValue' in props && useDefaultValue) {
    values = (0, _selectTools.toArray)(props.defaultValue);
  }

  if (!values.length) {
    return '';
  }

  var value = values[0];
  var label = value;

  if (props.labelInValue) {
    label = value.label;
  } else if (optionsInfo[(0, _selectTools.getMapKey)(value)]) {
    label = optionsInfo[(0, _selectTools.getMapKey)(value)].label;
  }

  if (label === undefined) {
    label = '';
  }

  return label;
};

OneSelect.getLabelFromOption = function (props, option) {
  return (0, _selectTools.getPropValue)(option, props.optionLabelProp);
};

OneSelect.getOptionsInfoFromProps = function (props, preState) {
  var options = OneSelect.getOptionsFromChildren(props.children);
  var optionsInfo = {};
  options.forEach(function (option) {
    var singleValue = (0, _selectTools.getValuePropValue)(option);
    var disabled = (0, _selectTools.getPropValue)(option, 'disabled');
    var exclusive = (0, _selectTools.getPropValue)(option, 'exclusive');
    optionsInfo[(0, _selectTools.getMapKey)(singleValue)] = {
      option: option,
      value: singleValue,
      disabled: disabled,
      exclusive: exclusive,
      label: OneSelect.getLabelFromOption(props, option),
      title: option.props.title
    };
  });

  if (preState) {
    // keep option info in pre state value.
    var oldOptionsInfo = preState.optionsInfo;
    var _value5 = preState.value;

    _value5.forEach(function (v) {
      var key = (0, _selectTools.getMapKey)(v);

      if (!optionsInfo[key] && oldOptionsInfo[key] !== undefined) {
        optionsInfo[key] = oldOptionsInfo[key];
      }
    });
  }

  return optionsInfo;
};

OneSelect.getValueFromProps = function (props, useDefaultValue) {
  var value = [];
  var multiple = (0, _selectTools.isMultipleOrTags)(props);

  if ('defaultValue' in props && useDefaultValue) {
    value = (0, _selectTools.toArray)(multiple && props.defaultValue == null ? [] : props.defaultValue);
  } else if ('value' in props) {
    value = (0, _selectTools.toArray)(multiple && props.value == null ? [] : props.value);
  }

  if (props.labelInValue) {
    value = value.map(function (v) {
      return v.key;
    });
  }

  return value;
};

module.exports = exports.default;
