// dls-illustrations-react@1.4.3/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dlsGraphics = require('dls-graphics');
var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};

  var target = _objectWithoutPropertiesLoose(source, excluded);

  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];

  if (_i == null) return;
  var _arr = [];
  var _n = true;
  var _d = false;

  var _s, _e;

  try {
    for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

var urlAlphabet = 'useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict';
var nanoid = function nanoid() {
  var size = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 21;
  var id = '';
  var i = size;

  while (i--) {
    id += urlAlphabet[Math.random() * 64 | 0];
  }

  return id;
};

var ESCAPE_MAP = {
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  '&': '&amp;'
};
function escapeHTML(html) {
  return html.replace(/[<>"&]/g, function (c) {
    return ESCAPE_MAP[c] || c;
  });
}
function uid() {
  return nanoid(5);
}
function updateId(content, id, instanceId) {
  var re = new RegExp("dls-".concat(id), 'g');
  return content.replace(re, "dls-".concat(id, "-").concat(instanceId));
}

var _excluded = ["class", "id"],
    _excluded2 = ["title", "className"];
function createSVG(name, _ref) {
  var contents = _ref.contents,
      _ref$attrs = _ref.attrs,
      classAttr = _ref$attrs["class"],
      id = _ref$attrs.id,
      attrs = _objectWithoutProperties(_ref$attrs, _excluded);

  var Illustration = /*#__PURE__*/React.forwardRef(function (_ref2, ref) {
    var title = _ref2.title,
        className = _ref2.className,
        props = _objectWithoutProperties(_ref2, _excluded2);

    var tabIndex = props.tabIndex;

    var _useState = React.useState(uid()),
        _useState2 = _slicedToArray(_useState, 1),
        instanceId = _useState2[0];

    var markup = {
      __html: (title ? "<title>".concat(escapeHTML(title), "</title>") : '') + updateId(contents, id, instanceId)
    };
    return /*#__PURE__*/React__default["default"].createElement("svg", _extends({
      className: [classAttr, className].filter(Boolean).join(' ') || null
    }, attrs, {
      focusable: tabIndex !== '0' ? false : null,
      dangerouslySetInnerHTML: markup,
      ref: ref
    }, props));
  });
  Illustration.displayName = name;
  return Illustration;
}

var IllustrationPartialBlankBrush = /*#__PURE__*/createSVG('IllustrationPartialBlankBrush', dlsGraphics.partialBlankBrush);

var IllustrationPartialBlank = /*#__PURE__*/createSVG('IllustrationPartialBlank', dlsGraphics.partialBlank);

var IllustrationPartialError = /*#__PURE__*/createSVG('IllustrationPartialError', dlsGraphics.partialError);

var IllustrationPartialForbidden = /*#__PURE__*/createSVG('IllustrationPartialForbidden', dlsGraphics.partialForbidden);

var IllustrationUnderReview = /*#__PURE__*/createSVG('IllustrationUnderReview', dlsGraphics.underReview);

var IllustrationBlank = /*#__PURE__*/createSVG('IllustrationBlank', dlsGraphics.blank);

var IllustrationClientError = /*#__PURE__*/createSVG('IllustrationClientError', dlsGraphics.clientError);

var IllustrationForbidden = /*#__PURE__*/createSVG('IllustrationForbidden', dlsGraphics.forbidden);

var IllustrationNoResults = /*#__PURE__*/createSVG('IllustrationNoResults', dlsGraphics.noResults);

var IllustrationNotFound = /*#__PURE__*/createSVG('IllustrationNotFound', dlsGraphics.notFound);

var IllustrationReviewError = /*#__PURE__*/createSVG('IllustrationReviewError', dlsGraphics.reviewError);

var IllustrationReviewPending = /*#__PURE__*/createSVG('IllustrationReviewPending', dlsGraphics.reviewPending);

var IllustrationReviewSuccess = /*#__PURE__*/createSVG('IllustrationReviewSuccess', dlsGraphics.reviewSuccess);

var IllustrationServerError = /*#__PURE__*/createSVG('IllustrationServerError', dlsGraphics.serverError);

var IllustrationImagePlaceholder = /*#__PURE__*/createSVG('IllustrationImagePlaceholder', dlsGraphics.imagePlaceholder);

var IllustrationLoading = /*#__PURE__*/createSVG('IllustrationLoading', dlsGraphics.loading);

var IllustrationSpotBarChart = /*#__PURE__*/createSVG('IllustrationSpotBarChart', dlsGraphics.spotBarChart);

var IllustrationSpotCleared = /*#__PURE__*/createSVG('IllustrationSpotCleared', dlsGraphics.spotCleared);

var IllustrationSpotFunnelChart = /*#__PURE__*/createSVG('IllustrationSpotFunnelChart', dlsGraphics.spotFunnelChart);

var IllustrationSpotGaugeChart = /*#__PURE__*/createSVG('IllustrationSpotGaugeChart', dlsGraphics.spotGaugeChart);

var IllustrationSpotImageError = /*#__PURE__*/createSVG('IllustrationSpotImageError', dlsGraphics.spotImageError);

var IllustrationSpotLineChart = /*#__PURE__*/createSVG('IllustrationSpotLineChart', dlsGraphics.spotLineChart);

var IllustrationSpotLoadError = /*#__PURE__*/createSVG('IllustrationSpotLoadError', dlsGraphics.spotLoadError);

var IllustrationSpotMap = /*#__PURE__*/createSVG('IllustrationSpotMap', dlsGraphics.spotMap);

var IllustrationSpotNetworkError = /*#__PURE__*/createSVG('IllustrationSpotNetworkError', dlsGraphics.spotNetworkError);

var IllustrationSpotNoAccess = /*#__PURE__*/createSVG('IllustrationSpotNoAccess', dlsGraphics.spotNoAccess);

var IllustrationSpotNoContent = /*#__PURE__*/createSVG('IllustrationSpotNoContent', dlsGraphics.spotNoContent);

var IllustrationSpotNoFavorites = /*#__PURE__*/createSVG('IllustrationSpotNoFavorites', dlsGraphics.spotNoFavorites);

var IllustrationSpotNoHistory = /*#__PURE__*/createSVG('IllustrationSpotNoHistory', dlsGraphics.spotNoHistory);

var IllustrationSpotNoImage = /*#__PURE__*/createSVG('IllustrationSpotNoImage', dlsGraphics.spotNoImage);

var IllustrationSpotNoInternet = /*#__PURE__*/createSVG('IllustrationSpotNoInternet', dlsGraphics.spotNoInternet);

var IllustrationSpotNoMessages = /*#__PURE__*/createSVG('IllustrationSpotNoMessages', dlsGraphics.spotNoMessages);

var IllustrationSpotNoResults = /*#__PURE__*/createSVG('IllustrationSpotNoResults', dlsGraphics.spotNoResults);

var IllustrationSpotNoStore = /*#__PURE__*/createSVG('IllustrationSpotNoStore', dlsGraphics.spotNoStore);

var IllustrationSpotNoVideo = /*#__PURE__*/createSVG('IllustrationSpotNoVideo', dlsGraphics.spotNoVideo);

var IllustrationSpotPieChart = /*#__PURE__*/createSVG('IllustrationSpotPieChart', dlsGraphics.spotPieChart);

var IllustrationSpotPresentation = /*#__PURE__*/createSVG('IllustrationSpotPresentation', dlsGraphics.spotPresentation);

var IllustrationSpotRadarChart = /*#__PURE__*/createSVG('IllustrationSpotRadarChart', dlsGraphics.spotRadarChart);

var IllustrationSpotRelationshipChart = /*#__PURE__*/createSVG('IllustrationSpotRelationshipChart', dlsGraphics.spotRelationshipChart);

var IllustrationSpotServerError = /*#__PURE__*/createSVG('IllustrationSpotServerError', dlsGraphics.spotServerError);

var IllustrationSpotTryLater = /*#__PURE__*/createSVG('IllustrationSpotTryLater', dlsGraphics.spotTryLater);

var IllustrationSpotVideoError = /*#__PURE__*/createSVG('IllustrationSpotVideoError', dlsGraphics.spotVideoError);

var IllustrationSpotWordCloud = /*#__PURE__*/createSVG('IllustrationSpotWordCloud', dlsGraphics.spotWordCloud);

exports.IllustrationBlank = IllustrationBlank;
exports.IllustrationClientError = IllustrationClientError;
exports.IllustrationForbidden = IllustrationForbidden;
exports.IllustrationImagePlaceholder = IllustrationImagePlaceholder;
exports.IllustrationLoading = IllustrationLoading;
exports.IllustrationNoResults = IllustrationNoResults;
exports.IllustrationNotFound = IllustrationNotFound;
exports.IllustrationPartialBlank = IllustrationPartialBlank;
exports.IllustrationPartialBlankBrush = IllustrationPartialBlankBrush;
exports.IllustrationPartialError = IllustrationPartialError;
exports.IllustrationPartialForbidden = IllustrationPartialForbidden;
exports.IllustrationReviewError = IllustrationReviewError;
exports.IllustrationReviewPending = IllustrationReviewPending;
exports.IllustrationReviewSuccess = IllustrationReviewSuccess;
exports.IllustrationServerError = IllustrationServerError;
exports.IllustrationSpotBarChart = IllustrationSpotBarChart;
exports.IllustrationSpotCleared = IllustrationSpotCleared;
exports.IllustrationSpotFunnelChart = IllustrationSpotFunnelChart;
exports.IllustrationSpotGaugeChart = IllustrationSpotGaugeChart;
exports.IllustrationSpotImageError = IllustrationSpotImageError;
exports.IllustrationSpotLineChart = IllustrationSpotLineChart;
exports.IllustrationSpotLoadError = IllustrationSpotLoadError;
exports.IllustrationSpotMap = IllustrationSpotMap;
exports.IllustrationSpotNetworkError = IllustrationSpotNetworkError;
exports.IllustrationSpotNoAccess = IllustrationSpotNoAccess;
exports.IllustrationSpotNoContent = IllustrationSpotNoContent;
exports.IllustrationSpotNoFavorites = IllustrationSpotNoFavorites;
exports.IllustrationSpotNoHistory = IllustrationSpotNoHistory;
exports.IllustrationSpotNoImage = IllustrationSpotNoImage;
exports.IllustrationSpotNoInternet = IllustrationSpotNoInternet;
exports.IllustrationSpotNoMessages = IllustrationSpotNoMessages;
exports.IllustrationSpotNoResults = IllustrationSpotNoResults;
exports.IllustrationSpotNoStore = IllustrationSpotNoStore;
exports.IllustrationSpotNoVideo = IllustrationSpotNoVideo;
exports.IllustrationSpotPieChart = IllustrationSpotPieChart;
exports.IllustrationSpotPresentation = IllustrationSpotPresentation;
exports.IllustrationSpotRadarChart = IllustrationSpotRadarChart;
exports.IllustrationSpotRelationshipChart = IllustrationSpotRelationshipChart;
exports.IllustrationSpotServerError = IllustrationSpotServerError;
exports.IllustrationSpotTryLater = IllustrationSpotTryLater;
exports.IllustrationSpotVideoError = IllustrationSpotVideoError;
exports.IllustrationSpotWordCloud = IllustrationSpotWordCloud;
exports.IllustrationUnderReview = IllustrationUnderReview;
exports.createSVG = createSVG;
