// @baidu/one-ui@4.41.1/es/components/empty/index.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import React, { forwardRef } from "react";
import { classnames } from '../../core/commonTools';
import { IllustrationSpotNoContent } from 'dls-illustrations-react';
import { withConfigConsumer } from '../providerConfig/context';

function renderPart(part, className) {
  if (part) {
    return /*#__PURE__*/React.createElement("div", {
      className: className
    }, part);
  }

  return null;
}
/**
 * @visibleName Empty
 */


export var EmptyInternal = /*#__PURE__*/forwardRef(function (_ref, ref) {
  var className = _ref.className,
      prefixCls = _ref.prefixCls,
      size = _ref.size,
      title = _ref.title,
      description = _ref.description,
      actions = _ref.actions,
      children = _ref.children,
      lightPrefix = _ref.lightPrefix,
      restProps = _objectWithoutPropertiesLoose(_ref, ["className", "prefixCls", "size", "title", "description", "actions", "children", "lightPrefix"]);

  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: classnames("" + prefixCls, prefixCls + "-" + size, className)
  }, restProps), renderPart(children, prefixCls + "-image"), renderPart(title, prefixCls + "-title"), renderPart(description, prefixCls + "-description"), renderPart(actions, prefixCls + "-actions"));
});
EmptyInternal.defaultProps = {
  prefixCls: 'one-empty',
  description: '暂无数据',
  size: 'medium',
  children: /*#__PURE__*/React.createElement(IllustrationSpotNoContent, null)
};
EmptyInternal.displayName = 'Empty';
var Empty = withConfigConsumer('empty')(EmptyInternal);
export default Empty;
export * from '../interface';
