// @baidu/one-ui@4.32.1/es/components/select/common/dropdownMenu.js
import _extends from "@babel/runtime/helpers/extends";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { Component, cloneElement } from 'react';
import { findDOMNode } from 'react-dom';
import toArray from 'rc-util/lib/Children/toArray';
import Menu from 'rc-menu';
import scrollIntoView from 'dom-scroll-into-view';
import raf from 'raf';
import { getSelectKeys, preventDefaultEvent, saveRef } from '../../../core/selectTools';
;

var DropdownMenu = /*#__PURE__*/function (_Component) {
  _inheritsLoose(DropdownMenu, _Component);

  function DropdownMenu(props) {
    var _this;

    _this = _Component.call(this, props) || this;
    _this.lastInputValue = void 0;
    _this.saveMenuRef = void 0;
    _this.menuRef = void 0;
    _this.saveManualRef = void 0;
    _this.manualRef = void 0;
    _this.firstActiveItem = void 0;
    _this.rafInstance = void 0;
    _this.lastVisible = void 0;

    _this.scrollActiveItemToView = function () {
      // scroll into view
      var itemComponent = findDOMNode(_this.firstActiveItem);
      var _this$props = _this.props,
          value = _this$props.value,
          visible = _this$props.visible,
          firstActiveValue = _this$props.firstActiveValue;

      if (!itemComponent || !visible) {
        return;
      }

      var scrollIntoViewOpts = {
        onlyScrollIfNeeded: true,
        alignWithTop: false
      };

      if ((!value || value.length === 0) && firstActiveValue) {
        scrollIntoViewOpts.alignWithTop = true;
      } // Delay to scroll since current frame item position is not ready when pre view is by filter


      _this.rafInstance = raf(function () {
        scrollIntoView(itemComponent, findDOMNode(_this.menuRef), scrollIntoViewOpts);
      });
    };

    _this.lastInputValue = props.inputValue;
    _this.saveMenuRef = saveRef(_assertThisInitialized(_this), 'menuRef');
    _this.saveManualRef = saveRef(_assertThisInitialized(_this), 'manualRef');
    return _this;
  }

  var _proto = DropdownMenu.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.scrollActiveItemToView();
    this.lastVisible = this.props.visible;
  };

  _proto.shouldComponentUpdate = function shouldComponentUpdate(nextProps) {
    if (!nextProps.visible) {
      this.lastVisible = false;
    } // freeze when hide


    return nextProps.visible || nextProps.inputValue !== this.props.inputValue;
  };

  _proto.componentDidUpdate = function componentDidUpdate(prevProps) {
    var props = this.props;

    if (!prevProps.visible && props.visible) {
      this.scrollActiveItemToView();
    }

    this.lastVisible = props.visible;
    this.lastInputValue = props.inputValue;
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    if (this.rafInstance && this.rafInstance.cancel) {
      this.rafInstance.cancel();
    }
  };

  _proto.renderMenu = function renderMenu() {
    var _this2 = this;

    var props = this.props;
    var menuItems = props.menuItems,
        menuItemSelectedIcon = props.menuItemSelectedIcon,
        defaultActiveFirstOption = props.defaultActiveFirstOption,
        value = props.value,
        prefixCls = props.prefixCls,
        multiple = props.multiple,
        onMenuSelect = props.onMenuSelect,
        onMenuDeselect = props.onMenuDeselect,
        inputValue = props.inputValue,
        firstActiveValue = props.firstActiveValue,
        backfillValue = props.backfillValue;

    if (menuItems && menuItems.length) {
      var menuProps;

      if (multiple) {
        menuProps = {
          onDeselect: onMenuDeselect,
          onSelect: onMenuSelect
        };
      } else {
        menuProps = {
          onClick: onMenuSelect
        };
      }

      var selectedKeys = getSelectKeys(menuItems, value);
      var clonedMenuItems = menuItems;

      if (selectedKeys.length || firstActiveValue) {
        if (props.visible && !this.lastVisible) {
          menuProps.activeKey = selectedKeys[0] || firstActiveValue;
        }

        var foundFirst = false; // set firstActiveItem via cloning menus
        // for scroll into view

        var clone = function clone(item) {
          if (!foundFirst && selectedKeys.indexOf(item.key) !== -1 || !foundFirst && !selectedKeys.length && firstActiveValue.indexOf(item.key) !== -1) {
            foundFirst = true;
            return /*#__PURE__*/cloneElement(item, {
              ref: function ref(_ref) {
                _this2.firstActiveItem = _ref;
              }
            });
          }

          return item;
        };

        clonedMenuItems = menuItems.map(function (item) {
          if (item.type.isMenuItemGroup) {
            var children = toArray(item.props.children).map(clone);
            return /*#__PURE__*/cloneElement(item, {}, children);
          }

          return clone(item);
        });
      } else {
        // Clear firstActiveItem when dropdown menu items was empty
        // Avoid `Unable to find node on an unmounted component`
        this.firstActiveItem = null;
      } // clear activeKey when inputValue change


      var lastValue = value && value[value.length - 1];

      if (inputValue !== this.lastInputValue && (!lastValue || lastValue !== backfillValue)) {
        menuProps.activeKey = '';
      }

      return /*#__PURE__*/React.createElement(Menu, _extends({
        ref: this.saveMenuRef,
        manualRef: this.saveManualRef,
        style: this.props.dropdownMenuStyle,
        defaultActiveFirst: defaultActiveFirstOption // @ts-ignore
        ,
        role: "listbox",
        itemIcon: multiple ? menuItemSelectedIcon : null,
        multiple: multiple
      }, menuProps, {
        selectedKeys: selectedKeys,
        prefixCls: prefixCls + "-menu"
      }), clonedMenuItems);
    }

    return null;
  };

  _proto.render = function render() {
    var renderMenu = this.renderMenu();
    var _this$props2 = this.props,
        prefixCls = _this$props2.prefixCls,
        footer = _this$props2.footer;
    return renderMenu ? /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-menu-container",
      onFocus: this.props.onPopupFocus,
      onMouseDown: preventDefaultEvent,
      onScroll: this.props.onPopupScroll
    }, renderMenu, footer ? /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-menu-footer"
    }, footer) : null) : null;
  };

  return DropdownMenu;
}(Component);

export { DropdownMenu as default };
