// @baidu/one-ui@4.32.1/es/core/cascaderPaneTools.js
import _pullAll from "lodash/pullAll";
import _uniq from "lodash/uniq";
// 将options转成对象
export var transOptionsToObject = function transOptionsToObject(options, fieldName, optionsObject, parent) {
  if (fieldName === void 0) {
    fieldName = 'value';
  }

  if (optionsObject === void 0) {
    optionsObject = {};
  }

  options.forEach(function (option) {
    var value = option[fieldName];
    optionsObject[value] = {
      value: value,
      parent: parent || '',
      label: option.label
    };

    if (option.disabled) {
      optionsObject[value].disabled = true;
    }

    if (option.children && option.children.length) {
      optionsObject[value].children = option.children.map(function (child) {
        return child[fieldName];
      });
      transOptionsToObject(option.children, 'value', optionsObject, value);
    }
  });
  return optionsObject;
}; // 获取选中状态

export var formatCheckedKeys = function formatCheckedKeys(optionsObject, checkedKeys, newCheckedKeys) {
  if (newCheckedKeys === void 0) {
    newCheckedKeys = [];
  }

  checkedKeys.forEach(function (checkedKey) {
    newCheckedKeys.push(checkedKey);

    if (optionsObject[checkedKey] && optionsObject[checkedKey].children) {
      formatCheckedKeys(optionsObject, optionsObject[checkedKey].children, newCheckedKeys);
    }
  });
  return newCheckedKeys;
}; // 获取选中后的半选状态的值

export var getIndeterminateKeys = function getIndeterminateKeys(optionsObject, targetKeys, indeterminates, checkedKeys) {
  if (indeterminates === void 0) {
    indeterminates = [];
  }

  if (checkedKeys === void 0) {
    checkedKeys = [];
  }

  targetKeys.forEach(function (checkedKey) {
    if (optionsObject[checkedKey] && optionsObject[checkedKey].parent) {
      // 存在父节点，遍历父节点
      var allChecked = true;
      var parentKey = optionsObject[checkedKey].parent;
      var children = optionsObject[parentKey].children;

      if (children.length === 1 && indeterminates.indexOf(children[0]) > -1) {
        // 只有一个子节点
        allChecked = false;
      } else {
        optionsObject[parentKey].children.forEach(function (child) {
          if (checkedKeys.indexOf(child) === -1) {
            allChecked = false;
          }
        });
      }

      if (!allChecked) {
        indeterminates.push(parentKey);
        getIndeterminateKeys(optionsObject, [parentKey], indeterminates, checkedKeys);
      } else if (allChecked) {
        // 移出半选
        if (indeterminates.indexOf(parentKey) > -1) {
          indeterminates.splice(indeterminates.indexOf(parentKey), 1);
        }

        getIndeterminateKeys(optionsObject, [parentKey], indeterminates, checkedKeys);
      }
    }
  });
  return indeterminates;
}; // 获取取消选中的半选

export var getCancelIndeterminateKeys = function getCancelIndeterminateKeys(optionsObject, targetKeys, indeterminates, checkedKeys) {
  if (indeterminates === void 0) {
    indeterminates = [];
  }

  if (checkedKeys === void 0) {
    checkedKeys = [];
  }

  targetKeys.forEach(function (checkedKey) {
    if (optionsObject[checkedKey] && optionsObject[checkedKey].parent) {
      var hasChecked = false;
      var allChecked = true;
      var parentKey = optionsObject[checkedKey].parent;
      optionsObject[parentKey].children.forEach(function (child) {
        if (checkedKeys.indexOf(child) > -1 || indeterminates.indexOf(child) > -1) {
          hasChecked = true;
        }

        if (checkedKeys.indexOf(child) === -1) {
          allChecked = false;
        }
      });

      if (hasChecked) {
        if (indeterminates.indexOf(parentKey) === -1 && allChecked === false) {
          indeterminates.push(parentKey);
        }

        getCancelIndeterminateKeys(optionsObject, [parentKey], indeterminates, checkedKeys);
      } else {
        if (indeterminates.indexOf(parentKey) > -1) {
          indeterminates.splice(indeterminates.indexOf(parentKey), 1);
        }

        getCancelIndeterminateKeys(optionsObject, [parentKey], indeterminates, checkedKeys);
      }
    }
  });
  return indeterminates;
}; // 获取父亲全选的节点

export var formatAllCheckedKeys = function formatAllCheckedKeys(optionsObject, checkedKey, alCheckedKeys, checkedKeys) {
  if (checkedKeys === void 0) {
    checkedKeys = [];
  }

  checkedKeys.push(checkedKey);
  var newCheckedKeys = [].concat(alCheckedKeys, [checkedKey]);

  if (optionsObject[checkedKey] && optionsObject[checkedKey].parent) {
    var allChecked = true;
    var parentKey = optionsObject[checkedKey].parent;
    optionsObject[parentKey].children.forEach(function (child) {
      if (newCheckedKeys.indexOf(child) === -1) {
        allChecked = false;
      }
    });

    if (allChecked) {
      formatAllCheckedKeys(optionsObject, parentKey, newCheckedKeys, checkedKeys);
    }
  }

  return checkedKeys;
}; // 遍历parent节点

export var formatParentKeys = function formatParentKeys(optionsObject, checkedKey, parentKeys) {
  if (parentKeys === void 0) {
    parentKeys = [];
  }

  if (optionsObject[checkedKey] && optionsObject[checkedKey].parent) {
    parentKeys.push(optionsObject[checkedKey].parent);
    formatParentKeys(optionsObject, optionsObject[checkedKey].parent, parentKeys);
  }

  return parentKeys;
}; // 删除选中的checked keys

export var formatDeleteCheckedKeys = function formatDeleteCheckedKeys(optionsObject, deleteKey, interCheckedKeys) {
  var deleteCheckedKeys = [].concat(formatCheckedKeys(optionsObject, [deleteKey]), formatParentKeys(optionsObject, deleteKey));
  return [].concat(_pullAll(interCheckedKeys, deleteCheckedKeys));
}; // 获取筛选的key

export var filterSearchValue = function filterSearchValue(flattenTrees, inputValue) {
  var newFlattenTrees = [];
  flattenTrees.forEach(function (flattenTree) {
    flattenTree.forEach(function (tree) {
      if (tree.label.indexOf(inputValue) > -1) {
        newFlattenTrees.push(flattenTree);
      }
    });
  });
  return newFlattenTrees;
}; // 展平树 - 单选的情况

export var flattenTree = function flattenTree(options, ancestor) {
  if (ancestor === void 0) {
    ancestor = [];
  }

  var flattenOptions = [];
  options.forEach(function (option) {
    var path = ancestor.concat(option);

    if (!option.children || !option.children.length) {
      flattenOptions.push(path);
    }

    if (option.children) {
      flattenOptions = flattenOptions.concat(flattenTree(option.children, path));
    }
  });
  return flattenOptions;
}; // 寻找祖先

export var findAncestor = function findAncestor(optionsObject, curKey, ancestor) {
  if (ancestor === void 0) {
    ancestor = [];
  }

  if (optionsObject[curKey] && optionsObject[curKey].children) {
    ancestor.unshift(optionsObject[curKey]);
  }

  if (optionsObject[curKey] && optionsObject[curKey].parent) {
    var parentKey = optionsObject[curKey].parent;
    ancestor.unshift(optionsObject[curKey]);
    ancestor.unshift(optionsObject[parentKey]);

    if (optionsObject[parentKey] && optionsObject[parentKey].parent) {
      findAncestor(optionsObject, parentKey, ancestor);
    }
  }

  return _uniq(ancestor);
}; // 展平多选树 - 多选的情况

export var flattenMultipleTree = function flattenMultipleTree(optionsObject, options, ancestor) {
  if (ancestor === void 0) {
    ancestor = [];
  }

  options.forEach(function (option) {
    if (option.children && option.children.length) {
      ancestor.push(findAncestor(optionsObject, option.value));
      flattenMultipleTree(optionsObject, option.children, ancestor);
    }
  });
  return ancestor;
};
export var getDisabledKeys = function getDisabledKeys(disabledKeys, optionsObject) {
  var newDisabledKeys = [].concat(disabledKeys);
  disabledKeys.forEach(function (key) {
    if (optionsObject[key] && optionsObject[key].children) {
      newDisabledKeys = [].concat(newDisabledKeys, getDisabledKeys(optionsObject[key].children, optionsObject));
    }
  });
  return newDisabledKeys;
}; // 获取所有disabled的key，如果该节点有child则，child也应该disabled

export var formatDisabledKeys = function formatDisabledKeys(optionsObject) {
  var disabledKeys = [];
  Object.keys(optionsObject).forEach(function (key) {
    if (optionsObject[key] && optionsObject[key].disabled) {
      disabledKeys.push(key);
    }
  });
  return _uniq(getDisabledKeys(disabledKeys, optionsObject));
};
export var filterDisabledParentCheckedKeys = function filterDisabledParentCheckedKeys(optionsObject, checkedKeys) {
  var removeFilters = [];
  checkedKeys.forEach(function (key) {
    if (optionsObject[key] && optionsObject[key].children && optionsObject[key].children.length) {
      optionsObject[key].children.forEach(function (child) {
        if (checkedKeys.indexOf(child) === -1) {
          removeFilters.push(key);
        }
      });
    }
  });
  return [].concat(_pullAll(checkedKeys, removeFilters));
};
