// @baidu/one-ui@4.32.1/lib/components/menu/common/menu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireDefault(require("react"));

var _miniStore = require("mini-store");

var _subPopupMenu = _interopRequireWildcard(require("./subPopupMenu"));

var _commonTools = require("../../../core/commonTools");

;

var Menu = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2["default"])(Menu, _React$Component);

  function Menu(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;
    _this.isRootMenu = void 0;
    _this.store = void 0;
    _this.innerMenu = void 0;

    _this.onSelect = function (selectInfo) {
      var _this$props = _this.props,
          selectable = _this$props.selectable,
          multiple = _this$props.multiple,
          onSelect = _this$props.onSelect;

      if (selectable) {
        var _this$store$getState = _this.store.getState(),
            selectedKeys = _this$store$getState.selectedKeys;

        var selectedKey = selectInfo.key;

        if (multiple) {
          if (!selectedKeys.includes(selectedKey)) {
            selectedKeys = selectedKeys.concat(selectedKey);
          }
        } else {
          selectedKeys = [selectedKey];
        }

        if (!('selectedKeys' in _this.props)) {
          _this.store.setState({
            selectedKeys: selectedKeys
          });
        }

        if (onSelect) {
          onSelect((0, _extends2["default"])({}, selectInfo, {
            selectedKeys: selectedKeys
          }));
        }
      }
    };

    _this.onKeyDown = function (e, callback) {
      _this.innerMenu.getWrappedInstance().onKeyDown(e, callback);
    };

    _this.onOpenChange = function (event) {
      var openKeys = _this.store.getState().openKeys.concat();

      var changed = false;

      var processSingle = function processSingle(e) {
        var oneChanged = false;

        if (e.open) {
          oneChanged = openKeys.indexOf(e.key) === -1;

          if (oneChanged) {
            openKeys.push(e.key);
          }
        } else {
          var index = openKeys.indexOf(e.key);
          oneChanged = index !== -1;

          if (oneChanged) {
            openKeys.splice(index, 1);
          }
        }

        changed = changed || oneChanged;
      };

      if (Array.isArray(event)) {
        // batch change call
        event.forEach(processSingle);
      } else {
        processSingle(event);
      }

      if (changed) {
        if (!('openKeys' in _this.props)) {
          _this.store.setState({
            openKeys: openKeys
          });
        }

        _this.props.onOpenChange(openKeys);
      }
    };

    _this.onDeselect = function (selectInfo) {
      var _this$props2 = _this.props,
          selectable = _this$props2.selectable,
          onDeselect = _this$props2.onDeselect;

      if (selectable) {
        var selectedKeys = _this.store.getState().selectedKeys.concat();

        var selectedKey = selectInfo.key;
        var index = selectedKeys.indexOf(selectedKey);

        if (index !== -1) {
          selectedKeys.splice(index, 1);
        }

        if (!('selectedKeys' in _this.props)) {
          _this.store.setState({
            selectedKeys: selectedKeys
          });
        }

        if (onDeselect) {
          onDeselect((0, _extends2["default"])({}, selectInfo, {
            selectedKeys: selectedKeys
          }));
        }
      }
    };

    _this.setInnerMenu = function (node) {
      _this.innerMenu = node;
    };

    _this.isRootMenu = true;
    var _selectedKeys = props.defaultSelectedKeys;
    var _openKeys = props.defaultOpenKeys;

    if ('selectedKeys' in props) {
      _selectedKeys = props.selectedKeys || [];
    }

    if ('openKeys' in props) {
      _openKeys = props.openKeys || [];
    }

    _this.store = (0, _miniStore.create)({
      selectedKeys: _selectedKeys,
      openKeys: _openKeys,
      activeKey: {
        '0-menu-': (0, _subPopupMenu.getActiveKey)(props, props.activeKey)
      }
    });
    return _this;
  }

  var _proto = Menu.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.updateMiniStore();
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    this.updateMiniStore();
  };

  _proto.updateMiniStore = function updateMiniStore() {
    if ('selectedKeys' in this.props) {
      this.store.setState({
        selectedKeys: this.props.selectedKeys || []
      });
    }

    if ('openKeys' in this.props) {
      this.store.setState({
        openKeys: this.props.openKeys || []
      });
    }
  };

  _proto.render = function render() {
    var _this$props3 = this.props,
        prefixCls = _this$props3.prefixCls,
        className = _this$props3.className,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props3, ["prefixCls", "className"]);
    var props = (0, _extends2["default"])({}, restProps, {
      prefixCls: prefixCls,
      className: (0, _commonTools.classnames)(className, prefixCls + "-root"),
      onOpenChange: this.onOpenChange,
      onDeselect: this.onDeselect,
      onSelect: this.onSelect,
      parentMenu: this
    });
    return /*#__PURE__*/_react["default"].createElement(_miniStore.Provider, {
      store: this.store
    }, /*#__PURE__*/_react["default"].createElement(_subPopupMenu["default"], (0, _extends2["default"])({}, props, {
      ref: this.setInnerMenu
    }), this.props.children));
  };

  return Menu;
}(_react["default"].Component);

Menu.defaultProps = {
  selectable: true,
  onClick: function onClick() {},
  onSelect: function onSelect() {},
  onOpenChange: function onOpenChange() {},
  onDeselect: function onDeselect() {},
  defaultSelectedKeys: [],
  defaultOpenKeys: [],
  subMenuOpenDelay: 0.1,
  subMenuCloseDelay: 0.1,
  triggerSubMenuAction: 'hover',
  prefixCls: 'one-menu',
  className: '',
  mode: 'vertical',
  builtinPlacements: {}
};
var _default = Menu;
exports["default"] = _default;
module.exports = exports.default;
