// @baidu/one-ui@4.32.1/es/components/layout/footer.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import { withConfigConsumer } from '../providerConfig/context';
import { classnames } from '../../core/commonTools';
import Context from './context';
var LayoutFooter = (_dec = withConfigConsumer('layout-footer'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(LayoutFooter, _PureComponent);

  function LayoutFooter() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = LayoutFooter.prototype;

  _proto.componentDidMount = function componentDidMount() {
    if (this.props.sticky) {
      var _this$context;

      (_this$context = this.context) == null ? void 0 : _this$context.registerStickyFooter();
    }
  };

  _proto.componentDidUpdate = function componentDidUpdate(prevProps) {
    var sticky = this.props.sticky;

    if (this.context && prevProps.sticky !== sticky) {
      sticky ? this.context.registerStickyFooter() : this.context.unregisterStickyFooter();
    }
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    if (this.props.sticky) {
      var _this$context2;

      (_this$context2 = this.context) == null ? void 0 : _this$context2.unregisterStickyFooter();
    }
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        className = _this$props.className,
        prefixCls = _this$props.prefixCls,
        children = _this$props.children,
        sticky = _this$props.sticky,
        restProps = _objectWithoutPropertiesLoose(_this$props, ["className", "prefixCls", "children", "sticky"]);

    return /*#__PURE__*/React.createElement("footer", _extends({
      className: classnames(prefixCls, className, (_classnames = {}, _classnames[prefixCls + "-sticky"] = sticky, _classnames))
    }, restProps), children);
  };

  return LayoutFooter;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-layout-footer',
  sticky: false
}, _class2.contextType = Context, _temp)) || _class);
export default LayoutFooter;
