// @baidu/one-ui@4.32.1/lib/components/toast/common/message.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _componentClasses = _interopRequireDefault(require("component-classes"));

var _dlsIconsReact = require("dls-icons-react");

var _loading = _interopRequireDefault(require("../../loading"));

var _commonTools = require("../../../core/commonTools");

var minHeightMap = {
  small: 37,
  medium: 42
};

var getIconInfo = function getIconInfo(type, prefixCls) {
  var map = {
    info: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconInfoCircleSolid, {
      className: prefixCls + "-info"
    }),
    warning: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconExclamationCircleSolid, {
      className: prefixCls + "-warning"
    }),
    success: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconCheckCircleSolid, {
      className: prefixCls + "-success"
    }),
    error: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconTimesCircleSolid, {
      className: prefixCls + "-error"
    }),
    loading: /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-loading"
    }, /*#__PURE__*/_react["default"].createElement(_loading["default"], null))
  };
  return map[type];
};

var Message = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Message, _PureComponent);

  function Message() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.messageRef = void 0;

    _this.componentDidMount = function () {
      var _this$props = _this.props,
          size = _this$props.size,
          prefixCls = _this$props.prefixCls;
      var height = _this.messageRef.offsetHeight;
      var minHeight = minHeightMap[size];

      if (height > minHeight) {
        var dom = (0, _componentClasses["default"])(_this.messageRef);
        dom.add(prefixCls + "-multiple-line");
      }
    };

    _this.saveRef = function (ref) {
      _this.messageRef = ref;
    };

    _this.render = function () {
      var _classnames;

      var _this$props2 = _this.props,
          prefixCls = _this$props2.prefixCls,
          type = _this$props2.type,
          title = _this$props2.title,
          size = _this$props2.size,
          content = _this$props2.content,
          showCloseIcon = _this$props2.showCloseIcon,
          onClose = _this$props2.onClose,
          target = _this$props2.target,
          instance = _this$props2.instance;
      var wrapClass = (0, _commonTools.classnames)((_classnames = {}, _classnames[prefixCls + "-" + type] = type, _classnames[prefixCls + "-header"] = !!title, _classnames[prefixCls + "-widthout-header"] = !title, _classnames[prefixCls + "-has-close-icon"] = !!showCloseIcon, _classnames), prefixCls + "-custom-content", prefixCls + "-" + size);

      function removeNotice() {
        if (instance) {
          instance.removeNotice(target);
        }

        onClose();
      }

      ;
      var iconNode = getIconInfo(type, prefixCls);
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: wrapClass,
        ref: _this.saveRef
      }, /*#__PURE__*/_react["default"].createElement("span", {
        className: prefixCls + "-icon"
      }, iconNode), /*#__PURE__*/_react["default"].createElement("div", {
        className: prefixCls + "-container"
      }, title ? /*#__PURE__*/_react["default"].createElement("div", {
        className: prefixCls + "-container-header"
      }, title) : null, content ? /*#__PURE__*/_react["default"].createElement("div", {
        className: prefixCls + "-container-content"
      }, content) : null), showCloseIcon && /*#__PURE__*/_react["default"].createElement("span", {
        className: prefixCls + "-close-icon"
      }, /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconTimes, {
        className: prefixCls + "-close-icon-svg",
        onClick: removeNotice
      })));
    };

    return _this;
  }

  return Message;
}(_react.PureComponent);

exports["default"] = Message;
Message.defaultProps = {
  size: 'medium',
  prefixCls: 'one-toast',
  type: 'info'
};
module.exports = exports.default;
