// @baidu/one-ui@4.41.1/es/components/collapse/panel.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _noop from "lodash/noop";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import { IconChevronDown, IconChevronRight } from 'dls-icons-react';
import Content from './common/content';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
;
var CollapsePanel = (_dec = withConfigConsumer('collapse'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(CollapsePanel, _PureComponent);

  function CollapsePanel() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;

    _this.onItemClick = function () {
      var _ref = _this.props,
          onItemClick = _ref.onItemClick,
          panelKey = _ref.panelKey,
          disabled = _ref.disabled;

      if (disabled) {
        return;
      }

      onItemClick(panelKey);
    };

    _this.onKeyPress = function (e) {
      if (e.key === 'Enter' || e.keyCode === 13 || e.which === 13 || e.key === ' ' || e.which === 32 || e.keyCode === 32) {
        // tab和确认
        _this.onItemClick();
      }
    };

    return _this;
  }

  var _proto = CollapsePanel.prototype;

  _proto.render = function render() {
    var _classnames, _classnames2;

    var _ref2 = this.props,
        className = _ref2.className,
        style = _ref2.style,
        prefixCls = _ref2.prefixCls,
        header = _ref2.header,
        headerClass = _ref2.headerClass,
        children = _ref2.children,
        isActive = _ref2.isActive,
        disabled = _ref2.disabled,
        destroyNotActivePanel = _ref2.destroyNotActivePanel,
        accordion = _ref2.accordion,
        renderDomWhenHide = _ref2.renderDomWhenHide,
        contentClassName = _ref2.contentClassName,
        gutter = _ref2.gutter,
        showExpandIcon = _ref2.showExpandIcon;
    var headerCls = classnames(prefixCls + "-item-header", (_classnames = {}, _classnames[headerClass] = headerClass, _classnames));
    var itemCls = classnames(prefixCls + "-item", (_classnames2 = {}, _classnames2[prefixCls + "-item-active"] = isActive, _classnames2[prefixCls + "-item-not-active"] = !isActive, _classnames2[prefixCls + "-item-disabled"] = disabled, _classnames2), className);
    var iconNode = isActive ? /*#__PURE__*/React.createElement(IconChevronDown, {
      className: prefixCls + "-angle-icon"
    }) : /*#__PURE__*/React.createElement(IconChevronRight, {
      className: prefixCls + "-angle-icon"
    });
    var itemStyle = style;

    if (gutter > 0) {
      itemStyle = _extends({}, style, {
        marginBottom: gutter
      });
    }

    return /*#__PURE__*/React.createElement("div", {
      className: itemCls,
      style: itemStyle
    }, /*#__PURE__*/React.createElement("div", {
      className: headerCls,
      onClick: this.onItemClick,
      onKeyPress: this.onKeyPress,
      "data-type": accordion ? 'accordion' : 'normal',
      tabIndex: disabled ? null : 0
    }, showExpandIcon && iconNode, /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-item-title"
    }, header || null)), /*#__PURE__*/React.createElement(Content, {
      prefixCls: prefixCls,
      isActive: isActive,
      destroyNotActivePanel: destroyNotActivePanel,
      renderDomWhenHide: renderDomWhenHide,
      contentClassName: contentClassName
    }, children));
  };

  return CollapsePanel;
}(PureComponent), _class2.defaultProps = {
  isActive: false,
  onItemClick: _noop,
  headerClass: '',
  prefixCls: 'one-collapse',
  disabled: false,
  destroyNotActivePanel: false,
  accordion: false,
  renderDomWhenHide: false,
  contentClassName: '',
  bordered: true,
  expandIconPosition: 'left'
}, _temp)) || _class);
export default CollapsePanel;
