// @baidu/one-ui@4.41.1/es/components/drawer/drawer.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/extends";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { Component } from 'react';
import createReactContext from 'create-react-context';
import { IconTimes } from 'dls-icons-react';
import Button from '../button';
import DrawerWrapper from './common/drawerWrapper';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
var DrawerContext = createReactContext(null);
;
var Drawer = (_dec = withConfigConsumer('drawer'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Drawer, _Component);

  function Drawer(props) {
    var _this;

    _this = _Component.call(this, props) || this;
    _this.parentDrawer = void 0;

    _this.componentDidUpdate = function (prevProps) {
      if (prevProps.visible !== _this.props.visible && _this.parentDrawer) {
        if (_this.props.visible) {
          _this.parentDrawer.push();
        } else {
          _this.parentDrawer.pull();
        }
      }
    };

    _this.onMaskClick = function (e) {
      if (!_this.props.maskClosable) {
        return;
      }

      _this.close(e);
    };

    _this.getDestoryOnClose = function () {
      return _this.props.destroyOnClose && !_this.props.visible;
    };

    _this.getPushTransform = function (placement) {
      if (placement === 'left' || placement === 'right') {
        return "translateX(" + (placement === 'left' ? 180 : -180) + "px)";
      }

      if (placement === 'top' || placement === 'bottom') {
        return "translateY(" + (placement === 'top' ? 180 : -180) + "px)";
      }
    };

    _this.getDrawerWrapperStyle = function () {
      var _this$props = _this.props,
          zIndex = _this$props.zIndex,
          placement = _this$props.placement,
          maskStyle = _this$props.maskStyle;
      return _this.state.push ? _extends({}, maskStyle, {
        zIndex: zIndex,
        transform: _this.getPushTransform(placement)
      }) : _extends({}, maskStyle, {
        zIndex: zIndex
      });
    };

    _this.close = function (e) {
      if (_this.props.visible !== undefined) {
        if (_this.props.onClose) {
          _this.props.onClose(e);
        }
      }
    };

    _this.push = function () {
      _this.setState({
        push: true
      });
    };

    _this.pull = function () {
      _this.setState({
        push: false
      });
    };

    _this.renderBody = function () {
      var _classnames, _classnames2;

      var isDestroyOnClose = _this.getDestoryOnClose();

      if (isDestroyOnClose) {
        return null;
      }

      var _this$props2 = _this.props,
          prefixCls = _this$props2.prefixCls,
          title = _this$props2.title,
          closable = _this$props2.closable,
          footer = _this$props2.footer,
          placement = _this$props2.placement,
          onOk = _this$props2.onOk,
          onCancel = _this$props2.onCancel,
          okText = _this$props2.okText,
          cancelText = _this$props2.cancelText,
          okProps = _this$props2.okProps,
          cancelProps = _this$props2.cancelProps,
          hideDefaultFooter = _this$props2.hideDefaultFooter,
          size = _this$props2.size,
          type = _this$props2.type,
          headerClassName = _this$props2.headerClassName;
      var containerStyle = placement === 'left' || placement === 'right' ? {
        overflow: 'auto',
        height: '100%'
      } : {}; // is have closer button

      var closer;

      if (closable) {
        closer = /*#__PURE__*/React.createElement(Button, {
          type: "text-aux",
          size: size,
          className: prefixCls + "-close",
          onClick: _this.close
        }, /*#__PURE__*/React.createElement(IconTimes, {
          className: prefixCls + "-close-icon"
        }));
      }

      var header;

      if (title) {
        header = /*#__PURE__*/React.createElement("div", {
          className: classnames(prefixCls + "-header", headerClassName)
        }, /*#__PURE__*/React.createElement("div", {
          className: prefixCls + "-header-container"
        }, /*#__PURE__*/React.createElement("div", {
          className: prefixCls + "-title"
        }, title)), closer);
      }

      var drawerFooter = !hideDefaultFooter ? [/*#__PURE__*/React.createElement(Button, _extends({
        key: "confirm",
        type: "primary",
        size: size,
        onClick: onOk
      }, okProps), okText), /*#__PURE__*/React.createElement(Button, _extends({
        key: "cancel",
        onClick: onCancel,
        size: size
      }, cancelProps), cancelText)] : null;

      if (footer && footer.length) {
        drawerFooter = footer;
      }

      var footerDom = null;

      if (drawerFooter) {
        footerDom = /*#__PURE__*/React.createElement("div", {
          className: prefixCls + "-footer"
        }, drawerFooter);
      }

      var bodyClass = classnames(prefixCls + "-body", (_classnames = {}, _classnames[prefixCls + "-body-" + placement] = true, _classnames[prefixCls + "-body-has-footer"] = footer && footer.length, _classnames));
      var drawerBodyClass = classnames(prefixCls + "-wrapper-body", prefixCls + "-wrapper-body-" + size, (_classnames2 = {}, _classnames2[prefixCls + "-wrapper-with-header"] = header, _classnames2[prefixCls + "-wrapper-with-close"] = closer, _classnames2[prefixCls + "-wrapper-type-basic"] = type === 'basic', _classnames2));
      return /*#__PURE__*/React.createElement("div", {
        className: drawerBodyClass,
        style: containerStyle
      }, /*#__PURE__*/React.createElement("div", {
        className: prefixCls + "-wrapper-body-container"
      }, header ? header : closer, /*#__PURE__*/React.createElement("div", {
        className: bodyClass,
        style: _this.props.style
      }, _this.props.children), footerDom));
    };

    _this.renderProvider = function (value) {
      var _this$props3 = _this.props,
          zIndex = _this$props3.zIndex,
          style = _this$props3.style,
          placement = _this$props3.placement,
          className = _this$props3.className,
          width = _this$props3.width,
          height = _this$props3.height,
          visible = _this$props3.visible,
          mask = _this$props3.mask,
          destroyOnClose = _this$props3.destroyOnClose,
          closable = _this$props3.closable,
          title = _this$props3.title,
          cancelProps = _this$props3.cancelProps,
          okProps = _this$props3.okProps,
          okText = _this$props3.okText,
          cancelText = _this$props3.cancelText,
          onOk = _this$props3.onOk,
          onCancel = _this$props3.onCancel,
          hideDefaultFooter = _this$props3.hideDefaultFooter,
          footer = _this$props3.footer,
          type = _this$props3.type,
          rest = _objectWithoutPropertiesLoose(_this$props3, ["zIndex", "style", "placement", "className", "width", "height", "visible", "mask", "destroyOnClose", "closable", "title", "cancelProps", "okProps", "okText", "cancelText", "onOk", "onCancel", "hideDefaultFooter", "footer", "type"]);

      var haveMask = mask ? '' : 'no-mask';
      _this.parentDrawer = value;
      var offsetStyle = {};

      if (placement === 'left' || placement === 'right') {
        offsetStyle.width = width;
      } else {
        offsetStyle.height = height;
      }

      return /*#__PURE__*/React.createElement(DrawerContext.Provider, {
        value: _assertThisInitialized(_this)
      }, /*#__PURE__*/React.createElement(DrawerWrapper, _extends({
        level: null,
        handler: false
      }, rest, offsetStyle, {
        open: visible,
        onClose: _this.onMaskClick,
        showMask: mask,
        placement: placement,
        style: _this.getDrawerWrapperStyle(),
        className: classnames(className, haveMask)
      }), _this.renderBody()));
    };

    _this.state = {
      push: false
    };
    return _this;
  }

  var _proto = Drawer.prototype;

  _proto.render = function render() {
    return /*#__PURE__*/React.createElement(DrawerContext.Consumer, null, this.renderProvider);
  };

  return Drawer;
}(Component), _class2.defaultProps = {
  prefixCls: 'one-drawer',
  width: 400,
  height: 256,
  closable: true,
  placement: 'right',
  maskClosable: true,
  mask: true,
  footer: null,
  hideDefaultFooter: true,
  onOk: function onOk() {},
  onCancel: function onCancel() {},
  okText: '确定',
  cancelText: '取消',
  okProps: {},
  cancelProps: {},
  size: 'medium',
  type: 'default',
  closeDrawerByClickBody: false
}, _temp)) || _class);
export default Drawer;
