// @baidu/one-ui@4.32.1/es/components/table/columnTitle.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

/**
 * @file column title
 * @author tangjinhua
 * @date 2021-09-17
 */
import React, { PureComponent } from 'react';
import Popover from '../popover';
;

var ColumnTitle = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(ColumnTitle, _PureComponent);

  function ColumnTitle() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.state = {
      showDesc: true
    };

    _this.hide = function () {
      _this.setState({
        showDesc: false
      });
    };

    _this.show = function () {
      _this.setState({
        showDesc: true
      });
    };

    return _this;
  }

  var _proto = ColumnTitle.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        desc = _this$props.desc,
        _this$props$title = _this$props.title,
        title = _this$props$title === void 0 ? null : _this$props$title;

    if (desc) {
      var content = desc;

      if ( /*#__PURE__*/React.isValidElement(desc)) {
        content = /*#__PURE__*/React.cloneElement(desc, {
          hide: this.hide,
          show: this.show
        });
      }

      return /*#__PURE__*/React.createElement(Popover, {
        content: content,
        size: "small",
        overlayStyle: {
          display: this.state.showDesc ? '' : 'none'
        }
      }, title);
    }

    return title;
  };

  return ColumnTitle;
}(PureComponent);

export { ColumnTitle as default };
