// @baidu/one-ui@4.32.1/src/components/tree/interface.ts
import {ReactElement, ReactNode} from "react";
import {BasePropsWithChildren} from "../interface"

export interface TreeProps extends BasePropsWithChildren {

    /**
     * 禁用
     */
    disabled?: boolean;

    /**
     * 尺寸
     */
    size?: 'small' | 'medium';

    /**
     * 是否展示多选框
     */
    checkable?: boolean;

    /**
     * 选项独立
     */
    checkStrictly?: boolean;

    /**
     * 默认已勾选项
     */
    defaultCheckedKeys?: string[] | {
        checked: string[];
        halfChecked?: string[];
    };

    /**
     * 已勾选项
     */
    checkedKeys?: string[] | {
        checked: string[];
        halfChecked?: string[];
    };

    /**
     * 勾选回调
     */
    onCheck?: (checkedKeys: string[], checkInfo: TreeNodeCheckInfo) => void;

    /**
     * 是否展示icon
     */
    showIcon?: boolean;

    /**
     * 自定义icon
     */
    icon?: ReactElement;

    /**
     * 默认展开全部
     */
    defaultExpandAll?: boolean;

    /**
     * 默认展开项
     */
    defaultExpandedKeys?: string[];

    /**
     * 展开项
     */
    expandedKeys?: string[];

    /**
     * 自动展开父级
     */
    autoExpandParent?: boolean;

    /**
     * 自动展开父级
     */
    defaultExpandParent?: boolean;

    /**
     * 展开回调
     */
    onExpand?: (expandedKeys: string[], info: {
        node: ReactElement;
        expanded: boolean;
        nativeEvent: MouseEvent;
    }) => void;

    /**
     * 默认已选项
     */
    defaultSelectedKeys?: string[];

    /**
     * 已选项
     */
    selectedKeys?: string[];

    /**
     * 是否多选
     */
    multiple?: boolean;

    /**
     * 是否可选
     */
    selectable?: boolean;

    /**
     * 选择回调
     */
    onSelect?: (selectedKeys: string[], selectInfo: TreeNodeSelectInfo) => void;

    /**
     * 加载数据
     */
    loadData?: (node, key: string, type: 'check' | 'expand' | 'select') => Promise<void>;

    /**
     * load完毕回调
     */
    onLoad?: (loadedKeys: string[], info) => void;

    /**
     * 已加载的key
     */
    loadedKeys?: string[];

    /**
     * 筛选节点
     */
    filterTreeNode?: (treeNode) => string | boolean;

    /**
     * 切换Icon
     */
    switcherIcon?: ReactElement | ((option: {isLeaf: boolean, loading: boolean}) => ReactElement);

    /**
     * 容器高度
     */
    parentContainerHeight?: number;

    /**
     * 节点高度
     */
    treeNodeHeight?: number;

    /**
     * 点击
     */
    onClick?: (e, treeNode) => void;

    /**
     * 双击
     */
    onDoubleClick?: (e, treeNode) => void;

    /**
     * MouseEnter
     */
    onMouseEnter?: (data: {event, node}) => void;

    /**
     * MouseLeave
     */
    onMouseLeave?: (data: {event, node}) => void;

    /**
     * RightClick
     */
    onRightClick?: (data: {event, node}) => void;

    /**
     * select同时进行check
     */
    checkActionEqualSelet?: boolean;

    /**
     * 滚动触发
     */
    onScroll?: (e) => void;

    /**
     * @deprecated
     */
    treeData?;
};

export interface TreeNodeProps extends BasePropsWithChildren {
    /**
     * 唯一标识
     */
    key?: string;

    /**
     * 标题
     */
    title?: ReactNode;

    /**
     * 禁用
     */
    disabled?: boolean;

    /**
     * disable时，禁止所有事件，如：点击
     */
    disabledAllWhenNodeDisabled?: boolean;

    /**
     * 可勾选
     */
    checkable?: boolean;

    /**
     * 禁用勾选框
     */
    disableCheckbox?: boolean;

    /**
     * 可选
     */
    selectable?: boolean;

    /**
     * 自定义图标
     */
    icon?: ReactElement | (() => ReactElement);

    /**
     * 是否为叶子节点
     */
    isLeaf?: boolean;

    /**
     * @internal
     */
    switcherIcon?: ReactElement | (() => ReactElement);

    /**
     * @internal
     */
    eventKey?: string;

    /**
     * @internal
     */
    expanded?: boolean;

    /**
     * @internal
     */
    selected?: boolean;

    /**
     * @internal
     */
    checked?: boolean;

    /**
     * @internal
     */
    loaded?: boolean;

    /**
     * @internal
     */
    loading?: boolean;

    /**
     * @internal
     */
    halfChecked?: boolean;

    /**
     * @internal
     */
    pos?: string;

    /**
     * @internal
     */
    size?: 'small' | 'medium';

    /**
     * @internal
     */
    level?: number;
};

export interface TreeNodeCheckInfo {

    /**
     * key
     */
    eventKey: string;

    /**
     * 勾选状态
     */
    checked: boolean;

    /**
     * @deprecated
     */
    node;

    /**
     * @deprecated
     */
    nativeEvent;

    /**
     * @deprecated
     */
    keyEntities;

    /**
     * @deprecated
     */
    event: 'check';

    /**
     * @deprecated
     */
    checkedNodes?;

    /**
     * @deprecated
     */
    checkedNodesPositions?;

    /**
     * @deprecated
     */
    halfCheckedKeys?: string[];
}

export interface TreeNodeSelectInfo {
    /**
     * key
     */
    eventKey: string;

    /**
     * 选择状态
     */
    selected: boolean;

    /**
     * @deprecated
     */
    node;

    /**
     * @deprecated
     */
    selectedNodes;

    /**
     * @deprecated
     */
    nativeEvent;

    /**
     * @deprecated
     */
    keyEntities;

    /**
     * @deprecated
     */
    event: 'select';

    /**
     * @deprecated
     */
    checkedKeys?: string[];

    /**
     * @deprecated
     */
    halfCheckedKeys?: string[];
}
