// @baidu/one-ui@4.41.1/es/components/tree/tree.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

/**
 * @file tree main
 * @author huangshiming
 * @date 2020-05-18
 */
import React, { isValidElement, PureComponent } from 'react';
import { classnames } from '../../core/commonTools';
import InnerVirtualTree from './common/virtualTree';
import InnerTree from './common/tree';
import { IconCaretRight } from 'dls-icons-react';
import IconLoading from '../loading';
import { withConfigConsumer } from '../providerConfig/context';
var Tree = (_dec = withConfigConsumer('tree'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Tree, _PureComponent);

  function Tree() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.treeRef = void 0;

    _this.renderSwitcherIcon = function (_ref) {
      var isLeaf = _ref.isLeaf,
          loading = _ref.loading;
      var _this$props = _this.props,
          prefixCls = _this$props.prefixCls,
          switcherIcon = _this$props.switcherIcon;

      if (loading) {
        return /*#__PURE__*/React.createElement(IconLoading, {
          size: "small",
          className: prefixCls + "-switcher-loading-icon"
        });
      }

      if (isLeaf) {
        return null;
      }

      var switcherCls = prefixCls + "-switcher-icon";

      if ( /*#__PURE__*/isValidElement(switcherIcon)) {
        var switcherOriginCls = switcherIcon.props && switcherIcon.props.className || '';
        return /*#__PURE__*/React.cloneElement(switcherIcon, {
          className: classnames(switcherOriginCls, switcherCls)
        });
      }

      return /*#__PURE__*/React.createElement(IconCaretRight, {
        className: prefixCls + "-node-switch-arrow"
      });
    };

    _this.setTreeRef = function (node) {
      _this.treeRef = node;
    };

    _this.getTreeInterState = function () {
      return _this.treeRef.state;
    };

    return _this;
  }

  var _proto = Tree.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props2 = this.props,
        prefixCls = _this$props2.prefixCls,
        className = _this$props2.className,
        showIcon = _this$props2.showIcon,
        size = _this$props2.size,
        switcherIcon = _this$props2.switcherIcon,
        parentContainerHeight = _this$props2.parentContainerHeight,
        children = _this$props2.children;
    var TreeComponent = !parentContainerHeight ? InnerTree : InnerVirtualTree;
    return /*#__PURE__*/React.createElement(TreeComponent, _extends({}, this.props, {
      ref: this.setTreeRef,
      className: classnames(className, (_classnames = {}, _classnames[prefixCls + "-icon-hide"] = !showIcon, _classnames), prefixCls + "-" + size),
      switcherIcon: typeof switcherIcon === 'function' ? switcherIcon : this.renderSwitcherIcon
    }), children);
  };

  return Tree;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-tree',
  checkable: false,
  showIcon: false,
  checkStrictly: false,
  size: 'medium'
}, _class2.TreeNode = void 0, _class2.VirtualTreeNode = void 0, _temp)) || _class);
export default Tree;
