// @baidu/one-ui@4.32.1/src/components/backTop/backTop.tsx
import React, {PureComponent} from 'react';
import {IconArrowToTop} from 'dls-icons-react';
import {noop} from 'lodash';
import Button from '../button';
import {classnames} from '../../core/commonTools';
import {BackTopProps} from './interface';
import {withConfigConsumer} from '../providerConfig/context';

const defaultVisibilityHeight = 400;

@withConfigConsumer('back-top')
class BackTop extends PureComponent<BackTopProps> {

    static defaultProps = {
        onClick: noop,
        visibilityHeight: defaultVisibilityHeight,
        prefixCls: 'one-back-top',
        className: ''
    }

    state = {
        visible : false
    };

    scrollEvent;

    static getDerivedStateFromProps = (nextProps, prevState) => {
        if ('visible' in nextProps
        && nextProps.visible !== prevState.visible) {
            return {
                visible: nextProps.visible
            };
        }
        return null;
    }

    componentDidMount() {
        const getTarget = this.props.target || this.getDefaultTarget;
        const target = getTarget();
        if (target && target.addEventListener && typeof target.addEventListener === 'function') {
            this.scrollEvent = target.addEventListener('scroll', this.handleScroll);
            this.handleScroll();
        }
    }

    componentWillUnmount() {
        if (this.scrollEvent) {
            // 初始化清除上一次监听的滚动元素
            this.scrollEvent.remove();
        }
    }

    onVisibleChange = visible => {
        if (!('visible' in this.props)) {
            this.setState({
                visible
            });
        }
    }

    getDefaultTarget = () => window;

    getScroll = target => {
        // 获取滚动高度
        if (typeof window === 'undefined') {
            return 0;
        }

        const prop = 'pageYOffset';
        const method = 'scrollTop';
        const isWindow = target === window;
        let ret = isWindow ? target[prop] : target[method];
        if (isWindow && typeof ret !== 'number') {
            ret = document.documentElement[method];
        }
        return ret;
    }

    handleScroll = () => {
        const {visibilityHeight, target = this.getDefaultTarget} = this.props;
        const scrollTop = this.getScroll(target());
        const visible = scrollTop > visibilityHeight;
        this.onVisibleChange(visible);
    };

    // 回到顶部
    scrollToTop = () => {
        const getTarget = this.props.target || this.getDefaultTarget;
        const targetNode = getTarget();
        if ('scrollTop' in targetNode) {
            targetNode.scrollTop = 0;
        }
        else {
            document.body.scrollTop = 0;
            document.documentElement.scrollTop = 0;
        }

        this.props.onClick();
    };

    render() {
        const {prefixCls, className, target} = this.props;
        const classes = classnames(prefixCls, className, {
            [`${prefixCls}-default`]: !target
        });
        return this.state.visible
            ? (
                <Button className={classes} onClick={this.scrollToTop} type="translucent" size="xsmall">
                    <IconArrowToTop className={`${prefixCls}-icon`} />
                </Button>
            )
            : null;
    }
}

export default BackTop;
