// @baidu/one-ui@4.32.1/lib/components/transfer/common/candidateFooterRender.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _button = _interopRequireDefault(require("../../button"));

;

var CommonFooterRender = function CommonFooterRender(props) {
  var candidateFooterProps = props.candidateFooterProps,
      treeName = props.treeName,
      size = props.size,
      prefixCls = props.prefixCls,
      candidateFooterRender = props.candidateFooterRender;

  if (typeof candidateFooterRender === 'function') {
    return candidateFooterRender(props);
  } else if (typeof candidateFooterRender === 'object') {
    return candidateFooterRender;
  }

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: prefixCls + "-select-footer"
  }, /*#__PURE__*/_react["default"].createElement(_button["default"], (0, _extends2["default"])({
    type: "text-strong",
    size: size
  }, candidateFooterProps), "+ \u65B0\u5EFA", treeName));
};

CommonFooterRender.defaultProps = {
  treeName: '',
  candidateFooterProps: {},
  size: 'medium'
};
var _default = CommonFooterRender;
exports["default"] = _default;
module.exports = exports.default;
