// @baidu/one-ui@4.32.1/lib/components/transfer/common/searchBoxRender.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _select = _interopRequireDefault(require("../../select"));

var _searchBox = _interopRequireDefault(require("../../searchBox"));

;
var Option = _select["default"].Option;

var CommonSearchRender = function CommonSearchRender(props) {
  var levelOptions = props.levelOptions,
      handleLevelChange = props.handleLevelChange,
      searchBoxProps = props.searchBoxProps,
      prefixCls = props.prefixCls,
      isShowLevelSelect = props.isShowLevelSelect,
      levelKey = props.levelKey,
      onSelectFocus = props.onSelectFocus,
      SearchBoxRender = props.SearchBoxRender,
      customSearchProps = (0, _objectWithoutPropertiesLoose2["default"])(props, ["levelOptions", "handleLevelChange", "searchBoxProps", "prefixCls", "isShowLevelSelect", "levelKey", "onSelectFocus", "SearchBoxRender"]);
  var selectProps = {
    mode: 'default',
    className: prefixCls + "-category"
  };

  if (levelKey != null) {
    selectProps.value = levelKey;
  }

  if (SearchBoxRender && typeof SearchBoxRender === 'function') {
    return /*#__PURE__*/_react["default"].createElement(SearchBoxRender, props);
  } else if ( /*#__PURE__*/(0, _react.isValidElement)(SearchBoxRender)) {
    return SearchBoxRender;
  }

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: prefixCls + "-search-box-bar"
  }, isShowLevelSelect ? /*#__PURE__*/_react["default"].createElement(_select["default"], (0, _extends2["default"])({
    defaultValue: levelKey || levelOptions[0] && levelOptions[0].value,
    onChange: handleLevelChange,
    onFocus: onSelectFocus
  }, selectProps), levelOptions.map(function (option) {
    return /*#__PURE__*/_react["default"].createElement(Option, {
      value: option.value,
      key: option.value
    }, option.label);
  })) : null, /*#__PURE__*/_react["default"].createElement(_searchBox["default"], (0, _extends2["default"])({}, searchBoxProps, customSearchProps)));
};

CommonSearchRender.defaultProps = {
  levelOptions: null,
  handleLevelChange: function handleLevelChange() {},
  searchBoxProps: {},
  isShowLevelSelect: false,
  levelKey: null,
  onSelectFocus: function onSelectFocus() {},
  prefixCls: 'one-transfer'
};
var _default = CommonSearchRender;
exports["default"] = _default;
module.exports = exports.default;
