// @baidu/one-ui@4.32.1/lib/components/dialog/actionButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _button = _interopRequireDefault(require("../button"));

;
;

var ActionButton = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(ActionButton, _Component);

  function ActionButton(props) {
    var _this;

    _this = _Component.call(this, props) || this;

    _this.onClick = function () {
      var _this$props = _this.props,
          actionFn = _this$props.actionFn,
          closeModal = _this$props.closeModal,
          triggerCancel = _this$props.triggerCancel;

      if (actionFn) {
        var ret;

        if (actionFn.length) {
          ret = actionFn(closeModal);
        } else {
          ret = actionFn();

          if (!ret) {
            closeModal({
              triggerCancel: triggerCancel
            });
          }
        }

        if (ret && ret.then) {
          _this.setState({
            loading: true
          });

          ret.then(function () {
            closeModal({
              triggerCancel: triggerCancel
            });
          }, function () {
            _this.setState({
              loading: false
            });
          });
        }
      } else {
        closeModal({
          triggerCancel: triggerCancel
        });
      }
    };

    _this.state = {
      loading: false
    };
    return _this;
  }

  var _proto = ActionButton.prototype;

  _proto.render = function render() {
    var _this$props2 = this.props,
        children = _this$props2.children,
        order = _this$props2.order,
        style = _this$props2.style,
        otherProps = _this$props2.otherProps,
        actionFn = _this$props2.actionFn,
        closeModal = _this$props2.closeModal,
        triggerCancel = _this$props2.triggerCancel,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props2, ["children", "order", "style", "otherProps", "actionFn", "closeModal", "triggerCancel"]);
    var loading = this.state.loading;
    return /*#__PURE__*/_react["default"].createElement(_button["default"], (0, _extends2["default"])({}, restProps, otherProps, {
      style: (0, _extends2["default"])({}, style, otherProps.style, {
        order: order
      }),
      onClick: this.onClick,
      loading: loading
    }), children);
  };

  return ActionButton;
}(_react.Component);

exports["default"] = ActionButton;
ActionButton.defaultProps = {
  otherProps: {},
  triggerCancel: true
};
module.exports = exports.default;
