// @baidu/one-ui@4.32.1/lib/components/dialog/confirm.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = confirm;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _dlsIconsReact = require("dls-icons-react");

var _dialog = _interopRequireDefault(require("./dialog"));

var _actionButton = _interopRequireDefault(require("./actionButton"));

var _commonTools = require("../../core/commonTools");

function confirm(config) {
  var _classnames2;

  var _config$prefixCls = config.prefixCls,
      prefixCls = _config$prefixCls === void 0 ? 'one-dialog-confirm' : _config$prefixCls,
      _config$buttonPositio = config.buttonPosition,
      buttonPosition = _config$buttonPositio === void 0 ? 'left' : _config$buttonPositio,
      _config$okCancel = config.okCancel,
      okCancel = _config$okCancel === void 0 ? true : _config$okCancel,
      content = config.content,
      okType = config.okType,
      onOk = config.onOk,
      _config$okProps = config.okProps,
      okProps = _config$okProps === void 0 ? {} : _config$okProps,
      _config$cancelProps = config.cancelProps,
      cancelProps = _config$cancelProps === void 0 ? {} : _config$cancelProps,
      onCancel = config.onCancel,
      _config$size = config.size,
      size = _config$size === void 0 ? 'medium' : _config$size,
      okText = config.okText,
      cancelType = config.cancelType,
      cancelText = config.cancelText,
      type = config.type,
      _config$className = config.className,
      className = _config$className === void 0 ? '' : _config$className,
      _config$title = config.title,
      title = _config$title === void 0 ? '提示' : _config$title,
      _config$buttonSize = config.buttonSize,
      buttonSize = _config$buttonSize === void 0 ? 'medium' : _config$buttonSize,
      _config$width = config.width,
      width = _config$width === void 0 ? 'small' : _config$width,
      okOrder = config.okOrder,
      cancelOrder = config.cancelOrder,
      icon = config.icon,
      iconType = config.iconType,
      _config$needCloseIcon = config.needCloseIcon,
      needCloseIcon = _config$needCloseIcon === void 0 ? false : _config$needCloseIcon;
  var maskClosable = config.maskClosable;
  var div = document.createElement('div');
  document.body.appendChild(div); // 默认为 false，保持旧版默认行为

  maskClosable = maskClosable === undefined ? false : maskClosable;

  var close = function close(params) {
    if (params === void 0) {
      params = {};
    }

    var unmountResult = _reactDom["default"].unmountComponentAtNode(div);

    if (unmountResult && div.parentNode) {
      div.parentNode.removeChild(div);
    }

    var triggerCancel = params.triggerCancel;

    if (onCancel && triggerCancel) {
      onCancel();
    }
  };

  var body = /*#__PURE__*/_react["default"].createElement("div", {
    className: prefixCls + "-body"
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: prefixCls + "-content"
  }, content));

  var footer = null;
  var okButtonProps = (0, _extends2["default"])({
    type: okType,
    actionFn: onOk,
    closeModal: close,
    autoFocus: true,
    order: okOrder,
    size: buttonSize,
    triggerCancel: false
  }, okProps);

  if (config.footer) {
    footer = config.footer;
  } else if (okCancel) {
    footer = [/*#__PURE__*/_react["default"].createElement(_actionButton["default"], (0, _extends2["default"])({
      key: "confirm"
    }, okButtonProps), okText), /*#__PURE__*/_react["default"].createElement(_actionButton["default"], (0, _extends2["default"])({
      key: "cancel",
      type: cancelType,
      closeModal: close,
      order: cancelOrder || null,
      size: buttonSize
    }, cancelProps), cancelText)];

    if (buttonPosition === 'right') {
      footer = [footer[1], footer[0]];
    }
  } else {
    footer = [/*#__PURE__*/_react["default"].createElement(_actionButton["default"], (0, _extends2["default"])({
      key: "confirm"
    }, okButtonProps), okText)];
  }

  var $title = title;
  var $icon = null;

  if (icon) {
    $icon = icon;
  } else if (iconType) {
    switch (iconType) {
      case 'success':
        $icon = /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconCheckCircle, null);
        break;

      case 'warning':
        $icon = /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconExclamationCircle, null);
        break;

      case 'info':
        $icon = /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconInfoCircle, null);
        break;

      case 'fail':
        $icon = /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconTimesCircle, null);
        break;

      default:
        break;
    }
  }

  if ($icon) {
    var _classnames;

    $title = /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-title-icon"
    }, /*#__PURE__*/_react["default"].createElement("span", {
      className: (0, _commonTools.classnames)(prefixCls + "-title-icon-container", (_classnames = {}, _classnames[prefixCls + "-title-icon-" + iconType] = !!iconType, _classnames[prefixCls + "-title-icon-custom"] = !!icon, _classnames))
    }, $icon), $title);
  }

  var classString = (0, _commonTools.classnames)(prefixCls, (_classnames2 = {}, _classnames2[prefixCls + "-" + type] = true, _classnames2[prefixCls + "-has-icon"] = !!$icon, _classnames2[prefixCls + "-order-reverse"] = cancelOrder && okOrder && cancelOrder < okOrder, _classnames2), className);

  _reactDom["default"].render( /*#__PURE__*/_react["default"].createElement(_dialog["default"], (0, _extends2["default"])({}, config, {
    className: classString,
    onCancel: close,
    visible: true,
    title: $title,
    maskClosable: maskClosable,
    size: size,
    footer: footer,
    width: width,
    buttonPosition: buttonPosition,
    needCloseIcon: needCloseIcon
  }), /*#__PURE__*/_react["default"].createElement("div", null, body)), div);

  return {
    destroy: close
  };
}

module.exports = exports.default;
