// @baidu/one-ui@4.32.1/lib/components/switch/common/switch.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _loading = _interopRequireDefault(require("../../loading"));

var _commonTools = require("../../../core/commonTools");

;

var InnerSwitch = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(InnerSwitch, _PureComponent);

  function InnerSwitch(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.switchRef = void 0;

    _this.onHandleClick = function (e) {
      var checked = _this.state.checked;
      var onClick = _this.props.onClick;
      var newChecked = !checked;

      _this.setChecked(newChecked, e);

      if (onClick) {
        onClick(newChecked, e);
      }
    };

    _this.onHandleKeyDown = function (e) {
      if (e.keyCode === 37) {
        // Left
        _this.setChecked(false, e);
      } else if (e.keyCode === 39) {
        // Right
        _this.setChecked(true, e);
      }
    };

    _this.saveRef = function (ref) {
      _this.switchRef = ref;
    };

    var _checked = false;

    if ('checked' in props) {
      _checked = !!props.checked;
    } else {
      _checked = !!props.defaultChecked;
    }

    _this.state = {
      checked: _checked
    };
    return _this;
  }

  var _proto = InnerSwitch.prototype;

  _proto.setChecked = function setChecked(checked, e) {
    var _this$props = this.props,
        disabled = _this$props.disabled,
        onChange = _this$props.onChange,
        loading = _this$props.loading;

    if (disabled || loading) {
      return;
    }

    if (!('checked' in this.props)) {
      this.setState({
        checked: checked
      });
    }

    if (onChange) {
      onChange(checked, e);
    }
  };

  _proto.focus = function focus() {
    if (this.switchRef) {
      this.switchRef.focus();
    }
  };

  _proto.blur = function blur() {
    if (this.switchRef) {
      this.switchRef.blur();
    }
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props2 = this.props,
        className = _this$props2.className,
        prefixCls = _this$props2.prefixCls,
        disabled = _this$props2.disabled,
        loadingIcon = _this$props2.loadingIcon,
        loading = _this$props2.loading,
        onChange = _this$props2.onChange,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props2, ["className", "prefixCls", "disabled", "loadingIcon", "loading", "onChange"]);
    var checked = this.state.checked;
    var switchClassName = (0, _commonTools.classnames)(prefixCls, (_classnames = {}, _classnames[prefixCls + "-checked"] = checked, _classnames[prefixCls + "-disabled"] = disabled || loading, _classnames), className);
    return /*#__PURE__*/_react["default"].createElement("button", (0, _extends2["default"])({}, restProps, {
      type: "button",
      role: "switch",
      "aria-checked": checked,
      disabled: disabled,
      className: switchClassName,
      onKeyDown: this.onHandleKeyDown,
      onClick: this.onHandleClick,
      ref: this.saveRef
    }), /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-thumb"
    }, loading ? loadingIcon || /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-loading-icon"
    }, /*#__PURE__*/_react["default"].createElement(_loading["default"], null)) : null));
  };

  return InnerSwitch;
}(_react.PureComponent);

InnerSwitch.defaultProps = {
  className: '',
  defaultChecked: false
};

InnerSwitch.getDerivedStateFromProps = function (nextProps) {
  var newState = {};

  if ('checked' in nextProps) {
    newState.checked = !!nextProps.checked;
  }

  return newState;
};

var _default = InnerSwitch;
exports["default"] = _default;
module.exports = exports.default;
