// @baidu/one-ui@4.32.1/src/components/dialog/confirm.tsx
import React from 'react';
import ReactDOM from 'react-dom';
import {
    IconCheckCircle,
    IconExclamationCircle,
    IconInfoCircle,
    IconTimesCircle
} from 'dls-icons-react';
import Dialog from './dialog';
import ActionButton from './actionButton';
import {classnames} from '../../core/commonTools';
import {DialogConfirmProps, DialogDestroyable} from './interface';

export default function confirm(config: DialogConfirmProps): DialogDestroyable {
    const {
        prefixCls = 'one-dialog-confirm',
        buttonPosition = 'left',
        okCancel = true,
        content,
        okType,
        onOk,
        okProps = {},
        cancelProps = {},
        onCancel,
        size = 'medium',
        okText,
        cancelType,
        cancelText,
        type,
        className = '',
        title = '提示',
        buttonSize = 'medium',
        width = 'small',
        okOrder,
        cancelOrder,
        icon,
        iconType,
        needCloseIcon = false
    } = config;
    let {
        maskClosable
    } = config;
    const div = document.createElement('div');
    document.body.appendChild(div);
    // 默认为 false，保持旧版默认行为
    maskClosable = maskClosable === undefined ? false : maskClosable;

    const close = (params: {triggerCancel?: boolean} = {}) => {
        const unmountResult = ReactDOM.unmountComponentAtNode(div);
        if (unmountResult && div.parentNode) {
            div.parentNode.removeChild(div);
        }
        const triggerCancel = params.triggerCancel;
        if (onCancel && triggerCancel) {
            onCancel();
        }
    };

    const body = (
        <div className={`${prefixCls}-body`}>
            <div className={`${prefixCls}-content`}>{content}</div>
        </div>
    );

    let footer = null;
    const okButtonProps = {
        type: okType,
        actionFn: onOk,
        closeModal: close,
        autoFocus: true,
        order: okOrder,
        size: buttonSize,
        triggerCancel: false,
        ...okProps
    };
    if (config.footer) {
        footer = config.footer;
    } else if (okCancel) {
        footer = [
            <ActionButton key="confirm" {...okButtonProps}>
                {okText}
            </ActionButton>,
            <ActionButton
                key="cancel"
                type={cancelType}
                closeModal={close}
                order={cancelOrder || null}
                size={buttonSize}
                {...cancelProps}
            >
                {cancelText}
            </ActionButton>
        ];
        if (buttonPosition === 'right') {
            footer = [footer[1], footer[0]];
        }
    }
    else {
        footer = [
            <ActionButton key="confirm" {...okButtonProps}>
                {okText}
            </ActionButton>
        ];
    }

    let $title = title;
    let $icon = null;
    if (icon) {
        $icon = icon;
    }
    else if (iconType) {
        switch (iconType) {
            case 'success':
                $icon = <IconCheckCircle />;
                break;
            case 'warning':
                $icon = <IconExclamationCircle />;
                break;
            case 'info':
                $icon = <IconInfoCircle />;
                break;
            case 'fail':
                $icon = <IconTimesCircle />;
                break;
            default:
                break;
        }
    }

    if ($icon) {
        $title = (
            <div className={`${prefixCls}-title-icon`}>
                <span className={classnames(`${prefixCls}-title-icon-container`, {
                    [`${prefixCls}-title-icon-${iconType}`]: !!iconType,
                    [`${prefixCls}-title-icon-custom`]: !!icon
                })}
                >
                    {$icon}
                </span>
                {$title}
            </div>
        );
    }

    const classString = classnames(prefixCls, {
        [`${prefixCls}-${type}`]: true,
        [`${prefixCls}-has-icon`]: !!$icon,
        [`${prefixCls}-order-reverse`]: cancelOrder && okOrder && cancelOrder < okOrder
    }, className);

    ReactDOM.render(
        <Dialog
            {...config}
            className={classString}
            onCancel={close}
            visible
            title={$title}
            maskClosable={maskClosable}
            size={size}
            footer={footer}
            width={width}
            buttonPosition={buttonPosition}
            needCloseIcon={needCloseIcon}
        >
            <div>
                {body}
            </div>
        </Dialog>,
        div);

    return {
        destroy: close
    };
}
