// @baidu/one-ui@4.32.1/src/components/dialog/interface.ts
import {CSSProperties, HTMLAttributes, ReactElement, ReactNode} from "react";
import Button, {ButtonProps, ButtonType} from "../button";
import {BasePropsWithChildren} from "../interface";

export interface DialogProps extends BasePropsWithChildren {

    /**
     * 弹窗是否可见
     */
    visible?: boolean;

    /**
     * 确认按钮是否为loading
     */
    confirmLoading?: boolean;

    /**
     * 弹窗标题
     */
    title?: ReactNode;

    /**
     * 弹窗点击按钮
     */
    onOk?(e): void;

    /**
     * 弹窗点击取消
     */
    onCancel?(e?): void;

    /**
     * 弹窗关闭后触发
     */
    afterClose?(): void;

    /**
     * footer为一个数组，数组里面是button
     */
    footer?: ReactElement[];

    /**
     * 确定按钮的文案
     */
    okText?: string;

    /**
     * 确定按钮的type
     */
    okType?: ButtonType;

    /**
     * 取消的文案
     */
    cancelText?: string;

    /**
     * 遮罩是否可以点击被关闭
     */
    maskClosable?: boolean;

    /**
     * 弹窗的类名自定义
     */
    wrapClassName?: string;

    /**
     * 弹窗挂载的位置
     */
    getContainer?: () => HTMLElement;

    /**
     * 弹窗的z-index
     */
    zIndex?: number;

    /**
     * 弹窗的尺寸
     */
    size?: 'small' | 'medium' | 'large';

    /**
     * 按钮的类型
     */
    buttonSize?: 'small' | 'medium' | 'large' | 'xsmall' | 'xlarge';

    /**
     * 弹窗是否居中显示
     */
    centered?: boolean;

    /**
     * 关闭以后弹窗是否销毁
     */
    destroyOnClose?: boolean;

    /**
     * 是否支持键盘esc关闭
     */
    keyboard?: boolean;

    /**
     * 是否展示遮罩
     */
    mask?: boolean;

    /**
     * 遮罩样式
     */
    maskStyle?: CSSProperties;

    /**
     * 遮罩属性
     */
    maskProps?: HTMLAttributes<HTMLElement>;

    /**
     * 容器样式
     */
    wrapStyle?: CSSProperties;

    /**
     * 容器属性
     */
    wrapProps?: HTMLAttributes<HTMLElement>;

    /**
     * 内容样式
     */
    bodyStyle?: CSSProperties;

    /**
     * 内容属性
     */
    bodyProps?: HTMLAttributes<HTMLElement>;

    /**
     * 按钮的位置
     */
    buttonPosition?: 'left' | 'center' | 'right';

    /**
     * 是否需要close的Icon
     */
    needCloseIcon?: boolean;

    /**
     * 确定按钮属性
     */
    okProps?: ButtonProps;

    /**
     * 取消按钮属性
     */
    cancelProps?: ButtonProps;

    /**
     * 确定 按钮的order
     */
    okOrder?: number;

    /**
     * 取消 按钮的order
     */
    cancelOrder?: number;

    /**
     * 弹窗宽度
     */
    width?: number | string;

    /**
     * 弹窗高度
     */
    height?: number | string;

    /**
     * 全屏
     */
    fullScreen?: boolean;

    /**
     * 点击X是否关闭弹窗
     */
    visibleControlledWhenClose?: boolean;

    /**
     * 挂在content上的classname
     */
    contentClassName?: string;

    /**
     * 遮罩样式
     */
    maskClassName?: string;

    /**
     * 强制渲染
     */
    forceRender?: boolean;
};

export interface DialogConfirmProps extends DialogProps {
    /**
     * 是否展示默认 确定 取消
     */
    okCancel?: boolean;

    /**
     * 内容
     */
    content?: ReactNode;

    /**
     * 确定按钮类型
     */
    okType?: ButtonType;

    /**
     * 取消按钮类型
     */
    cancelType?: ButtonType;

    /**
     * 类型
     */
    type?: 'confirm' | 'alert';

    /**
     * 自定义图标
     */
    icon?: ReactElement;

    /**
     * 图标类型
     */
    iconType?: 'info' | 'success' | 'warning' | 'fail'
};

export type DialogAlertProps = DialogConfirmProps;

export type DialogDestroyable = {
    destroy(params?: {triggerCancel?: boolean}): void
}
