// @baidu/one-ui@4.32.1/src/components/switch/switch.tsx
import React, {PureComponent} from 'react';
import omit from 'omit.js';
import CommonSwitch from './common/switch';
import {classnames} from '../../core/commonTools';
import {SwitchProps} from './interface';
import {withConfigConsumer} from '../providerConfig/context';

const defaultSize = 'medium';

@withConfigConsumer('switch')
class Switch extends PureComponent<SwitchProps> {

    static defaultProps = {
        prefixCls: 'one-switch',
        loading: false,
        size: defaultSize
    };

    static displayName = 'Switch';

    rcSwitch;

    focus() {
        this.rcSwitch.focus();
    }

    blur() {
        this.rcSwitch.blur();
    }

    saveSwitch = node => {
        this.rcSwitch = node;
    }

    render() {
        const {
            prefixCls, loading, className = '', size: rawSize
        } = this.props;

        const size = ['xsmall', 'small', 'medium'].includes(rawSize) ? rawSize : defaultSize;

        const classes = classnames(className, {
            [`${prefixCls}-${size}`]: size,
            [`${prefixCls}-loading`]: loading
        });
        const otherProps = {
            className: classes,
            ref: this.saveSwitch,
            loading
        };
        return (
            <CommonSwitch
                {...omit(this.props, ['size'])}
                {...otherProps}
            />
        );
    }
}

export default Switch;
