// @baidu/one-ui@4.32.1/lib/components/alert/page.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _partial2 = _interopRequireDefault(require("lodash/partial"));

var _react = _interopRequireWildcard(require("react"));

var _warning = _interopRequireDefault(require("warning"));

var _dlsIconsReact = require("dls-icons-react");

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

var AlertPage = (_dec = (0, _context.withConfigConsumer)('alert-page'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(AlertPage, _PureComponent);

  function AlertPage(_props) {
    var _this;

    _this = _PureComponent.call(this, _props) || this;
    _this.alertPageRef = void 0;

    _this.onClose = function (closeFunc, index, e) {
      // 全局的
      _this.props.onClose(index); // 每一个alert自身的onClose


      if (typeof closeFunc === 'function') {
        closeFunc(e);
      }

      if (!('dataSource' in _this.props)) {
        // 移除
        var dataSource = _this.state.dataSource;
        dataSource.splice(index, 1);

        _this.setState({
          dataSource: [].concat(dataSource)
        });
      }

      var activeIndex = _this.state.activeIndex;

      if (activeIndex > 0) {
        activeIndex = activeIndex - 1;
      }

      if (!_this.isActiveIndexControlled()) {
        _this.setState({
          activeIndex: activeIndex
        });
      }

      _this.props.onPrevChange(activeIndex);
    };

    _this.getItems = function () {
      var dataSource = [].concat(_this.state.dataSource);
      var children = dataSource.map(function (child, index) {
        if (! /*#__PURE__*/(0, _react.isValidElement)(child)) {
          return null;
        }

        var prefixCls = _this.props.prefixCls;
        var key = child.key || String(index);
        var props = (0, _extends2["default"])({
          key: key
        }, child.props, {
          onClose: (0, _partial2["default"])(_this.onClose, child.props.onClose, index),
          visible: true
        });
        return /*#__PURE__*/_react["default"].createElement("span", {
          className: prefixCls + "-slick-item",
          key: key
        }, /*#__PURE__*/(0, _react.cloneElement)(child, props));
      });
      return children;
    };

    _this.saveRef = function (ref) {
      _this.alertPageRef = ref;
    };

    _this.prevChange = function () {
      var current = _this.state.activeIndex;

      if (current === 0) {
        return;
      }

      var prevSlide = current - 1;

      if (!_this.isActiveIndexControlled()) {
        _this.setState({
          activeIndex: prevSlide
        });
      }

      _this.props.onPrevChange(prevSlide);
    };

    _this.nextChange = function (childrenLength) {
      var current = _this.state.activeIndex;

      if (current === childrenLength - 1) {
        return;
      }

      var nextSlide = current + 1;

      if (!_this.isActiveIndexControlled()) {
        _this.setState({
          activeIndex: nextSlide
        });
      }

      _this.props.onNextChange(nextSlide);
    };

    var _dataSource = _props.dataSource,
        defaultDataSource = _props.defaultDataSource,
        slider = _props.slider,
        _activeIndex = _props.activeIndex,
        initialSlide = _props.initialSlide,
        defaultActiveIndex = _props.defaultActiveIndex;

    if ('initialSlide' in _props) {
      (0, _warning["default"])(false, 'The `initialSlide` prop of `Alert.Page` is deprecated, use `defaultActiveIndex` instead');
    }

    if ('slider' in _props) {
      (0, _warning["default"])(false, 'The `slider` prop of `Alert.Page` is deprecated, use `activeIndex` instead');
    }

    _this.state = {
      dataSource: _dataSource || defaultDataSource || [],
      activeIndex: defaultActiveIndex || initialSlide || 0
    };
    return _this;
  }

  AlertPage.getDerivedStateFromProps = function getDerivedStateFromProps(nextProps, prevState) {
    var newState = {};

    if ('dataSource' in nextProps) {
      newState.dataSource = [].concat(nextProps.dataSource);
    }

    if ('slider' in nextProps && nextProps.slider !== prevState.activeIndex) {
      newState.activeIndex = nextProps.slider;
    }

    if ('activeIndex' in nextProps && nextProps.activeIndex !== prevState.activeIndex) {
      newState.activeIndex = nextProps.activeIndex;
    }

    return newState;
  };

  var _proto = AlertPage.prototype;

  _proto.isActiveIndexControlled = function isActiveIndexControlled() {
    return 'slider' in this.props || 'activeIndex' in this.props;
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        className = _this$props.className,
        style = _this$props.style,
        size = _this$props.size;
    var _this$state = this.state,
        dataSource = _this$state.dataSource,
        activeIndex = _this$state.activeIndex;

    if (!dataSource.length) {
      return null;
    }

    var children = this.getItems();
    var single = dataSource.length === 1;
    var alertClassNames = (0, _commonTools.classnames)(prefixCls, className, prefixCls + "-" + size, (_classnames = {}, _classnames[prefixCls + "-first-page"] = activeIndex === 0, _classnames[prefixCls + "-last-page"] = activeIndex === children.length - 1, _classnames[prefixCls + "-single"] = single, _classnames));

    if (!children[activeIndex]) {
      return null;
    }

    return /*#__PURE__*/_react["default"].createElement("div", {
      ref: this.saveRef,
      className: alertClassNames,
      style: style
    }, children[activeIndex], !single && /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-count"
    }, /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconChevronLeft, {
      className: prefixCls + "-count-prev",
      onClick: this.prevChange
    }), /*#__PURE__*/_react["default"].createElement("span", null, activeIndex + 1, "/", children.length), /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconChevronRight, {
      className: prefixCls + "-count-next",
      onClick: (0, _partial2["default"])(this.nextChange, children.length)
    })));
  };

  return AlertPage;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-alert-page',
  defaultActiveIndex: 0,
  size: 'medium',
  onClose: _noop2["default"],
  onPrevChange: _noop2["default"],
  onNextChange: _noop2["default"]
}, _temp)) || _class);
var _default = AlertPage;
exports["default"] = _default;
module.exports = exports.default;
