// @baidu/one-ui@4.32.1/src/components/input/interface.ts
import {ReactNode, ChangeEvent, Ref} from "react";
import {BaseProps, BasePropsWithChildren} from "../interface";

export interface InputProps extends BaseProps,
    Omit<JSX.IntrinsicElements['input'], 'ref' | 'size' | 'onChange' | 'prefix'> {

        /**
         * 默认值
         */
        defaultValue?: string;

        /**
         * 值，受控
         */
        value?: string;

        /**
         * 值变化时回调
         */
        onChange?: (e: {
            value: string,
            errorMessage: string,
            event: ChangeEvent<HTMLInputElement>,
            target: HTMLInputElement
        }) => void;

        /**
         * 回车时回调
         */
         onPressEnter?(e: {
            value: string,
            errorMessage: string
        }): void;

        /**
         * 尺寸
         */
        size?: 'xsmall' | 'small' | 'medium' | 'large';

        /**
         * 类型
         */
        type?: 'normal' | 'inline' | React.InputHTMLAttributes<HTMLInputElement>['type'];

        /**
         * 是否禁用
         */
        disabled?: boolean;

        /**
         * 是否只读
         */
        readOnly?: boolean;

        /**
         * 宽度
         */
        width?: number | string;

        /**
         * 占位提示
         */
        placeholder?: string;

        /**
         * 计数方式，当为cn时，汉字记为2个字符；en时，汉字记为1个字符。默认为cn
         */
        countMode?: 'cn' | 'en';

        /**
         * 自定义计数方式
         *
         * @param value 当前值
         */
        getLength?(value: string): number;

        /**
         * 不参与计数字符
         */
        filterArray?: string[];

        /**
         * 是否展示清除Icon
         */
        showClear?: boolean;

        /**
         * 点击清除Icon回调
         */
        onClear?(): void;

        /**
         * 组合框，左侧元素
         */
        addonBefore?: ReactNode;

        /**
         * 组合框，右侧元素
         */
        addonAfter?: ReactNode;

        /**
         * 前缀 Icon
         */
        prefix?: ReactNode;

        /**
         * 后缀 Icon
         */
        suffix?: ReactNode;

        /**
         * 错误信息；当不传此值时，会内部校验是否非空、最短字符、最长字符；当传此值时，错误信息以此值为准
         */
        errorMessage?: ReactNode;

        /**
         * 是否展示错误信息，如果为false，当errorMessage存在时，仅有外框标红，但不显示错误信息
         */
        showErrorMessage?: boolean;

        /**
         * 错误位置
         * @deprecated
         */
        errorLocation?: 'right' | 'layer' | 'bottom';

        /**
         * input的ref
         */
        inputRef?: Ref<HTMLInputElement>;

        /**
         * 必填
         */
        isRequired?: boolean;

        /**
         * 必填，错误信息
         */
        requiredErrorMessage?: ReactNode;

        /**
         * 最大字符长度
         */
        maxLen?: number;

        /**
         * 最大字符长度，错误信息
         */
        maxLenErrorMessage?: ReactNode;

        /**
         * 最小字符长度
         */
        minLen?: number;

        /**
         * 最小字符长度，错误信息
         */
        minLenErrorMessage?: ReactNode;

        /**
         * 错误信息是不是html
         *
         * @deprecated
         * @internal
         */
        isErrorHTML?: boolean;

        /**
         * 可直接挂载在input上的属性，比较粗暴，不建议使用
         *
         * @deprecated
         */
        originInputProps?: JSX.IntrinsicElements['input'];
    };

    export type InputGroupProps = BasePropsWithChildren;
