// @baidu/one-ui@4.41.1/es/components/layout/layout.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { PureComponent, isValidElement } from 'react';
import { withConfigConsumer } from '../providerConfig/context';
import { classnames } from '../../core/commonTools';
import LayoutSidebar from './sidebar';
import Context from './context';
;
var Layout = (_dec = withConfigConsumer('layout'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Layout, _PureComponent);

  function Layout() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.state = {
      sidebarCount: 0,
      hasStickyFooter: false
    };

    _this.registerSidebar = function () {
      _this.setState(function (_ref) {
        var sidebarCount = _ref.sidebarCount;
        return {
          sidebarCount: sidebarCount + 1
        };
      });
    };

    _this.unregisterSidebar = function () {
      _this.setState(function (_ref2) {
        var sidebarCount = _ref2.sidebarCount;
        return {
          sidebarCount: sidebarCount - 1
        };
      });
    };

    _this.registerStickyFooter = function () {
      _this.setState({
        hasStickyFooter: true
      });
    };

    _this.unregisterStickyFooter = function () {
      _this.setState({
        hasStickyFooter: false
      });
    };

    _this.contextValue = {
      registerSidebar: _this.registerSidebar,
      unregisterSidebar: _this.unregisterSidebar,
      registerStickyFooter: _this.registerStickyFooter,
      unregisterStickyFooter: _this.unregisterStickyFooter
    };
    return _this;
  }

  var _proto = Layout.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        className = _this$props.className,
        prefixCls = _this$props.prefixCls,
        children = _this$props.children,
        direction = _this$props.direction,
        lightPrefix = _this$props.lightPrefix,
        restProps = _objectWithoutPropertiesLoose(_this$props, ["className", "prefixCls", "children", "direction", "lightPrefix"]);

    var _this$state = this.state,
        sidebarCount = _this$state.sidebarCount,
        hasStickyFooter = _this$state.hasStickyFooter;
    var hasSidebar = React.Children.toArray(children).filter(isValidElement).some(function (child) {
      return child.type === LayoutSidebar;
    }) || sidebarCount > 0;
    var rowDirection = direction === 'row' || hasSidebar;
    return /*#__PURE__*/React.createElement(Context.Provider, {
      value: this.contextValue
    }, /*#__PURE__*/React.createElement("div", _extends({
      className: classnames(prefixCls, className, (_classnames = {}, _classnames[prefixCls + "-row-direction"] = rowDirection, _classnames[prefixCls + "-with-sidebar"] = hasSidebar, _classnames[prefixCls + "-with-sticky-footer"] = hasStickyFooter, _classnames))
    }, restProps), children));
  };

  return Layout;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-layout',
  direction: 'column'
}, _class2.Sidebar = void 0, _class2.Header = void 0, _class2.Content = void 0, _class2.Footer = void 0, _temp)) || _class);
export default Layout;
