// @baidu/one-ui@4.41.1/es/components/checkbox/button.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import { withConfigConsumer } from '../providerConfig/context';
import Checkbox from './checkbox';
import Context from './context';
var CheckboxButton = (_dec = withConfigConsumer('checkbox-button'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(CheckboxButton, _PureComponent);

  function CheckboxButton() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = CheckboxButton.prototype;

  _proto.render = function render() {
    var _this = this;

    return /*#__PURE__*/React.createElement(Context.Consumer, null, function (context) {
      return _this.renderCheckbox(context);
    });
  };

  _proto.renderCheckbox = function renderCheckbox(context) {
    var _this$props = this.props,
        type = _this$props.type,
        props = _objectWithoutPropertiesLoose(_this$props, ["type"]);

    var buttonType = type || (context == null ? void 0 : context.type) || 'strong';
    return /*#__PURE__*/React.createElement(Checkbox, _extends({}, props, {
      mode: "strong",
      buttonType: buttonType
    }));
  };

  return CheckboxButton;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-checkbox-button'
}, _temp)) || _class);
export default CheckboxButton;
