// @baidu/one-ui@4.32.1/src/components/toast/index.tsx
import React from 'react';
import Notification from 'rc-notification';
import MessageContainer, {MessageProps} from './common/message';
import {ToastConfigOptions, ToastProps, ToastSize} from './interface';

let defaultDuration = 3;
let defaultTop: string | number = '10%';
let messageInstance;
let key = 1;
let prefixCls = 'one-toast';
let getContainer;
let messageSize: ToastSize = 'medium';

const getMessageInstance = callback => {
    if (messageInstance) {
        callback(messageInstance);
        return;
    }
    Notification.newInstance({
        prefixCls,
        transitionName: 'move-up',
        style: {top: defaultTop}, // 覆盖原来的样式
        getContainer
    }, instance => {
        if (messageInstance) {
            callback(messageInstance);
            return;
        }
        messageInstance = instance;
        callback(instance);
    });
};

const notice = (props: ToastProps) => {
    const duration = props.duration !== undefined ? props.duration : defaultDuration;
    const target = key++;
    const closePromise = new Promise(resolve => {
        const callback = () => {
            if (typeof props.onClose === 'function') {
                props.onClose();
            }
            return resolve(true);
        };
        getMessageInstance(instance => {
            const {title, content, type, style, showCloseIcon = true} = props;
            const size = props.size || messageSize;
            const messageProps: MessageProps = {
                prefixCls,
                type,
                title,
                size,
                content,
                showCloseIcon,
                target,
                instance,
                onClose: callback
            };
            instance.notice({
                key: target,
                duration,
                style,
                content: (
                    <MessageContainer {...messageProps} />
                ),
                onClose: callback
            });
        });
    });
    const result = () => {
        if (messageInstance) {
            messageInstance.removeNotice(target);
        }
    };
    result.then = (filled, rejected) => closePromise.then(filled, rejected);
    result.promise = closePromise;
    return result;
};

export default {
    info(props: ToastProps) {
        return notice({
            type: 'info',
            ...props
        });
    },
    success(props: ToastProps) {
        return notice({
            type: 'success',
            ...props
        });
    },
    error(props: ToastProps) {
        return notice({
            type: 'error',
            ...props
        });
    },
    warning(props: ToastProps) {
        return notice({
            type: 'warning',
            ...props
        });
    },
    loading(props: ToastProps) {
        return notice({
            type: 'loading',
            ...props
        });
    },
    config(options: ToastConfigOptions) {
        if (options.top !== undefined) {
            defaultTop = options.top;
            messageInstance = null; // delete messageInstance for new defaultTop
        }
        if (options.duration !== undefined) {
            defaultDuration = options.duration;
        }
        if (options.prefixCls !== undefined) {
            prefixCls = options.prefixCls;
        }
        if (options.getContainer !== undefined) {
            getContainer = options.getContainer;
        }
        if (options.size !== undefined) {
            messageSize = options.size;
        }
    },
    destroy() {
        if (messageInstance) {
            messageInstance.destroy();
            messageInstance = null;
        }
    },
    /**
     * @internal 仅限特殊场合，不得用于业务
     */
    Element: MessageContainer
};

export * from './interface';
