// @baidu/one-ui@4.32.1/es/components/tag/group.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _pull from "lodash/pull";
import _partial from "lodash/partial";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import Tag from './tag';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
;
var TagGroup = (_dec = withConfigConsumer('tag'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(TagGroup, _PureComponent);

  function TagGroup(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.isUniqueMode = function () {
      return _this.props.mode === 'unique';
    };

    _this.onChangeGroup = function (value, checked) {
      var selectedValue = [].concat(_this.state.value);

      if (checked) {
        // 选中
        if (_this.isUniqueMode()) {
          // 单选模式
          selectedValue = [value];
        } else {
          selectedValue.push(value);
        }
      } else if (_this.isUniqueMode()) {
        // 没选中，单选模式
        selectedValue = [];
      } else {
        _pull(selectedValue, value);
      }

      if (!('value' in _this.props)) {
        _this.setState({
          value: selectedValue
        });
      }

      var onChange = _this.props.onChange;
      onChange(selectedValue);
    };

    _this.state = {
      value: props.defaultValue || props.value || [],
      prevProps: props
    };
    return _this;
  }

  var _proto = TagGroup.prototype;

  _proto.render = function render() {
    var _this2 = this;

    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        className = _this$props.className,
        dataSource = _this$props.dataSource,
        size = _this$props.size;
    var value = this.state.value;
    var selectedValue = this.isUniqueMode() ? value && value.length && [value[0]] || [] : value;
    var groupTagClassName = classnames(className, prefixCls + "-group-wrapper");
    return /*#__PURE__*/React.createElement("div", {
      className: groupTagClassName
    }, dataSource.map(function (tag, index) {
      var value = tag.value,
          label = tag.label,
          _tag$tagProps = tag.tagProps,
          tagProps = _tag$tagProps === void 0 ? {} : _tag$tagProps;
      var checked = selectedValue.indexOf(value) > -1;
      var onClose = tagProps.onClose;

      if (onClose) {
        tagProps.onClose = _partial(onClose, value);
      }

      return /*#__PURE__*/React.createElement(Tag, _extends({
        key: index,
        size: size,
        checkable: true,
        checked: checked,
        onChange: _partial(_this2.onChangeGroup, value)
      }, tagProps), label);
    }));
  };

  return TagGroup;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-tag',
  mode: 'unique',
  dataSource: [],
  className: '',
  onChange: function onChange() {},
  size: 'medium'
}, _class2.getDerivedStateFromProps = function (nextProps, prevState) {
  var newState = {
    prevProps: nextProps
  };
  var prevProps = prevState.prevProps;

  if ('value' in nextProps && nextProps.value !== prevState.value) {
    newState.value = nextProps.value;
  }

  if ('dataSource' in nextProps && nextProps.dataSource !== prevProps.dataSource) {
    newState.dataSource = nextProps.dataSource;
  }

  return newState;
}, _temp)) || _class);
export default TagGroup;
