// @baidu/one-ui@4.41.1/es/components/sortable/sortable.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _noop from "lodash/noop";

var _dec, _class, _class2, _temp;

/**
 * @file 可拖拽排序组件
 * @author tangjinhua
 * @date 2021/06/30
 */
import React, { PureComponent } from 'react';
import { classnames, isSafari as checkIsSafari, isInsideTransformedContainer } from '../../core/commonTools';
import { ReactSortable } from 'react-sortablejs';
import cloneElement from 'clone-element';
import { withConfigConsumer } from '../providerConfig/context';
var isSafari = checkIsSafari();
var Sortable = (_dec = withConfigConsumer('sortable'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Sortable, _PureComponent);

  function Sortable() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.state = {
      list: []
    };
    _this.offsetX = 0;
    _this.offsetY = 0;

    _this.handleOnChange = function (event) {
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          before = _this$props.before;

      if (!onChange) {
        return;
      }

      var oldIndex = event.oldIndex,
          newIndex = event.newIndex;
      var headOffset = before ? 0 : 1;

      if (!before) {
        oldIndex += headOffset;
        newIndex += headOffset;
      }

      var list = _this.state.list.slice();

      if (oldIndex === newIndex || newIndex === list.length - 1 // tail
      ) {
        return;
      }

      var oldOption = list[oldIndex];
      list.splice(oldIndex, 1);
      list.splice(newIndex, 0, oldOption);
      onChange(list.map(function (item) {
        return item.option;
      }).slice(1, -1), oldIndex - headOffset, newIndex - headOffset);
    };

    _this.setDragImage = function (dataTransfer, ele) {
      if (isSafari && isInsideTransformedContainer(ele)) {
        var eleImage = cloneElement(ele);
        eleImage.style.position = 'fixed';
        eleImage.style.left = '0px';
        eleImage.style.top = '0px';
        document.body.append(eleImage);
        requestAnimationFrame(function () {
          return document.body.removeChild(eleImage);
        });

        if (dataTransfer && dataTransfer.setDragImage) {
          dataTransfer.setDragImage(eleImage, _this.offsetX, _this.offsetY);
        }
      }
    };

    _this.setOffset = function (_ref) {
      var _ref$originalEvent = _ref.originalEvent,
          offsetX = _ref$originalEvent.offsetX,
          offsetY = _ref$originalEvent.offsetY;
      _this.offsetX = offsetX;
      _this.offsetY = offsetY;
    };

    _this.handleOnStart = function () {
      _this.toggleDraggingClassName(true);

      var onStart = _this.props.onStart;

      if (onStart) {
        onStart();
      }
    };

    _this.handleOnEnd = function (event) {
      _this.toggleDraggingClassName(false);

      _this.handleOnChange(event);

      var onEnd = _this.props.onEnd;

      if (onEnd) {
        onEnd();
      }
    };

    return _this;
  }

  Sortable.getDerivedStateFromProps = function getDerivedStateFromProps(props) {
    var list = props.options.map(function (option) {
      return {
        option: option
      };
    }); // 数据适配`before`与`after`节点

    list.unshift({
      filtered: true
    });
    list.push({
      filtered: true
    });
    return {
      list: list
    };
  };

  var _proto = Sortable.prototype;

  _proto.componentWillUnmount = function componentWillUnmount() {
    // 拖拽过程中组件可能卸载，补充移除样式
    this.toggleDraggingClassName(false);
  };

  _proto.toggleDraggingClassName = function toggleDraggingClassName(add) {
    var classList = document.documentElement.classList;
    var className = this.props.prefixCls + "-dragging";

    if (add) {
      classList.add(className);
    } else {
      classList.remove(className);
    }
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props2 = this.props,
        prefixCls = _this$props2.prefixCls,
        className = _this$props2.className,
        options = _this$props2.options,
        renderOption = _this$props2.renderOption,
        useHandle = _this$props2.useHandle,
        onChange = _this$props2.onChange,
        before = _this$props2.before,
        after = _this$props2.after,
        lightPrefix = _this$props2.lightPrefix,
        restProps = _objectWithoutPropertiesLoose(_this$props2, ["prefixCls", "className", "options", "renderOption", "useHandle", "onChange", "before", "after", "lightPrefix"]);

    return /*#__PURE__*/React.createElement(ReactSortable, _extends({}, restProps, {
      key: lightPrefix,
      className: classnames(prefixCls, className, (_classnames = {}, _classnames[prefixCls + "-use-handle"] = useHandle, _classnames)),
      handle: useHandle ? "." + prefixCls + "-handle" : null,
      list: this.state.list,
      setData: this.setDragImage,
      onChoose: this.setOffset,
      onStart: this.handleOnStart,
      onEnd: this.handleOnEnd,
      ghostClass: classnames(prefixCls + "-ghost"),
      dragClass: classnames(prefixCls + "-drag"),
      chosenClass: classnames(prefixCls + "-chosen"),
      draggable: "." + prefixCls + "-draggable",
      setList: _noop
    }), /*#__PURE__*/React.createElement(React.Fragment, null, before), options.map(function (option, index) {
      var optionElement = renderOption(option, index);
      return /*#__PURE__*/React.cloneElement(optionElement, {
        className: classnames(prefixCls + "-draggable", optionElement.props.className)
      });
    }), /*#__PURE__*/React.createElement(React.Fragment, null, after));
  };

  return Sortable;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-sortable',
  options: [],
  renderOption: function renderOption(option) {},
  useHandle: false,
  // 未放开属性
  animation: 200,
  easing: 'cubic-bezier(.25, .1, .25, 1)'
}, _class2.Handle = void 0, _temp)) || _class);
export default Sortable;
