// @baidu/one-ui@4.41.1/es/components/transfer/common/selectedPane.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

/**
 * @author huangshiming
 * @file selected pane
 * @date 2020-05-16
 */
import React, { PureComponent } from 'react';
import SelectedItemRender from './itemRender';
import Tree from '../../tree';
import Button from '../../button';
import { IconTimes } from 'dls-icons-react';
import { formatSelectedList } from '../../../core/transferTools';
import Stack from '../../stack';
import Message from '../../message';
var TreeNode = Tree.TreeNode,
    VirtualTreeNode = Tree.VirtualTreeNode;

var SelectedPane = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(SelectedPane, _PureComponent);

  function SelectedPane() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = SelectedPane.prototype;

  _proto.renderNode = function renderNode(key, entities) {
    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        allDataMap = _this$props.allDataMap,
        selectedItemProps = _this$props.selectedItemProps,
        SelectedItem = _this$props.SelectedItem,
        onDelete = _this$props.onDelete,
        virtualListHeight = _this$props.virtualListHeight;
    var CustomTreeNode = virtualListHeight ? VirtualTreeNode : TreeNode;
    var selectedItem = allDataMap[key] || {};
    var treeNodeKey = selectedItem.key;
    var disabled = selectedItem.disabled;
    var children = entities[key] && entities[key].children;
    var renderItem;

    if (SelectedItem && typeof SelectedItem === 'function') {
      var CustomSelectedItemRender = SelectedItem;

      var selectedProps = _extends({}, selectedItem, {
        itemKey: treeNodeKey,
        prefixCls: prefixCls
      }, selectedItemProps);

      delete selectedProps.CandidateItem;
      renderItem = /*#__PURE__*/React.createElement(CustomSelectedItemRender, selectedProps);
    } else if (SelectedItem && typeof SelectedItem === 'object') {
      renderItem = SelectedItem;
    } else {
      renderItem = /*#__PURE__*/React.createElement(SelectedItemRender, {
        prefixCls: prefixCls,
        title: selectedItem.title
      });
    }

    var treeNodeProps = _extends({}, selectedItem, {
      disabled: disabled,
      title: /*#__PURE__*/React.createElement("div", {
        className: prefixCls + "-select-selected"
      }, renderItem, /*#__PURE__*/React.createElement(Button, {
        disabled: disabled,
        type: "text-aux",
        className: prefixCls + "-delete-item",
        onClick: function onClick(e) {
          return onDelete(key, e);
        }
      }, /*#__PURE__*/React.createElement(IconTimes, null)))
    });

    if (children) {
      return /*#__PURE__*/React.createElement(CustomTreeNode, _extends({}, treeNodeProps, {
        key: treeNodeKey
      }), this.selectedNodeTrees(children, entities));
    }

    return /*#__PURE__*/React.createElement(CustomTreeNode, _extends({}, treeNodeProps, {
      key: treeNodeKey
    }));
  };

  _proto.selectedNodeTrees = function selectedNodeTrees(topLevel, entities) {
    var _this = this;

    return topLevel.map(function (key) {
      return _this.renderNode(key, entities);
    });
  };

  _proto.render = function render() {
    var _this$props2 = this.props,
        selectedList = _this$props2.selectedList,
        parentRelationMap = _this$props2.parentRelationMap,
        expandedSelectedKeys = _this$props2.expandedSelectedKeys,
        onExpand = _this$props2.onExpand,
        onSelect = _this$props2.onSelect,
        prefixCls = _this$props2.prefixCls,
        getSelectedTreeRef = _this$props2.getSelectedTreeRef,
        virtualListHeight = _this$props2.virtualListHeight,
        size = _this$props2.size,
        allDataMap = _this$props2.allDataMap,
        selectedEmpty = _this$props2.selectedEmpty;

    var _formatSelectedList = formatSelectedList(selectedList, parentRelationMap, allDataMap),
        entities = _formatSelectedList.entities,
        topLevel = _formatSelectedList.topLevel;

    var selectedTreeProps = {
      expandedKeys: (expandedSelectedKeys || []).map(function (key) {
        return "" + key;
      }),
      onExpand: onExpand,
      onSelect: onSelect,
      className: prefixCls + "-selected-tree",
      parentContainerHeight: virtualListHeight || null,
      selectedKeys: [],
      size: size
    };
    return selectedList.length ? /*#__PURE__*/React.createElement(Tree, _extends({}, selectedTreeProps, {
      ref: getSelectedTreeRef
    }), this.selectedNodeTrees(topLevel, entities)) : /*#__PURE__*/React.createElement(Stack, {
      justify: "center",
      className: prefixCls + "-empty"
    }, /*#__PURE__*/React.createElement(Message, {
      type: "aux",
      display: "simple",
      size: size
    }, selectedEmpty));
  };

  return SelectedPane;
}(PureComponent);

export default SelectedPane;
