// @baidu/one-ui@4.32.1/lib/components/slider/slider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends3 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _Slider = _interopRequireDefault(require("rc-slider/lib/Slider"));

var _Range = _interopRequireDefault(require("rc-slider/lib/Range"));

var _Handle = _interopRequireDefault(require("rc-slider/lib/Handle"));

var _tooltip = _interopRequireDefault(require("../tooltip"));

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

var Slider = (_dec = (0, _context.withConfigConsumer)('slider'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Slider, _PureComponent);

  function Slider(_props) {
    var _this;

    _this = _PureComponent.call(this, _props) || this;
    _this.tooltipRef = void 0;
    _this.rcSlider = void 0;

    _this.toggleTooltipVisible = function (index, visible) {
      if (_this.props.disabled) {
        return;
      }

      _this.setState(function (_ref) {
        var _extends2;

        var visibles = _ref.visibles;
        return {
          visibles: (0, _extends3["default"])({}, visibles, (_extends2 = {}, _extends2[index] = visible, _extends2))
        };
      });
    };

    _this.getTooltipRef = function (ref) {
      _this.tooltipRef = ref;
    };

    _this.onChange = function (props) {
      _this.props.onChange(props);

      if (_this.tooltipRef && _this.tooltipRef.trigger && _this.tooltipRef.trigger.forcePopupAlign && typeof _this.tooltipRef.trigger.forcePopupAlign === 'function') {
        _this.tooltipRef.trigger.forcePopupAlign();
      }
    };

    _this.saveSlider = function (node) {
      _this.rcSlider = node;
    };

    _this.state = {
      visibles: {}
    };
    return _this;
  }

  var _proto = Slider.prototype;

  _proto.handleWithTooltip = function handleWithTooltip(_ref2) {
    var _this2 = this;

    var value = _ref2.value,
        dragging = _ref2.dragging,
        index = _ref2.index,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_ref2, ["value", "dragging", "index"]);
    var _this$props = this.props,
        tipFormatter = _this$props.tipFormatter,
        tooltipVisible = _this$props.tooltipVisible,
        tooltipPlacement = _this$props.tooltipPlacement,
        getTooltipPopupContainer = _this$props.getTooltipPopupContainer;
    var visibles = this.state.visibles;
    var isTipFormatter = tipFormatter ? visibles[index] || dragging : false;
    var visible = tooltipVisible || tooltipVisible === undefined && isTipFormatter;
    return /*#__PURE__*/_react["default"].createElement(_tooltip["default"], {
      title: tipFormatter ? tipFormatter(value) : '',
      visible: visible,
      placement: tooltipPlacement,
      transitionName: "zoom-down",
      key: index,
      getPopupContainer: getTooltipPopupContainer,
      type: "dark",
      toolTipRef: this.getTooltipRef
    }, /*#__PURE__*/_react["default"].createElement(_Handle["default"], (0, _extends3["default"])({}, restProps, {
      value: value,
      onMouseEnter: function onMouseEnter() {
        return _this2.toggleTooltipVisible(index, true);
      },
      onMouseLeave: function onMouseLeave() {
        return _this2.toggleTooltipVisible(index, false);
      },
      onChange: this.onChange
    })));
  };

  _proto.focus = function focus() {
    this.rcSlider.focus();
  };

  _proto.blur = function blur() {
    this.rcSlider.blur();
  };

  _proto.render = function render() {
    var _classnames,
        _this3 = this;

    var _this$props2 = this.props,
        customizePrefixCls = _this$props2.prefixCls,
        size = _this$props2.size,
        customizeTooltipPrefixCls = _this$props2.tooltipPrefixCls,
        range = _this$props2.range,
        readOnly = _this$props2.readOnly,
        disabled = _this$props2.disabled,
        className = _this$props2.className,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props2, ["prefixCls", "size", "tooltipPrefixCls", "range", "readOnly", "disabled", "className"]);
    var classNames = (0, _commonTools.classnames)(className, customizePrefixCls + "-" + size, (_classnames = {}, _classnames[customizePrefixCls + "-readOnly"] = readOnly, _classnames));

    if (range) {
      return /*#__PURE__*/_react["default"].createElement(_Range["default"], (0, _extends3["default"])({}, restProps, {
        ref: this.saveSlider,
        handle: function handle(info) {
          return _this3.handleWithTooltip(info);
        },
        prefixCls: customizePrefixCls,
        className: classNames,
        tooltipPrefixCls: customizeTooltipPrefixCls,
        disabled: disabled || readOnly,
        onChange: this.onChange
      }));
    }

    return /*#__PURE__*/_react["default"].createElement(_Slider["default"], (0, _extends3["default"])({}, restProps, {
      ref: this.saveSlider,
      handle: function handle(info) {
        return _this3.handleWithTooltip(info);
      },
      prefixCls: customizePrefixCls,
      className: classNames,
      tooltipPrefixCls: customizeTooltipPrefixCls,
      disabled: disabled || readOnly,
      onChange: this.onChange
    }));
  };

  return Slider;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-slider',
  defaultValue: 0,
  disabled: false,
  readOnly: false,
  dots: false,
  included: true,
  max: 100,
  min: 0,
  range: false,
  step: 1,
  onChange: function onChange() {},
  onAfterChange: function onAfterChange() {},
  tipFormatter: function tipFormatter(value) {
    return value.toString();
  },
  tooltipPlacement: 'top',
  getTooltipPopupContainer: function getTooltipPopupContainer() {
    return document.body;
  },
  size: 'medium'
}, _temp)) || _class);
var _default = Slider;
exports["default"] = _default;
module.exports = exports.default;
