// @baidu/one-ui@4.41.1/lib/components/uploader/common/fileAnchor.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../../core/commonTools");

var _button = _interopRequireDefault(require("../../button"));

var FileAnchor = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(FileAnchor, _PureComponent);

  function FileAnchor() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = FileAnchor.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        onClick = _this$props.onClick,
        loading = _this$props.loading,
        prefixCls = _this$props.prefixCls,
        disabled = _this$props.disabled,
        helperText = _this$props.helperText,
        maxSize = _this$props.maxSize,
        helperTextPosition = _this$props.helperTextPosition,
        pickerIcon = _this$props.pickerIcon,
        pickerText = _this$props.pickerText;
    var buttonClassName = prefixCls + "-anchor-file-button";
    var buttonProps = {
      className: buttonClassName,
      onClick: onClick,
      loading: loading,
      disabled: disabled,
      icon: pickerIcon || 'upload'
    };
    var fileAnchorClassName = (0, _commonTools.classnames)(prefixCls + "-anchor-file-container", (_classnames = {}, _classnames[prefixCls + "-anchor-file-container-" + helperTextPosition] = helperTextPosition === 'right' || helperTextPosition === 'bottom', _classnames));
    var helperTextNode;

    if (helperText === undefined) {
      helperTextNode = defaultHelperText(maxSize); // 兼容之前实现
    } else if (typeof helperText === 'function') {
      helperTextNode = helperText();
    } else {
      helperTextNode = helperText;
    }

    return /*#__PURE__*/_react["default"].createElement("span", {
      className: fileAnchorClassName
    }, /*#__PURE__*/_react["default"].createElement(_button["default"], buttonProps, pickerText || '文件上传'), helperTextNode ? /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-helper-text"
    }, helperTextNode) : null);
  };

  return FileAnchor;
}(_react.PureComponent);

exports["default"] = FileAnchor;
FileAnchor.defaultProps = {
  helperTextPosition: 'right'
};

function defaultHelperText(maxSize) {
  if (maxSize) {
    return '文件大小不超过' + maxSize / (1024 * 1024) + 'MB';
  }

  return null;
}

module.exports = exports.default;
