// @baidu/one-ui@4.41.1/lib/components/uploader/common/imageAnchor.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _loading = _interopRequireDefault(require("../../loading"));

var _multipleAnchor = _interopRequireDefault(require("./multipleAnchor"));

var _commonTools = require("../../../core/commonTools");

var _mediaIcon = _interopRequireDefault(require("./mediaIcon"));

var ImageAnchor = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(ImageAnchor, _PureComponent);

  function ImageAnchor() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = ImageAnchor.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        onClick = _this$props.onClick,
        loading = _this$props.loading,
        prefixCls = _this$props.prefixCls,
        disabled = _this$props.disabled,
        realType = _this$props.realType,
        entries = _this$props.entries,
        pickerIcon = _this$props.pickerIcon,
        pickerText = _this$props.pickerText,
        theme = _this$props.theme;
    var buttonClassName = (0, _commonTools.classnames)(prefixCls + "-anchor-image-button", (_classnames = {}, _classnames[prefixCls + "-anchor-image-button-loading"] = loading, _classnames[prefixCls + "-anchor-image-button-disabled"] = disabled, _classnames));
    var buttonProps = {
      className: buttonClassName,
      onClick: disabled || loading ? _noop2["default"] : onClick
    };
    var showTexts = {
      image: '图片',
      video: '视频',
      media: '多媒体'
    };

    if (loading) {
      return /*#__PURE__*/_react["default"].createElement("div", buttonProps, /*#__PURE__*/_react["default"].createElement(_loading["default"], {
        type: "strong"
      }));
    }

    if (entries && entries.length) {
      return /*#__PURE__*/_react["default"].createElement(_multipleAnchor["default"], this.props);
    }

    return /*#__PURE__*/_react["default"].createElement("div", buttonProps, /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-add-icon"
    }, pickerIcon || /*#__PURE__*/_react["default"].createElement(_mediaIcon["default"], {
      theme: theme,
      type: realType,
      className: prefixCls + "-" + realType + "-add-icon"
    })), /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-anchor-image-button-text"
    }, pickerText || "\u4E0A\u4F20" + showTexts[realType]));
  };

  return ImageAnchor;
}(_react.PureComponent);

exports["default"] = ImageAnchor;
module.exports = exports.default;
