// @baidu/one-ui@4.41.1/lib/components/embedded/embedded.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _dlsIconsReact = require("dls-icons-react");

var _button = _interopRequireDefault(require("../button"));

var _commonTools = require("../../core/commonTools");

/**
 * @deprecated
 */
var Embedded = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Embedded, _PureComponent);

  function Embedded(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.onClose = function () {
      _this.setState({
        visible: false
      });

      _this.props.onClose();
    };

    _this.renderFooter = function () {
      var _this$props = _this.props,
          footer = _this$props.footer,
          onOk = _this$props.onOk,
          onCancel = _this$props.onCancel,
          okProps = _this$props.okProps,
          cancelProps = _this$props.cancelProps,
          okOrder = _this$props.okOrder,
          cancelOrder = _this$props.cancelOrder,
          buttonSize = _this$props.buttonSize,
          size = _this$props.size,
          okText = _this$props.okText,
          cancelText = _this$props.cancelText,
          hideDefaultFooter = _this$props.hideDefaultFooter;

      if (footer) {
        return footer;
      }

      var okStyle = okProps.style ? (0, _extends2["default"])({}, okProps.style, {
        order: okOrder
      }) : {
        order: okOrder
      };
      var cancelStyle = cancelProps.style ? (0, _extends2["default"])({}, cancelProps.style, {
        order: cancelOrder
      }) : {
        order: cancelOrder
      };
      var defaultFooter = [/*#__PURE__*/_react["default"].createElement(_button["default"], (0, _extends2["default"])({
        size: buttonSize || size,
        key: "confirm",
        onClick: onOk,
        type: "primary"
      }, okProps, {
        style: okStyle
      }), okText), /*#__PURE__*/_react["default"].createElement(_button["default"], (0, _extends2["default"])({
        size: buttonSize || size,
        onClick: onCancel,
        key: "cancel",
        type: "normal"
      }, cancelProps, {
        style: cancelStyle
      }), cancelText)];
      return !hideDefaultFooter ? defaultFooter : null;
    };

    _this.state = {
      visible: true
    };
    return _this;
  }

  var _proto = Embedded.prototype;

  _proto.render = function render() {
    var _this$props2 = this.props,
        title = _this$props2.title,
        prefixCls = _this$props2.prefixCls,
        className = _this$props2.className,
        closable = _this$props2.closable,
        children = _this$props2.children,
        size = _this$props2.size,
        position = _this$props2.position;
    var visible = this.state.visible;

    if ('visible' in this.props) {
      visible = this.props.visible;
    }

    if (!visible) {
      return null;
    }

    var footer = this.renderFooter();
    var embeddedClassNames = (0, _commonTools.classnames)(prefixCls, className, prefixCls + "-" + size);
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: embeddedClassNames
    }, closable ? /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-close",
      onClick: this.onClose
    }, /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconTimes, {
      className: prefixCls + "-close-icon"
    })) : null, /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-header"
    }, title), /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-body"
    }, children), footer ? /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-footer " + prefixCls + "-" + position
    }, footer) : null);
  };

  return Embedded;
}(_react.PureComponent);

exports["default"] = Embedded;
Embedded.propTypes = {
  /** 面板关闭时触发 */
  onClose: _propTypes["default"].func,

  /** 传入底部的button数组 */
  footer: _propTypes["default"].array,

  /** 面板的标题 */
  title: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** 面板是否可以关闭 */
  closable: _propTypes["default"].bool,

  /** 类名前缀 */
  prefixCls: _propTypes["default"].string,

  /** 自定义类名 */
  className: _propTypes["default"].string,
  children: _propTypes["default"].node,

  /** 面板是否可见 */
  visible: _propTypes["default"].bool,
  onOk: _propTypes["default"].func,
  onCancel: _propTypes["default"].func,
  okProps: _propTypes["default"].object,
  cancelProps: _propTypes["default"].object,
  okOrder: _propTypes["default"].number,
  cancelOrder: _propTypes["default"].number,

  /** 按钮的类型 */
  buttonSize: _propTypes["default"].oneOf(['small', 'medium', 'large', 'xsmall', 'xlarge']),
  size: _propTypes["default"].oneOf(['small', 'medium']),
  okText: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),
  cancelText: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),
  hideDefaultFooter: _propTypes["default"].bool,
  position: _propTypes["default"].oneOf(['left', 'center', 'right'])
};
Embedded.defaultProps = {
  footer: null,
  closable: true,
  prefixCls: 'one-embedded',
  className: '',
  onOk: function onOk() {},
  onCancel: function onCancel() {},
  okProps: {},
  cancelProps: {},
  okOrder: 1,
  cancelOrder: 2,
  size: 'medium',
  okText: '确定',
  cancelText: '取消',
  hideDefaultFooter: true,
  position: 'left'
};
module.exports = exports.default;
