// @baidu/one-ui@4.32.1/src/components/transfer/common/title.tsx
/**
 * @file 穿梭框的title
 * @author huangshiming
 * @date 2020-05-16
 */
import React, {ReactNode} from 'react';

interface TitleProps {
    title?: ReactNode;
    treeName?: ReactNode;
    selectedNum?: number;
    maxSelectedNum?: number;
    showSelectedNum?: boolean;
    showCandidateNum?: boolean;
    prefixCls?: string;
    candidateNum?: number;
};

const TitleRender = (props: TitleProps) => {
    const {
        title,
        treeName,
        selectedNum,
        maxSelectedNum,
        showSelectedNum,
        showCandidateNum,
        prefixCls,
        candidateNum
    } = props;
    // candidateNum 为备选区展示的个数，selectedNum为已选区展示的个数
    let titleDetail = (
        <span>
            {title}{treeName}
        </span>
    );
    let numberDom;
    if (selectedNum && showSelectedNum) {
        numberDom = (
            <span>
                ({selectedNum}{maxSelectedNum ? `/${maxSelectedNum}` : ''})
            </span>
        );
    }
    if (candidateNum && showCandidateNum) {
        numberDom = (
            <span>
                ({candidateNum > 0 ? candidateNum : 0})
            </span>
        );
    }
    return (
        <span className={`${prefixCls}-pane-title-text`} title={typeof title === 'string' ? title : undefined}>
            <span className={`${prefixCls}-pane-title-text-text`}>{titleDetail}</span>
            {
                numberDom ? <span className={`${prefixCls}-pane-title-text-number`}>
                    {numberDom}
                </span> : null
            }
        </span>
    );
};

export default TitleRender;
