// @baidu/one-ui@4.41.1/es/components/select/singleSelect.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _map from "lodash/map";
import React, { PureComponent } from 'react';
import { classnames } from '../../core/commonTools';
import Dropdown from '../dropdown';
import Menu from '../menu';
import { IconChevronRight } from 'dls-icons-react';
// import Icon from '../icon';
var MenuItem = Menu.Item;
var SubMenu = Menu.SubMenu;
var Divider = Menu.Divider;
;

var SingleSelect = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(SingleSelect, _PureComponent);

  function SingleSelect(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.onChange = function (e) {
      var onChange = _this.props.onChange;

      if (onChange) {
        onChange(e);
      }

      _this.setState({
        isExpend: false,
        value: e.key
      });
    };

    _this.getSubMenuItem = function (option, index) {
      var disabled = option.disabled,
          value = option.value,
          label = option.label;
      var menuItem = {
        key: value,
        disabled: disabled || false
      };

      if (option.divider) {
        return /*#__PURE__*/React.createElement(Divider, {
          key: index + "-divider"
        });
      }

      if (option.children && option.children.length) {
        return /*#__PURE__*/React.createElement(SubMenu, _extends({}, menuItem, {
          title: label
        }), option.children.map(function (child) {
          return _this.getSubMenuItem(child);
        }));
      }

      return /*#__PURE__*/React.createElement(MenuItem, menuItem, label);
    };

    _this.getDropdownOverlay = function () {
      var _this$props = _this.props,
          options = _this$props.options,
          dropdownHeight = _this$props.dropdownHeight;
      var value = _this.state.value;

      if ('value' in _this.props) {
        value = _this.props.value;
      }

      var menu = /*#__PURE__*/React.createElement(Menu, {
        selectable: true,
        selectedKeys: ["" + value],
        style: {
          height: dropdownHeight
        },
        onClick: _this.onChange,
        mode: "vertical"
      }, options.map(function (option, index) {
        return _this.getSubMenuItem(option, index);
      }));
      return menu;
    };

    _this.renderValueByOptions = function (options, value, label) {
      if (label === void 0) {
        label = '';
      }

      _map(options, function (option) {
        if ((typeof value === 'number' || typeof option.value === 'number') && +option.value === +value) {
          label = option.label;
        } else if (option.value === value) {
          label = option.label;
        } else if (option.children) {
          label = _this.renderValueByOptions(option.children, value, label);
        }
      });

      return label;
    };

    _this.getSelectedLabel = function () {
      var value = _this.state.value;

      if ('value' in _this.props) {
        value = _this.props.value;
      }

      var _this$props2 = _this.props,
          options = _this$props2.options,
          selectorName = _this$props2.selectorName;

      var label = _this.renderValueByOptions(options, value);

      return label || selectorName;
    };

    _this.dropdownVisibleChange = function (visible) {
      var onVisibleChange = _this.props.onVisibleChange;

      if (onVisibleChange) {
        onVisibleChange(visible);
      }

      _this.setState({
        isExpend: visible
      });
    };

    _this.state = {
      isExpend: false,
      value: props.value || ''
    };
    return _this;
  }

  var _proto = SingleSelect.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props3 = this.props,
        trigger = _this$props3.trigger,
        placement = _this$props3.placement,
        getPopupContainer = _this$props3.getPopupContainer,
        disabled = _this$props3.disabled,
        visible = _this$props3.visible,
        prefixCls = _this$props3.prefixCls,
        style = _this$props3.style;
    var overlay = this.getDropdownOverlay();
    var dropdownProps = {
      overlay: overlay,
      trigger: disabled ? [] : [trigger],
      onVisibleChange: this.dropdownVisibleChange,
      placement: placement,
      getPopupContainer: getPopupContainer
    };

    if ('visible' in this.props) {
      // @ts-ignore
      dropdownProps.visible = visible;
    }

    var isExpend = this.state.isExpend;
    var classes = classnames(prefixCls, (_classnames = {}, _classnames[prefixCls + "-open"] = isExpend || visible, _classnames[prefixCls + "-disabled"] = disabled, _classnames));
    return /*#__PURE__*/React.createElement(Dropdown, dropdownProps, /*#__PURE__*/React.createElement("span", {
      className: classes,
      style: style
    }, /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-text"
    }, this.getSelectedLabel()), /*#__PURE__*/React.createElement(IconChevronRight, null)));
  };

  return SingleSelect;
}(PureComponent);

SingleSelect.defaultProps = {
  prefixCls: 'one-single-select',
  disabled: false,
  dropdownHeight: 'auto',
  style: {},
  trigger: 'hover',
  selectorName: '请选择...'
};
export { SingleSelect as default };
