// @baidu/one-ui@4.32.1/src/components/button/interface.ts
import {MouseEvent, ReactNode} from 'react';
import {BasePropsWithChildren} from '../interface';

export type ButtonType =
    'normal' | 'basic' | 'strong' | 'primary' | 'translucent' | 'link'
    | 'text' | 'text-strong' | 'text-aux'
    | 'ghost' | 'ghost-strong' | 'ghost-aux' | 'ghost-reverse';

export type ButtonSize = 'large' | 'medium' | 'small' | 'xsmall' | 'xlarge';

export interface ButtonProps extends BasePropsWithChildren,
    Omit<JSX.IntrinsicElements['button'], 'type' | 'ref'> {
        /**
         * 普通按钮、基础按钮、加强按钮、重要按钮、半透明按钮、文字按钮、加强文字按钮、弱文字按钮
         */
        type?: ButtonType;
        /**
         * 名称
         */
        name?: string;
        /**
         * 尺寸
         */
        size?: ButtonSize;
        /**
         * 自定义icon
         */
        icon?: ReactNode;
        /**
         * 设置 button 原生的 type 值
         */
        htmlType?: 'submit' | 'button' | 'reset';
        /**
         * 禁用
         */
        disabled?: boolean;
        /**
         * 只读状态
         */
        readOnly?: boolean;
        /**
         * @deprecated
         */
        readonly?: boolean;
        /**
         * click 事件的回调
         */
        onClick?(event: MouseEvent): void;
        /**
         * loading 状态
         */
        loading?: boolean;
    };
