// @baidu/one-ui@4.32.1/src/components/timePicker/timePicker.tsx
import React, {PureComponent} from 'react';
import Trigger from 'rc-trigger';
import moment, {Moment} from 'moment';
import {generateOptions, toNearestValidTime, noop} from '../../core/timePickerTools';
import TimeInput from './common/timeInput';
import TimePickerPanel from './common/timePickerPanel';
import placements from './common/placements';
import {classnames} from '../../core/commonTools';
import {TimePickerProps} from './interface';
import {withConfigConsumer} from '../providerConfig/context';

export const getFormat = props => {
    const {format, showHour, showMinute, showSecond, use12Hours} = props;
    if (format) {
        return format;
    }

    if (use12Hours) {
        const fmtString = [showHour ? 'h' : '', showMinute ? 'mm' : '', showSecond ? 'ss' : '']
            .filter(item => !!item)
            .join(':');

        return fmtString.concat(' a');
    }

    return [showHour ? 'HH' : '', showMinute ? 'mm' : '', showSecond ? 'ss' : '']
        .filter(item => !!item)
        .join(':');
};

export const momentValue = (value, props) => {
    return value ? moment(value, getFormat(props), true) : null;
};

interface TimePickerState {
    value?: Moment;
    open?: boolean;
    dropdownWidth?: number;
};

@withConfigConsumer('time-picker')
class TimePicker extends PureComponent<TimePickerProps, TimePickerState> {

    static defaultProps = {
        prefixCls: 'one-time-picker',
        inputReadOnly: false,
        popupClassName: '',
        popupStyle: {},
        align: {
            ignoreShake: true
        },
        defaultOpenValue: moment(),
        format: 'HH:mm:ss',
        allowEmpty: true,
        showHour: true,
        showMinute: true,
        showSecond: true,
        disabledHours: noop,
        disabledMinutes: noop,
        disabledSeconds: noop,
        hideDisabledOptions: false,
        transitionName: 'one-transition-slide-down',
        placement: 'bottomLeft',
        onChange: noop,
        onAmPmChange: noop,
        onOpen: noop,
        onClose: noop,
        onFocus: noop,
        onBlur: noop,
        onKeyDown: noop,
        use12Hours: false,
        errorMessage: '',
        errorLocation: 'bottom',
        size: 'medium',
        width: 150
    };

    constructor(props) {
        super(props);
        const {open, value} = props;
        this.state = {
            open,
            value: momentValue(value, props),
            dropdownWidth: null
        };
    }

    timePickerTargetRef;
    panelInstance;

    componentDidMount() {
        this.setDropdownWidth();
    }

    componentDidUpdate() {
        this.setDropdownWidth();
    }

    static getDerivedStateFromProps(nextProps) {
        const {value, open} = nextProps;
        const newState: TimePickerState = {};
        if ('value' in nextProps) {
            newState.value = momentValue(value, nextProps);
        }
        if (open !== undefined) {
            newState.open = open;
        }
        return newState;
    }

    setDropdownWidth = () => {
        const width = this.timePickerTargetRef.offsetWidth;
        if (width !== this.state.dropdownWidth) {
            this.setState({dropdownWidth: width});
        }
    }

    saveRef = ref => {
        this.timePickerTargetRef = ref;
    }

    onPanelChange = value => {
        this.setValue(value);
    };

    onAmPmChange = ampm => {
        this.props.onAmPmChange(ampm);
    };

    onClear = event => {
        event.stopPropagation();
        this.setValue(null);
        this.setOpen(false);
    };

    onVisibleChange = open => {
        this.setOpen(open);
    };

    onEsc = () => {
        this.setOpen(false);
    };

    onKeyDown = e => {
        if (e.keyCode === 40) {
            this.setOpen(true);
        }
        this.props.onKeyDown(e);
    };

    onFocus = () => {
        this.props.onFocus();
    }

    onBlur = () => {
        this.props.onBlur();
    }

    getPlaceHolder = () => {
        const {placeholder, showHour, showMinute, showSecond} = this.props;
        if ('placeholder' in this.props) {
            return placeholder;
        }
        return [showHour ? '小时' : '', showMinute ? '分钟' : '', showSecond ? '秒钟' : '']
            .filter(item => !!item)
            .join(':');
    }

    setValue = value => {
        if (!('value' in this.props)) {
            this.setState({
                value
            });
        }
        let formatValue = value;
        if (value && typeof value === 'object' && value.format) {
            formatValue = value.format(getFormat(this.props));
        }
        this.props.onChange(formatValue);
    }

    setOpen = open => {
        const {onOpen, onClose} = this.props;
        const {open: currentOpen} = this.state;
        if (currentOpen !== open) {
            if (!('open' in this.props)) {
                this.setState({open});
            }
            if (open) {
                onOpen({open});
            }
            else {
                onClose({open});
            }
        }
    }

    getPopupClassName = () => {
        const {showHour, showMinute, showSecond, use12Hours, prefixCls, popupClassName} = this.props;
        let className = popupClassName;
        // Keep it for old compatibility
        if ((!showHour || !showMinute || !showSecond) && !use12Hours) {
            className += ` ${prefixCls}-panel-narrow`;
        }
        let selectColumnCount = 0;
        if (showHour) {
            selectColumnCount += 1;
        }
        if (showMinute) {
            selectColumnCount += 1;
        }
        if (showSecond) {
            selectColumnCount += 1;
        }
        if (use12Hours) {
            selectColumnCount += 1;
        }
        className += ` ${prefixCls}-panel-column-${selectColumnCount}`;
        return className;
    }

    getOptions = () => {
        const {
            disabledMinutes,
            disabledSeconds,
            hideDisabledOptions,
            hourStep,
            minuteStep,
            secondStep
        } = this.props;
        const value = this.state.value;
        const disabledHourOptions = this.disabledHours();
        const disabledMinuteOptions = disabledMinutes(value ? value.hour() : null);
        const disabledSecondOptions = disabledSeconds(
            value ? value.hour() : null,
            value ? value.minute() : null,
        );
        const hourOptions = generateOptions(24, disabledHourOptions, hideDisabledOptions, hourStep);
        const minuteOptions = generateOptions(
            60,
            disabledMinuteOptions,
            hideDisabledOptions,
            minuteStep,
        );
        const secondOptions = generateOptions(
            60,
            disabledSecondOptions,
            hideDisabledOptions,
            secondStep,
        );
        return {hourOptions, minuteOptions, secondOptions};
    }

    getPanelElement = () => {
        const {
            prefixCls,
            disabledMinutes,
            disabledSeconds,
            hideDisabledOptions,
            showHour,
            showMinute,
            showSecond,
            defaultOpenValue,
            use12Hours,
            size
        } = this.props;
        const value = this.state.value;
        const {hourOptions, minuteOptions, secondOptions} = this.getOptions();
        return (
            <TimePickerPanel
                prefixCls={`${prefixCls}-panel`}
                ref={ref => {
                    this.panelInstance = ref;
                }}
                className={`${prefixCls}-panel-${size}`}
                value={value}
                onChange={this.onPanelChange}
                onAmPmChange={this.onAmPmChange}
                defaultOpenValue={defaultOpenValue}
                showHour={showHour}
                showMinute={showMinute}
                showSecond={showSecond}
                onEsc={this.onEsc}
                format={getFormat(this.props)}
                disabledHours={this.disabledHours}
                disabledMinutes={disabledMinutes}
                disabledSeconds={disabledSeconds}
                use12Hours={use12Hours}
                hourOptions={hourOptions}
                minuteOptions={minuteOptions}
                secondOptions={secondOptions}
                isAM={this.isAM()}
            />
        );
    }

    disabledHours = () => {
        const {use12Hours, disabledHours} = this.props;
        let disabledOptions = disabledHours();
        if (use12Hours && Array.isArray(disabledOptions)) {
            if (this.isAM()) {
                disabledOptions = disabledOptions.filter(h => h < 12).map(h => (h === 0 ? 12 : h));
            }
            else {
                disabledOptions = disabledOptions.map(h => (h === 12 ? 12 : h - 12));
            }
        }
        return disabledOptions;
    };

    isAM() {
        const defaultOpenValue = this.props.defaultOpenValue;
        const value = this.state.value;
        const realValue = value || defaultOpenValue;
        return realValue.hour() >= 0 && realValue.hour() < 12;
    }

    render() {
        const {
            prefixCls,
            placement,
            align,
            disabled,
            style,
            className,
            getPopupContainer,
            name,
            inputReadOnly,
            popupStyle,
            defaultOpenValue,
            disabledMinutes,
            disabledSeconds,
            allowEmpty,
            errorMessage,
            errorLocation,
            size,
            width,
            transitionName
        } = this.props;
        const {open, value} = this.state;
        const popupClassName = this.getPopupClassName();

        const {hourOptions, minuteOptions, secondOptions} = this.getOptions();
        const validDefaultOpenValue = toNearestValidTime(
            defaultOpenValue,
            hourOptions,
            minuteOptions,
            secondOptions,
        );
        const errorClass = `${prefixCls}-error`;
        const errorProps = {
            className: classnames(
                errorClass,
                `${errorClass}-${errorLocation}`
            )
        };
        const timePickerClassName = classnames(
            prefixCls,
            className,
            `${prefixCls}-${size}`,
            {
                [`${prefixCls}-disabled`]: disabled,
                [`${prefixCls}-open`]: open
            }
        );
        const timePickerStyle = {
            ...popupStyle
        };
        const dropdownWidth = this.state.dropdownWidth;
        if (dropdownWidth) {
            timePickerStyle.minWidth = `${dropdownWidth}px`;
        }

        const timePickerElement = (
            <span
                className={timePickerClassName}
            >
                <Trigger
                    prefixCls={`${prefixCls}-panel`}
                    popupClassName={popupClassName}
                    popupStyle={timePickerStyle}
                    popup={this.getPanelElement()}
                    popupAlign={align}
                    builtinPlacements={placements}
                    popupPlacement={placement}
                    popupTransitionName={transitionName}
                    action={disabled ? [] : ['click']}
                    destroyPopupOnHide
                    getPopupContainer={getPopupContainer}
                    popupVisible={open}
                    onPopupVisibleChange={this.onVisibleChange}
                >
                    <span className={`${prefixCls}-input-container`} ref={this.saveRef}>
                        <TimeInput
                            style={style}
                            prefixCls={prefixCls}
                            defaultOpenValue={validDefaultOpenValue}
                            value={value}
                            onEsc={this.onEsc}
                            format={getFormat(this.props)}
                            placeholder={this.getPlaceHolder()}
                            hourOptions={hourOptions}
                            minuteOptions={minuteOptions}
                            secondOptions={secondOptions}
                            disabledHours={this.disabledHours}
                            disabledMinutes={disabledMinutes}
                            disabledSeconds={disabledSeconds}
                            onChange={this.onPanelChange}
                            allowEmpty={allowEmpty}
                            onKeyDown={this.onKeyDown}
                            inputReadOnly={inputReadOnly}
                            onFocus={this.onFocus}
                            onBlur={this.onBlur}
                            disabled={disabled}
                            name={name}
                            onClear={this.onClear}
                            errorMessage={errorMessage}
                            size={size}
                            width={width}
                        />
                    </span>
                </Trigger>
                {errorMessage ? <div {...errorProps}>{errorMessage}</div> : null}
            </span>
        );
        return timePickerElement;
    }
}

export default TimePicker;
