// @baidu/one-ui@4.32.1/src/components/datePicker/common/monthAndYearPanel.tsx
import React, {PureComponent} from 'react';
import {connect} from 'mini-store';
import YearPanel from './yearPanel';
import MonthPanel from './monthPanel';
import {DatePickerStore} from '../datePicker';

interface PanelRenderProps {
    prefixCls: string,
    panelType: string,
    endDatePanelType: string,
    multiple: boolean,
    type: string
};

class PanelRender extends PureComponent<PanelRenderProps> {

    render() {
        const {prefixCls, panelType, endDatePanelType, multiple = false, type} = this.props;
        if (!multiple && panelType === 'date') {
            return null;
        }
        if (multiple && type === 'prevMultiple' && panelType === 'date') {
            return null;
        }
        if (multiple && type === 'nextMultiple' && endDatePanelType === 'date') {
            return null;
        }
        return (
            <div className={`${prefixCls}-panel-container`}>
                <YearPanel {...this.props} />
                <MonthPanel {...this.props} />
            </div>
        );
    }
}

export default connect((state: DatePickerStore) => {
    return {
        panelType: state.panelType,
        endDatePanelType: state.endDatePanelType
    };
})(PanelRender);
