// @baidu/one-ui@4.32.1/src/components/datePicker/common/shortCut.tsx
import React, {PureComponent} from 'react';
import {partial} from 'lodash';
import dayjs from 'dayjs';
import {
    getShortCutDate,
    getTimeTramp
} from '../../../core/datePickerTools';

interface ShortCutProps {
    shortcuts?: {
        label: string;
        from: number | {
            startOf: 'month' | 'week' | 'day',
            months?: number
            weeks?: number
            days?: number
        };
        to: number | {
            startOf: 'month' | 'week' | 'day',
            months?: number
            weeks?: number
            days?: number
        };
    }[],
    prefixCls?: string,
    dateFormat?: string,
    onChange(value: string[]): void
}

export default class ShortCut extends PureComponent<ShortCutProps> {
    static defaultProps = {
        onChange() {}
    }

    onClickShortCut = (index: number) => {
        const {shortcuts, dateFormat, onChange} = this.props;
        const shortcut = shortcuts[index];
        const {
            beginDate,
            endDate
        } = getShortCutDate(shortcut, dateFormat);
        let currentBeginDate = beginDate;
        let currentEndDate = endDate;
        if (getTimeTramp(currentBeginDate) > getTimeTramp(currentEndDate)) {
            currentBeginDate = endDate;
            currentEndDate = beginDate;
        }
        onChange([
            dayjs(new Date(currentBeginDate)).format(dateFormat),
            dayjs(new Date(currentEndDate)).format(dateFormat)
        ]);
    }

    render() {
        const {prefixCls, shortcuts} = this.props;
        return (
            <div className={`${prefixCls}-short-cut`}>
                {
                    shortcuts.map((shortcut, index) => {
                        return (
                            <div
                                className={`${prefixCls}-short-cut-item`}
                                onClick={partial(this.onClickShortCut, index)}
                                key={index}
                            >
                                {shortcut.label}
                            </div>
                        );
                    })
                }
            </div>
        );
    }
}
