// @baidu/one-ui@4.32.1/src/components/datePicker/common/titleHeader.tsx
import React, {PureComponent} from 'react';
import {connect} from 'mini-store';
import {partial} from 'lodash';
import {
    IconChevronDoubleLeft,
    IconChevronDoubleRight,
    IconChevronLeft,
    IconChevronRight,
    IconChevronDown,
    IconChevronUp
} from 'dls-icons-react';
import {
    getDetailDate,
    monthDayRange,
    getTimeTramp
} from '../../../core/datePickerTools';
import {DatePickerStore} from '../datePicker';


interface TitleHeaderProps {
    store: Object,
    prefixCls: string,
    showYear: number,
    showMonth: number,
    panelType: string,
    validateMinDate: string,
    validateMaxDate: string,
    endDateShowYear: number,
    endDateShowMonth: number,
    endDatePanelType: string,
    multiple: boolean,
    isMonthRender: boolean
}

interface TitleHeaderState {
    endDateShowYear?: number;
    showYear?: number;
    endDateShowMonth?: number;
    showMonth?: number;
    endDatePanelType?: 'month' | 'date';
    panelType?: 'year' | 'month' | 'date';
}

type ChangeType = 'single' | 'multipleNext' | 'multiplePrev';

class TitleHeader extends PureComponent<TitleHeaderProps> {

    static defaultProps = {
        multiple: false
    }

    constructor(props: TitleHeaderProps) {
        super(props);
        this.store = this.props.store;
    }
    store;
    onGoToPrevYear = (type: ChangeType) => {
        const {showYear, endDateShowYear} = this.props;
        const currentYear = type === 'multipleNext' ? endDateShowYear : showYear;
        const prevYear = currentYear - 1;
        const newState: TitleHeaderState = {};
        if (type === 'multipleNext') {
            newState.endDateShowYear = prevYear;
        }
        else {
            newState.showYear = prevYear;
        }
        this.store.setState(newState);
    }

    onGoToPrevMonth = (type: ChangeType) => {
        const {showMonth, endDateShowMonth, showYear, endDateShowYear} = this.props;
        const currentMonth = type === 'multipleNext' ? endDateShowMonth : showMonth;
        let currentYear = type === 'multipleNext' ? endDateShowYear : showYear;
        let prevMonth = currentMonth - 1;
        if (currentMonth === 1) {
            prevMonth = 12;
            currentYear--;
        }
        const newState: TitleHeaderState = {};
        if (type === 'multipleNext') {
            newState.endDateShowMonth = prevMonth;
            newState.endDateShowYear = currentYear;
        }
        else {
            newState.showMonth = prevMonth;
            newState.showYear = currentYear;
        }
        this.store.setState(newState);
    }

    onGoToNextYear = (type: ChangeType) => {
        const {showYear, endDateShowYear} = this.props;
        const currentYear = type === 'multipleNext' ? endDateShowYear : showYear;
        const nextYear = currentYear + 1;
        const newState: TitleHeaderState = {};
        if (type === 'multipleNext') {
            newState.endDateShowYear = nextYear;
        }
        else {
            newState.showYear = nextYear;
        }
        this.store.setState(newState);
    }

    onGoToNextMonth = (type: ChangeType) => {
        const {showMonth, endDateShowMonth, showYear, endDateShowYear} = this.props;
        const currentMonth = type === 'multipleNext' ? endDateShowMonth : showMonth;
        let currentYear = type === 'multipleNext' ? endDateShowYear : showYear;
        let nextMonth = currentMonth + 1;
        if (currentMonth === 12) {
            nextMonth = 1;
            currentYear++;
        }
        const newState: TitleHeaderState = {};
        if (type === 'multipleNext') {
            newState.endDateShowMonth = nextMonth;
            newState.endDateShowYear = currentYear;
        }
        else {
            newState.showMonth = nextMonth;
            newState.showYear = currentYear;
        }
        this.store.setState(newState);
    }

    togglePanel = (type: ChangeType) => {
        const newState: TitleHeaderState = {};
        if (type === 'multipleNext') {
            newState.endDatePanelType = this.props.endDatePanelType === 'date' ? 'month' : 'date';
        }
        else if (!this.props.isMonthRender) {
            newState.panelType = this.props.panelType === 'date' ? 'month' : 'date';
        }
        else {
            newState.panelType = this.props.panelType === 'month' ? 'year' : 'month';
        }
        this.store.setState(newState);
    }

    showChangeYearIcon = (type: ChangeType) => {
        const {
            validateMinDate,
            validateMaxDate,
            showYear,
            endDateShowYear
        } = this.props;
        const minYear = getDetailDate(validateMinDate).fullYear;
        const maxYear = getDetailDate(validateMaxDate).fullYear;
        let showPrevYearIcon;
        let showNextYearIcon;
        if (type === 'single') {
            showPrevYearIcon = minYear < showYear;
            showNextYearIcon = maxYear > showYear;
        }
        else if (type === 'multiplePrev') {
            // 多选中的第一个面板
            showNextYearIcon = showYear < endDateShowYear;
            showPrevYearIcon = minYear < showYear;
        }
        else if (type === 'multipleNext') {
            // 多选中的第二个面板
            showNextYearIcon = maxYear > endDateShowYear;
            showPrevYearIcon = endDateShowYear > showYear;
        }
        return {
            showPrevYearIcon,
            showNextYearIcon
        };
    }

    showChangeMonthIcon = (type: ChangeType) => {
        const {
            validateMinDate,
            validateMaxDate,
            showYear,
            showMonth,
            endDateShowYear,
            endDateShowMonth
        } = this.props;
        const minDate = getDetailDate(validateMinDate);
        const maxDate = getDetailDate(validateMaxDate);
        let firstDayInCurrentMonth = `${showYear}/${showMonth}/01`;
        let lastDayInCurrentMonth = `${showYear}/${showMonth}/${monthDayRange(showYear)[showMonth - 1]}`;
        if (type === 'multipleNext') {
            firstDayInCurrentMonth = `${endDateShowYear}/${endDateShowMonth}/01`;
            lastDayInCurrentMonth
                = `${endDateShowYear}/${endDateShowMonth}/${monthDayRange(endDateShowYear)[endDateShowMonth - 1]}`;
        }
        const minDateStamp = getTimeTramp(`${minDate.fullYear}/${minDate.fullMonth}/01`);
        const maxDateStamp
            = getTimeTramp(
                `${maxDate.fullYear}/${maxDate.fullMonth}/${monthDayRange(maxDate.fullYear)[maxDate.fullMonth - 1]}`
            );
        let showPrevMonthIcon;
        let showNextMonthIcon;
        if (type === 'single') {
            showPrevMonthIcon = getTimeTramp(firstDayInCurrentMonth) > minDateStamp;
            showNextMonthIcon = getTimeTramp(lastDayInCurrentMonth) < maxDateStamp;
        }
        else if (type === 'multiplePrev') {
            // 多选面板中的第一个面板
            showPrevMonthIcon = getTimeTramp(firstDayInCurrentMonth) > minDateStamp;
            let endDate = `${endDateShowYear}/${endDateShowMonth - 1}/01`;
            if (endDateShowMonth === 1) {
                endDate = `${endDateShowYear - 1}/12/01`;
            }
            showNextMonthIcon = getTimeTramp(firstDayInCurrentMonth) < getTimeTramp(endDate);
        }
        else if (type === 'multipleNext') {
            // 多选面板的第二个面板
            let firstDate = `${showYear}/${showMonth + 1}/01`;
            if (showMonth === 12) {
                firstDate = `${endDateShowYear + 1}/01/01`;
            }
            showPrevMonthIcon = getTimeTramp(firstDayInCurrentMonth) > getTimeTramp(firstDate);
            showNextMonthIcon = getTimeTramp(lastDayInCurrentMonth) < maxDateStamp;
        }
        return {
            showPrevMonthIcon,
            showNextMonthIcon
        };
    }

    renderItem = (type: ChangeType = 'single') => {
        const {
            prefixCls,
            showYear,
            showMonth,
            endDateShowYear,
            endDateShowMonth,
            endDatePanelType,
            panelType,
            isMonthRender
        } = this.props;
        let currentPanelType = panelType;
        let currentShowYear = showYear;
        let currentShowMonth = showMonth;
        if (type === 'multipleNext') {
            currentPanelType = endDatePanelType;
            currentShowYear = endDateShowYear;
            currentShowMonth = endDateShowMonth;
        }
        const classNames = `${prefixCls}-title-header`;
        let iconType = currentPanelType === 'date'
            ? <IconChevronDown className={`${prefixCls}-title-arrow-down`} />
            : <IconChevronUp className={`${prefixCls}-title-arrow-up`} />;
        if (isMonthRender) {
            iconType = currentPanelType === 'month'
                ? <IconChevronDown className={`${prefixCls}-title-arrow-down`} />
                : <IconChevronUp className={`${prefixCls}-title-arrow-up`} />;
        }
        const showYearIcon = this.showChangeYearIcon(type);
        const showMonthIcon = this.showChangeMonthIcon(type);
        return (
            <div className={classNames}>
                {
                    showYearIcon.showPrevYearIcon ? (
                        <span
                            className={`${classNames}-link-year ${classNames}-link-year-prev`}
                            onClick={partial(this.onGoToPrevYear, type)}
                        >
                            <IconChevronDoubleLeft />
                        </span>
                    ) : null
                }
                {
                    showMonthIcon.showPrevMonthIcon && !isMonthRender ? (
                        <span
                            className={`${classNames}-link-month ${classNames}-link-month-prev`}
                            onClick={partial(this.onGoToPrevMonth, type)}
                        >
                            <IconChevronLeft />
                        </span>
                    ) : null
                }
                <span className={`${classNames}-content`} onClick={partial(this.togglePanel, type)}>
                    <span>
                        {`${currentShowYear}年`}
                        {!isMonthRender ? `${currentShowMonth}月` : null}
                    </span>
                    {iconType}
                </span>
                {
                    showYearIcon.showNextYearIcon ? (
                        <span
                            className={`${classNames}-link-month ${classNames}-link-month-next`}
                            onClick={partial(this.onGoToNextYear, type)}
                        >
                            <IconChevronDoubleRight />
                        </span>
                    ) : null
                }
                {
                    showMonthIcon.showNextMonthIcon && !isMonthRender ? (
                        <span
                            className={`${classNames}-link-year ${classNames}-link-year-next`}
                            onClick={partial(this.onGoToNextMonth, type)}
                        >
                            <IconChevronRight />
                        </span>
                    ) : null
                }
            </div>
        );
    }

    renderMultipleHeader = () => {
        const prefixCls = this.props.prefixCls;
        return (
            <div className={`${prefixCls}-multiple-header`}>
                <span>{this.renderItem('multiplePrev')}</span>
                <span>{this.renderItem('multipleNext')}</span>
            </div>
        );
    }

    render() {
        const multiple = this.props.multiple;
        return multiple ? this.renderMultipleHeader() : this.renderItem();
    }
}

export default connect((state: DatePickerStore) => {
    return {
        showYear: state.showYear,
        showMonth: state.showMonth,
        panelType: state.panelType,
        validateMinDate: state.validateMinDate,
        validateMaxDate: state.validateMaxDate,
        endDateShowYear: state.endDateShowYear,
        endDateShowMonth: state.endDateShowMonth,
        endDatePanelType: state.endDatePanelType
    };
})(TitleHeader);
