// @baidu/one-ui@4.32.1/lib/components/dialog/common/dialogWrapper.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _Portal = _interopRequireDefault(require("rc-util/lib/Portal"));

var _dialog = _interopRequireDefault(require("./dialog"));

var DialogWrap = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(DialogWrap, _Component);

  function DialogWrap() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;
    _this._component = null;

    _this.renderComponent = function () {};

    _this.removeContainer = function () {};

    _this.saveDialog = function (node) {
      _this._component = node;
    };

    _this.getComponent = function (extra) {
      if (extra === void 0) {
        extra = {};
      }

      return /*#__PURE__*/_react["default"].createElement(_dialog["default"], (0, _extends2["default"])({
        ref: _this.saveDialog
      }, _this.props, extra, {
        key: "dialog"
      }));
    };

    _this.getContainer = function () {
      var container = document.createElement('div');

      if (_this.props.getContainer) {
        _this.props.getContainer().appendChild(container);
      } else {
        document.body.appendChild(container);
      }

      return container;
    };

    return _this;
  }

  var _proto = DialogWrap.prototype;

  _proto.shouldComponentUpdate = function shouldComponentUpdate(_ref) {
    var visible = _ref.visible,
        forceRender = _ref.forceRender;
    return !!(this.props.visible || visible) || this.props.forceRender || forceRender;
  };

  _proto.render = function render() {
    var _this$props = this.props,
        visible = _this$props.visible,
        forceRender = _this$props.forceRender;
    var portal = null;

    if (visible || forceRender || this._component) {
      portal = /*#__PURE__*/_react["default"].createElement(_Portal["default"], {
        getContainer: this.getContainer
      }, this.getComponent());
    }

    return portal;
  };

  return DialogWrap;
}(_react.Component);

DialogWrap.defaultProps = {
  visible: false,
  forceRender: false
};
var _default = DialogWrap;
exports["default"] = _default;
module.exports = exports.default;
