// @baidu/one-ui@4.32.1/lib/components/switch/switch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _omit = _interopRequireDefault(require("omit.js"));

var _switch = _interopRequireDefault(require("./common/switch"));

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

var defaultSize = 'medium';
var Switch = (_dec = (0, _context.withConfigConsumer)('switch'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Switch, _PureComponent);

  function Switch() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.rcSwitch = void 0;

    _this.saveSwitch = function (node) {
      _this.rcSwitch = node;
    };

    return _this;
  }

  var _proto = Switch.prototype;

  _proto.focus = function focus() {
    this.rcSwitch.focus();
  };

  _proto.blur = function blur() {
    this.rcSwitch.blur();
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        loading = _this$props.loading,
        _this$props$className = _this$props.className,
        className = _this$props$className === void 0 ? '' : _this$props$className,
        rawSize = _this$props.size;
    var size = ['xsmall', 'small', 'medium'].includes(rawSize) ? rawSize : defaultSize;
    var classes = (0, _commonTools.classnames)(className, (_classnames = {}, _classnames[prefixCls + "-" + size] = size, _classnames[prefixCls + "-loading"] = loading, _classnames));
    var otherProps = {
      className: classes,
      ref: this.saveSwitch,
      loading: loading
    };
    return /*#__PURE__*/_react["default"].createElement(_switch["default"], (0, _extends2["default"])({}, (0, _omit["default"])(this.props, ['size']), otherProps));
  };

  return Switch;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-switch',
  loading: false,
  size: defaultSize
}, _class2.displayName = 'Switch', _temp)) || _class);
var _default = Switch;
exports["default"] = _default;
module.exports = exports.default;
