// @baidu/one-ui@4.32.1/es/components/datePicker/datePicker.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _noop from "lodash/noop";
import _isEqual from "lodash/isEqual";
import _isEmpty from "lodash/isEmpty";
import _isArray from "lodash/isArray";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import { Provider, create } from 'mini-store';
import { IconCalendar, IconTimesCircle } from 'dls-icons-react';
import { getDetailDate, getTodayDetail, transDateFormat, validateData, formatButtonText } from '../../core/datePickerTools';
import TitleHeader from './common/titleHeader';
import DayItemRender from './common/dayItemRender';
import MonthAndYearPanel from './common/monthAndYearPanel';
import { classnames } from '../../core/commonTools';
import Input from './common/input';
import Layer from '../overlay';
import Button from '../button';
import { align } from './util';
import { withConfigConsumer } from '../providerConfig/context';
import { getValidWeekDates } from './util';

var normalizeValue = function normalizeValue(value) {
  if (_isEmpty(value)) {
    return [];
  }

  if (_isArray(value)) {
    return value;
  }

  return [value];
};

var updateValue = function updateValue(value) {
  var date = value[0] ? getDetailDate(value[0]) : getTodayDetail();
  return {
    _value: value,
    showYear: date.fullYear,
    showMonth: date.fullMonth
  };
};

var DatePicker = (_dec = withConfigConsumer('date-picker'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(DatePicker, _PureComponent);

  function DatePicker(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.store = void 0;

    _this.onMouseEnter = function (value) {
      _this.setState({
        hoverDate: value
      });
    };

    _this.onMouseLeave = function () {
      _this.setState({
        hoverDate: ''
      });
    };

    _this.componentDidUpdate = function (prevProps) {
      var newState = {};

      if ('value' in _this.props && !_isEqual(_this.props.value, prevProps.value)) {
        var _value = normalizeValue(_this.props.value); // 控制一些组件内部属性


        newState = _extends({}, newState, updateValue(_value));
      }

      if ('validateMinDate' in _this.props && _this.props.validateMinDate !== prevProps.validateMinDate) {
        newState = _extends({}, newState, {
          validateMinDate: transDateFormat(_this.props.validateMinDate) || validateData().validateMinDate
        });
      }

      if ('validateMaxDate' in _this.props && _this.props.validateMaxDate !== prevProps.validateMaxDate) {
        newState = _extends({}, newState, {
          validateMaxDate: transDateFormat(_this.props.validateMaxDate) || validateData().validateMaxDate
        });
      }

      _this.store.setState(newState);
    };

    _this.onSelectDay = function (value, colsePanel) {
      if (colsePanel === void 0) {
        colsePanel = true;
      }

      var _this$props = _this.props,
          onChange = _this$props.onChange,
          validator = _this$props.validator,
          dateFormat = _this$props.dateFormat,
          validateMaxDate = _this$props.validateMaxDate,
          validateMinDate = _this$props.validateMinDate,
          mode = _this$props.mode;
      var newState = {
        hoverDate: ''
      };
      var date = Array.isArray(value) ? value.filter(Boolean)[0] : value;
      var currentValue = mode === 'week' ? getValidWeekDates(date, dateFormat, validateMinDate, validateMaxDate) : normalizeValue(date);

      if (!('value' in _this.props)) {
        // 非受控
        newState.currentValue = currentValue;
      }

      _this.store.setState(updateValue(currentValue));

      if (!('visible' in _this.props) && colsePanel) {
        // 非受控
        newState.visible = false;
      }

      var validatorError = '';
      validatorError = validator && typeof validator === 'function' ? validator(mode === 'week' ? currentValue : date) : '';

      if (!validatorError) {
        onChange(mode === 'week' ? currentValue : date);
        newState.errorMessage = '';
      } else {
        newState.errorMessage = validatorError;
      }

      _this.setState(newState);
    };

    _this.renderDatePickerBody = function () {
      var _this$props2 = _this.props,
          prefixCls = _this$props2.prefixCls,
          size = _this$props2.size;
      var visible = _this.state.visible;
      var datePickerPanelClassName = classnames(prefixCls + "-panel", prefixCls + "-panel-" + size);
      return /*#__PURE__*/React.createElement(Provider, {
        store: _this.store
      }, /*#__PURE__*/React.createElement("div", {
        className: datePickerPanelClassName
      }, /*#__PURE__*/React.createElement(Input, _extends({}, _this.props, {
        visible: visible,
        onChange: _this.onSelectDay
      })), /*#__PURE__*/React.createElement(TitleHeader, _this.props), /*#__PURE__*/React.createElement(DayItemRender, _extends({}, _this.props, {
        onMouseEnter: _this.onMouseEnter,
        onMouseLeave: _this.onMouseLeave,
        hoverDate: _this.state.hoverDate,
        onChange: _this.onSelectDay
      })), /*#__PURE__*/React.createElement(MonthAndYearPanel, _this.props)));
    };

    _this.onLayerVisibleChange = function (visible) {
      _this.props.onVisibleChange(visible);

      if (!('visible' in _this.props)) {
        _this.setState({
          visible: visible
        });
      }
    };

    _this.onOpenDatePicker = function (e) {
      var _this$props3 = _this.props,
          onClickButton = _this$props3.onClickButton,
          value = _this$props3.value;
      var formatValue;

      if ('value' in _this.props) {
        formatValue = normalizeValue(value);
      } else {
        formatValue = _this.state.currentValue;
      } // 每次打开的时候，重置一下面板的type，并且保证当前value与state的value保持一致


      _this.store.setState(_extends({
        panelType: 'date'
      }, updateValue(formatValue)));

      _this.onLayerVisibleChange(true);

      onClickButton(e);
    };

    _this.onDetete = function (e) {
      e.stopPropagation();
      e.preventDefault();

      _this.props.onDelete();

      if (!('value' in _this.props)) {
        _this.setState({
          currentValue: []
        });

        _this.store.setState({
          _value: []
        });
      }
    };

    var defaultValue = props.defaultValue,
        _value2 = props.value,
        _visible = props.visible,
        defaultVisible = props.defaultVisible;

    var _value3 = normalizeValue(defaultValue || _value2);

    _this.state = {
      currentValue: _value3,
      visible: defaultVisible || _visible || false,
      errorMessage: '',
      hoverDate: ''
    };
    _this.store = create(_extends({}, updateValue(_value3), {
      // 面板展示 date or month - 选择日期 or 月份
      panelType: 'date',
      validateMinDate: transDateFormat(props.validateMinDate) || validateData().validateMinDate,
      validateMaxDate: transDateFormat(props.validateMaxDate) || validateData().validateMaxDate
    }));
    return _this;
  }

  var _proto = DatePicker.prototype;

  _proto.render = function render() {
    var _classnames, _classnames2;

    var _this$props4 = this.props,
        prefixCls = _this$props4.prefixCls,
        className = _this$props4.className,
        disabled = _this$props4.disabled,
        getPopupContainer = _this$props4.getPopupContainer,
        dateFormat = _this$props4.dateFormat,
        popupPlacement = _this$props4.popupPlacement,
        size = _this$props4.size,
        customButtonTitle = _this$props4.customButtonTitle,
        showDeleteIcon = _this$props4.showDeleteIcon,
        placeholder = _this$props4.placeholder,
        style = _this$props4.style,
        value = _this$props4.value;
    var _this$state = this.state,
        currentValue = _this$state.currentValue,
        visible = _this$state.visible,
        errorMessage = _this$state.errorMessage;
    var formatValue;

    if ('value' in this.props) {
      formatValue = normalizeValue(value);
    } else {
      formatValue = currentValue;
    }

    var buttonText = formatButtonText({
      value: formatValue,
      dateFormat: dateFormat
    });
    var text = customButtonTitle || buttonText || placeholder;
    var datePickerClassNames = classnames(prefixCls, className, (_classnames = {}, _classnames[prefixCls + "-empty"] = !buttonText, _classnames[prefixCls + "-disabled"] = disabled, _classnames[prefixCls + "-showDeleteIcon"] = showDeleteIcon, _classnames));
    var layerProps = {
      trigger: disabled ? '' : 'click',
      visible: visible,
      onVisibleChange: this.onLayerVisibleChange,
      overlay: this.renderDatePickerBody(),
      dropdownMatchSelectWidth: false,
      getPopupContainer: getPopupContainer,
      popupPlacement: popupPlacement,
      popupClassName: classnames(prefixCls + "-overlay", prefixCls + "-overlay-" + size),
      align: align
    };
    var textClassName = classnames(prefixCls + "-title-text", (_classnames2 = {}, _classnames2[prefixCls + "-title-text-empty"] = _isEmpty(formatValue), _classnames2));
    return /*#__PURE__*/React.createElement("div", {
      className: datePickerClassNames,
      style: style
    }, /*#__PURE__*/React.createElement(Layer, layerProps, /*#__PURE__*/React.createElement(Button, {
      className: prefixCls + "-title",
      disabled: disabled,
      size: size,
      type: "basic",
      onClick: this.onOpenDatePicker
    }, /*#__PURE__*/React.createElement("span", {
      className: textClassName
    }, text), /*#__PURE__*/React.createElement(IconCalendar, {
      className: prefixCls + "-button-icon " + prefixCls + "-button-icon-calendar"
    }), /*#__PURE__*/React.createElement(IconTimesCircle, {
      className: prefixCls + "-button-icon " + prefixCls + "-button-icon-close",
      onClick: this.onDetete
    }))), errorMessage ? /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-error-message"
    }, errorMessage) : null);
  };

  return DatePicker;
}(PureComponent), _class2.defaultProps = {
  dateFormat: 'YYYY/MM/DD',
  prefixCls: 'one-date-picker',
  disabled: false,
  size: 'medium',
  onVisibleChange: _noop,
  onChange: _noop,
  onClickButton: _noop,
  getPopupContainer: function getPopupContainer() {
    return document.body;
  },
  onDelete: _noop,
  showDeleteIcon: false,
  placeholder: '请选择日期'
}, _class2.RangePicker = void 0, _class2.MonthPicker = void 0, _class2.getDerivedStateFromProps = function (nextProps) {
  var newState = {};

  if ('value' in nextProps) {
    newState.currentValue = normalizeValue(nextProps.value);
  }

  if ('visible' in nextProps) {
    newState.visible = nextProps.visible;
  }

  return newState;
}, _temp)) || _class);
export default DatePicker;
