// @baidu/one-ui@4.32.1/src/components/alert/alert.tsx
import React, {PureComponent} from 'react';
import classes from 'component-classes';
import {noop} from 'lodash';
import {
    IconTimes,
    IconCheckCircleSolid,
    IconExclamationCircleSolid,
    IconTimesCircleSolid,
    IconInfoCircleSolid
} from 'dls-icons-react';
import {classnames} from '../../core/commonTools';
import IconSvg from '../iconSvg';
import {withConfigConsumer} from '../providerConfig/context';
import {AlertProps} from './interface';
import AlertPage from './page';
import omit from 'omit.js';

const defaultHeight = 32;

interface AlertState {
    visible?: boolean;
}

@withConfigConsumer('alert')
class Alert extends PureComponent<AlertProps, AlertState> {
    static defaultProps = {
        onClose: noop,
        prefixCls: 'one-alert',
        type: 'info',
        size: 'medium'
    };

    static Page: typeof AlertPage;

    constructor(props) {
        super(props);
        this.state = {
            visible: props.visible || true
        };
    }

    alertContentRef;
    alertRef;

    componentDidMount = () => {
        this.setDomClassByHeight();
    }

    componentDidUpdate = () => {
        this.setDomClassByHeight();
    }

    setDomClassByHeight = () => {
        const alertContentRef = this.alertContentRef;
        const alertRef = this.alertRef;
        if (!alertContentRef || !alertRef) {
            return null;
        }
        const height = alertContentRef.offsetHeight;
        const multipleClassName = `${this.props.prefixCls}-multiple`;
        const singleClassName = `${this.props.prefixCls}-single`;
        const dom = classes(alertRef);
        if (height > defaultHeight) {
            dom.remove(singleClassName);
            dom.add(multipleClassName);
        }
        else if (dom.has(multipleClassName)) {
            dom.remove(multipleClassName);
            dom.add(singleClassName);
        }
    }

    static getDerivedStateFromProps = (nextProps, prevState) => {
        if ('visible' in nextProps && nextProps.visible !== prevState.visible) {
            return {visible: nextProps.visible};
        }
        return null;
    }

    onHandleClose = e => {
        e.preventDefault();
        if (!('visble' in this.props)) {
            this.setState({
                visible: false
            });
        }
        const onClose = this.props.onClose;
        if (onClose) {
            onClose(e);
        }
    }

    saveRef = type => ref => {
        this[type] = ref;
    }

    render() {
        const {className, prefixCls, showIcon, icon, type, content, closable, title, size, ...rest} = this.props;
        const visible = this.state.visible;
        if (!visible) {
            return null;
        }
        let iconNode = null;
        if (!icon) {
            switch (type) {
                case 'success':
                    iconNode = <IconCheckCircleSolid className={`${prefixCls}-title-icon`} />;
                    break;
                case 'info':
                    iconNode = <IconInfoCircleSolid className={`${prefixCls}-title-icon`} />;
                    break;
                case 'error':
                    iconNode = <IconTimesCircleSolid className={`${prefixCls}-title-icon`} />;
                    break;
                case 'warning':
                    iconNode = <IconExclamationCircleSolid className={`${prefixCls}-title-icon`} />;
                    break;
                default:
                    iconNode = null;
            }
        }
        else if (typeof icon === 'string') {
            iconNode = <IconSvg className={`${prefixCls}-title-icon`} type={icon} />;
        }
        else if (React.isValidElement(icon)) {
            iconNode = icon;
        }
        const alertCls = classnames(
            prefixCls,
            `${prefixCls}-${type}`,
            `${prefixCls}-${size}`,
            {
                [`${prefixCls}-no-title`]: !title,
                [`${prefixCls}-with-title`]: !!title,
                [`${prefixCls}-show-icon`]: showIcon,
                [`${prefixCls}-has-close-icon`]: closable
            },
            className,
        );
        const closeIcon = closable ? (
            <span onClick={this.onHandleClose} className={`${prefixCls}-close-icon`}>
                <IconTimes />
            </span>
        ) : null;
        return (
            <div
                className={alertCls}
                ref={this.saveRef('alertRef')}
                {...omit(rest, ['visible', 'onClose'])}
            >
                {
                    showIcon
                        ? <span className={`${prefixCls}-icon`}>{iconNode}</span>
                        : null
                }
                <div>
                    {
                        title ? <div className={`${prefixCls}-title`}>{title}</div> : null
                    }
                    <div className={`${prefixCls}-content`} ref={this.saveRef('alertContentRef')}>{content}</div>
                </div>
                {closeIcon}
            </div>
        );
    }
}

export default Alert;
