// @baidu/one-ui@4.32.1/src/components/alert/interface.ts
import {ReactNode, ReactComponentElement} from 'react';
import {BaseProps} from '../interface';
import Alert from './alert';
export interface AlertProps extends BaseProps {
    /**
     * 类型: 成功 | 通知 | 警告 | 错误
     */
    type?: 'success' | 'info' | 'warning' | 'error';
    /**
     * 尺寸
     */
    size?: 'small' | 'medium';
    /**
     * 标题
     */
    title?: ReactNode;
    /**
     * 内容
     */
    content?: ReactNode;
    /**
     * 是否可被关闭
     */
    closable?: boolean;
    /**
     * 是否可见
     */
    visible?: boolean;
    /**
     * 是否展示Icon
     */
    showIcon?: boolean;
    /**
     * 展示的Icon（string类型则使用IconSvg）
     */
    icon?: ReactNode;
    /**
     * 关闭回调
     */
    onClose?(e: MouseEvent): void;
}

export interface AlertPageProps extends BaseProps {
    /**
     * 尺寸
     */
    size?: 'small' | 'medium';
    /**
     * 数据源（非受控场景）
     */
    defaultDataSource?: ReactComponentElement<typeof Alert, AlertProps>[];
    /**
     * 数据源（受控场景）
     */
    dataSource?: ReactComponentElement<typeof Alert, AlertProps>[];
    /**
     * 初始页（非受控场景）
     */
    defaultActiveIndex?: number;
    /**
     * 当前页（受控场景）
     */
    activeIndex?: number;
    /**
     * @deprecated v4.2.0 use defaultActiveIndex
     */
    initialSlide?: number;
    /**
     * @deprecated v4.2.0 use activeIndex
     */
    slider?: number;
    /**
     * 往前翻页的回调
     */
    onPrevChange?(index: number): void;
    /**
     * 往后翻页的回调
     */
    onNextChange?(index: number): void;
    /**
     * 关闭回调
     */
    onClose?(index: number): void;
}
