// @baidu/one-ui@4.32.1/src/components/overlay/interface.ts
import {ReactElement, ReactNode} from "react";
import {BaseProps} from "../interface";
import {ButtonProps} from "../button";

export interface OverlayProps extends BaseProps {

    /**
     * 触发方式
     */
    trigger?: 'hover' | 'click' | '';

    /**
     * 位置
     */
    popupPlacement?: 'bottomLeft' | 'bottomRight' | 'topLeft' | 'topRight';

    /**
     * 是否禁用，默认false
     */
    disabled?: boolean;

    /**
     * 默认是否可视
     */
    visible?: boolean;

    /**
     * 弹层内容
     */
    overlay?: ReactNode;

    /**
     * 弹层样式
     */
    overlayClassName?: string;

    /**
     * 使用标准触发的按钮的文案
     */
    header?: string;

    /**
     * visible变化的时候触发的函数
     */
    onVisibleChange?(visible: boolean);

    /**
     * 使用让popLayer与按钮同宽
     */
    dropdownMatchSelectWidth?: boolean;

    /**
     * popLayer 挂载的函数
     */
    getPopupContainer?: (triggerNode: HTMLElement) => HTMLElement;

    /**
     * 触发元素
     */
    children?: ReactElement;

    /**
     * 按钮属性
     */
    buttonProps?: ButtonProps;

    /**
     * @deprecated
     * @internal
     */
    align?;

    /**
     * @internal
     */
    popupClassName?: string;
};
