// @baidu/one-ui@4.41.1/lib/components/progress/common/circle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _rcProgress = require("rc-progress");

var Circle = function Circle(props) {
  var prefixCls = props.prefixCls,
      percent = props.percent,
      strokeLinecap = props.strokeLinecap,
      strokeColor = props.strokeColor,
      strokeWidth = props.strokeWidth,
      width = props.width,
      trailColor = props.trailColor,
      children = props.children;
  var circleStyle = {
    width: width,
    height: width
  };
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: prefixCls + "-inner",
    style: circleStyle
  }, /*#__PURE__*/_react["default"].createElement(_rcProgress.Circle, {
    percent: percent,
    strokeWidth: strokeWidth,
    trailWidth: strokeWidth,
    strokeColor: strokeColor,
    strokeLinecap: strokeLinecap,
    trailColor: trailColor,
    prefixCls: prefixCls
  }), children);
};

Circle.defaultProps = {
  prefixCls: 'one-progress',
  percent: 0,
  strokeLinecap: 'round',
  strokeColor: '',
  strokeWidth: 0,
  trailColor: '#e2e6f0'
};
var _default = Circle;
exports["default"] = _default;
module.exports = exports.default;
