// @baidu/one-ui@4.41.1/es/components/switch/switch.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import omit from 'omit.js';
import CommonSwitch from './common/switch';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
var defaultSize = 'medium';
var Switch = (_dec = withConfigConsumer('switch'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Switch, _PureComponent);

  function Switch() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.rcSwitch = void 0;

    _this.saveSwitch = function (node) {
      _this.rcSwitch = node;
    };

    return _this;
  }

  var _proto = Switch.prototype;

  _proto.focus = function focus() {
    this.rcSwitch.focus();
  };

  _proto.blur = function blur() {
    this.rcSwitch.blur();
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        loading = _this$props.loading,
        _this$props$className = _this$props.className,
        className = _this$props$className === void 0 ? '' : _this$props$className,
        rawSize = _this$props.size;
    var size = ['xsmall', 'small', 'medium'].includes(rawSize) ? rawSize : defaultSize;
    var classes = classnames(className, (_classnames = {}, _classnames[prefixCls + "-" + size] = size, _classnames[prefixCls + "-loading"] = loading, _classnames));
    var otherProps = {
      className: classes,
      ref: this.saveSwitch,
      loading: loading
    };
    return /*#__PURE__*/React.createElement(CommonSwitch, _extends({}, omit(this.props, ['size']), otherProps));
  };

  return Switch;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-switch',
  loading: false,
  size: defaultSize
}, _class2.displayName = 'Switch', _temp)) || _class);
export default Switch;
