// @baidu/one-ui@4.32.1/lib/components/table/common/dragLine.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../../core/commonTools");

var DragLine = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(DragLine, _PureComponent);

  function DragLine() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.state = {
      dragLeft: 0,
      showLine: false
    };

    _this.setShowLine = function (showLine) {
      _this.setState({
        showLine: showLine
      });
    };

    _this.setDragLeft = function (dragLeft) {
      _this.setState({
        dragLeft: dragLeft
      });
    };

    return _this;
  }

  var _proto = DragLine.prototype;

  _proto.render = function render() {
    var _classnames;

    var prefixCls = this.props.prefixCls;
    var _this$state = this.state,
        dragLeft = _this$state.dragLeft,
        showLine = _this$state.showLine;
    return /*#__PURE__*/_react["default"].createElement("div", {
      key: "drag-line",
      style: {
        left: dragLeft
      },
      className: (0, _commonTools.classnames)(prefixCls + "-drag-line", (_classnames = {}, _classnames[prefixCls + "-drag-line-show"] = showLine, _classnames))
    });
  };

  return DragLine;
}(_react.PureComponent);

var _default = DragLine;
exports["default"] = _default;
module.exports = exports.default;
