// @baidu/one-ui@4.32.1/src/components/select/checkboxText.tsx
import React, {PureComponent} from 'react';
import SearchText from './searchText';
import Checkbox from '../checkbox';
import Tooltip, {TooltipPlacement} from '../tooltip';

interface CheckboxTextState {
    searchValue?: string;
    source?: Array<string | number>;
    checked?: boolean;
};

interface CheckboxTextProps {
    value?: string | number;
    searchValue?: string;
    source?: Array<string | number>;
    checked?: boolean;
    text?: string;
    label?: string;
    disabled?: boolean;
    disabledReason?: string;
    disablePlacement?: TooltipPlacement;
};

class CheckboxText extends PureComponent<CheckboxTextProps, CheckboxTextState> {
    static isCheckboxText = true;

    static defaultProps = {
        searchValue: '',
        source: [],
        disablePlacement: 'right',
        disabled: false
    }

    constructor(props) {
        super(props);
        this.state = {
            searchValue: props.searchValue,
            source: props.source,
            checked: false
        };
    }

    static getDerivedStateFromProps = (nextProps, prevState) => {
        const searchValue = prevState.searchValue;
        const newState: CheckboxTextState = {};
        if ('searchValue' in nextProps && searchValue !== nextProps.searchValue) {
            newState.searchValue = nextProps.searchValue;
        }
        if ('source' in nextProps) {
            newState.source = nextProps.source;
        }
        if ('checked' in nextProps) {
            newState.checked = nextProps.checked;
        }
        return newState;
    }

    render() {
        const {text, value, label, disabled, disabledReason, disablePlacement} = this.props;
        const searchLabel = text || label;
        const {searchValue, source} = this.state;
        const searchTextProps = {
            text: searchLabel,
            showSearch: true,
            searchValue
        };
        let checked = false;
        if (source.indexOf(value) > -1) {
            checked = true;
        }
        if ('checked' in this.props) {
            checked = this.state.checked;
        }
        const checkboxTextDom = (
            <span>
                <Checkbox checked={checked} disabled={disabled} />
                <SearchText
                    {...searchTextProps}
                    disabled={disabled}
                />
            </span>
        );
        return (disabled && disabledReason) ? (
            <Tooltip title={disabledReason} placement={disablePlacement}>
                {checkboxTextDom}
            </Tooltip>
        ) : checkboxTextDom;
    }
}

export default CheckboxText;
