// @baidu/one-ui@4.41.1/es/components/overlay/content.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
import { classnames } from '../../core/commonTools';
;

var Content = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Content, _PureComponent);

  function Content() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Content.prototype;

  _proto.componentDidUpdate = function componentDidUpdate() {
    var trigger = this.props.trigger;

    if (trigger) {
      trigger.forcePopupAlign();
    }
  };

  _proto.render = function render() {
    var _this$props = this.props,
        overlay = _this$props.overlay,
        prefixCls = _this$props.prefixCls,
        id = _this$props.id,
        className = _this$props.className,
        style = _this$props.style;
    return /*#__PURE__*/React.createElement("div", {
      className: classnames(prefixCls + "-inner", className),
      id: id,
      style: style
    }, typeof overlay === 'function' ? overlay() : overlay);
  };

  return Content;
}(PureComponent);

export { Content as default };
