// @baidu/one-ui@4.32.1/src/components/select/common/popSelectTrigger.tsx
import React, {PureComponent, ReactNode} from 'react';
import Trigger from 'rc-trigger';
import Content from './popContent';

Trigger.displayName = 'Trigger';
const BUILT_IN_PLACEMENTS = {
    bottomLeft: {
        points: ['tl', 'bl'],
        offset: [0, 4],
        overflow: {
            adjustX: 1,
            adjustY: 1
        }
    },
    topLeft: {
        points: ['bl', 'tl'],
        offset: [0, -4],
        overflow: {
            adjustX: 1,
            adjustY: 1
        }
    },
    bottomRight: {
        points: ['tr', 'br'],
        offset: [0, 4],
        overflow: {
            adjustX: 1,
            adjustY: 1
        }
    },
    topRight: {
        points: ['br', 'tr'],
        offset: [0, -4],
        overflow: {
            adjustX: 1,
            adjustY: 1
        }
    }
};


interface PopSelectTriggerProps {
    trigger: any;
    children: any;
    defaultVisible?: boolean;
    visible: boolean;
    transitionName: string;
    animation?: any;
    onVisibleChange(visible: boolean);
    afterVisibleChange?();
    overlay: ReactNode | (() => ReactNode);
    overlayStyle: object;
    overlayClassName: string;
    prefixCls: string;
    mouseEnterDelay: number;
    mouseLeaveDelay: number;
    getPopupContainer(): HTMLElement;
    destroyPopUpOnHide: boolean;
    align: object;
    id?: string;
    dropdownMatchSelectWidth: boolean;
    popupPlacement: string;
};

interface PopSelectTriggerState {
    dropdownWidth: number;
};

class PopSelectTrigger extends PureComponent<PopSelectTriggerProps, PopSelectTriggerState> {

    static defaultProps = {
        prefixCls: 'one-select-pop',
        mouseEnterDelay: 0,
        destroyPopUpOnHide: false,
        mouseLeaveDelay: 0.1,
        align: {},
        trigger: ['hover'],
        dropdownMatchSelectWidth: true,
        popupPlacement: 'bottomLeft'
    };

    constructor(props) {
        super(props);
        this.state = {
            dropdownWidth: null
        };
    }

    trigger;
    root;

    componentDidMount() {
        this.setDropdownWidth();
    }

    componentDidUpdate() {
        this.setDropdownWidth();
    }

    setDropdownWidth = () => {
        const width = this.root.offsetWidth;
        if (width !== this.state.dropdownWidth) {
            this.setState({dropdownWidth: width});
        }
    }

    getPopupElement = () => {
        const {overlay, prefixCls, id} = this.props;
        return (
            <Content
                key="content"
                trigger={this.trigger}
                prefixCls={prefixCls}
                id={id}
                overlay={overlay}
            />
        );
    }

    getPopupDomNode() {
        return this.trigger.getPopupDomNode();
    }

    saveTrigger = node => {
        this.trigger = node;
    }

    saveRoot = node => {
        this.root = node
    };

    render() {
        const {
            overlayClassName, trigger,
            mouseEnterDelay, mouseLeaveDelay,
            overlayStyle, prefixCls,
            children, onVisibleChange, afterVisibleChange,
            transitionName, animation,
            align,
            destroyPopUpOnHide,
            defaultVisible, getPopupContainer,
            dropdownMatchSelectWidth,
            popupPlacement,
            ...restProps
        } = this.props;
        const extraProps = {...restProps};
        if ('visible' in this.props) {
            // @ts-ignore
            extraProps.popupVisible = this.props.visible;
        }
        const widthProp = dropdownMatchSelectWidth ? 'width' : 'minWidth';
        const popupStyle = {...overlayStyle};
        if (this.state.dropdownWidth) {
            popupStyle[widthProp] = `${this.state.dropdownWidth}px`;
        }
        return (
            <Trigger
                popupClassName={overlayClassName}
                ref={this.saveTrigger}
                prefixCls={prefixCls}
                popup={this.getPopupElement}
                action={trigger}
                builtinPlacements={BUILT_IN_PLACEMENTS}
                popupPlacement={popupPlacement}
                popupAlign={align}
                getPopupContainer={getPopupContainer}
                onPopupVisibleChange={onVisibleChange}
                afterPopupVisibleChange={afterVisibleChange}
                popupTransitionName={transitionName}
                popupAnimation={animation}
                defaultPopupVisible={defaultVisible}
                destroyPopupOnHide={destroyPopUpOnHide}
                mouseLeaveDelay={mouseLeaveDelay}
                popupStyle={popupStyle}
                mouseEnterDelay={mouseEnterDelay}

                {...extraProps}
            >
                {React.cloneElement(children, {ref: this.saveRoot})}
            </Trigger>
        );
    }
}

export default PopSelectTrigger;
