// @baidu/one-ui@4.41.1/lib/components/select/searchText.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _flattenDeep2 = _interopRequireDefault(require("lodash/flattenDeep"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../core/commonTools");

;
;

var SearchText = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(SearchText, _PureComponent);

  function SearchText(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.formatTextWithColor = function (text, searchValue) {
      var pivolIndex = text.indexOf(searchValue);
      var prefixCls = _this.props.prefixCls;

      if (pivolIndex < 0) {
        return /*#__PURE__*/_react["default"].createElement("span", null, text);
      }

      var textArray = (0, _flattenDeep2["default"])(text.split(searchValue).map(function (node, index) {
        return index === 0 ? [/*#__PURE__*/_react["default"].createElement("span", {
          key: index
        }, node)] : [/*#__PURE__*/_react["default"].createElement("span", {
          key: index + "-highlight",
          className: prefixCls + "-search-text-highlight"
        }, searchValue), /*#__PURE__*/_react["default"].createElement("span", {
          key: index
        }, node)];
      }));

      var strDom = /*#__PURE__*/_react["default"].createElement("span", null, textArray.map(function (node) {
        return node;
      }));

      return strDom;
    };

    _this.state = {
      searchValue: props.searchValue || ''
    };
    return _this;
  }

  var _proto = SearchText.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        text = _this$props.text,
        showSearch = _this$props.showSearch,
        className = _this$props.className,
        prefixCls = _this$props.prefixCls,
        disabled = _this$props.disabled;
    var searchValue = this.state.searchValue;
    var searchTextClassName = (0, _commonTools.classnames)((_classnames = {}, _classnames[prefixCls + "-search-text-disabled"] = disabled, _classnames), className);

    if (!showSearch || !searchValue || typeof text !== 'string') {
      return /*#__PURE__*/_react["default"].createElement("span", {
        className: className
      }, text);
    }

    return /*#__PURE__*/_react["default"].createElement("span", {
      className: searchTextClassName
    }, this.formatTextWithColor(text, searchValue));
  };

  return SearchText;
}(_react.PureComponent);

SearchText.defaultProps = {
  className: '',
  disabled: false,
  prefixCls: 'one-select'
};

SearchText.getDerivedStateFromProps = function (nextProps, prevState) {
  var searchValue = nextProps.searchValue;
  var currentSearchValue = prevState.searchValue;

  if (searchValue !== currentSearchValue) {
    return {
      searchValue: searchValue
    };
  }

  return null;
};

var _default = SearchText;
exports["default"] = _default;
module.exports = exports.default;
