// @baidu/one-ui@4.41.1/lib/components/link/navLink.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _textLink = _interopRequireDefault(require("./textLink"));

/**
 * @file 支持react-router的link
 * @author huangshiming@baidu.com
 * @date 2020-11-26
 */
function isModifiedEvent(event) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
}

var NavLink = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var target = props.target,
      onClick = props.onClick,
      navigate = props.navigate,
      lightPrefix = props.lightPrefix,
      restProps = (0, _objectWithoutPropertiesLoose2["default"])(props, ["target", "onClick", "navigate", "lightPrefix"]);

  var handleOnClick = function handleOnClick(event) {
    try {
      if (onClick) {
        onClick(event);
      }
    } catch (ex) {
      event.preventDefault();
      throw ex;
    }

    if (!event.defaultPrevented && event.button === 0 && (!target || target === '_self') && !isModifiedEvent(event)) {
      event.preventDefault();
      navigate();
    }
  };

  return /*#__PURE__*/_react["default"].createElement(_textLink["default"], (0, _extends2["default"])({}, restProps, {
    target: target,
    isAtag: true,
    ref: ref,
    onClick: handleOnClick
  }));
});
var _default = NavLink;
exports["default"] = _default;
module.exports = exports.default;
