// @baidu/one-ui@4.41.1/lib/components/link/textLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

var Link = (_dec = (0, _context.withConfigConsumer)('text-link'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Link, _PureComponent);

  function Link() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;

    _this.onClick = function () {
      var _this$props = _this.props,
          toUrl = _this$props.toUrl,
          target = _this$props.target,
          disabled = _this$props.disabled;

      if (disabled) {
        return;
      }

      window.open(toUrl, target);
    };

    return _this;
  }

  var _proto = Link.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props2 = this.props,
        type = _this$props2.type,
        className = _this$props2.className,
        prefixCls = _this$props2.prefixCls,
        children = _this$props2.children,
        disabled = _this$props2.disabled,
        size = _this$props2.size,
        isAtag = _this$props2.isAtag,
        target = _this$props2.target,
        toUrl = _this$props2.toUrl,
        lightPrefix = _this$props2.lightPrefix,
        rest = (0, _objectWithoutPropertiesLoose2["default"])(_this$props2, ["type", "className", "prefixCls", "children", "disabled", "size", "isAtag", "target", "toUrl", "lightPrefix"]);
    var linkClassName = (0, _commonTools.classnames)(prefixCls, className, prefixCls + "-" + size, (_classnames = {}, _classnames[prefixCls + "-" + type] = type, _classnames[prefixCls + "-disabled"] = disabled, _classnames[prefixCls + "-is-a-tag"] = isAtag, _classnames));

    if (isAtag) {
      return disabled ? /*#__PURE__*/_react["default"].createElement("a", (0, _extends2["default"])({}, rest, {
        className: linkClassName
      }), children) : /*#__PURE__*/_react["default"].createElement("a", (0, _extends2["default"])({}, rest, {
        className: linkClassName,
        target: target,
        href: toUrl
      }), children);
    }

    return /*#__PURE__*/_react["default"].createElement("span", (0, _extends2["default"])({}, rest, {
      className: linkClassName,
      onClick: this.onClick
    }), children);
  };

  return Link;
}(_react.PureComponent), _class2.defaultProps = {
  type: 'normal',
  target: '_self',
  size: 'medium',
  prefixCls: 'one-text-link',
  disabled: false,
  isAtag: false
}, _class2.NavLink = void 0, _temp)) || _class);
var _default = Link;
exports["default"] = _default;
module.exports = exports.default;
