// @baidu/one-ui@4.41.1/es/components/breadcrumb/breadcrumb.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { PureComponent, cloneElement } from 'react';
import { IconChevronRight } from 'dls-icons-react';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
var Breadcrumb = (_dec = withConfigConsumer('breadcrumb'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Breadcrumb, _PureComponent);

  function Breadcrumb() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Breadcrumb.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        separator = _this$props.separator,
        prefixCls = _this$props.prefixCls,
        style = _this$props.style,
        className = _this$props.className,
        children = _this$props.children,
        size = _this$props.size,
        type = _this$props.type;
    var crumbs = React.Children.map(children, function (element, index) {
      if (! /*#__PURE__*/React.isValidElement(element)) {
        return element;
      }

      return /*#__PURE__*/cloneElement(element, {
        separator: separator,
        key: index
      });
    });
    var breadcrumbCls = classnames(prefixCls, prefixCls + "-" + size, className, prefixCls + "-" + type);
    return /*#__PURE__*/React.createElement("div", {
      className: breadcrumbCls,
      style: style
    }, crumbs);
  };

  return Breadcrumb;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-breadcrumb',
  separator: /*#__PURE__*/React.createElement(IconChevronRight, null),
  size: 'medium',
  type: 'normal'
}, _class2.Item = void 0, _temp)) || _class);
export default Breadcrumb;
