// @baidu/one-ui@4.32.1/lib/components/form/formFieldGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = exports.FormFieldGroup = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../core/commonTools");

var _context = require("./context");

/**
 * @visibleName FormFieldGroup
 */
var FormFieldGroup = function FormFieldGroup(_ref) {
  var _classnames;

  var children = _ref.children,
      className = _ref.className,
      prefixCls = _ref.prefixCls,
      layout = _ref.layout,
      labelAlign = _ref.labelAlign,
      labelPosition = _ref.labelPosition,
      density = _ref.density,
      size = _ref.size,
      props = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["children", "className", "prefixCls", "layout", "labelAlign", "labelPosition", "density", "size"]);
  var formContext = (0, _react.useContext)(_context.FormContext);
  return /*#__PURE__*/_react["default"].createElement(_context.FormContext.Provider, {
    value: (0, _extends2["default"])({}, formContext, {
      labelAlign: labelAlign,
      labelPosition: labelPosition
    })
  }, /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
    className: (0, _commonTools.classnames)(prefixCls + "-field-group", className, prefixCls + "-" + layout, prefixCls + "-" + size, prefixCls + "-type-" + density, (_classnames = {}, _classnames[prefixCls + "-label-position-" + labelPosition] = labelPosition, _classnames))
  }, props), children));
};

exports.FormFieldGroup = FormFieldGroup;
FormFieldGroup.defaultProps = {
  prefixCls: 'one-form',
  layout: 'default',
  labelAlign: 'left',
  labelPosition: 'side',
  density: 'compact'
};
var _default = FormFieldGroup;
exports["default"] = _default;
