// @baidu/one-ui@4.32.1/src/components/badge/badge.tsx
import React, {PureComponent, ReactNode} from 'react';
import omit from 'omit.js';
import {classnames} from '../../core/commonTools';
import {BadgeProps} from './interface';
import {withConfigConsumer} from '../providerConfig/context';

interface BadgeState {
    count?: number;
    visible?: boolean;
};

@withConfigConsumer('badge')
class Badge extends PureComponent<BadgeProps, BadgeState> {

    static defaultProps = {
        prefixCls: 'one-badge',
        count: null,
        showZero: false,
        isDot: false,
        overflowCount: 99,
        visible: true,
        textContent: ''
    }

    constructor(props) {
        super(props);
        this.state = {
            count: props.count,
            visible: props.visible
        };
    }

    static getDerivedStateFromProps = (nextProps, prevState) => {
        const newState: BadgeState = {};
        if ('visible' in nextProps && nextProps.visible !== prevState.visible) {
            newState.visible = nextProps.visible;
        }
        if ('count' in nextProps && nextProps.count !== prevState.count) {
            newState.count = nextProps.count;
        }
        return newState;
    }

    render = () => {
        const {
            showZero,
            prefixCls,
            overflowCount,
            className,
            style,
            children,
            isDot,
            type,
            text,
            offset,
            color,
            textContent,
            ...restProps
        } = this.props;
        const {visible, count} = this.state;
        let displayCount = +count > +overflowCount ? `${overflowCount}+` : count;
        const isZero = displayCount === 0;
        const isDotMode = (isDot && !isZero) || type;
        // 小圆点状态下，不展示数字
        if (isDotMode) {
            displayCount = '';
        }
        const isEmpty = (displayCount === null || displayCount === undefined || displayCount === '') && !textContent;
        const hidden = ((isEmpty || (isZero && !showZero)) && !isDotMode) || !visible;
        const typeCls = classnames({
            [`${prefixCls}-type-dot`]: !!type,
            [`${prefixCls}-type-${type}`]: !!type
        });
        const numberBoxCls = classnames({
            [`${prefixCls}-dot`]: isDotMode,
            [`${prefixCls}-count`]: !isDotMode,
            [`${prefixCls}-type-${type}`]: !!type,
            [`${prefixCls}-hidden`]: hidden
        });
        const badgeCls = classnames(className, prefixCls, {
            [`${prefixCls}-type`]: !!type,
            [`${prefixCls}-not-a-wrapper`]: !children
        });
        let styleWithOffset = offset ? {
            right: -offset[0],
            marginTop: offset[1],
            ...style
        } : style;
        if (color) {
            styleWithOffset = {
                ...styleWithOffset,
                backgroundColor: color
            };
        }
        const divProps = omit(restProps, [
            'visible'
        ]);
        if (!children && type) {
            const colorStyle = color ? {backgroundColor: color} : undefined;
            return (
                <span {...divProps} className={badgeCls} style={style}>
                    <span className={typeCls} style={colorStyle} />
                    <span className={`${prefixCls}-type-text`}>{text}</span>
                </span>
            );
        }
        let showContent: ReactNode = displayCount;
        // 如果有文字形式，数字形式默认失效
        if (textContent) {
            showContent = textContent;
        }
        return (
            <span {...divProps} className={badgeCls}>
                {children}
                <span className={numberBoxCls} style={styleWithOffset}>
                    {showContent}
                </span>
            </span>
        );
    }
}

export default Badge;
