// @baidu/one-ui@4.41.1/es/components/toast/index.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import Notification from 'rc-notification';
import MessageContainer from './common/message';
var defaultDuration = 3;
var defaultTop = '10%';
var key = 1;
var prefixCls = 'one-toast';
var defaultContainer;
var messageSize = 'medium';
var DEFAULT_CONTAINER = new String('__ONE__TOAST__DEFAULT__');
var instancesMap = new WeakMap();
var instances = new Set();

var getMessageInstance = function getMessageInstance(container, containerKey, callback) {
  var instance = instancesMap.get(containerKey);

  if (instance) {
    callback(instance);
    return;
  }

  Notification.newInstance({
    prefixCls: prefixCls,
    transitionName: 'move-up',
    style: {
      top: defaultTop
    },
    // 覆盖原来的样式
    getContainer: container
  }, function (newInstance) {
    var instance = instancesMap.get(containerKey);

    if (instance) {
      callback(instance);
      return;
    }

    instancesMap.set(containerKey, newInstance);
    instances.add(newInstance);
    callback(newInstance);
  });
};

var notice = function notice(props) {
  var duration = props.duration !== undefined ? props.duration : defaultDuration;
  var target = key++;
  var container = props.getContainer || defaultContainer;
  var containerKey = container ? container() : DEFAULT_CONTAINER;
  var closePromise = new Promise(function (resolve) {
    var callback = function callback() {
      if (typeof props.onClose === 'function') {
        props.onClose();
      }

      return resolve(true);
    };

    getMessageInstance(container, containerKey, function (instance) {
      var title = props.title,
          content = props.content,
          type = props.type,
          style = props.style,
          _props$showCloseIcon = props.showCloseIcon,
          showCloseIcon = _props$showCloseIcon === void 0 ? true : _props$showCloseIcon;
      var size = props.size || messageSize;
      var messageProps = {
        prefixCls: prefixCls,
        type: type,
        title: title,
        size: size,
        content: content,
        showCloseIcon: showCloseIcon,
        target: target,
        instance: instance,
        onClose: callback
      };
      instance.notice({
        key: target,
        duration: duration,
        style: style,
        content: /*#__PURE__*/React.createElement(MessageContainer, messageProps),
        onClose: callback
      });
    });
  });

  var result = function result() {
    var instance = instancesMap.get(containerKey);

    if (instance) {
      instance.removeNotice(target);
    }
  };

  result.then = function (filled, rejected) {
    return closePromise.then(filled, rejected);
  };

  result.promise = closePromise;
  return result;
};

export default {
  info: function info(props) {
    return notice(_extends({
      type: 'info'
    }, props));
  },
  success: function success(props) {
    return notice(_extends({
      type: 'success'
    }, props));
  },
  error: function error(props) {
    return notice(_extends({
      type: 'error'
    }, props));
  },
  warning: function warning(props) {
    return notice(_extends({
      type: 'warning'
    }, props));
  },
  loading: function loading(props) {
    return notice(_extends({
      type: 'loading'
    }, props));
  },
  config: function config(options) {
    if (options.top !== undefined) {
      defaultTop = options.top;
      instancesMap = new WeakMap();
      instances.clear();
    }

    if (options.duration !== undefined) {
      defaultDuration = options.duration;
    }

    if (options.prefixCls !== undefined) {
      prefixCls = options.prefixCls;
    }

    if (options.getContainer !== undefined) {
      defaultContainer = options.getContainer;
    }

    if (options.size !== undefined) {
      messageSize = options.size;
    }
  },
  destroy: function destroy() {
    instances.forEach(function (instance) {
      instance.destroy();
    });
    instances.clear();
    instancesMap = new WeakMap();
  },

  /**
   * @internal 仅限特殊场合，不得用于业务
   */
  Element: MessageContainer
};
export * from './interface';
