// @baidu/one-ui@4.32.1/es/components/form/common/createFieldsStore.js
import _extends from "@babel/runtime/helpers/extends";
import _set from "lodash/set";
import createFormField, { isFormField } from './createFormField';
import { hasRules, flattenFields, getMessageStrs, startsWith } from '../../../core/formTools';

function partOf(a, b) {
  return b.indexOf(a) === 0 && ['.', '['].indexOf(b[a.length]) !== -1;
}

function internalFlattenFields(fields) {
  return flattenFields(fields, function (_, node) {
    return isFormField(node);
  }, 'You must wrap field data with `createFormField`.');
}

export var FieldsStore = /*#__PURE__*/function () {
  function FieldsStore(_fields) {
    var _this = this;

    this.fields = void 0;
    this.fieldsMeta = void 0;

    this.setFieldsInitialValue = function (initialValues) {
      var flattenedInitialValues = _this.flattenRegisteredFields(initialValues);

      var fieldsMeta = _this.fieldsMeta;
      Object.keys(flattenedInitialValues).forEach(function (name) {
        if (fieldsMeta[name]) {
          _this.setFieldMeta(name, _extends({}, _this.getFieldMeta(name), {
            initialValue: flattenedInitialValues[name]
          }));
        }
      });
    };

    this.getAllValues = function () {
      var fieldsMeta = _this.fieldsMeta,
          fields = _this.fields;
      return Object.keys(fieldsMeta).reduce(function (acc, name) {
        return _set(acc, name, _this.getValueFromFields(name, fields));
      }, {});
    };

    this.getFieldsValue = function (names) {
      return _this.getNestedFields(names, _this.getFieldValue);
    };

    this.getFieldValue = function (name) {
      var fields = _this.fields;
      return _this.getNestedField(name, function (fullName) {
        return _this.getValueFromFields(fullName, fields);
      });
    };

    this.getFieldMessage = function (name, type) {
      var messages = _this.getNestedField(name, function (fullName) {
        return getMessageStrs(_this.getFieldMember(fullName, type));
      });

      return messages;
    };

    this.getFieldsError = function (names) {
      return _this.getNestedFields(names, _this.getFieldError);
    };

    this.getFieldError = function (name) {
      return _this.getFieldMessage(name, 'errors');
    };

    this.getFieldsWarning = function (names) {
      return _this.getNestedFields(names, _this.getFieldWarning);
    };

    this.getFieldWarning = function (name) {
      return _this.getFieldMessage(name, 'warnings');
    };

    this.getFieldsSuccess = function (names) {
      return _this.getNestedFields(names, _this.getFieldSuccess);
    };

    this.getFieldSuccess = function (name) {
      return _this.getFieldMessage(name, 'successes');
    };

    this.isFieldValidating = function (name) {
      return _this.getFieldMember(name, 'validating');
    };

    this.isFieldsValidating = function (ns) {
      var names = ns || _this.getValidFieldsName();

      return names.some(function (n) {
        return _this.isFieldValidating(n);
      });
    };

    this.isFieldTouched = function (name) {
      return _this.getFieldMember(name, 'touched');
    };

    this.isFieldsTouched = function (ns) {
      var names = ns || _this.getValidFieldsName();

      return names.some(function (n) {
        return _this.isFieldTouched(n);
      });
    };

    this.fields = internalFlattenFields(_fields);
    this.fieldsMeta = {};
  }

  var _proto = FieldsStore.prototype;

  _proto.updateFields = function updateFields(fields) {
    this.fields = internalFlattenFields(fields);
  };

  _proto.flattenRegisteredFields = function flattenRegisteredFields(fields) {
    var validFieldsName = this.getAllFieldsName();
    return flattenFields(fields, function (path) {
      return validFieldsName.indexOf(path) >= 0;
    }, 'You cannot set a form field before rendering a field associated with the value.');
  };

  _proto.setFields = function setFields(fields) {
    var _this2 = this;

    var fieldsMeta = this.fieldsMeta;

    var nowFields = _extends({}, this.fields, fields);

    var nowValues = {};
    Object.keys(fieldsMeta).forEach(function (f) {
      nowValues[f] = _this2.getValueFromFields(f, nowFields);
    });
    Object.keys(nowValues).forEach(function (f) {
      var value = nowValues[f];

      var fieldMeta = _this2.getFieldMeta(f);

      if (fieldMeta && fieldMeta.normalize) {
        var nowValue = fieldMeta.normalize(value, _this2.getValueFromFields(f, _this2.fields), nowValues);

        if (nowValue !== value) {
          nowFields[f] = _extends({}, nowFields[f], {
            value: nowValue
          });
        }
      }
    });
    this.fields = nowFields;
  };

  _proto.resetFields = function resetFields(ns) {
    var fields = this.fields;
    var names = ns ? this.getValidFieldsFullName(ns) : this.getAllFieldsName();
    return names.reduce(function (acc, name) {
      var field = fields[name];

      if (field && 'value' in field) {
        acc[name] = {};
      }

      return acc;
    }, {});
  };

  _proto.setFieldMeta = function setFieldMeta(name, meta) {
    this.fieldsMeta[name] = meta;
  };

  _proto.setFieldsAsDirty = function setFieldsAsDirty() {
    var _this3 = this;

    Object.keys(this.fields).forEach(function (name) {
      var field = _this3.fields[name];
      var fieldMeta = _this3.fieldsMeta[name];

      if (field && fieldMeta && hasRules(fieldMeta.validate)) {
        _this3.fields[name] = _extends({}, field, {
          dirty: true
        });
      }
    });
  };

  _proto.getFieldMeta = function getFieldMeta(name) {
    this.fieldsMeta[name] = this.fieldsMeta[name] || {};
    return this.fieldsMeta[name];
  };

  _proto.getValueFromFields = function getValueFromFields(name, fields) {
    var field = fields[name];

    if (field && 'value' in field) {
      return field.value;
    }

    var fieldMeta = this.getFieldMeta(name);
    return fieldMeta && fieldMeta.initialValue;
  };

  _proto.getValidFieldsName = function getValidFieldsName() {
    var _this4 = this;

    var fieldsMeta = this.fieldsMeta;
    return fieldsMeta ? Object.keys(fieldsMeta).filter(function (name) {
      return !_this4.getFieldMeta(name).hidden;
    }) : [];
  };

  _proto.getAllFieldsName = function getAllFieldsName() {
    var fieldsMeta = this.fieldsMeta;
    return fieldsMeta ? Object.keys(fieldsMeta) : [];
  };

  _proto.getValidFieldsFullName = function getValidFieldsFullName(maybePartialName) {
    var maybePartialNames = Array.isArray(maybePartialName) ? maybePartialName : [maybePartialName];
    return this.getValidFieldsName().filter(function (fullName) {
      return maybePartialNames.some(function (partialName) {
        return fullName === partialName || startsWith(fullName, partialName) && ['.', '['].indexOf(fullName[partialName.length]) >= 0;
      });
    });
  };

  _proto.getFieldValuePropValue = function getFieldValuePropValue(fieldMeta) {
    var _ref;

    var name = fieldMeta.name,
        getValueProps = fieldMeta.getValueProps,
        valuePropName = fieldMeta.valuePropName;
    var field = this.getField(name);
    var fieldValue = 'value' in field ? field.value : fieldMeta.initialValue;

    if (getValueProps) {
      return getValueProps(fieldValue);
    }

    return _ref = {}, _ref[valuePropName] = fieldValue, _ref;
  };

  _proto.getField = function getField(name) {
    return _extends({}, this.fields[name], {
      name: name
    });
  };

  _proto.getNotCollectedFields = function getNotCollectedFields() {
    var _this5 = this;

    var fieldsName = this.getValidFieldsName();
    return fieldsName.filter(function (name) {
      return !_this5.fields[name];
    }).map(function (name) {
      return {
        name: name,
        dirty: false,
        value: _this5.getFieldMeta(name).initialValue
      };
    }).reduce(function (acc, field) {
      return _set(acc, field.name, createFormField(field));
    }, {});
  };

  _proto.getNestedAllFields = function getNestedAllFields() {
    var _this6 = this;

    return Object.keys(this.fields).reduce(function (acc, name) {
      return _set(acc, name, createFormField(_this6.fields[name]));
    }, this.getNotCollectedFields());
  };

  _proto.getFieldMember = function getFieldMember(name, member) {
    return this.getField(name)[member];
  };

  _proto.getNestedFields = function getNestedFields(names, getter) {
    var fields = names || this.getValidFieldsName();
    return fields.reduce(function (acc, f) {
      return _set(acc, f, getter(f));
    }, {});
  };

  _proto.getNestedField = function getNestedField(name, getter) {
    var fullNames = this.getValidFieldsFullName(name);

    if (fullNames.length === 0 || fullNames.length === 1 && fullNames[0] === name // Name already is full name.
    ) {
      return getter(name);
    }

    var isArrayValue = fullNames[0][name.length] === '[';
    var suffixNameStartIndex = isArrayValue ? name.length : name.length + 1;
    return fullNames.reduce(function (acc, fullName) {
      return _set(acc, fullName.slice(suffixNameStartIndex), getter(fullName));
    }, isArrayValue ? [] : {});
  };

  // @private
  // BG: `a` and `a.b` cannot be use in the same form
  _proto.isValidNestedFieldName = function isValidNestedFieldName(name) {
    var names = this.getAllFieldsName();
    return names.every(function (n) {
      return !partOf(n, name) && !partOf(name, n);
    });
  };

  _proto.clearField = function clearField(name) {
    delete this.fields[name];
    delete this.fieldsMeta[name];
  };

  return FieldsStore;
}();
export default function createFieldsStore(fields) {
  return new FieldsStore(fields);
}
