// @baidu/one-ui@4.32.1/es/components/uploader/common/imageAnchor.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _noop from "lodash/noop";
import React, { PureComponent } from 'react';
import Loading from '../../loading';
import MultipleAnchor from './multipleAnchor';
import { classnames } from '../../../core/commonTools';
import MediaIcon from './mediaIcon';

var ImageAnchor = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(ImageAnchor, _PureComponent);

  function ImageAnchor() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = ImageAnchor.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        onClick = _this$props.onClick,
        loading = _this$props.loading,
        prefixCls = _this$props.prefixCls,
        disabled = _this$props.disabled,
        realType = _this$props.realType,
        entries = _this$props.entries,
        pickerIcon = _this$props.pickerIcon,
        pickerText = _this$props.pickerText,
        theme = _this$props.theme;
    var buttonClassName = classnames(prefixCls + "-anchor-image-button", (_classnames = {}, _classnames[prefixCls + "-anchor-image-button-loading"] = loading, _classnames[prefixCls + "-anchor-image-button-disabled"] = disabled, _classnames));
    var buttonProps = {
      className: buttonClassName,
      onClick: disabled || loading ? _noop : onClick
    };
    var showTexts = {
      image: '图片',
      video: '视频',
      media: '多媒体'
    };

    if (loading) {
      return /*#__PURE__*/React.createElement("div", buttonProps, /*#__PURE__*/React.createElement(Loading, {
        type: "strong"
      }));
    }

    if (entries && entries.length) {
      return /*#__PURE__*/React.createElement(MultipleAnchor, this.props);
    }

    return /*#__PURE__*/React.createElement("div", buttonProps, /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-add-icon"
    }, pickerIcon || /*#__PURE__*/React.createElement(MediaIcon, {
      theme: theme,
      type: realType,
      className: prefixCls + "-" + realType + "-add-icon"
    })), /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-anchor-image-button-text"
    }, pickerText || "\u4E0A\u4F20" + showTexts[realType]));
  };

  return ImageAnchor;
}(PureComponent);

export { ImageAnchor as default };
