// @baidu/one-ui@4.32.1/lib/components/checkbox/button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _context = require("../providerConfig/context");

var _checkbox = _interopRequireDefault(require("./checkbox"));

var _context2 = _interopRequireDefault(require("./context"));

var _dec, _class, _class2, _temp;

var CheckboxButton = (_dec = (0, _context.withConfigConsumer)('checkbox-button'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(CheckboxButton, _PureComponent);

  function CheckboxButton() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = CheckboxButton.prototype;

  _proto.render = function render() {
    var _this = this;

    return /*#__PURE__*/_react["default"].createElement(_context2["default"].Consumer, null, function (context) {
      return _this.renderCheckbox(context);
    });
  };

  _proto.renderCheckbox = function renderCheckbox(context) {
    var _this$props = this.props,
        type = _this$props.type,
        props = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, ["type"]);
    var buttonType = type || (context == null ? void 0 : context.type) || 'strong';
    return /*#__PURE__*/_react["default"].createElement(_checkbox["default"], (0, _extends2["default"])({}, props, {
      mode: "strong",
      buttonType: buttonType
    }));
  };

  return CheckboxButton;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-checkbox-button'
}, _temp)) || _class);
var _default = CheckboxButton;
exports["default"] = _default;
module.exports = exports.default;
