// @baidu/one-ui@4.41.1/es/components/datePicker/rangePicker.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _noop from "lodash/noop";

var _dec, _class, _class2, _temp;

import React, { PureComponent, isValidElement } from 'react';
import { IconCalendar, IconTimesCircle } from 'dls-icons-react';
import { Provider, create } from 'mini-store';
import { transDateFormat, validateData, formatButtonText, formatInitialRangeDateInfo } from '../../core/datePickerTools';
import TitleHeader from './common/titleHeader';
import RangeRender from './common/rangeRender';
import Input from './common/input';
import Layer from '../overlay';
import Button from '../button';
import { classnames } from '../../core/commonTools';
import ShortCut from './common/shortCut';
import { align } from './util';
import { withConfigConsumer } from '../providerConfig/context';
var RangePicker = (_dec = withConfigConsumer('date-picker'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(RangePicker, _PureComponent);

  function RangePicker(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.store = void 0;

    _this.componentDidUpdate = function (prevProps) {
      var newState = {};

      if ('value' in _this.props) {
        var _this$props = _this.props,
            value = _this$props.value,
            validateMaxDate = _this$props.validateMaxDate;

        var _formatInitialRangeDa = formatInitialRangeDateInfo(value, validateMaxDate),
            beginDateYear = _formatInitialRangeDa.beginDateYear,
            beginDateMonth = _formatInitialRangeDa.beginDateMonth,
            endDateYear = _formatInitialRangeDa.endDateYear,
            endDateMonth = _formatInitialRangeDa.endDateMonth; // 控制一些组件内部属性


        newState = _extends({}, newState, {
          _value: value || [],
          showYear: beginDateYear,
          showMonth: beginDateMonth,
          endDateShowYear: endDateYear,
          endDateShowMonth: endDateMonth
        });
      }

      if ('validateMinDate' in _this.props && _this.props.validateMinDate !== prevProps.validateMinDate) {
        newState = _extends({}, newState, {
          validateMinDate: transDateFormat(_this.props.validateMinDate) || validateData().validateMinDate
        });
      }

      if ('validateMaxDate' in _this.props && _this.props.validateMaxDate !== prevProps.validateMaxDate) {
        newState = _extends({}, newState, {
          validateMaxDate: transDateFormat(_this.props.validateMaxDate) || validateData().validateMaxDate
        });
      }

      _this.store.setState(newState);
    };

    _this.onSelectDay = function (value, colsePanel) {
      if (colsePanel === void 0) {
        colsePanel = true;
      }

      var _this$props2 = _this.props,
          onChange = _this$props2.onChange,
          validateMaxDate = _this$props2.validateMaxDate,
          validator = _this$props2.validator;
      var newState = {};

      if (!('value' in _this.props)) {
        // 非受控
        newState.currentValue = value;
      }

      var _formatInitialRangeDa2 = formatInitialRangeDateInfo(value, validateMaxDate),
          beginDateYear = _formatInitialRangeDa2.beginDateYear,
          beginDateMonth = _formatInitialRangeDa2.beginDateMonth,
          endDateYear = _formatInitialRangeDa2.endDateYear,
          endDateMonth = _formatInitialRangeDa2.endDateMonth;

      _this.store.setState({
        _value: value,
        showYear: beginDateYear,
        showMonth: beginDateMonth,
        endDateShowYear: endDateYear,
        endDateShowMonth: endDateMonth
      });

      if (!('visible' in _this.props) && colsePanel) {
        // 非受控
        newState.visible = false;
      }

      var validatorError = '';
      validatorError = validator && typeof validator === 'function' ? validator(value) : '';

      if (!validatorError) {
        onChange(value);
        newState.errorMessage = '';
      } else {
        newState.errorMessage = validatorError;
      }

      _this.setState(newState);
    };

    _this.renderMultipleDatePickerBody = function () {
      var _this$props3 = _this.props,
          prefixCls = _this$props3.prefixCls,
          shortcuts = _this$props3.shortcuts,
          size = _this$props3.size,
          hideInput = _this$props3.hideInput;
      var visible = _this.state.visible;
      var datePickerContainerClassName = classnames(prefixCls + "-panel", prefixCls + "-multiple", prefixCls + "-panel-" + size, hideInput && prefixCls + "-panel-hide-input");
      return /*#__PURE__*/React.createElement(Provider, {
        store: _this.store
      }, /*#__PURE__*/React.createElement("div", {
        className: datePickerContainerClassName
      }, shortcuts && shortcuts.length ? /*#__PURE__*/React.createElement(ShortCut, _extends({}, _this.props, {
        onChange: _this.onSelectDay
      })) : null, /*#__PURE__*/React.createElement("div", null, !hideInput && /*#__PURE__*/React.createElement(Input, _extends({}, _this.props, {
        visible: visible,
        onChange: _this.onSelectDay,
        multiple: true
      })), /*#__PURE__*/React.createElement(TitleHeader, _extends({}, _this.props, {
        multiple: true
      })), /*#__PURE__*/React.createElement(RangeRender, _extends({}, _this.props, {
        multiple: true,
        onChange: _this.onSelectDay,
        visible: visible
      })))));
    };

    _this.onLayerVisibleChange = function (visible) {
      _this.props.onVisibleChange(visible);

      if (!('visible' in _this.props)) {
        _this.setState({
          visible: visible
        });
      }
    };

    _this.onOpenDatePicker = function (e) {
      var _this$props4 = _this.props,
          onClickButton = _this$props4.onClickButton,
          validateMaxDate = _this$props4.validateMaxDate;
      var formatValue;

      if ('value' in _this.props) {
        formatValue = _this.props.value || [];
      } else {
        formatValue = _this.state.currentValue;
      }

      var _formatInitialRangeDa3 = formatInitialRangeDateInfo(formatValue, validateMaxDate),
          beginDateYear = _formatInitialRangeDa3.beginDateYear,
          beginDateMonth = _formatInitialRangeDa3.beginDateMonth,
          endDateYear = _formatInitialRangeDa3.endDateYear,
          endDateMonth = _formatInitialRangeDa3.endDateMonth; // 每次打开的时候，重置一下面板的type，并且保证当前value与state的value保持一致


      _this.store.setState({
        panelType: 'date',
        endDatePanelType: 'date',
        _value: formatValue,
        showYear: beginDateYear,
        showMonth: beginDateMonth,
        endDateShowYear: endDateYear,
        endDateShowMonth: endDateMonth
      });

      _this.onLayerVisibleChange(true);

      onClickButton(e);
    };

    _this.onDetete = function (e) {
      e.stopPropagation();
      e.preventDefault();

      _this.props.onDelete();

      if (!('value' in _this.props)) {
        _this.setState({
          currentValue: []
        });

        _this.store.setState({
          _value: []
        });
      }
    };

    var defaultValue = props.defaultValue,
        _value2 = props.value,
        _visible = props.visible,
        defaultVisible = props.defaultVisible,
        _validateMaxDate = props.validateMaxDate;

    var _value = defaultValue || _value2 || [];

    var _formatInitialRangeDa4 = formatInitialRangeDateInfo(_value, _validateMaxDate),
        _beginDateYear = _formatInitialRangeDa4.beginDateYear,
        _beginDateMonth = _formatInitialRangeDa4.beginDateMonth,
        _endDateYear = _formatInitialRangeDa4.endDateYear,
        _endDateMonth = _formatInitialRangeDa4.endDateMonth;

    _this.state = {
      currentValue: _value,
      visible: defaultVisible || _visible || false
    };
    _this.store = create({
      _value: _value,
      showYear: _beginDateYear,
      showMonth: _beginDateMonth,
      endDateShowYear: _endDateYear,
      endDateShowMonth: _endDateMonth,
      // 面板展示 date or month - 选择日期 or 月份
      panelType: 'date',
      endDatePanelType: 'date',
      validateMinDate: transDateFormat(props.validateMinDate) || validateData().validateMinDate,
      validateMaxDate: transDateFormat(props.validateMaxDate) || validateData().validateMaxDate
    });
    return _this;
  }

  var _proto = RangePicker.prototype;

  _proto.render = function render() {
    var _classnames, _classnames2, _classnames3;

    var _this$props5 = this.props,
        prefixCls = _this$props5.prefixCls,
        className = _this$props5.className,
        disabled = _this$props5.disabled,
        getPopupContainer = _this$props5.getPopupContainer,
        dateFormat = _this$props5.dateFormat,
        shortcuts = _this$props5.shortcuts,
        popupPlacement = _this$props5.popupPlacement,
        size = _this$props5.size,
        customButtonTitle = _this$props5.customButtonTitle,
        showDeleteIcon = _this$props5.showDeleteIcon,
        placeholder = _this$props5.placeholder,
        style = _this$props5.style,
        hideInput = _this$props5.hideInput,
        children = _this$props5.children;
    var _this$state = this.state,
        currentValue = _this$state.currentValue,
        visible = _this$state.visible,
        errorMessage = _this$state.errorMessage;
    var formatValue;

    if ('value' in this.props) {
      formatValue = this.props.value;
    } else {
      formatValue = currentValue;
    }

    var buttonText = formatButtonText({
      value: formatValue,
      dateFormat: dateFormat
    });
    var text = customButtonTitle || buttonText || /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-range-default-text"
    }, /*#__PURE__*/React.createElement("span", null, placeholder[0]), /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-range-default-text-connector"
    }, "~"), /*#__PURE__*/React.createElement("span", null, placeholder[1]));
    var datePickerClassNames = classnames(prefixCls, className, (_classnames = {}, _classnames[prefixCls + "-empty"] = !buttonText, _classnames[prefixCls + "-disabled"] = disabled, _classnames[prefixCls + "-showDeleteIcon"] = showDeleteIcon, _classnames));

    var layerProps = _extends({
      trigger: disabled ? '' : 'click',
      visible: visible,
      onVisibleChange: this.onLayerVisibleChange,
      overlay: this.renderMultipleDatePickerBody(),
      dropdownMatchSelectWidth: false,
      getPopupContainer: getPopupContainer,
      popupClassName: classnames(prefixCls + "-overlay", !hideInput && prefixCls + "-overlay-" + size, (_classnames2 = {}, _classnames2[prefixCls + "-overlay-has-shortcuts"] = shortcuts && shortcuts.length, _classnames2)),
      popupPlacement: popupPlacement
    }, hideInput ? {} : {
      align: align
    });

    var textClassName = classnames(prefixCls + "-title-text", (_classnames3 = {}, _classnames3[prefixCls + "-title-text-empty"] = !buttonText, _classnames3));
    return /*#__PURE__*/React.createElement("div", {
      className: datePickerClassNames,
      style: style
    }, /*#__PURE__*/React.createElement(Layer, layerProps, children ? /*#__PURE__*/isValidElement(children) ? children : /*#__PURE__*/React.createElement("span", null, children) : /*#__PURE__*/React.createElement(Button, {
      className: prefixCls + "-title",
      disabled: disabled,
      size: size,
      onClick: this.onOpenDatePicker,
      type: "basic"
    }, /*#__PURE__*/React.createElement("span", {
      className: textClassName
    }, text), /*#__PURE__*/React.createElement(IconCalendar, {
      className: prefixCls + "-button-icon " + prefixCls + "-button-icon-calendar"
    }), /*#__PURE__*/React.createElement(IconTimesCircle, {
      className: prefixCls + "-button-icon " + prefixCls + "-button-icon-close",
      onClick: this.onDetete
    }))), errorMessage ? /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-error-message"
    }, errorMessage) : null);
  };

  return RangePicker;
}(PureComponent), _class2.defaultProps = {
  dateFormat: 'YYYY/MM/DD',
  prefixCls: 'one-date-picker',
  className: '',
  disabled: false,
  size: 'medium',
  onVisibleChange: _noop,
  onChange: _noop,
  onClickButton: _noop,
  shortcuts: null,
  getPopupContainer: function getPopupContainer() {
    return document.body;
  },
  showDeleteIcon: false,
  onDelete: _noop,
  placeholder: ['开始日期', '结束日期'],
  mode: 'date'
}, _class2.getDerivedStateFromProps = function (nextProps) {
  var newState = {};

  if ('value' in nextProps) {
    newState.currentValue = nextProps.value;
  }

  if ('visible' in nextProps) {
    newState.visible = nextProps.visible;
  }

  return newState;
}, _temp)) || _class);
export default RangePicker;
