// @baidu/one-ui@4.41.1/lib/components/card/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {
  CardInternal: true
};
exports["default"] = exports.CardInternal = void 0;

var _extends3 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _context = require("../providerConfig/context");

var _interface = require("./interface");

Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  exports[key] = _interface[key];
});

var CardHeader = function CardHeader(_ref) {
  var prefixCls = _ref.prefixCls,
      children = _ref.children,
      extra = _ref.extra;

  if (children || extra) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-header"
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-header"
    }, children), extra ? /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-header-extra"
    }, extra) : null);
  }

  return null;
};

var CardBody = function CardBody(_ref2) {
  var prefixCls = _ref2.prefixCls,
      children = _ref2.children;

  if (children) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-body"
    }, children);
  }

  return null;
};

var CardFooter = function CardFooter(_ref3) {
  var prefixCls = _ref3.prefixCls,
      children = _ref3.children;

  if (children) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-footer"
    }, children);
  }

  return null;
};
/**
 * @visibleName Card
 */


var CardInternal = /*#__PURE__*/(0, _react.forwardRef)(function (_ref4, ref) {
  var _classNames, _extends2;

  var prefixCls = _ref4.prefixCls,
      children = _ref4.children,
      className = _ref4.className,
      header = _ref4.header,
      headerExtra = _ref4.headerExtra,
      footer = _ref4.footer,
      bordered = _ref4.bordered,
      headerLine = _ref4.headerLine,
      footerLine = _ref4.footerLine,
      size = _ref4.size,
      styleProp = _ref4.style,
      borderRadius = _ref4.borderRadius,
      shadow = _ref4.shadow,
      lightPrefix = _ref4.lightPrefix,
      restProps = (0, _objectWithoutPropertiesLoose2["default"])(_ref4, ["prefixCls", "children", "className", "header", "headerExtra", "footer", "bordered", "headerLine", "footerLine", "size", "style", "borderRadius", "shadow", "lightPrefix"]);
  var rootClass = (0, _classnames["default"])(prefixCls, className, (_classNames = {}, _classNames[prefixCls + "-bordered"] = bordered, _classNames[prefixCls + "-shadow-" + shadow] = shadow, _classNames[prefixCls + "-header-line"] = headerLine, _classNames[prefixCls + "-footer-line"] = footerLine, _classNames[prefixCls + "-" + size] = size, _classNames[prefixCls + "-border-radius-" + borderRadius] = typeof borderRadius === 'string', _classNames));
  var style = typeof borderRadius === 'number' ? (0, _extends3["default"])((_extends2 = {}, _extends2["--" + (prefixCls + "-border-radius")] = borderRadius + "px", _extends2), styleProp) : styleProp;
  return /*#__PURE__*/_react["default"].createElement("div", (0, _extends3["default"])({
    className: rootClass
  }, restProps, {
    style: style,
    ref: ref
  }), /*#__PURE__*/_react["default"].createElement(CardHeader, {
    prefixCls: prefixCls,
    extra: headerExtra
  }, header), /*#__PURE__*/_react["default"].createElement(CardBody, {
    prefixCls: prefixCls
  }, children), /*#__PURE__*/_react["default"].createElement(CardFooter, {
    prefixCls: prefixCls
  }, footer));
});
exports.CardInternal = CardInternal;
CardInternal.defaultProps = {
  prefixCls: 'one-card',
  size: 'medium',
  bordered: true,
  headerLine: true,
  footerLine: true
};
CardInternal.displayName = 'Card';
var Card = (0, _context.withConfigConsumer)('card')(CardInternal);
var _default = Card;
exports["default"] = _default;
