// @baidu/one-ui@4.32.1/lib/components/cascaderPane/input.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _input = _interopRequireDefault(require("../input"));

;

var CascaderInput = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(CascaderInput, _PureComponent);

  function CascaderInput(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.onInputChange = function (e) {
      var value = e.value;

      if (!('value' in _this.props)) {
        _this.setState({
          value: value
        });
      }

      _this.props.onInputChange(e);
    };

    _this.state = {
      value: props.value || props.defaultValue || undefined
    };
    return _this;
  }

  CascaderInput.getDerivedStateFromProps = function getDerivedStateFromProps(nextProps) {
    var newState = {};

    if ('value' in nextProps) {
      newState.value = nextProps.value;
    }

    return newState;
  };

  var _proto = CascaderInput.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        searchProps = _this$props.searchProps,
        width = _this$props.width,
        inputType = _this$props.inputType;
    var value = this.state.value;
    return /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement(_input["default"], (0, _extends2["default"])({}, searchProps, {
      value: value,
      onChange: this.onInputChange,
      width: width,
      type: inputType
    })));
  };

  return CascaderInput;
}(_react.PureComponent);

CascaderInput.defaultProps = {
  searchProps: {
    placeholder: '请输入...'
  },
  onInputChange: function onInputChange() {},
  width: 360
};
var _default = CascaderInput;
exports["default"] = _default;
module.exports = exports.default;
