// @baidu/one-ui@4.32.1/lib/components/cascaderPane/menu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _last2 = _interopRequireDefault(require("lodash/last"));

var _react = _interopRequireWildcard(require("react"));

var _column = _interopRequireDefault(require("./column"));

/**
 * @file 菜单
 */
var Menu = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Menu, _PureComponent);

  function Menu() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Menu.prototype;

  _proto.prefix = function prefix(name) {
    return this.props.prefixCls + "-" + name;
  };

  _proto.render = function render() {
    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        size = _this$props.size,
        showCheckbox = _this$props.showCheckbox,
        style = _this$props.style,
        CustomItemRender = _this$props.CustomItemRender,
        firstColumnGroup = _this$props.firstColumnGroup,
        expandIcon = _this$props.expandIcon,
        loadingIcon = _this$props.loadingIcon,
        value = _this$props.value,
        onSelect = _this$props.onSelect,
        expandTrigger = _this$props.expandTrigger,
        columnHeader = _this$props.columnHeader,
        columnFooter = _this$props.columnFooter,
        activeValue = _this$props.activeValue,
        onHover = _this$props.onHover,
        onCheck = _this$props.onCheck,
        width = _this$props.width,
        columnWidth = _this$props.columnWidth,
        columns = _this$props.columns,
        forwardRef = _this$props.forwardRef,
        renderOption = _this$props.renderOption,
        loadingMap = _this$props.loadingMap;
    var menuStyle = (0, _extends2["default"])({}, style);

    if (width) {
      menuStyle.width = width;
    }

    return /*#__PURE__*/_react["default"].createElement("div", {
      className: this.prefix('menus'),
      style: menuStyle,
      ref: forwardRef
    }, this.renderHeader(), /*#__PURE__*/_react["default"].createElement("div", {
      className: this.prefix('menus-container')
    }, columns.map(function (options, index) {
      var lastOne = columns.length - 1 === index;
      return /*#__PURE__*/_react["default"].createElement(_column["default"], {
        prefixCls: prefixCls,
        key: index,
        size: size,
        index: index,
        width: lastOne ? undefined : columnWidth,
        options: options,
        value: (0, _last2["default"])(value),
        valuePath: value,
        activeValuePath: activeValue,
        expandIcon: expandIcon,
        loadingIcon: loadingIcon,
        loadingMap: loadingMap,
        onSelect: onSelect,
        onHover: onHover,
        onCheck: onCheck,
        showCheckbox: showCheckbox,
        expandTrigger: expandTrigger,
        firstColumnGroup: index === 0 && firstColumnGroup,
        CustomItemRender: index === 0 && CustomItemRender,
        header: columnHeader,
        footer: columnFooter,
        renderOption: renderOption
      });
    })), this.renderFooter());
  };

  _proto.renderHeader = function renderHeader() {
    var header = this.props.header;

    if (typeof header === 'function') {
      header = header();
    }

    if (header) {
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: this.prefix('header')
      }, header);
    }

    return null;
  };

  _proto.renderFooter = function renderFooter() {
    var footer = this.props.footer;

    if (typeof footer === 'function') {
      footer = footer();
    }

    if (footer) {
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: this.prefix('footer')
      }, footer);
    }

    return null;
  };

  return Menu;
}(_react.PureComponent);

exports["default"] = Menu;
module.exports = exports.default;
