// @baidu/one-ui@4.32.1/src/components/timePicker/common/timeSelecter.tsx
import React, {createRef, MouseEventHandler, PureComponent} from 'react';
import {scrollTo} from '../../../core/timePickerTools';
import {classnames} from '../../../core/commonTools';

interface TimeSelectorProps {
    prefixCls: string,
    options: {value: string, disabled?: boolean}[],
    selectedIndex: number,
    type: string,
    onSelect(type: string, value: number),
    onMouseEnter: MouseEventHandler
};

export default class TimeSelecter extends PureComponent<TimeSelectorProps> {

    state = {
        active: false
    };

    root = createRef<HTMLDivElement>();

    componentDidMount() {
        // jump to selected option
        this.scrollToSelected(0);
    }

    componentDidUpdate(prevProps) {
        // smooth scroll to selected option
        if (prevProps.selectedIndex !== this.props.selectedIndex) {
            this.scrollToSelected(120);
        }
    }

    onSelect = value => {
        const {onSelect, type} = this.props;
        onSelect(type, value);
    };

    getOptions = () => {
        const {options, selectedIndex, prefixCls} = this.props;
        return options.map((item, index) => {
            const cls = classnames({
                [`${prefixCls}-select-option-selected`]: selectedIndex === index,
                [`${prefixCls}-select-option-disabled`]: item.disabled
            });
            const onClick = item.disabled
                ? undefined
                : () => {
                    this.onSelect(item.value);
                };
            return (
                <li onClick={onClick} className={cls} key={index}>
                    {item.value}
                </li>
            );
        });
    }

    handleMouseEnter = e => {
        this.setState({active: true});
        this.props.onMouseEnter(e);
    };

    handleMouseLeave = () => {
        this.setState({active: false});
    };

    list: HTMLUListElement;

    saveList = node => {
        this.list = node;
    };

    scrollToSelected = duration => {
        // move to selected item
        if (!this.list) {
            return;
        }
        let index = this.props.selectedIndex;
        if (index < 0) {
            index = 0;
        }
        const topOption = this.list.children[index] as HTMLLIElement;
        const to = topOption.offsetTop;
        scrollTo(this.root.current, to, duration);
    }

    render() {
        const {prefixCls, options} = this.props;
        if (options.length === 0) {
            return null;
        }
        const cls = classnames(
            `${prefixCls}-select`,
            {
                [`${prefixCls}-select-active`]: this.state.active
            }
        );
        return (
            <div
                className={cls}
                onMouseEnter={this.handleMouseEnter}
                onMouseLeave={this.handleMouseLeave}
                ref={this.root}
            >
                <ul ref={this.saveList}>{this.getOptions()}</ul>
            </div>
        );
    }
}
