// @baidu/one-ui@4.41.1/es/components/uploader/common/fileItem.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
import { IconFile, IconTrash } from 'dls-icons-react';
import Progress from '../../progress';
import Button from '../../button';
import { originStatus } from '../../../core/uploaderTools';
import { classnames } from '../../../core/commonTools';

var FileItem = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(FileItem, _PureComponent);

  function FileItem() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = FileItem.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        status = _this$props.status,
        name = _this$props.name,
        progressStep = _this$props.progressStep,
        errorMessage = _this$props.errorMessage,
        onRemove = _this$props.onRemove,
        index = _this$props.index,
        className = _this$props.className,
        controls = _this$props.controls,
        file = _this$props.file;
    var itemClassNames = classnames(prefixCls + "-file-item", prefixCls + "-file-item-" + status, className);
    var actions = [{
      key: 'delete',
      icon: /*#__PURE__*/React.createElement(IconTrash, null),
      onClick: function onClick(file, event) {
        return onRemove(index, event);
      }
    }];

    if (typeof controls === 'function') {
      actions = controls(file, actions) || [];
    }

    return /*#__PURE__*/React.createElement("div", {
      className: itemClassNames
    }, /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-file-item-container"
    }, /*#__PURE__*/React.createElement(IconFile, {
      className: prefixCls + "-file-item-file-icon"
    }), /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-file-item-label"
    }, name), actions.map(function (_ref) {
      var visible = _ref.visible,
          disabled = _ref.disabled,
          key = _ref.key,
          icon = _ref.icon,
          _onClick = _ref.onClick;

      if (visible === false) {
        return null;
      }

      return /*#__PURE__*/React.createElement(Button, {
        key: key,
        type: "text-aux",
        disabled: disabled,
        className: prefixCls + "-file-item-control",
        onClick: function onClick(event) {
          return _onClick(file, event);
        }
      }, icon);
    })), status === originStatus.UPLOADING ? /*#__PURE__*/React.createElement(Progress, {
      size: "small",
      percent: progressStep,
      showInfo: false,
      className: prefixCls + "-file-item-progress"
    }) : null, status === originStatus.ERROR && errorMessage && errorMessage.length ? /*#__PURE__*/React.createElement("div", {
      className: prefixCls + "-file-item-message"
    }, errorMessage.join('，')) : null);
  };

  return FileItem;
}(PureComponent);

export { FileItem as default };
