// @baidu/one-ui@4.41.1/lib/components/breadcrumb/breadcrumbItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _dlsIconsReact = require("dls-icons-react");

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

var BreadcrumbItem = (_dec = (0, _context.withConfigConsumer)('breadcrumb'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(BreadcrumbItem, _PureComponent);

  function BreadcrumbItem() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = BreadcrumbItem.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        separator = _this$props.separator,
        children = _this$props.children,
        onClick = _this$props.onClick,
        disabled = _this$props.disabled,
        className = _this$props.className,
        lightPrefix = _this$props.lightPrefix,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, ["prefixCls", "separator", "children", "onClick", "disabled", "className", "lightPrefix"]);
    var link;
    var linkClassName = (0, _commonTools.classnames)((_classnames = {}, _classnames[prefixCls + "-link-disabled"] = disabled, _classnames[prefixCls + "-link"] = !disabled, _classnames), className);
    var linkProps = (0, _extends2["default"])({
      className: linkClassName
    }, restProps);

    if ('href' in this.props) {
      if (disabled) {
        delete linkProps.href;
      }

      link = /*#__PURE__*/_react["default"].createElement("a", linkProps, children);
    } else {
      link = /*#__PURE__*/_react["default"].createElement("span", linkProps, children);
    }

    if (children) {
      return /*#__PURE__*/_react["default"].createElement("span", {
        onClick: onClick
      }, link, /*#__PURE__*/_react["default"].createElement("span", {
        className: prefixCls + "-separator"
      }, separator));
    }

    return null;
  };

  return BreadcrumbItem;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-breadcrumb',
  separator: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconChevronRight, null),
  onClick: _noop2["default"],
  disabled: false
}, _temp)) || _class);
var _default = BreadcrumbItem;
exports["default"] = _default;
module.exports = exports.default;
