// @baidu/one-ui@4.41.1/lib/components/breadcrumb/breadcrumb.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _dlsIconsReact = require("dls-icons-react");

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

var Breadcrumb = (_dec = (0, _context.withConfigConsumer)('breadcrumb'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Breadcrumb, _PureComponent);

  function Breadcrumb() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Breadcrumb.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        separator = _this$props.separator,
        prefixCls = _this$props.prefixCls,
        style = _this$props.style,
        className = _this$props.className,
        children = _this$props.children,
        size = _this$props.size,
        type = _this$props.type;

    var crumbs = _react["default"].Children.map(children, function (element, index) {
      if (! /*#__PURE__*/_react["default"].isValidElement(element)) {
        return element;
      }

      return /*#__PURE__*/(0, _react.cloneElement)(element, {
        separator: separator,
        key: index
      });
    });

    var breadcrumbCls = (0, _commonTools.classnames)(prefixCls, prefixCls + "-" + size, className, prefixCls + "-" + type);
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: breadcrumbCls,
      style: style
    }, crumbs);
  };

  return Breadcrumb;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-breadcrumb',
  separator: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconChevronRight, null),
  size: 'medium',
  type: 'normal'
}, _class2.Item = void 0, _temp)) || _class);
var _default = Breadcrumb;
exports["default"] = _default;
module.exports = exports.default;
