// @baidu/one-ui@4.41.1/lib/components/message/message.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../core/commonTools");

var _popover = _interopRequireDefault(require("../popover"));

var _dlsIconsReact = require("dls-icons-react");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

var typedIcons = {
  info: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconInfoCircleSolid, null),
  aux: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconInfoCircleSolid, null),
  error: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconExclamationCircleSolid, null),
  warning: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconExclamationTriangleSolid, null),
  success: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconCheckCircleSolid, null)
};
var Message = (_dec = (0, _context.withConfigConsumer)('message'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Message, _PureComponent);

  function Message() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Message.prototype;

  _proto.renderRoot = function renderRoot(iconNode, contentNode) {
    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        className = _this$props.className,
        type = _this$props.type,
        size = _this$props.size,
        display = _this$props.display,
        children = _this$props.children,
        icon = _this$props.icon,
        lightPrefix = _this$props.lightPrefix,
        resetProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, ["prefixCls", "className", "type", "size", "display", "children", "icon", "lightPrefix"]);
    return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
      className: (0, _commonTools.classnames)(prefixCls, prefixCls + "-" + size, prefixCls + "-type-" + type, prefixCls + "-display-" + display, className)
    }, resetProps), iconNode, display !== 'popup' ? contentNode : null);
  };

  _proto.renderIcon = function renderIcon(contentNode) {
    var _this$props2 = this.props,
        icon = _this$props2.icon,
        type = _this$props2.type,
        display = _this$props2.display,
        prefixCls = _this$props2.prefixCls;

    if (display === 'simple') {
      return null;
    }

    var iconNode = icon || typedIcons[type];

    if (display === 'popup') {
      iconNode = /*#__PURE__*/_react["default"].createElement(_popover["default"], {
        placement: "top",
        content: contentNode
      }, iconNode);
    }

    return /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-icon"
    }, iconNode);
  };

  _proto.render = function render() {
    var _this$props3 = this.props,
        prefixCls = _this$props3.prefixCls,
        children = _this$props3.children;

    var contentNode = /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-content"
    }, children);

    var iconNode = this.renderIcon(contentNode);
    return this.renderRoot(iconNode, contentNode);
  };

  return Message;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-message',
  type: 'info',
  size: 'medium',
  display: 'normal'
}, _temp)) || _class);
var _default = Message;
exports["default"] = _default;
module.exports = exports.default;
