// @baidu/one-ui@4.32.1/es/components/datePicker/common/input.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _partial from "lodash/partial";
import React, { PureComponent } from 'react';
import { connect } from 'mini-store';
import dayjs from 'dayjs';
import { formatMultipleDate, getTimeStamp, getTimeTramp, monthDayRange } from '../../../core/datePickerTools';
import { getValidDate, getValidWeekDates } from '../util';

var Input = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Input, _PureComponent);

  function Input(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.onChangeInput = function (type, e) {
      var value = e.target.value;
      var _this$props = _this.props,
          dateFormat = _this$props.dateFormat,
          onChange = _this$props.onChange,
          multiple = _this$props.multiple,
          validateMaxDate = _this$props.validateMaxDate,
          validateMinDate = _this$props.validateMinDate,
          inputType = _this$props.inputType,
          validateDisabled = _this$props.validateDisabled,
          mode = _this$props.mode;
      var _this$state = _this.state,
          endDate = _this$state.endDate,
          beginDate = _this$state.beginDate;
      var reg = /^[\d/.///-]*$/;

      if (!reg.test(value)) {
        return;
      }

      var newState = {};

      if (type === 'beginDate') {
        newState = {
          beginDate: value
        };
      } else if (type === 'endDate') {
        newState = {
          endDate: value
        };
      }

      _this.setState(newState);

      if (validateDisabled && typeof validateDisabled === 'function') {
        if (validateDisabled({
          dayItem: value,
          timeStamp: getTimeTramp(value),
          getTimeStamp: getTimeTramp
        })) {
          return;
        }
      }

      if (newState.endDate) {
        var endDateTimeTramp = getTimeTramp(newState.endDate);

        if (endDateTimeTramp < getTimeTramp(validateMinDate) || endDateTimeTramp > getTimeTramp(validateMaxDate)) {
          return;
        }
      }

      if (newState.beginDate) {
        var beginDateTimeTramp = getTimeTramp(newState.beginDate);

        if (beginDateTimeTramp < getTimeTramp(validateMinDate) || beginDateTimeTramp > getTimeTramp(validateMaxDate)) {
          return;
        }
      }

      value = value.replace(/\./g, '/');
      var timeArray = value.split('/');

      if (timeArray.length === 1) {
        timeArray = value.split('-');
      }

      if (getTimeStamp(value) && inputType !== 'month' && timeArray.length === 3 && ("" + timeArray[0]).length === 4 && ("" + timeArray[1]).length === 2 && ("" + timeArray[2]).length === 2 && monthDayRange(timeArray[0])[+timeArray[1] - 1] && +timeArray[2] <= monthDayRange(timeArray[0])[+timeArray[1] - 1]) {
        // 合法时间戳的话，并且日期为标准的XXXX/XX/XX格式，则实时修改
        var formatDate = dayjs(new Date(value)).format(dateFormat);

        if (!multiple && mode !== 'week' && getTimeTramp(value) <= getTimeTramp(validateMaxDate) && getTimeTramp(value) >= getTimeTramp(validateMinDate)) {
          onChange(formatDate, false);
        } else if (type === 'beginDate' && getTimeTramp(value) < getTimeTramp(endDate) && getTimeTramp(value) >= getTimeTramp(validateMinDate)) {
          if (mode === 'week' && !multiple) {
            onChange(getValidWeekDates(value, dateFormat, validateMinDate, validateMaxDate), false);
          } else {
            var endDateFormat = endDate ? dayjs(new Date(endDate)).format(dateFormat) : '';
            onChange([getValidDate(formatDate, dateFormat, validateMinDate, validateMaxDate, mode, false), endDateFormat], false);
          }
        } else if (type === 'endDate' && getTimeTramp(value) > getTimeTramp(beginDate) && getTimeTramp(value) <= getTimeTramp(validateMaxDate)) {
          if (mode === 'week' && !multiple) {
            onChange(getValidWeekDates(value, dateFormat, validateMinDate, validateMaxDate), false);
          } else {
            var beginDateFormat = beginDate ? dayjs(new Date(beginDate)).format(dateFormat) : '';
            onChange([beginDateFormat, getValidDate(formatDate, dateFormat, validateMinDate, validateMaxDate, mode, true)], false);
          }
        }
      }

      if (inputType === 'month' && timeArray.length === 2 && ("" + timeArray[0]).length === 4 && ("" + timeArray[1]).length === 2 && +timeArray[1] < 13 && getTimeTramp(value) <= getTimeTramp(validateMaxDate) && getTimeTramp(value) >= getTimeTramp(validateMinDate)) {
        var _formatDate = dayjs(new Date(value)).format(dateFormat);

        onChange(_formatDate, false);
      }
    };

    _this.renderInput = function () {
      var inputProps = {
        value: _this.state.beginDate,
        onChange: _partial(_this.onChangeInput, 'beginDate'),
        placeholder: '请输入日期'
      };
      return /*#__PURE__*/React.createElement("div", {
        className: _this.props.prefixCls + "-input-container"
      }, /*#__PURE__*/React.createElement("input", inputProps));
    };

    _this.renderMultipleInput = function () {
      var prefixCls = _this.props.prefixCls;
      var _this$state2 = _this.state,
          beginDate = _this$state2.beginDate,
          endDate = _this$state2.endDate;
      var beginInputProps = {
        value: beginDate || '',
        onChange: _partial(_this.onChangeInput, 'beginDate'),
        placeholder: '请输入日期'
      };
      var endInputProps = {
        value: endDate || '',
        onChange: _partial(_this.onChangeInput, 'endDate'),
        placeholder: '请输入日期'
      };
      var containerClassNames = prefixCls + "-input-multiple";
      return /*#__PURE__*/React.createElement("div", {
        className: containerClassNames
      }, /*#__PURE__*/React.createElement("div", {
        className: _this.props.prefixCls + "-input-container"
      }, /*#__PURE__*/React.createElement("input", beginInputProps)), /*#__PURE__*/React.createElement("span", null, " ~ "), /*#__PURE__*/React.createElement("div", {
        className: _this.props.prefixCls + "-input-container"
      }, /*#__PURE__*/React.createElement("input", endInputProps)));
    };

    var _value = props.value,
        _dateFormat = props.dateFormat;

    var _formatMultipleDate = formatMultipleDate(_value, _dateFormat),
        _beginDate = _formatMultipleDate.beginDate,
        _endDate = _formatMultipleDate.endDate;

    _this.state = {
      beginDate: _beginDate,
      endDate: _endDate,
      prevProps: _this.props
    };
    return _this;
  }

  var _proto = Input.prototype;

  _proto.render = function render() {
    var _this$props2 = this.props,
        prefixCls = _this$props2.prefixCls,
        multiple = _this$props2.multiple,
        mode = _this$props2.mode;
    var inputClassNames = prefixCls + "-input";
    return /*#__PURE__*/React.createElement("div", {
      className: inputClassNames
    }, multiple || mode === 'week' ? this.renderMultipleInput() : this.renderInput());
  };

  return Input;
}(PureComponent);

Input.defaultProps = {
  onChange: function onChange() {},
  multiple: false
};

Input.getDerivedStateFromProps = function (nextProps, prevState) {
  var prevProps = prevState.prevProps;

  if ('value' in nextProps && nextProps.value !== prevProps.value || nextProps.visible !== prevProps.visible) {
    var _value2 = nextProps.value,
        dateFormat = nextProps.dateFormat;

    var _formatMultipleDate2 = formatMultipleDate(_value2, dateFormat),
        beginDate = _formatMultipleDate2.beginDate,
        endDate = _formatMultipleDate2.endDate;

    return {
      beginDate: beginDate,
      endDate: endDate,
      prevProps: nextProps
    };
  }

  return null;
};

export default connect(function (state) {
  return {
    value: state._value,
    validateMinDate: state.validateMinDate,
    validateMaxDate: state.validateMaxDate
  };
})(Input);
