// @baidu/one-ui@4.32.1/es/components/datePicker/common/rangeRender.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import _extends from "@babel/runtime/helpers/extends";
import _partial from "lodash/partial";
import React, { PureComponent } from 'react';
import { connect } from 'mini-store';
import dayjs from 'dayjs';
import MonthAndYearPanel from './monthAndYearPanel';
import RangeDayRender from './rangeDayRender';
import { getTimeTramp, getDetailDate } from '../../../core/datePickerTools';
import en from 'dayjs/locale/en';
import { getValidDate } from '../util';
dayjs.locale(_extends({}, en, {
  weekStart: 1
}));
;
;

var RangeRender = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(RangeRender, _PureComponent);

  function RangeRender(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.store = void 0;

    _this.onPickerDay = function (type, value, readOnly) {
      if (readOnly === void 0) {
        readOnly = false;
      }

      var _this$state = _this.state,
          step = _this$state.step,
          beginDate = _this$state.beginDate;
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          dateFormat = _this$props.dateFormat,
          showYear = _this$props.showYear,
          showMonth = _this$props.showMonth,
          endDateShowYear = _this$props.endDateShowYear,
          endDateShowMonth = _this$props.endDateShowMonth,
          validateMinDate = _this$props.validateMinDate,
          validateMaxDate = _this$props.validateMaxDate,
          mode = _this$props.mode;

      if (step === 0) {
        // 表示开始选择
        _this.setState({
          step: 1,
          endDate: '',
          beginDate: getValidDate(value, dateFormat, validateMinDate, validateMaxDate, mode, false)
        }); // 如果点击了readOnly部分的话，修改当前面板展示的年月


        if (readOnly) {
          var _getDetailDate = getDetailDate(value),
              fullYear = _getDetailDate.fullYear,
              fullMonth = _getDetailDate.fullMonth;

          var newState = {};
          var currentFirstDate = getTimeTramp(fullYear + "/" + fullMonth + "/01");

          if (type === 'prevMultiple' && currentFirstDate !== getTimeTramp(endDateShowYear + "/" + endDateShowMonth + "/01")) {
            newState = {
              showYear: fullYear,
              showMonth: fullMonth
            };
          } else if (type === 'nextMultiple' && currentFirstDate !== getTimeTramp(showYear + "/" + showMonth + "/01")) {
            newState = {
              endDateShowYear: fullYear,
              endDateShowMonth: fullMonth
            };
          }

          _this.store.setState(newState);
        }
      } else if (step === 1) {
        var currentBeginDate = beginDate;
        var currentEndDate = getValidDate(value, dateFormat, validateMinDate, validateMaxDate, mode, true);

        if (getTimeTramp(currentBeginDate) > getTimeTramp(currentEndDate)) {
          currentBeginDate = getValidDate(value, dateFormat, validateMinDate, validateMaxDate, mode, false);
          currentEndDate = getValidDate(beginDate, dateFormat, validateMinDate, validateMaxDate, mode, true);
        }

        _this.setState({
          step: 0,
          endDate: currentEndDate,
          beginDate: currentBeginDate,
          hoverDate: ''
        });

        onChange([dayjs(new Date(currentBeginDate)).format(dateFormat), dayjs(new Date(currentEndDate)).format(dateFormat)]);
      }
    };

    _this.onMouseEnter = function (value) {
      _this.setState({
        hoverDate: value
      });
    };

    _this.onMouseLeave = function () {
      _this.setState({
        hoverDate: ''
      });
    };

    var currentDate = props.currentDate;
    _this.store = _this.props.store;
    _this.state = {
      beginDate: currentDate[0] || '',
      endDate: currentDate[1] || '',
      prevProps: props,
      step: 0,
      hoverDate: ''
    };
    return _this;
  }

  RangeRender.getDerivedStateFromProps = function getDerivedStateFromProps(nextProps, prevState) {
    var prevProps = prevState.prevProps;
    var newState = {};

    if ('currentDate' in nextProps && nextProps.currentDate !== prevProps.currentDate || prevProps.visible !== nextProps.visible) {
      newState = _extends({}, newState, {
        beginDate: nextProps.currentDate[0] || '',
        endDate: nextProps.currentDate[1] || '',
        prevProps: nextProps,
        step: 0
      });
    }

    return newState;
  };

  var _proto = RangeRender.prototype;

  _proto.getDate = function getDate(date, endOfWeek) {
    var _this$props2 = this.props,
        mode = _this$props2.mode,
        dateFormat = _this$props2.dateFormat,
        validateMinDate = _this$props2.validateMinDate,
        validateMaxDate = _this$props2.validateMaxDate;
    var maxDateTime = validateMaxDate ? getTimeTramp(validateMaxDate) : validateMaxDate;
    var minDateTime = validateMinDate ? getTimeTramp(validateMinDate) : validateMinDate;
    var validTime;

    if (mode === 'week') {
      if (endOfWeek) {
        var dateTime = getTimeTramp(dayjs(date).endOf('week').startOf('day'));
        validTime = maxDateTime && dateTime > maxDateTime ? maxDateTime : dateTime;
      } else {
        var _dateTime = getTimeTramp(dayjs(date).startOf('week'));

        validTime = minDateTime && _dateTime > minDateTime ? _dateTime : minDateTime;
      }
    } else {
      validTime = getTimeTramp(date);
    }

    return dayjs(validTime).format(dateFormat);
  };

  _proto.render = function render() {
    var containerClassName = this.props.prefixCls + "-range";
    var _this$state2 = this.state,
        beginDate = _this$state2.beginDate,
        endDate = _this$state2.endDate,
        hoverDate = _this$state2.hoverDate;
    var currentDate = [beginDate, endDate];
    return /*#__PURE__*/React.createElement("div", {
      className: containerClassName
    }, /*#__PURE__*/React.createElement("div", {
      className: containerClassName + "-item"
    }, /*#__PURE__*/React.createElement(MonthAndYearPanel, _extends({}, this.props, {
      type: "prevMultiple"
    })), /*#__PURE__*/React.createElement(RangeDayRender, _extends({}, this.props, {
      type: "prevMultiple",
      currentDate: currentDate,
      onChange: _partial(this.onPickerDay, 'prevMultiple'),
      onMouseEnter: this.onMouseEnter,
      onMouseLeave: this.onMouseLeave,
      hoverDate: hoverDate
    }))), /*#__PURE__*/React.createElement("div", {
      className: containerClassName + "-item"
    }, /*#__PURE__*/React.createElement(MonthAndYearPanel, _extends({}, this.props, {
      type: "nextMultiple"
    })), /*#__PURE__*/React.createElement(RangeDayRender, _extends({}, this.props, {
      type: "nextMultiple",
      currentDate: currentDate,
      onChange: _partial(this.onPickerDay, 'nextMultiple'),
      onMouseEnter: this.onMouseEnter,
      onMouseLeave: this.onMouseLeave,
      hoverDate: hoverDate
    }))));
  };

  return RangeRender;
}(PureComponent);

RangeRender.defaultProps = {
  onChange: function onChange() {}
};
export default connect(function (state) {
  return {
    currentDate: state._value,
    endDateShowYear: state.endDateShowYear,
    endDateShowMonth: state.endDateShowMonth,
    showYear: state.showYear,
    showMonth: state.showMonth
  };
})(RangeRender);
