// less-plugin-dls@11.11.0/variables.palette.ai.js
export const dlsColorBrand = '#4d79ff'
export const dlsColorBrand0 = '#fff'
export const dlsColorBrand1 = '#f2f7ff'
export const dlsColorBrand2 = '#e3edff'
export const dlsColorBrand3 = '#c4daff'
export const dlsColorBrand4 = '#99beff'
export const dlsColorBrand5 = '#729cfe'
export const dlsColorBrand6 = '#4d79ff'
export const dlsColorBrand7 = '#3a5bfd'
export const dlsColorBrand8 = '#333fe6'
export const dlsColorBrand9 = '#0f2dbd'
export const dlsColorBrand10 = '#1c244a'
export const dlsColorBrand11 = '#000'
export const dlsColorInfo = '#4d79ff'
export const dlsColorInfo0 = '#fff'
export const dlsColorInfo1 = '#f2f7ff'
export const dlsColorInfo2 = '#e3edff'
export const dlsColorInfo3 = '#c4daff'
export const dlsColorInfo4 = '#99beff'
export const dlsColorInfo5 = '#729cfe'
export const dlsColorInfo6 = '#4d79ff'
export const dlsColorInfo7 = '#3a5bfd'
export const dlsColorInfo8 = '#333fe6'
export const dlsColorInfo9 = '#0f2dbd'
export const dlsColorInfo10 = '#1c244a'
export const dlsColorInfo11 = '#000'
export const dlsColorSuccess = '#3ac222'
export const dlsColorSuccess0 = '#fff'
export const dlsColorSuccess1 = '#eeffeb'
export const dlsColorSuccess2 = '#e3ffde'
export const dlsColorSuccess3 = '#cfffc7'
export const dlsColorSuccess4 = '#b2ffa5'
export const dlsColorSuccess5 = '#8cff78'
export const dlsColorSuccess6 = '#59f53d'
export const dlsColorSuccess7 = '#3ac222'
export const dlsColorSuccess8 = '#19a500'
export const dlsColorSuccess9 = '#137e00'
export const dlsColorSuccess10 = '#0c4e00'
export const dlsColorSuccess11 = '#000'
export const dlsColorWarning = '#fea800'
export const dlsColorWarning0 = '#fff'
export const dlsColorWarning1 = '#fff8eb'
export const dlsColorWarning2 = '#fff4dd'
export const dlsColorWarning3 = '#ffebc4'
export const dlsColorWarning4 = '#ffdf9f'
export const dlsColorWarning5 = '#ffce6f'
export const dlsColorWarning6 = '#ffba32'
export const dlsColorWarning7 = '#fea800'
export const dlsColorWarning8 = '#d88f00'
export const dlsColorWarning9 = '#a56d00'
export const dlsColorWarning10 = '#664300'
export const dlsColorWarning11 = '#000'
export const dlsColorError = '#f53f3f'
export const dlsColorError0 = '#fff'
export const dlsColorError1 = '#ffebeb'
export const dlsColorError2 = '#ffe2e2'
export const dlsColorError3 = '#ffd3d3'
export const dlsColorError4 = '#ffbcbc'
export const dlsColorError5 = '#ff9e9e'
export const dlsColorError6 = '#ff7878'
export const dlsColorError7 = '#f53f3f'
export const dlsColorError8 = '#d01b1b'
export const dlsColorError9 = '#9f0000'
export const dlsColorError10 = '#620000'
export const dlsColorError11 = '#000'
export const dlsColorGray0 = '#fff'
export const dlsColorGray1 = '#f6f7fa'
export const dlsColorGray2 = '#ebedf5'
export const dlsColorGray3 = '#e2e6f0'
export const dlsColorGray4 = '#d3d9e6'
export const dlsColorGray5 = '#c5ccdb'
export const dlsColorGray6 = '#a8b0bf'
export const dlsColorGray7 = '#848b99'
export const dlsColorGray8 = '#545b66'
export const dlsColorGray9 = '#191b1e'
export const dlsColorGray10 = '#0a0b0d'
export const dlsColorGray11 = '#000'
export const dlsColorTranslucent0 = 'transparent'
export const dlsColorTranslucent1 = 'rgba(110, 156, 247, 0.08)'
export const dlsColorTranslucent2 = 'rgba(82, 129, 224, 0.12)'
export const dlsColorTranslucent3 = 'rgba(133, 163, 224, 0.25)'
export const dlsColorTranslucent4 = 'rgba(41, 82, 163, 0.2)'
export const dlsColorTranslucent5 = 'rgba(60, 83, 133, 0.3)'
export const dlsColorTranslucent6 = 'rgba(83, 98, 128, 0.5)'
export const dlsColorTranslucent7 = 'rgba(80, 90, 107, 0.8)'
export const dlsColorTranslucent8 = 'rgba(45, 52, 64, 0.8)'
export const dlsColorTranslucent9 = 'rgba(13, 15, 18, 0.95)'
export const dlsColorTranslucent10 = 'rgba(4, 4, 5, 0.95)'
export const dlsColorTranslucent11 = '#000'
