// @baidu/one-ui@4.41.1/lib/components/transfer/common/candidateTitle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _title = _interopRequireDefault(require("./title"));

var _button = _interopRequireDefault(require("../../button"));

var _tooltip = _interopRequireDefault(require("../../tooltip"));

var _transferTools = require("../../../core/transferTools");

/**
 * @file candidate render
 * @author huangshiming
 * @date 2020-05-16
 */
;

var CandidateTitle = function CandidateTitle(props) {
  var prefixCls = props.prefixCls,
      maxSelectedNum = props.maxSelectedNum,
      allDataMap = props.allDataMap,
      candidateList = props.candidateList,
      selectedList = props.selectedList,
      treeName = props.treeName,
      size = props.size,
      onSelectAll = props.onSelectAll,
      showSelectAll = props.showSelectAll,
      showCandidateNum = props.showCandidateNum,
      CustomCandidateOperation = props.CustomCandidateOperation;
  var candidateNum = ((0, _transferTools.getCandidateNumber)(allDataMap, candidateList) || []).length;
  var crossLength = 0;
  candidateList.forEach(function (candidate) {
    var strSelectedList = selectedList.map(function (item) {
      return "" + item;
    });

    if (strSelectedList.indexOf("" + candidate) > -1) {
      crossLength++;
    }
  });
  var selectedListLength = selectedList.length;
  var disabledAddAll = maxSelectedNum && maxSelectedNum <= selectedListLength || maxSelectedNum && candidateNum > maxSelectedNum || maxSelectedNum && maxSelectedNum < selectedListLength + candidateNum - crossLength;

  var AddAllButton = /*#__PURE__*/_react["default"].createElement(_button["default"], {
    type: "text-strong",
    size: size,
    className: prefixCls + "-add-all-button",
    disabled: disabledAddAll,
    onClick: onSelectAll
  }, "\u5168\u9009");

  var titleRender;

  if (props.CandidateTitleRender && typeof props.CandidateTitleRender === 'function') {
    var customRenderProps = {
      title: '可选',
      showCandidateNum: showCandidateNum,
      showSelectedNum: false,
      unSelectedNum: candidateNum,
      selectedNum: selectedListLength
    };
    var CustomCandidateTitleRender = props.CandidateTitleRender;
    titleRender = /*#__PURE__*/_react["default"].createElement(CustomCandidateTitleRender, (0, _extends2["default"])({}, props, customRenderProps));
  } else if (props.CandidateTitleRender && typeof props.CandidateTitleRender === 'object') {
    titleRender = props.CandidateTitleRender;
  } else {
    titleRender = /*#__PURE__*/_react["default"].createElement(_title["default"], (0, _extends2["default"])({}, props, {
      title: "\u53EF\u9009",
      showCandidateNum: showCandidateNum,
      showSelectedNum: false,
      candidateNum: candidateNum,
      selectedNum: selectedListLength
    }));
  }

  var operation;

  if (CustomCandidateOperation) {
    operation = CustomCandidateOperation;
  } else {
    operation = !showSelectAll ? null : disabledAddAll ? /*#__PURE__*/_react["default"].createElement(_tooltip["default"], {
      title: "\u53EF\u9009" + treeName + "\u6570\u91CF\u8D85\u8FC7\u53EF\u6DFB\u52A0\u6570\u91CF\u7684\u4E0A\u9650\uFF0C\u65E0\u6CD5\u6267\u884C\u5168\u90E8\u6DFB\u52A0",
      placement: "rightBottom"
    }, AddAllButton) : AddAllButton;
  }

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: prefixCls + "-pane-title"
  }, titleRender, operation);
};

var _default = CandidateTitle;
exports["default"] = _default;
module.exports = exports.default;
