// @baidu/one-ui@4.41.1/lib/components/transfer/common/selectedTitle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _title = _interopRequireDefault(require("./title"));

var _button = _interopRequireDefault(require("../../button"));

var _transferTools = require("../../../core/transferTools");

/**
 * @file selected title
 * @author huangshiming
 * @date 2020-05-16
 */
;

var getSelectedNum = function getSelectedNum(allDataMap, selectedList) {
  var list = [];

  var mergeChildren = function mergeChildren(children) {
    children.forEach(function (id) {
      var _allDataMap$key;

      var key = String(id);
      var children = (_allDataMap$key = allDataMap[key]) == null ? void 0 : _allDataMap$key.children;
      var hasChildren = children && children.length;

      if (hasChildren) {
        mergeChildren(children);
      }

      if (!hasChildren && !list.includes(key)) {
        list.push(key);
      }
    });
  };

  mergeChildren(selectedList);
  return list.length;
};

var SelectedTitle = function SelectedTitle(props) {
  var prefixCls = props.prefixCls,
      size = props.size,
      selectedList = props.selectedList,
      treeName = props.treeName,
      maxSelectedNum = props.maxSelectedNum,
      showSelectedNum = props.showSelectedNum,
      onDeleteAll = props.onDeleteAll,
      showCandidateNum = props.showCandidateNum,
      parentRelationMap = props.parentRelationMap,
      allDataMap = props.allDataMap,
      SelectedTitleButtonRender = props.SelectedTitleButtonRender,
      CustomSelectedOperation = props.CustomSelectedOperation,
      SelectedTitleRender = props.SelectedTitleRender;
  var selectedTitleProps = (0, _extends2["default"])({}, props, {
    prefixCls: prefixCls,
    treeName: treeName,
    title: '已选',
    selectedNum: getSelectedNum(allDataMap, selectedList || []),
    maxSelectedNum: maxSelectedNum,
    showSelectedNum: showSelectedNum,
    showCandidateNum: showCandidateNum,
    SelectedTitleButtonRender: SelectedTitleButtonRender
  });
  var titleRender;

  if (typeof SelectedTitleRender === 'function') {
    var CustomSelectedTitle = SelectedTitleRender;
    titleRender = /*#__PURE__*/_react["default"].createElement(CustomSelectedTitle, selectedTitleProps);
  } else if (SelectedTitleRender && typeof SelectedTitleRender === 'object') {
    titleRender = SelectedTitleRender;
  } else {
    titleRender = /*#__PURE__*/_react["default"].createElement(_title["default"], selectedTitleProps);
  }

  var buttonDisabled = (selectedList || []).length <= 0;
  var disabledKeys = [];
  selectedList.forEach(function (key) {
    if ((0, _transferTools.getCheckedKeyIsDisabledByParentRelationMap)(key, parentRelationMap, allDataMap)) {
      disabledKeys.push(key);
    }
  });

  if (disabledKeys.length === selectedList.length && disabledKeys.length > 0) {
    buttonDisabled = true;
  }

  var buttonRender;
  var buttonProps = {
    className: prefixCls + "-delete-all",
    onClick: onDeleteAll,
    type: 'text-strong',
    disabled: buttonDisabled,
    size: size
  };

  if (SelectedTitleButtonRender) {
    buttonRender = /*#__PURE__*/_react["default"].createElement(SelectedTitleButtonRender, buttonProps);
  } else if (CustomSelectedOperation) {
    buttonRender = CustomSelectedOperation;
  } else {
    buttonRender = /*#__PURE__*/_react["default"].createElement(_button["default"], buttonProps, "\u6E05\u7A7A");
  }

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: prefixCls + "-pane-title"
  }, titleRender, buttonRender);
};

var _default = SelectedTitle;
exports["default"] = _default;
module.exports = exports.default;
