// @baidu/one-ui@4.32.1/es/components/cascaderPane/input.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
import Input from '../input';
;

var CascaderInput = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(CascaderInput, _PureComponent);

  function CascaderInput(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.onInputChange = function (e) {
      var value = e.value;

      if (!('value' in _this.props)) {
        _this.setState({
          value: value
        });
      }

      _this.props.onInputChange(e);
    };

    _this.state = {
      value: props.value || props.defaultValue || undefined
    };
    return _this;
  }

  CascaderInput.getDerivedStateFromProps = function getDerivedStateFromProps(nextProps) {
    var newState = {};

    if ('value' in nextProps) {
      newState.value = nextProps.value;
    }

    return newState;
  };

  var _proto = CascaderInput.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        searchProps = _this$props.searchProps,
        width = _this$props.width,
        inputType = _this$props.inputType;
    var value = this.state.value;
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Input, _extends({}, searchProps, {
      value: value,
      onChange: this.onInputChange,
      width: width,
      type: inputType
    })));
  };

  return CascaderInput;
}(PureComponent);

CascaderInput.defaultProps = {
  searchProps: {
    placeholder: '请输入...'
  },
  onInputChange: function onInputChange() {},
  width: 360
};
export default CascaderInput;
