// @baidu/one-ui@4.32.1/src/components/uploader/common/fileAnchor.tsx
import React, {PureComponent} from 'react';
import {classnames} from '../../../core/commonTools';
import Button from '../../button';
import {UploaderFileAnchorProps} from '../interface';

export default class FileAnchor extends PureComponent<UploaderFileAnchorProps> {
    static defaultProps = {
        helperTextPosition: 'right'
    }

    render() {
        const {
            onClick,
            loading,
            prefixCls,
            disabled,
            helperText,
            maxSize,
            helperTextPosition,
            pickerIcon,
            pickerText
        } = this.props;
        const buttonClassName = `${prefixCls}-anchor-file-button`;
        const buttonProps = {
            className: buttonClassName,
            onClick,
            loading,
            disabled,
            icon: pickerIcon || 'upload'
        };
        const fileAnchorClassName = classnames(
            `${prefixCls}-anchor-file-container`,
            {
                [`${prefixCls}-anchor-file-container-${helperTextPosition}`]:
                    helperTextPosition === 'right' || helperTextPosition === 'bottom'
            }
        );

        let helperTextNode;
        if (helperText === undefined) {
            helperTextNode = defaultHelperText(maxSize); // 兼容之前实现
        }
        else if (typeof helperText === 'function') {
            helperTextNode = helperText();
        }
        else {
            helperTextNode = helperText;
        }

        return (
            <span className={fileAnchorClassName}>
                <Button {...buttonProps}>
                    {pickerText || '文件上传'}
                </Button>
                {
                    helperTextNode
                        ? <span className={`${prefixCls}-helper-text`}>{helperTextNode}</span>
                        : null
                }
            </span>
        );
    }
}

function defaultHelperText(maxSize: number) {
    if (maxSize) {
        return '文件大小不超过' + maxSize / (1024 * 1024) + 'MB';
    }
    return null;
}
