// @baidu/one-ui@4.32.1/src/components/uploader/common/imageAnchor.tsx
import React, {PureComponent} from 'react';
import {noop} from 'lodash';
import Loading from '../../loading';
import MultipleAnchor from './multipleAnchor';
import {classnames} from '../../../core/commonTools';
import {UploaderImageAnchorProps} from '../interface';
import MediaIcon from './mediaIcon';

export default class ImageAnchor extends PureComponent<UploaderImageAnchorProps> {

    render() {
        const {
            onClick,
            loading,
            prefixCls,
            disabled,
            realType,
            entries,
            pickerIcon,
            pickerText,
            theme
        } = this.props;
        const buttonClassName = classnames(
            `${prefixCls}-anchor-image-button`,
            {
                [`${prefixCls}-anchor-image-button-loading`]: loading,
                [`${prefixCls}-anchor-image-button-disabled`]: disabled
            }
        );
        const buttonProps = {
            className: buttonClassName,
            onClick: (disabled || loading) ? noop : onClick
        };
        const showTexts = {
            image: '图片',
            video: '视频',
            media: '多媒体'
        };
        if (loading) {
            return (
                <div {...buttonProps}>
                    <Loading type="strong" />
                </div>
            );
        }
        if (entries && entries.length) {
            return <MultipleAnchor {...this.props} />;
        }
        return (
            <div {...buttonProps}>
                <div className={`${prefixCls}-add-icon`}>
                    {
                        pickerIcon
                        || (
                            <MediaIcon
                                theme={theme}
                                type={realType}
                                className={`${prefixCls}-${realType}-add-icon`}
                            />
                        )
                    }
                </div>
                <div className={`${prefixCls}-anchor-image-button-text`}>
                    {pickerText || `上传${showTexts[realType]}`}
                </div>
            </div>
        );
    }
}
