// @baidu/one-ui@4.32.1/es/components/select/checkboxText.js
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
import SearchText from './searchText';
import Checkbox from '../checkbox';
import Tooltip from '../tooltip';
;
;

var CheckboxText = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(CheckboxText, _PureComponent);

  function CheckboxText(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.state = {
      searchValue: props.searchValue,
      source: props.source,
      checked: false
    };
    return _this;
  }

  var _proto = CheckboxText.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        text = _this$props.text,
        value = _this$props.value,
        label = _this$props.label,
        disabled = _this$props.disabled,
        disabledReason = _this$props.disabledReason,
        disablePlacement = _this$props.disablePlacement;
    var searchLabel = text || label;
    var _this$state = this.state,
        searchValue = _this$state.searchValue,
        source = _this$state.source;
    var searchTextProps = {
      text: searchLabel,
      showSearch: true,
      searchValue: searchValue
    };
    var checked = false;

    if (source.indexOf(value) > -1) {
      checked = true;
    }

    if ('checked' in this.props) {
      checked = this.state.checked;
    }

    var checkboxTextDom = /*#__PURE__*/React.createElement("span", null, /*#__PURE__*/React.createElement(Checkbox, {
      checked: checked,
      disabled: disabled
    }), /*#__PURE__*/React.createElement(SearchText, _extends({}, searchTextProps, {
      disabled: disabled
    })));
    return disabled && disabledReason ? /*#__PURE__*/React.createElement(Tooltip, {
      title: disabledReason,
      placement: disablePlacement
    }, checkboxTextDom) : checkboxTextDom;
  };

  return CheckboxText;
}(PureComponent);

CheckboxText.isCheckboxText = true;
CheckboxText.defaultProps = {
  searchValue: '',
  source: [],
  disablePlacement: 'right',
  disabled: false
};

CheckboxText.getDerivedStateFromProps = function (nextProps, prevState) {
  var searchValue = prevState.searchValue;
  var newState = {};

  if ('searchValue' in nextProps && searchValue !== nextProps.searchValue) {
    newState.searchValue = nextProps.searchValue;
  }

  if ('source' in nextProps) {
    newState.source = nextProps.source;
  }

  if ('checked' in nextProps) {
    newState.checked = nextProps.checked;
  }

  return newState;
};

export default CheckboxText;
