// @baidu/one-ui@4.41.1/lib/components/radio/button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _radio = _interopRequireDefault(require("./radio"));

var _context = _interopRequireDefault(require("./context"));

var _context2 = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

var RadioButton = (_dec = (0, _context2.withConfigConsumer)('radio-button'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(RadioButton, _PureComponent);

  function RadioButton() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = RadioButton.prototype;

  _proto.render = function render() {
    var _this = this;

    return /*#__PURE__*/_react["default"].createElement(_context["default"].Consumer, null, function (context) {
      return _this.renderRadio(context);
    });
  };

  _proto.renderRadio = function renderRadio(context) {
    var _this$props = this.props,
        type = _this$props.type,
        props = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, ["type"]);
    return /*#__PURE__*/_react["default"].createElement(_radio["default"], (0, _extends2["default"])({}, props, {
      type: type || (context == null ? void 0 : context.type) || 'strong'
    }));
  };

  return RadioButton;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-radio-button',
  size: 'medium'
}, _temp)) || _class);
var _default = RadioButton;
exports["default"] = _default;
module.exports = exports.default;
