// @baidu/one-ui@4.32.1/lib/components/checkbox/common/commonCheckbox.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../../core/commonTools");

var _dlsIconsReact = require("dls-icons-react");

var Checkbox = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Checkbox, _PureComponent);

  function Checkbox(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.input = void 0;

    _this.saveInput = function (node) {
      _this.input = node;
    };

    _this.onHandleChange = function (e) {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          onChange = _this$props.onChange;

      if (disabled) {
        return;
      }

      var checked = e.target.checked;

      if (!('checked' in _this.props)) {
        _this.setState({
          checked: checked
        });
      }

      if (onChange) {
        onChange({
          target: (0, _extends2["default"])({}, _this.props, {
            checked: checked
          }),
          stopPropagation: function stopPropagation() {
            e.stopPropagation();
          },
          preventDefault: function preventDefault() {
            e.preventDefault();
          },
          nativeEvent: e.nativeEvent
        });
      }
    };

    var _checked = 'checked' in props ? props.checked : props.defaultChecked;

    _this.state = {
      checked: _checked
    };
    return _this;
  }

  var _proto = Checkbox.prototype;

  _proto.focus = function focus() {
    this.input.focus();
  };

  _proto.blur = function blur() {
    this.input.blur();
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props2 = this.props,
        prefixCls = _this$props2.prefixCls,
        className = _this$props2.className,
        style = _this$props2.style,
        type = _this$props2.type,
        mode = _this$props2.mode,
        disabled = _this$props2.disabled,
        onClick = _this$props2.onClick,
        onFocus = _this$props2.onFocus,
        onBlur = _this$props2.onBlur,
        autoFocus = _this$props2.autoFocus,
        indeterminate = _this$props2.indeterminate,
        inputOriginProps = _this$props2.inputOriginProps,
        name = _this$props2.name;
    var checked = this.state.checked;
    var classString = (0, _commonTools.classnames)(prefixCls, className, (_classnames = {}, _classnames[prefixCls + "-checked"] = checked, _classnames[prefixCls + "-disabled"] = disabled, _classnames));
    return /*#__PURE__*/_react["default"].createElement("span", {
      className: classString,
      style: style
    }, /*#__PURE__*/_react["default"].createElement("input", (0, _extends2["default"])({
      type: type,
      disabled: disabled,
      className: prefixCls + "-input",
      checked: !!checked,
      onClick: onClick,
      onFocus: onFocus,
      onBlur: onBlur,
      onChange: this.onHandleChange,
      autoFocus: autoFocus,
      ref: this.saveInput,
      name: name
    }, inputOriginProps)), /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-inner"
    }, type === 'checkbox' && indeterminate && mode !== 'strong' ? /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconMinus, null) : null, type === 'checkbox' && checked && mode !== 'strong' ? /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconCheck, null) : null));
  };

  return Checkbox;
}(_react.PureComponent);

Checkbox.defaultProps = {
  type: 'checkbox',
  defaultChecked: false,
  onFocus: _noop2["default"],
  onBlur: _noop2["default"],
  onChange: _noop2["default"]
};

Checkbox.getDerivedStateFromProps = function (nextProps) {
  if ('checked' in nextProps) {
    return {
      checked: nextProps.checked
    };
  }

  return null;
};

var _default = Checkbox;
exports["default"] = _default;
module.exports = exports.default;
