// @baidu/one-ui@4.41.1/lib/components/transition/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _reactTransitionGroup = require("react-transition-group");

var _commonTools = require("../../core/commonTools");

var Transition = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Transition, _PureComponent);

  function Transition() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Transition.prototype;

  _proto.getDurations = function getDurations(timeout) {
    if (typeof timeout === 'number') {
      return {
        "in": timeout + 'ms',
        out: timeout + 'ms'
      };
    }

    return {
      "in": timeout.enter + 'ms',
      out: timeout.exit + 'ms'
    };
  };

  _proto.render = function render() {
    var _this$props = this.props,
        show = _this$props["in"],
        appear = _this$props.appear,
        timeout = _this$props.timeout,
        children = _this$props.children,
        name = _this$props.name,
        prefixCls = _this$props.prefixCls,
        className = _this$props.className,
        unmountOnExit = _this$props.unmountOnExit,
        style = _this$props.style,
        lightPrefix = _this$props.lightPrefix,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, ["in", "appear", "timeout", "children", "name", "prefixCls", "className", "unmountOnExit", "style", "lightPrefix"]);
    var durations = this.getDurations(timeout);
    var transitionStyle = (0, _extends2["default"])({
      '--dls-transition-duration-in': durations["in"],
      '--dls-transition-duration-out': durations.out
    }, style);
    var cls = name === 'move' ? (0, _commonTools.classnames)(prefixCls + "-" + name, className) : className;
    return /*#__PURE__*/_react["default"].createElement(_reactTransitionGroup.CSSTransition, (0, _extends2["default"])({
      "in": show,
      appear: appear,
      timeout: timeout,
      unmountOnExit: unmountOnExit,
      classNames: prefixCls + "-" + name
    }, restProps), /*#__PURE__*/_react["default"].isValidElement(children) ? /*#__PURE__*/_react["default"].cloneElement(children, {
      className: (0, _commonTools.classnames)(children.props.className, cls),
      style: (0, _extends2["default"])({}, children.props.style, transitionStyle)
    }) : children);
  };

  return Transition;
}(_react.PureComponent);

Transition.defaultProps = {
  prefixCls: 'one-transition',
  "in": true,
  appear: true,
  timeout: {
    appear: 100,
    enter: 100,
    exit: 100
  },
  unmountOnExit: false
};
var _default = Transition;
exports["default"] = _default;
module.exports = exports.default;
