// @baidu/one-ui@4.32.1/src/components/progress/common/circle.tsx
import React, {ReactNode} from 'react';
import {Circle as RCCircle} from 'rc-progress';

type CircleProps = {
    prefixCls: string,
    percent: number,
    strokeLinecap: 'round' | 'square',
    strokeColor: string,
    strokeWidth: number,
    width: number | string,
    trailColor: string,
    children: ReactNode
};

const Circle = (props: CircleProps) => {
    const {
        prefixCls,
        percent,
        strokeLinecap,
        strokeColor,
        strokeWidth,
        width,
        trailColor,
        children
    } = props;
    const circleStyle = {
        width,
        height: width
    };
    return (
        <div className={`${prefixCls}-inner`} style={circleStyle}>
            <RCCircle
                percent={percent}
                strokeWidth={strokeWidth}
                trailWidth={strokeWidth}
                strokeColor={strokeColor}
                strokeLinecap={strokeLinecap}
                trailColor={trailColor}
                prefixCls={prefixCls}
            />
            {children}
        </div>
    );
};

Circle.defaultProps = {
    prefixCls: 'one-progress',
    percent: 0,
    strokeLinecap: 'round',
    strokeColor: '',
    strokeWidth: 0,
    trailColor: '#e2e6f0'
};

export default Circle;
