// @baidu/one-ui@4.32.1/lib/core/openAnimation.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _cssAnimation = _interopRequireDefault(require("css-animation"));

var _getRequestAnimationFrame = _interopRequireWildcard(require("./getRequestAnimationFrame"));

var reqAnimFrame = (0, _getRequestAnimationFrame["default"])();

var animate = function animate(node, show, done, prefixCls) {
  if (prefixCls === void 0) {
    prefixCls = 'one';
  }

  var height;
  var requestAnimationFrameId;
  return (0, _cssAnimation["default"])(node, prefixCls + "-motion-collapse", {
    start: function start() {
      if (!show) {
        node.style.height = node.offsetHeight + "px";
        node.style.opacity = 1;
      } else {
        height = node.offsetHeight;
        node.style.height = 0;
        node.style.opacity = 0;
      }
    },
    active: function active() {
      if (requestAnimationFrameId) {
        (0, _getRequestAnimationFrame.cancelRequestAnimationFrame)(requestAnimationFrameId);
      }

      requestAnimationFrameId = reqAnimFrame(function () {
        node.style.height = (show ? height : 0) + "px";
        node.style.opacity = show ? 1 : 0;
      });
    },
    end: function end() {
      if (requestAnimationFrameId) {
        (0, _getRequestAnimationFrame.cancelRequestAnimationFrame)(requestAnimationFrameId);
      }

      node.style.height = '';
      node.style.opacity = '';
      done();
    }
  });
};

var animation = {
  enter: function enter(node, done) {
    return animate(node, true, done);
  },
  leave: function leave(node, done) {
    return animate(node, false, done);
  },
  appear: function appear(node, done) {
    return animate(node, true, done);
  }
};
var _default = animation;
exports["default"] = _default;
module.exports = exports.default;
