// @baidu/one-ui@4.41.1/lib/core/dragTools.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.onMouseDown = exports.removeDocumentListener = exports.getPositionFromEvent = void 0;

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var getPositionFromEvent = function getPositionFromEvent(event) {
  var _event$pageX = event.pageX,
      pageX = _event$pageX === void 0 ? 0 : _event$pageX,
      _event$pageY = event.pageY,
      pageY = _event$pageY === void 0 ? 0 : _event$pageY;
  return {
    x: pageX,
    y: pageY
  };
};

exports.getPositionFromEvent = getPositionFromEvent;

var removeDocumentListener = function removeDocumentListener() {
  document.onmousemove = null;
  document.onmouseup = null;
};

exports.removeDocumentListener = removeDocumentListener;

var onMouseDown = function onMouseDown(_ref) {
  var event = _ref.event,
      onDragged = _ref.onDragged,
      onMoving = _ref.onMoving,
      onDropped = _ref.onDropped;
  var sourcePos = {};
  var startDragFlag = false;
  event.persist();
  event.stopPropagation();
  event.preventDefault();

  document.onmousemove = function (moveEvent) {
    moveEvent.preventDefault();

    if (!startDragFlag) {
      if ((0, _isFunction2["default"])(onDragged)) {
        var _position = getPositionFromEvent(event);

        onDragged({
          ele: event.target,
          position: _position
        });
        sourcePos = _position;
      }

      startDragFlag = true;
    }

    var position = getPositionFromEvent(moveEvent); // $emit isMoving

    if ((0, _isFunction2["default"])(onMoving)) {
      onMoving({
        ele: event.target,
        position: position
      });
    }
  };

  document.onmouseup = function (upEvent) {
    var position = getPositionFromEvent(upEvent);
    var _sourcePos = sourcePos,
        x = _sourcePos.x,
        y = _sourcePos.y;
    var isEqualSourcePos = +position.x === +x && +position.y === +y; // $emit isDropped

    if ((0, _isFunction2["default"])(onDropped) && startDragFlag) {
      onDropped({
        ele: upEvent.target,
        position: position,
        isEqualSourcePos: isEqualSourcePos
      });
    }

    startDragFlag = false;
    removeDocumentListener();
  };
};

exports.onMouseDown = onMouseDown;
