// @baidu/one-ui@4.41.1/lib/core/tableTools.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.normalizeColumns = normalizeColumns;
exports.measureScrollbar = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

/**
 * @file 表格的工具方法
 */
var scrollbarMeasure = {
  position: 'absolute',
  top: '-9999px',
  width: '50px',
  height: '50px',
  overflow: 'scroll'
};

var measureScrollbar = function measureScrollbar(direction) {
  if (direction === void 0) {
    direction = 'vertical';
  }

  if (typeof document === 'undefined' || typeof window === 'undefined') {
    return 0;
  }

  var scrollDiv = document.createElement('div');
  Object.keys(scrollbarMeasure).forEach(function (scrollProp) {
    scrollDiv.style[scrollProp] = scrollbarMeasure[scrollProp];
  });
  document.body.appendChild(scrollDiv);
  var size = 0;

  if (direction === 'vertical') {
    size = scrollDiv.offsetWidth - scrollDiv.clientWidth;
  } else if (direction === 'horizontal') {
    size = scrollDiv.offsetHeight - scrollDiv.clientHeight;
  }

  document.body.removeChild(scrollDiv);
  return size;
};

exports.measureScrollbar = measureScrollbar;

function normalizeColumns(elements) {
  var columns = [];

  _react["default"].Children.forEach(elements, function (element) {
    if (! /*#__PURE__*/_react["default"].isValidElement(element)) {
      return;
    }

    var column = (0, _extends2["default"])({}, element.props);

    if (element.key) {
      column.key = element.key;
    }

    if (element.type && element.type.__ONE_TABLE_COLUMN_GROUP) {
      column.children = normalizeColumns(column.children);
    }

    columns.push(column);
  });

  return columns;
}
