// @baidu/one-ui@4.32.1/lib/components/steps/steps.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../core/commonTools");

var _dlsIconsReact = require("dls-icons-react");

var _context = require("../providerConfig/context");

var _dec, _class, _class2, _temp;

function getStatus(stepIndex, stepStatus, currentIndex, currentStatus) {
  if (stepStatus) {
    return stepStatus;
  }

  if (stepIndex === currentIndex) {
    return currentStatus;
  } else if (stepIndex < currentIndex) {
    return 'finish';
  }

  return 'wait';
}

var Steps = (_dec = (0, _context.withConfigConsumer)('steps'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  (0, _inheritsLoose2["default"])(Steps, _Component);

  function Steps() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = Steps.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        style = _this$props.style,
        className = _this$props.className,
        children = _this$props.children,
        direction = _this$props.direction,
        labelPlacement = _this$props.labelPlacement,
        showTipWhenHover = _this$props.showTipWhenHover,
        status = _this$props.status,
        size = _this$props.size,
        type = _this$props.type,
        current = _this$props.current,
        initialStep = _this$props.initialStep,
        onClickStep = _this$props.onClickStep,
        onChange = _this$props.onChange,
        theme = _this$props.theme,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, ["prefixCls", "style", "className", "children", "direction", "labelPlacement", "showTipWhenHover", "status", "size", "type", "current", "initialStep", "onClickStep", "onChange", "theme"]);
    var icons = {
      finish: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconCheck, {
        className: prefixCls + "-check"
      }),
      error: /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconTimes, {
        className: prefixCls + "-close"
      })
    };
    var normalizedLabelPlacement = type === 'dot' ? 'vertical' : labelPlacement;

    var filteredChildren = _react["default"].Children.toArray(children).filter(function (child) {
      return !!child;
    });

    var stepClxs = (0, _commonTools.classnames)(prefixCls, prefixCls + "-" + direction, prefixCls + "-type-" + type, className, (_classnames = {}, _classnames[prefixCls + "-" + size] = size, _classnames[prefixCls + "-label-" + normalizedLabelPlacement] = direction === 'horizontal', _classnames));
    return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
      className: stepClxs,
      style: style
    }, restProps), _react.Children.map(filteredChildren, function (child, index) {
      if (! /*#__PURE__*/(0, _react.isValidElement)(child)) {
        return child;
      }

      var currentStepNumber = initialStep + index;
      var childProps = (0, _extends2["default"])({
        stepNumber: "" + (currentStepNumber + 1),
        prefixCls: prefixCls,
        icons: icons,
        showTipWhenHover: showTipWhenHover,
        onClickStep: onClickStep,
        onChange: onChange,
        currentStep: currentStepNumber === current,
        type: type
      }, child.props);
      var nextIndex = index + 1;
      var nextChild = filteredChildren[nextIndex];

      if ( /*#__PURE__*/(0, _react.isValidElement)(nextChild)) {
        var nextStatus = getStatus(nextIndex, nextChild.props.status, current, status);
        childProps.className = (0, _commonTools.classnames)(prefixCls + "-next-" + nextStatus, childProps.className);
      }

      if (!childProps.status) {
        childProps.status = getStatus(index, childProps.status, current, status);
      }

      return /*#__PURE__*/(0, _react.cloneElement)(child, childProps);
    }));
  };

  return Steps;
}(_react.Component), _class2.Step = void 0, _class2.defaultProps = {
  prefixCls: 'one-steps',
  current: 0,
  direction: 'horizontal',
  labelPlacement: 'horizontal',
  type: 'default',
  initialStep: 0,
  status: 'process',
  size: 'medium',
  showTipWhenHover: true
}, _temp)) || _class);
var _default = Steps;
exports["default"] = _default;
module.exports = exports.default;
