// @baidu/one-ui@4.41.1/lib/components/loading/loading.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _dlsGraphics = require("dls-graphics");

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _transition = _interopRequireDefault(require("../transition"));

var _dec, _class, _class2, _temp;

var Loading = (_dec = (0, _context.withConfigConsumer)('loading'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Loading, _PureComponent);

  function Loading() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Loading.prototype;

  _proto.isNestedPattern = function isNestedPattern() {
    return !!(this.props && this.props.children);
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        className = _this$props.className,
        size = _this$props.size,
        loading = _this$props.loading,
        style = _this$props.style,
        tip = _this$props.tip,
        type = _this$props.type,
        children = _this$props.children,
        textDirection = _this$props.textDirection,
        CustomIconNode = _this$props.CustomIconNode;
    var classes = (0, _commonTools.classnames)(prefixCls, (_classnames = {}, _classnames[prefixCls + "-" + size] = size, _classnames[prefixCls + "-visible"] = loading, _classnames[prefixCls + "-show-text"] = !!tip, _classnames[prefixCls + "-type-" + type] = type, _classnames), prefixCls + "-" + textDirection, className);
    var attrs = _dlsGraphics.loading.attrs,
        contents = _dlsGraphics.loading.contents;
    var loadingSvgProps = (0, _extends2["default"])({}, attrs);
    delete loadingSvgProps["class"];
    var loadingSvgClassName = (0, _commonTools.classnames)(attrs["class"], prefixCls + "-loading-icon");

    var loadingElement = /*#__PURE__*/_react["default"].createElement("div", {
      className: classes,
      style: style
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-icon-element"
    }, CustomIconNode || /*#__PURE__*/_react["default"].createElement("svg", (0, _extends2["default"])({}, loadingSvgProps, {
      dangerouslySetInnerHTML: {
        __html: contents
      },
      className: loadingSvgClassName
    })), tip ? /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-text"
    }, tip) : null));

    if (this.isNestedPattern()) {
      var _classnames2;

      var containerClassName = (0, _commonTools.classnames)(prefixCls + "-container", (_classnames2 = {}, _classnames2[prefixCls + "-blur"] = loading, _classnames2));
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: (0, _commonTools.classnames)(prefixCls + "-nested-loading", prefixCls + "-nested-loading-" + type)
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: containerClassName
      }, children), loading ? /*#__PURE__*/_react["default"].createElement(_transition["default"], {
        name: "fade",
        "in": true,
        appear: true,
        timeout: 500
      }, loadingElement) : null);
    }

    return loadingElement;
  };

  return Loading;
}(_react.PureComponent), _class2.defaultProps = {
  loading: true,
  prefixCls: 'one-loading',
  className: '',
  size: 'medium',
  style: {},
  tip: '',
  type: 'normal',
  textDirection: 'horizontal',
  CustomIconNode: null
}, _class2.Bar = void 0, _temp)) || _class);
var _default = Loading;
exports["default"] = _default;
module.exports = exports.default;
