// @baidu/one-ui@4.32.1/es/components/steps/steps.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

/**
 * @file 步骤条
 * @author huangshiming
 */
import React, { cloneElement, Children, Component, isValidElement } from 'react';
import { classnames } from '../../core/commonTools';
import { IconCheck, IconTimes } from 'dls-icons-react';
import { withConfigConsumer } from '../providerConfig/context';

function getStatus(stepIndex, stepStatus, currentIndex, currentStatus) {
  if (stepStatus) {
    return stepStatus;
  }

  if (stepIndex === currentIndex) {
    return currentStatus;
  } else if (stepIndex < currentIndex) {
    return 'finish';
  }

  return 'wait';
}

var Steps = (_dec = withConfigConsumer('steps'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Steps, _Component);

  function Steps() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = Steps.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        style = _this$props.style,
        className = _this$props.className,
        children = _this$props.children,
        direction = _this$props.direction,
        labelPlacement = _this$props.labelPlacement,
        showTipWhenHover = _this$props.showTipWhenHover,
        status = _this$props.status,
        size = _this$props.size,
        type = _this$props.type,
        current = _this$props.current,
        initialStep = _this$props.initialStep,
        onClickStep = _this$props.onClickStep,
        onChange = _this$props.onChange,
        theme = _this$props.theme,
        restProps = _objectWithoutPropertiesLoose(_this$props, ["prefixCls", "style", "className", "children", "direction", "labelPlacement", "showTipWhenHover", "status", "size", "type", "current", "initialStep", "onClickStep", "onChange", "theme"]);

    var icons = {
      finish: /*#__PURE__*/React.createElement(IconCheck, {
        className: prefixCls + "-check"
      }),
      error: /*#__PURE__*/React.createElement(IconTimes, {
        className: prefixCls + "-close"
      })
    };
    var normalizedLabelPlacement = type === 'dot' ? 'vertical' : labelPlacement;
    var filteredChildren = React.Children.toArray(children).filter(function (child) {
      return !!child;
    });
    var stepClxs = classnames(prefixCls, prefixCls + "-" + direction, prefixCls + "-type-" + type, className, (_classnames = {}, _classnames[prefixCls + "-" + size] = size, _classnames[prefixCls + "-label-" + normalizedLabelPlacement] = direction === 'horizontal', _classnames));
    return /*#__PURE__*/React.createElement("div", _extends({
      className: stepClxs,
      style: style
    }, restProps), Children.map(filteredChildren, function (child, index) {
      if (! /*#__PURE__*/isValidElement(child)) {
        return child;
      }

      var currentStepNumber = initialStep + index;

      var childProps = _extends({
        stepNumber: "" + (currentStepNumber + 1),
        prefixCls: prefixCls,
        icons: icons,
        showTipWhenHover: showTipWhenHover,
        onClickStep: onClickStep,
        onChange: onChange,
        currentStep: currentStepNumber === current,
        type: type
      }, child.props);

      var nextIndex = index + 1;
      var nextChild = filteredChildren[nextIndex];

      if ( /*#__PURE__*/isValidElement(nextChild)) {
        var nextStatus = getStatus(nextIndex, nextChild.props.status, current, status);
        childProps.className = classnames(prefixCls + "-next-" + nextStatus, childProps.className);
      }

      if (!childProps.status) {
        childProps.status = getStatus(index, childProps.status, current, status);
      }

      return /*#__PURE__*/cloneElement(child, childProps);
    }));
  };

  return Steps;
}(Component), _class2.Step = void 0, _class2.defaultProps = {
  prefixCls: 'one-steps',
  current: 0,
  direction: 'horizontal',
  labelPlacement: 'horizontal',
  type: 'default',
  initialStep: 0,
  status: 'process',
  size: 'medium',
  showTipWhenHover: true
}, _temp)) || _class);
export default Steps;
