// @baidu/one-ui@4.41.1/lib/components/overlay/buildInPlacements.js
"use strict";

exports.__esModule = true;
exports.BUILT_IN_PLACEMENTS = void 0;

/**
 * @file overlay的placements
 * @author huangshiming
 * @date 2020-05-04
 */
var BUILT_IN_PLACEMENTS = {
  left: {
    points: ['cr', 'cl'],
    offset: [-4, 0],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  right: {
    points: ['cl', 'cr'],
    offset: [4, 0],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  top: {
    points: ['bc', 'tc'],
    offset: [0, -4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  bottom: {
    points: ['tc', 'bc'],
    offset: [0, 4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  bottomLeft: {
    points: ['tl', 'bl'],
    offset: [0, 4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  topLeft: {
    points: ['bl', 'tl'],
    offset: [0, -4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  bottomRight: {
    points: ['tr', 'br'],
    offset: [0, 4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  topRight: {
    points: ['br', 'tr'],
    offset: [0, -4],
    overflow: {
      adjustX: 1,
      adjustY: 1
    }
  },
  leftTop: {
    points: ['tr', 'cl'],
    offset: [-4, 0]
  },
  rightTop: {
    points: ['tl', 'cr'],
    offset: [4, 0]
  },
  rightBottom: {
    points: ['bl', 'cr'],
    offset: [4, 0]
  },
  leftBottom: {
    points: ['br', 'cl'],
    offset: [-4, 0]
  }
};
exports.BUILT_IN_PLACEMENTS = BUILT_IN_PLACEMENTS;
