// @baidu/one-ui@4.41.1/lib/components/overlay/content.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../core/commonTools");

;

var Content = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Content, _PureComponent);

  function Content() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = Content.prototype;

  _proto.componentDidUpdate = function componentDidUpdate() {
    var trigger = this.props.trigger;

    if (trigger) {
      trigger.forcePopupAlign();
    }
  };

  _proto.render = function render() {
    var _this$props = this.props,
        overlay = _this$props.overlay,
        prefixCls = _this$props.prefixCls,
        id = _this$props.id,
        className = _this$props.className,
        style = _this$props.style;
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: (0, _commonTools.classnames)(prefixCls + "-inner", className),
      id: id,
      style: style
    }, typeof overlay === 'function' ? overlay() : overlay);
  };

  return Content;
}(_react.PureComponent);

exports["default"] = Content;
module.exports = exports.default;
