// @baidu/one-ui@4.41.1/lib/components/overlay/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
var _exportNames = {};
exports["default"] = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _dlsIconsReact = require("dls-icons-react");

var _button = _interopRequireDefault(require("../button"));

var _innerOverlay = _interopRequireDefault(require("./innerOverlay"));

var _commonTools = require("../../core/commonTools");

var _context = require("../providerConfig/context");

var _interface = require("./interface");

Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  exports[key] = _interface[key];
});

var _dec, _class, _class2, _temp;

var Overlay = (_dec = (0, _context.withConfigConsumer)('overlay'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Overlay, _PureComponent);

  function Overlay(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.onVisibleChange = function (visible) {
      if (!('visible' in _this.props)) {
        _this.setState({
          visible: visible
        });
      }

      var onVisibleChange = _this.props.onVisibleChange;

      if (onVisibleChange) {
        onVisibleChange(visible);
      }
    };

    _this.renderHeader = function (prefixCls) {
      var _classnames;

      var _this$props = _this.props,
          children = _this$props.children,
          disabled = _this$props.disabled,
          header = _this$props.header,
          buttonProps = _this$props.buttonProps;
      var visible = _this.state.visible;

      if (children) {
        return children;
      }

      var headerClassName = prefixCls + "-header";
      var classes = (0, _commonTools.classnames)(headerClassName, (_classnames = {}, _classnames[headerClassName + "-open"] = visible, _classnames[headerClassName + "-disabled"] = disabled, _classnames));
      return /*#__PURE__*/_react["default"].createElement("span", {
        className: classes
      }, /*#__PURE__*/_react["default"].createElement(_button["default"], (0, _extends2["default"])({
        icon: visible ? /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconChevronUp, null) : /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconChevronDown, null),
        disabled: disabled,
        type: "basic"
      }, buttonProps), header));
    };

    _this.state = {
      visible: props.visible,
      prevProps: props
    };
    return _this;
  }

  Overlay.getDerivedStateFromProps = function getDerivedStateFromProps(nextProps, prevState) {
    var newState = {
      prevProps: nextProps
    };

    if ('visible' in nextProps && nextProps.visible !== prevState.prevProps.visible) {
      newState.visible = nextProps.visible;
    }

    return newState;
  };

  var _proto = Overlay.prototype;

  _proto.render = function render() {
    var _classnames2;

    var _this$props2 = this.props,
        prefixCls = _this$props2.prefixCls,
        lightPrefix = _this$props2.lightPrefix,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props2, ["prefixCls", "lightPrefix"]);
    var children = this.renderHeader(prefixCls);
    var overlayProps = (0, _extends2["default"])({}, restProps, {
      visible: this.state.visible,
      onVisibleChange: this.onVisibleChange,
      prefixCls: prefixCls
    });
    var childrenClassName = (0, _commonTools.classnames)(children.props && children.props.className || '', (_classnames2 = {}, _classnames2[prefixCls + "-open"] = !!this.state.visible, _classnames2[prefixCls + "-close"] = !this.state.visible, _classnames2));
    return /*#__PURE__*/_react["default"].createElement(_innerOverlay["default"], overlayProps, /*#__PURE__*/_react["default"].cloneElement(children, {
      className: childrenClassName
    }));
  };

  return Overlay;
}(_react.PureComponent), _class2.defaultProps = {
  prefixCls: 'one-overlay',
  trigger: 'hover',
  disabled: false,
  header: '',
  onVisibleChange: function onVisibleChange() {},
  dropdownMatchSelectWidth: true,
  buttonProps: {},
  transitionName: 'one-transition-slide-down'
}, _temp)) || _class);
;
var _default = Overlay;
exports["default"] = _default;
module.exports = exports.default;
