// @baidu/one-ui@4.41.1/lib/components/overlay/innerOverlay.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _rcTrigger = _interopRequireDefault(require("rc-trigger"));

var _content = _interopRequireDefault(require("./content"));

var _buildInPlacements = require("./buildInPlacements");

/**
 * @file inner overlay
 * @author huangshiming
 * @date 2020-05-04
 */
;

var InnerOverlay = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(InnerOverlay, _PureComponent);

  function InnerOverlay(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;
    _this.trigger = void 0;
    _this.root = void 0;

    _this.setDropdownWidth = function () {
      var width = _reactDom["default"].findDOMNode((0, _assertThisInitialized2["default"])(_this)).offsetWidth;

      if (width !== _this.state.dropdownWidth) {
        _this.setState({
          dropdownWidth: width
        });
      }
    };

    _this.getPopupElement = function () {
      var _this$props = _this.props,
          overlay = _this$props.overlay,
          prefixCls = _this$props.prefixCls,
          id = _this$props.id,
          overlayClassName = _this$props.overlayClassName,
          dropdownMatchSelectWidth = _this$props.dropdownMatchSelectWidth,
          overlayStyle = _this$props.overlayStyle;
      var widthProp = dropdownMatchSelectWidth ? 'width' : 'minWidth';
      var style = (0, _extends2["default"])({}, overlayStyle);

      if (_this.state.dropdownWidth) {
        style[widthProp] = _this.state.dropdownWidth;
      }

      return /*#__PURE__*/_react["default"].createElement(_content["default"], {
        key: "content",
        trigger: _this.trigger,
        prefixCls: prefixCls,
        id: id,
        overlay: overlay,
        className: overlayClassName,
        style: style
      });
    };

    _this.saveTrigger = function (node) {
      _this.trigger = node;
    };

    _this.saveRoot = function (node) {
      _this.root = node;
    };

    _this.state = {
      dropdownWidth: null
    };
    return _this;
  }

  var _proto = InnerOverlay.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.setDropdownWidth();
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    this.setDropdownWidth();
  };

  _proto.getPopupDomNode = function getPopupDomNode() {
    return this.trigger.getPopupDomNode();
  };

  _proto.render = function render() {
    var _this$props2 = this.props,
        overlayClassName = _this$props2.overlayClassName,
        trigger = _this$props2.trigger,
        mouseEnterDelay = _this$props2.mouseEnterDelay,
        mouseLeaveDelay = _this$props2.mouseLeaveDelay,
        overlayStyle = _this$props2.overlayStyle,
        prefixCls = _this$props2.prefixCls,
        children = _this$props2.children,
        onVisibleChange = _this$props2.onVisibleChange,
        afterVisibleChange = _this$props2.afterVisibleChange,
        transitionName = _this$props2.transitionName,
        animation = _this$props2.animation,
        align = _this$props2.align,
        destroyPopUpOnHide = _this$props2.destroyPopUpOnHide,
        defaultVisible = _this$props2.defaultVisible,
        getPopUpContainer = _this$props2.getPopUpContainer,
        dropdownMatchSelectWidth = _this$props2.dropdownMatchSelectWidth,
        popupPlacement = _this$props2.popupPlacement,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props2, ["overlayClassName", "trigger", "mouseEnterDelay", "mouseLeaveDelay", "overlayStyle", "prefixCls", "children", "onVisibleChange", "afterVisibleChange", "transitionName", "animation", "align", "destroyPopUpOnHide", "defaultVisible", "getPopUpContainer", "dropdownMatchSelectWidth", "popupPlacement"]);
    var extraProps = (0, _extends2["default"])({}, restProps);

    if ('visible' in this.props) {
      // @ts-ignore
      extraProps.popupVisible = this.props.visible;
    }

    return /*#__PURE__*/_react["default"].createElement(_rcTrigger["default"], (0, _extends2["default"])({
      ref: this.saveTrigger,
      prefixCls: prefixCls,
      popup: this.getPopupElement,
      action: trigger,
      builtinPlacements: _buildInPlacements.BUILT_IN_PLACEMENTS,
      popupPlacement: popupPlacement,
      popupAlign: align,
      getPopupContainer: getPopUpContainer,
      onPopupVisibleChange: onVisibleChange,
      afterPopupVisibleChange: afterVisibleChange,
      popupTransitionName: transitionName,
      popupAnimation: animation,
      defaultPopupVisible: defaultVisible,
      destroyPopupOnHide: destroyPopUpOnHide,
      mouseLeaveDelay: mouseLeaveDelay,
      mouseEnterDelay: mouseEnterDelay
    }, extraProps), children);
  };

  return InnerOverlay;
}(_react.PureComponent);

InnerOverlay.defaultProps = {
  prefixCls: 'one-overlay',
  mouseEnterDelay: 0,
  destroyPopUpOnHide: false,
  mouseLeaveDelay: 0.1,
  align: {
    ignoreShake: true
  },
  trigger: ['hover'],
  dropdownMatchSelectWidth: true,
  popupPlacement: 'bottomLeft'
};
var _default = InnerOverlay;
exports["default"] = _default;
module.exports = exports.default;
