// @baidu/one-ui@4.32.1/es/components/menu/menuItem.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
import InnerMenuItem from './common/menuItem';
import Tooltip from '../tooltip';
import Context from './context';

var MenuItem = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(MenuItem, _PureComponent);

  function MenuItem() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = MenuItem.prototype;

  _proto.render = function render() {
    var inlineCollapsed = this.context.inlineCollapsed;
    var _this$props = this.props,
        children = _this$props.children,
        icon = _this$props.icon,
        level = _this$props.level,
        prefixCls = _this$props.prefixCls;
    var collapsed = icon && inlineCollapsed && level === 1;
    var item = /*#__PURE__*/React.createElement(InnerMenuItem, this.props, /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-item-span"
    }, icon && /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-item-custom-icon"
    }, icon), !collapsed && children));

    if (!(inlineCollapsed && level === 1)) {
      return item;
    }

    return /*#__PURE__*/React.createElement(Tooltip, {
      title: /*#__PURE__*/React.createElement("span", {
        className: prefixCls + "-inline-collapsed-tooltip-title"
      }, children),
      placement: "right",
      overlayClassName: prefixCls + "-inline-collapsed-tooltip"
    }, item);
  };

  return MenuItem;
}(PureComponent);

MenuItem.isMenuItem = true;
MenuItem.contextType = Context;
MenuItem.defaultProps = {
  prefixCls: 'one-menu'
};
export default MenuItem;
