// @baidu/one-ui@4.41.1/lib/components/config.js
"use strict";

exports.__esModule = true;
exports.THEME_LIGHT_AI_PREFIX = exports.THEME_LIGHT_AI = exports.THEME_LIGHT_D22 = exports.THEME_LIGHT_D20 = exports.THEME = exports.COMPONENT_INVALID = exports.COMPONENT_FOCUS = exports.COMPONENT_MAIN = exports.PREFIX = void 0;
// 组件前缀
var PREFIX = 'one'; // 组件可见边界，通常用来设置border(原组件设计边界不清，所以增加该标记)

exports.PREFIX = PREFIX;

var COMPONENT_MAIN = function COMPONENT_MAIN(prefix) {
  return (prefix || PREFIX) + '-main';
}; // 焦点


exports.COMPONENT_MAIN = COMPONENT_MAIN;

var COMPONENT_FOCUS = function COMPONENT_FOCUS(prefix) {
  return (prefix || PREFIX) + '-focus';
}; // 错误


exports.COMPONENT_FOCUS = COMPONENT_FOCUS;

var COMPONENT_INVALID = function COMPONENT_INVALID(prefix) {
  return (prefix || PREFIX) + '-invalid';
}; // 主题


exports.COMPONENT_INVALID = COMPONENT_INVALID;

var THEME = function THEME(prefix) {
  return (prefix || PREFIX) + '-theme';
}; // 默认主题


exports.THEME = THEME;
var THEME_LIGHT_D20 = 'light-d20'; // 无边框d22主题

exports.THEME_LIGHT_D20 = THEME_LIGHT_D20;
var THEME_LIGHT_D22 = 'light-d22'; // AI主题

exports.THEME_LIGHT_D22 = THEME_LIGHT_D22;
var THEME_LIGHT_AI = 'light-ai'; // AI主题样式前缀(不直接用light-ai是因为可能会和AI组件重名)

exports.THEME_LIGHT_AI = THEME_LIGHT_AI;
var THEME_LIGHT_AI_PREFIX = 'one-ai';
exports.THEME_LIGHT_AI_PREFIX = THEME_LIGHT_AI_PREFIX;
