// @baidu/one-ui@4.32.1/src/components/tag/editableGroup.tsx
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {partial} from 'lodash';
import {IconPlus} from 'dls-icons-react';
import Tag from './tag';
import Tooltip from '../tooltip';
import Input from '../input';
import {classnames} from '../../core/commonTools';
import {TagDataItem, TagEditableGroupProps} from './interface';
import {withConfigConsumer} from '../providerConfig/context';

const inputWidth = 58;

interface TagEditableGroupState {
    dataSource?: TagDataItem[];
    inputValue?: string;
    inputVisible?: boolean;
};

@withConfigConsumer('tag')
class TagEditableGroup extends PureComponent<TagEditableGroupProps, TagEditableGroupState> {

    static defaultProps = {
        prefixCls: 'one-tag',
        className: '',
        size: 'medium',
        onClose() {},
        onInputConfirm() {}
    }

    constructor(props) {
        super(props);
        this.state = {
            dataSource: props.defaultDataSource || props.dataSource || [],
            inputVisible: false,
            inputValue: ''
        };
    }

    static getDerivedStateFromProps = (nextProps, prevState) => {
        if ('dataSource' in nextProps
        && nextProps.dataSource !== prevState.dataSource) {
            return {dataSource: nextProps.dataSource};
        }
        return null;
    }

    handleClose = value => {
        let dataSource = this.state.dataSource;
        const onClose = this.props.onClose;
        onClose(dataSource, value);
        if (!('dataSource' in this.props)) {
            dataSource = dataSource.filter(tag => tag.value !== value);
            this.setState({
                dataSource
            });
        }
    };

    input;

    showInput = () => {
        this.setState({inputVisible: true}, () => this.input.focus());
    };

    handleInputChange = e => {
        this.setState({inputValue: e.value});
    };

    handleInputConfirm = () => {
        const {dataSource, inputValue} = this.state;
        const onInputConfirm = this.props.onInputConfirm;
        onInputConfirm(dataSource, inputValue);
        const newState: TagEditableGroupState = {
            inputVisible: false,
            inputValue: ''
        };
        if (!('dataSource' in this.props)) {
            let currentTags = dataSource;
            if (inputValue) {
                currentTags = [...currentTags, {
                    label: inputValue,
                    value: inputValue
                }];
            }
            newState.dataSource = currentTags;
        }
        this.setState(newState);
    };

    saveInputRef = input => {
        this.input = input;
    };

    render() {
        const {className, prefixCls, size} = this.props;
        const {dataSource, inputVisible, inputValue} = this.state;
        const groupTagClassName = classnames(className, `${prefixCls}-group-wrapper`);
        return (
            <div className={groupTagClassName}>
                {dataSource.map((tag, index) => {
                    const {label, value, tagProps} = tag;
                    const isLongTag = typeof label === 'string' && label.length > 20;
                    const tagElem = (
                        <Tag
                            key={index}
                            closable
                            onClose={partial(this.handleClose, value)}
                            noClosed
                            {...tagProps}
                        >
                            {isLongTag ? `${label.slice(0, 20)}...` : label}
                        </Tag>
                    );
                    return isLongTag ? (
                        <Tooltip title={label} key={index}>
                            {tagElem}
                        </Tooltip>
                    ) : (
                        tagElem
                    );
                })}
                {inputVisible && (
                    <Input
                        inputRef={this.saveInputRef}
                        size={size}
                        style={{width: inputWidth}}
                        value={inputValue}
                        onChange={this.handleInputChange}
                        onBlur={this.handleInputConfirm}
                        onPressEnter={this.handleInputConfirm}
                    />
                )}
                {!inputVisible && (
                    <Tag onClick={this.showInput} className={`${prefixCls}-add-tag`}>
                        <IconPlus className={`${prefixCls}-add-icon`} />
                        添加
                    </Tag>
                )}
            </div>
        );
    }
}

export default TagEditableGroup;
