// @baidu/one-ui@4.41.1/lib/components/menu/menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../core/commonTools");

var _warning = _interopRequireDefault(require("warning"));

var _menuItem = _interopRequireDefault(require("./menuItem"));

var _divider = _interopRequireDefault(require("./divider"));

var _menuItemGroup = _interopRequireDefault(require("./menuItemGroup"));

var _subMenu = _interopRequireDefault(require("./subMenu"));

var _menu = _interopRequireDefault(require("./common/menu"));

var _context = require("../providerConfig/context");

var _context2 = _interopRequireDefault(require("./context"));

var _dec, _class, _class2, _temp;

var horizontalMap = {
  1: 'large',
  2: 'medium',
  3: 'small'
};
var Menu = (_dec = (0, _context.withConfigConsumer)('menu'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Menu, _PureComponent);

  function Menu(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.getContext = function () {
      var _this$props = _this.props,
          inlineCollapsed = _this$props.inlineCollapsed,
          inlineIndent = _this$props.inlineIndent,
          inlinePaddingLeft = _this$props.inlinePaddingLeft;
      return {
        inlineCollapsed: inlineCollapsed,
        inlineIndent: inlineIndent,
        inlinePaddingLeft: inlinePaddingLeft
      };
    };

    _this.setOpenKeys = function (openKeys) {
      if (!('openKeys' in _this.props)) {
        _this.setState({
          openKeys: openKeys
        });
      }
    };

    _this.getRealMenuMode = function () {
      var _this$props2 = _this.props,
          inlineCollapsed = _this$props2.inlineCollapsed,
          mode = _this$props2.mode;

      if (_this.state.switchModeFromInline && inlineCollapsed) {
        return 'inline';
      }

      return inlineCollapsed ? 'vertical' : mode;
    };

    _this.getInlineCollapsed = function () {
      var inlineCollapsed = _this.props.inlineCollapsed;
      return inlineCollapsed;
    };

    _this.handleOpenChange = function (openKeys) {
      _this.setOpenKeys(openKeys);

      var onOpenChange = _this.props.onOpenChange;

      if (onOpenChange) {
        onOpenChange(openKeys);
      }
    };

    _this.handleClick = function (e) {
      _this.handleOpenChange([]);

      var onClick = _this.props.onClick;

      if (onClick) {
        onClick(e);
      }
    };

    (0, _warning["default"])(!('inlineCollapsed' in props && props.mode !== 'inline'), '`inlineCollapsed` should only be used when Menu\'s `mode` is inline.');

    var _openKeys;

    if ('defaultOpenKeys' in props) {
      _openKeys = props.defaultOpenKeys;
    } else if ('openKeys' in props) {
      _openKeys = props.openKeys;
    }

    _this.state = {
      openKeys: _openKeys || []
    };
    return _this;
  }

  Menu.getDerivedStateFromProps = function getDerivedStateFromProps(props, state) {
    var mode = props.mode,
        openKeys = props.openKeys,
        inlineCollapsed = props.inlineCollapsed;
    var newState = {
      mode: mode,
      inlineCollapsed: inlineCollapsed
    };

    if (state.mode === 'inline' && mode !== 'inline') {
      newState.switchModeFromInline = true;
    }

    if ('openKeys' in props) {
      newState.openKeys = openKeys;
      return newState;
    }

    if (inlineCollapsed && !state.inlineCollapsed) {
      newState.switchModeFromInline = !inlineCollapsed;
      newState.inlineOpenKeys = state.openKeys;
      newState.openKeys = [];
    }

    if (!inlineCollapsed && state.inlineCollapsed) {
      newState.openKeys = state.inlineOpenKeys || [];
      newState.inlineOpenKeys = [];
    }

    return newState;
  };

  var _proto = Menu.prototype;

  _proto.render = function render() {
    var _classnames, _classnames2;

    var _this$props3 = this.props,
        prefixCls = _this$props3.prefixCls,
        className = _this$props3.className,
        mode = _this$props3.mode,
        style = _this$props3.style,
        needBorder = _this$props3.needBorder,
        arrowPosition = _this$props3.arrowPosition,
        size = _this$props3.size,
        type = _this$props3.type,
        menuLevel = _this$props3.menuLevel,
        inlinePaddingLeft = _this$props3.inlinePaddingLeft,
        inlineCollapsed = _this$props3.inlineCollapsed,
        lightPrefix = _this$props3.lightPrefix,
        restProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props3, ["prefixCls", "className", "mode", "style", "needBorder", "arrowPosition", "size", "type", "menuLevel", "inlinePaddingLeft", "inlineCollapsed", "lightPrefix"]);
    var menuMode = this.getRealMenuMode();
    var menuClassName = (0, _commonTools.classnames)(prefixCls + "-light", (_classnames = {}, _classnames[prefixCls + "-inline-collapsed"] = inlineCollapsed, _classnames[prefixCls + "-container-border"] = mode === 'inline' && needBorder, _classnames[prefixCls + "-container-arrow-left"] = arrowPosition === 'left', _classnames[prefixCls + "-type-basic"] = type === 'basic', _classnames));
    var currentSize = size || horizontalMap[menuLevel > 3 ? 3 : menuLevel] || 'medium';
    var menuBoxCls = (0, _commonTools.classnames)(prefixCls + "-" + mode + "-box", prefixCls + "-" + mode + "-" + currentSize, (_classnames2 = {}, _classnames2[prefixCls + "-" + mode + "-container-collapsed"] = inlineCollapsed, _classnames2), className);
    var menuProps = {
      openKeys: this.state.openKeys,
      onOpenChange: this.handleOpenChange,
      className: menuClassName,
      mode: menuMode,
      prefixCls: prefixCls
    };

    if (menuMode !== 'inline') {
      menuProps.onClick = this.handleClick;
    }

    return /*#__PURE__*/_react["default"].createElement(_context2["default"].Provider, {
      value: this.getContext()
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: menuBoxCls,
      style: style
    }, /*#__PURE__*/_react["default"].createElement(_menu["default"], (0, _extends2["default"])({}, restProps, menuProps))));
  };

  return Menu;
}(_react.PureComponent), _class2.Divider = _divider["default"], _class2.Item = _menuItem["default"], _class2.SubMenu = _subMenu["default"], _class2.ItemGroup = _menuItemGroup["default"], _class2.defaultProps = {
  mode: 'vertical',
  type: 'strong',
  menuLevel: 2,
  // 默认是二级导航
  needBorder: false,
  popupClassName: '',
  arrowPosition: 'right',
  inlineIndent: 16,
  inlinePaddingLeft: 20,
  prefixCls: 'one-menu'
}, _temp)) || _class);
var _default = Menu;
exports["default"] = _default;
module.exports = exports.default;
