// @baidu/one-ui@4.41.1/lib/components/menu/menuItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _menuItem = _interopRequireDefault(require("./common/menuItem"));

var _tooltip = _interopRequireDefault(require("../tooltip"));

var _context = _interopRequireDefault(require("./context"));

var MenuItem = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(MenuItem, _PureComponent);

  function MenuItem() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = MenuItem.prototype;

  _proto.render = function render() {
    var inlineCollapsed = this.context.inlineCollapsed;
    var _this$props = this.props,
        children = _this$props.children,
        icon = _this$props.icon,
        level = _this$props.level,
        prefixCls = _this$props.prefixCls;
    var collapsed = icon && inlineCollapsed && level === 1;

    var item = /*#__PURE__*/_react["default"].createElement(_menuItem["default"], this.props, /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-item-span"
    }, icon && /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-item-custom-icon"
    }, icon), !collapsed && children));

    if (!(inlineCollapsed && level === 1)) {
      return item;
    }

    return /*#__PURE__*/_react["default"].createElement(_tooltip["default"], {
      title: /*#__PURE__*/_react["default"].createElement("span", {
        className: prefixCls + "-inline-collapsed-tooltip-title"
      }, children),
      placement: "right",
      overlayClassName: prefixCls + "-inline-collapsed-tooltip"
    }, item);
  };

  return MenuItem;
}(_react.PureComponent);

MenuItem.isMenuItem = true;
MenuItem.contextType = _context["default"];
MenuItem.defaultProps = {
  prefixCls: 'one-menu'
};
var _default = MenuItem;
exports["default"] = _default;
module.exports = exports.default;
