// @baidu/one-ui@4.32.1/src/components/drawer/common/drawerChild.tsx
import getScrollBarSize from 'rc-util/lib/getScrollBarSize';
import KeyCode from 'rc-util/lib/KeyCode';
import React from 'react';
import omit from 'omit.js';
import {
    addEventListener,
    dataToArray,
    getTouchParentScroll,
    removeEventListener,
    transitionEnd,
    isNumeric
} from '../../../core/drawerTools';
import {
    classnames,
    windowIsUndefined,
    passiveSupported
} from '../../../core/commonTools';

const currentDrawer = {};


import {DrawerWrapperProps} from './drawerWrapper';

interface DrawerChildProps extends Omit<DrawerWrapperProps, 'defaultOpen' | 'wrapperClassName' | 'forceRender'> {
    getOpenCount?(): number;
    switchScrollingEffect?;
    scrollLocker?;
};

class DrawerChild extends React.Component<DrawerChildProps> {

    constructor(props) {
        super(props);
        this.state = {
            _self: this
        };
    }

    static getDerivedStateFromProps = (props, {prevProps, _self}) => {
        const nextState = {
            prevProps: props
        };
        if (prevProps !== undefined) {
            const {placement, level} = props;
            if (placement !== prevProps.placement) {
                // test 的 bug, 有动画过场，删除 dom
                _self.contentDom = null;
            }
            if (level !== prevProps.level) {
                _self.getLevelDom(props);
            }
        }
        return nextState;
    }

    passive;
    drawerId;
    dom;
    handlerDom;
    maskDom;
    contentDom;
    levelDom;
    contentWrapper;
    startPos;

    handleClickBody = e => {
        const onClose = this.props.onClose;
        if (onClose) {
            onClose(e);
        }
    }

    componentDidMount() {
        if (!windowIsUndefined) {
            this.passive = passiveSupported ? {passive: false} : false;
            if (this.props.closeDrawerByClickBody && this.props.open) {
                window.addEventListener('click', this.handleClickBody);
            }
        }
        const open = this.props.open;
        this.drawerId = `drawer_id_${
            Number(
                (Date.now() + Math.random()).toString().replace('.', Math.round(Math.random() * 9).toString()),
            ).toString(16)
        }`;
        this.getLevelDom(this.props);
        if (open) {
            currentDrawer[this.drawerId] = open;
            // 默认打开状态时推出 level;
            this.openLevelTransition();
            this.forceUpdate(() => {
                this.domFocus();
            });
        }
    }

    componentDidUpdate(prevProps) {
        const open = this.props.open;
        if (open !== prevProps.open) {
            if (open) {
                this.domFocus();
            }
            currentDrawer[this.drawerId] = !!open;
            this.openLevelTransition();
        }
    }

    componentWillUnmount() {
        const {getOpenCount, open} = this.props;
        delete currentDrawer[this.drawerId];

        if (!windowIsUndefined) {
            window.removeEventListener('click', this.handleClickBody);
        }

        if (open) {
            this.setLevelTransform(false);
            document.body.style.touchAction = '';
        }
        if (typeof getOpenCount === 'function' && !getOpenCount()) {
            document.body.style.overflow = '';
            document.body.style.width = '';
        }
    }

    domFocus = () => {
        if (this.dom) {
            this.dom.focus();
        }
    }

    removeStartHandler = e => {
        if (e.touches.length > 1) {
            return;
        }
        this.startPos = {
            x: e.touches[0].clientX,
            y: e.touches[0].clientY
        };
    }

    removeMoveHandler = e => {
        if (e.changedTouches.length > 1) {
            return;
        }
        const currentTarget = e.currentTarget;
        const differX = e.changedTouches[0].clientX - this.startPos.x;
        const differY = e.changedTouches[0].clientY - this.startPos.y;
        if (
            currentTarget === this.maskDom
            || currentTarget === this.handlerDom
            || (currentTarget === this.contentDom && getTouchParentScroll(
                currentTarget, e.target, differX, differY)
            )
        ) {
            e.preventDefault();
        }
    }

    transitionEnd = e => {
        const dom = e.target;
        removeEventListener(dom, transitionEnd, this.transitionEnd);
        dom.style.transition = '';
    }

    onKeyDown = e => {
        if (e.keyCode === KeyCode.ESC) {
            const onClose = this.props.onClose;
            e.stopPropagation();
            if (onClose) {
                onClose(e);
            }
        }
    }

    onWrapperTransitionEnd = e => {
        const {open, afterVisibleChange} = this.props;
        if (e.target === this.contentWrapper && e.propertyName.match(/transform$/)) {
            this.dom.style.transition = '';
            if (!open && this.getCurrentDrawerSome()) {
                document.body.style.overflowX = '';
                if (this.maskDom) {
                    this.maskDom.style.left = '';
                    this.maskDom.style.width = '';
                }
            }
            if (afterVisibleChange) {
                afterVisibleChange(!!open);
            }
        }
    }

    openLevelTransition = () => {
        const {open, width, height} = this.props;
        const {isHorizontal, placementName} = this.getHorizontalBoolAndPlacementName();
        const contentValue = this.contentDom
            ? this.contentDom.getBoundingClientRect()[isHorizontal ? 'width' : 'height']
            : 0;
        const value = (isHorizontal ? width : height) || contentValue;
        this.setLevelAndScrolling(open, placementName, value);
    }

    setLevelTransform = (
        open,
        placementName?,
        value?,
        right?
    ) => {
        const {placement, duration, ease, showMask} = this.props;
        // router 切换时可能会导至页面失去滚动条，所以需要时时获取。
        this.levelDom.forEach(dom => {
            dom.style.transition = `transform ${duration} ${ease}`;
            addEventListener(dom, transitionEnd, this.transitionEnd);
            let levelValue = open ? value : 0;
            const $value = typeof levelValue === 'number' ? `${levelValue}px` : levelValue;
            let placementPos = placement === 'left' || placement === 'top' ? $value : `-${$value}`;
            placementPos = showMask && placement === 'right' && right
                ? `calc(${placementPos} + ${right}px)` : placementPos;
            dom.style.transform = levelValue ? `${placementName}(${placementPos})` : '';
        });
    }

    setLevelAndScrolling = (
        open,
        placementName,
        value
    ) => {
        const onChange = this.props.onChange;
        if (!windowIsUndefined) {
            const right = document.body.scrollHeight > (
                window.innerHeight || document.documentElement.clientHeight
            ) && window.innerWidth > document.body.offsetWidth ? getScrollBarSize(true) : 0;
            this.setLevelTransform(open, placementName, value, right);
            this.toggleScrollingToDrawerAndBody(right);
        }
        if (onChange) {
            onChange(open);
        }
    }

    toggleScrollingToDrawerAndBody = right => {
        const {getOpenCount, getContainer, showMask, open} = this.props;
        const container = getContainer && getContainer();
        const openCount = getOpenCount && getOpenCount();
        // 处理 body 滚动
        if (container && container.parentNode === document.body && showMask) {
            const eventArray = ['touchstart'];
            const domArray = [document.body, this.maskDom, this.handlerDom, this.contentDom];
            if (open && document.body.style.overflow !== 'hidden') {
                if (openCount === 1) {
                    document.body.style.overflow = 'hidden';
                    document.body.style.width = `calc(100% - ${right}px)`;
                }
                document.body.style.touchAction = 'none';
                // 手机禁滚
                domArray.forEach((item, i) => {
                    if (!item) {
                        return;
                    }
                    addEventListener(
                        item,
                        eventArray[i] || 'touchmove',
                        i ? this.removeMoveHandler : this.removeStartHandler,
                        this.passive,
                    );
                });
            }
            else if (this.getCurrentDrawerSome()) {
                // 没有弹框的状态下清除 overflow;
                if (!openCount) {
                    document.body.style.overflow = '';
                    document.body.style.width = '';
                }
                document.body.style.touchAction = '';
                // 恢复事件
                domArray.forEach((item, i) => {
                    if (!item) {
                        return;
                    }
                    removeEventListener(
                        item,
                        eventArray[i] || 'touchmove',
                        i ? this.removeMoveHandler : this.removeStartHandler,
                        this.passive,
                    );
                });
            }
        }
    }

    getCurrentDrawerSome = () => !Object.keys(currentDrawer).some(key => currentDrawer[key]);

    getLevelDom = ({level, getContainer}: {level, getContainer?}) => {
        if (windowIsUndefined) {
            return;
        }
        const container = getContainer && getContainer();
        const parent = container ? container.parentNode : null;
        this.levelDom = [];
        if (level === 'all') {
            const children = parent ? Array.prototype.slice.call(parent.children) : [];
            children.forEach(child => {
                if (
                    child.nodeName !== 'SCRIPT' && child.nodeName !== 'STYLE'
                    && child.nodeName !== 'LINK' && child !== container
                ) {
                    this.levelDom.push(child);
                }
            });
        }
        else if (level) {
            dataToArray(level).forEach(key => {
                document.querySelectorAll(key).forEach(item => {
                    this.levelDom.push(item);
                });
            });
        }
    }

    getHorizontalBoolAndPlacementName = () => {
        const placement = this.props.placement;
        const isHorizontal = placement === 'left' || placement === 'right';
        const placementName = `translate${isHorizontal ? 'X' : 'Y'}`;
        return {
            isHorizontal,
            placementName
        };
    }

    onClick = e => {
        e.stopPropagation();
    }

    render() {
        const {
            className,
            children,
            style,
            width,
            height,
            open: $open,
            prefixCls,
            placement,
            level,
            ease,
            duration,
            getContainer,
            handler,
            onChange,
            afterVisibleChange,
            showMask,
            maskClosable,
            maskStyle,
            drawerStyle,
            onClose,
            onHandleClick,
            keyboard,
            getOpenCount,
            ...props
        } = this.props;
        // 首次渲染都将是关闭状态。
        const open = this.dom ? $open : false;
        const wrapperClassName = classnames(prefixCls, {
            [`${prefixCls}-${placement}`]: true,
            [`${prefixCls}-open`]: open,
            [className || '']: !!className,
            'no-mask': !showMask
        });

        const placementName = this.getHorizontalBoolAndPlacementName().placementName;
        // 百分比与像素动画不同步，第一次打用后全用像素动画。
        // const defaultValue = !this.contentDom || !level ? '100%' : `${value}px`;
        const placementPos = placement === 'left' || placement === 'top' ? '-100%' : '100%';
        const transform = open ? '' : `${placementName}(${placementPos})`;

        const handlerChildren = handler && React.cloneElement(handler, {
            onClick: e => {
                if (handler.props.onClick) {
                    handler.props.onClick();
                }
                if (onHandleClick) {
                    onHandleClick(e);
                }
            },
            ref: c => {
                this.handlerDom = c;
            }
        });
        return (
            <div
                {...omit(props, ['switchScrollingEffect', 'closeDrawerByClickBody', 'scrollLocker'])}
                tabIndex={-1}
                className={wrapperClassName}
                style={style}
                ref={c => {
                    this.dom = c;
                }}
                onKeyDown={open && keyboard ? this.onKeyDown : undefined}
                onTransitionEnd={this.onWrapperTransitionEnd}
                onClick={this.onClick}
            >
                {showMask && (
                    <div
                        className={`${prefixCls}-mask`}
                        onClick={maskClosable ? onClose : undefined}
                        style={maskStyle}
                        ref={c => {
                            this.maskDom = c;
                        }}
                    />
                )}
                <div
                    className={`${prefixCls}-content-wrapper`}
                    style={{
                        transform,
                        msTransform: transform,
                        width: isNumeric(width) ? `${width}px` : width,
                        height: isNumeric(height) ? `${height}px` : height,
                        ...drawerStyle
                    }}
                    ref={c => {
                        this.contentWrapper = c;
                    }}
                >
                    <div
                        className={`${prefixCls}-content`}
                        ref={c => {
                            this.contentDom = c;
                        }}
                        onTouchStart={open && showMask ? this.removeStartHandler : undefined} // 跑用例用
                        onTouchMove={open && showMask ? this.removeMoveHandler : undefined} // 跑用例用
                    >
                        {children}
                    </div>
                    {handlerChildren}
                </div>
            </div>
        );
    }
}

export default DrawerChild;
