// @baidu/one-ui@4.41.1/lib/components/tree/common/treeNode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _commonTools = require("../../../core/commonTools");

var _treeTools = require("../../../core/treeTools");

var _checkbox = _interopRequireDefault(require("../../checkbox"));

var _searchText = _interopRequireDefault(require("../../select/searchText"));

var _childrenTools = _interopRequireDefault(require("../../../core/childrenTools"));

var _context = _interopRequireDefault(require("./context"));

var ICON_OPEN = 'open';
var ICON_CLOSE = 'close';
var defaultTitle = '---';

var TreeNode = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(TreeNode, _PureComponent);

  function TreeNode() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.selectHandle = void 0;

    _this.onSelectorClick = function (e) {
      if (_this.isDisabled() && _this.props.disabledAllWhenNodeDisabled) {
        return;
      }

      var onNodeClick = _this.context.rcTree.onNodeClick;
      onNodeClick(e, (0, _assertThisInitialized2["default"])(_this));

      if (_this.isSelectable()) {
        _this.onSelect(e);
      } else {
        _this.onCheck(e);
      }
    };

    _this.onSelectorDoubleClick = function (e) {
      if (_this.isDisabled() && _this.props.disabledAllWhenNodeDisabled) {
        return;
      }

      var onNodeDoubleClick = _this.context.rcTree.onNodeDoubleClick;
      onNodeDoubleClick(e, (0, _assertThisInitialized2["default"])(_this));
    };

    _this.onSelect = function (e) {
      if (_this.isDisabled()) {
        return;
      }

      var onNodeSelect = _this.context.rcTree.onNodeSelect; // e.preventDefault();

      onNodeSelect(e, (0, _assertThisInitialized2["default"])(_this));
    };

    _this.onCheck = function (e) {
      if (_this.isDisabled()) {
        return;
      }

      var _this$props = _this.props,
          disableCheckbox = _this$props.disableCheckbox,
          checked = _this$props.checked;
      var _this$context$rcTree = _this.context.rcTree,
          checkable = _this$context$rcTree.checkable,
          onNodeCheck = _this$context$rcTree.onNodeCheck;

      if (!checkable || disableCheckbox) {
        return;
      }

      e.preventDefault();
      var targetChecked = !checked;
      onNodeCheck(e, (0, _assertThisInitialized2["default"])(_this), targetChecked);
    };

    _this.onMouseEnter = function (e) {
      var onNodeMouseEnter = _this.context.rcTree.onNodeMouseEnter;
      onNodeMouseEnter(e, (0, _assertThisInitialized2["default"])(_this));
    };

    _this.onMouseLeave = function (e) {
      var onNodeMouseLeave = _this.context.rcTree.onNodeMouseLeave;
      onNodeMouseLeave(e, (0, _assertThisInitialized2["default"])(_this));
    };

    _this.onContextMenu = function (e) {
      var onNodeContextMenu = _this.context.rcTree.onNodeContextMenu;
      onNodeContextMenu(e, (0, _assertThisInitialized2["default"])(_this));
    };

    _this.onExpand = function (e) {
      if (_this.isDisabled() && _this.props.disabledAllWhenNodeDisabled) {
        return;
      }

      var onNodeExpand = _this.context.rcTree.onNodeExpand;
      onNodeExpand(e, (0, _assertThisInitialized2["default"])(_this));
    };

    _this.setSelectHandle = function (node) {
      _this.selectHandle = node;
    };

    _this.getNodeChildren = function () {
      var _this$props2 = _this.props,
          children = _this$props2.children,
          disableCheckbox = _this$props2.disableCheckbox,
          disabled = _this$props2.disabled;
      var originList = (0, _childrenTools["default"])(children).map(function (child) {
        // 父存在disabled的话，所有的子都是disabled checkbox
        var props = (0, _extends2["default"])({}, child.props);

        if (disableCheckbox) {
          props.disableCheckbox = disableCheckbox;
        }

        if (disabled) {
          props.disabled = disabled;
        }

        return /*#__PURE__*/(0, _react.cloneElement)(child, props);
      }).filter(function (node) {
        return node;
      });
      var targetList = originList.filter(_treeTools.isTreeNode);

      if (originList.length !== targetList.length) {
        (0, _treeTools.warnOnlyTreeNode)();
      }

      return targetList;
    };

    _this.getNodeState = function (isLeaf) {
      var expanded = _this.props.expanded;

      if (isLeaf) {
        return null;
      }

      return expanded ? ICON_OPEN : ICON_CLOSE;
    };

    _this.isLeaf = function (hasChildren) {
      var _this$props3 = _this.props,
          isLeaf = _this$props3.isLeaf,
          loaded = _this$props3.loaded;
      var loadData = _this.context.rcTree.loadData;

      if (isLeaf === false) {
        return false;
      }

      return isLeaf || !loadData && !hasChildren || loadData && loaded && !hasChildren;
    };

    _this.isDisabled = function () {
      var disabled = _this.props.disabled;
      var treeDisabled = _this.context.rcTree.disabled; // Follow the logic of Selectable

      if (disabled === false) {
        return false;
      }

      return !!(treeDisabled || disabled);
    };

    _this.syncLoadData = function (props) {
      var expanded = props.expanded,
          loading = props.loading,
          loaded = props.loaded;
      var _this$context$rcTree2 = _this.context.rcTree,
          loadData = _this$context$rcTree2.loadData,
          onNodeLoad = _this$context$rcTree2.onNodeLoad;

      if (!loadData || !expanded || loaded || loading) {
        return;
      }

      var hasChildren = _this.getNodeChildren().length > 0; // read from state to avoid loadData at same time

      if (!hasChildren && !_this.isLeaf(hasChildren)) {
        // We needn't reload data when has children in sync logic
        // It's only needed in node expanded
        onNodeLoad((0, _assertThisInitialized2["default"])(_this));
      }
    };

    _this.renderSwitcher = function (isLeaf) {
      var _this$props4 = _this.props,
          switcherIconFromProps = _this$props4.switcherIcon,
          expanded = _this$props4.expanded;
      var _this$context$rcTree3 = _this.context.rcTree,
          prefixCls = _this$context$rcTree3.prefixCls,
          switcherIconFromCtx = _this$context$rcTree3.switcherIcon;
      var switcherIcon = switcherIconFromProps || switcherIconFromCtx;

      if (isLeaf) {
        return /*#__PURE__*/_react["default"].createElement("span", {
          className: (0, _commonTools.classnames)(prefixCls + "-switcher", prefixCls + "-switcher-noop")
        }, typeof switcherIcon === 'function' ? /*#__PURE__*/_react["default"].createElement(switcherIcon, (0, _extends2["default"])({}, _this.props, {
          isLeaf: true
        })) : switcherIcon);
      }

      var switcherCls = (0, _commonTools.classnames)(prefixCls + "-switcher", prefixCls + "-switcher_" + (expanded ? ICON_OPEN : ICON_CLOSE));
      return /*#__PURE__*/_react["default"].createElement("span", {
        onClick: _this.onExpand,
        className: switcherCls
      }, typeof switcherIcon === 'function' ? /*#__PURE__*/_react["default"].createElement(switcherIcon, (0, _extends2["default"])({}, _this.props, {
        isLeaf: false
      })) : switcherIcon);
    };

    _this.renderCheckbox = function () {
      var _this$props5 = _this.props,
          checked = _this$props5.checked,
          halfChecked = _this$props5.halfChecked,
          disableCheckbox = _this$props5.disableCheckbox,
          size = _this$props5.size;
      var _this$context$rcTree4 = _this.context.rcTree,
          prefixCls = _this$context$rcTree4.prefixCls,
          checkable = _this$context$rcTree4.checkable;

      var disabled = _this.isDisabled();

      var currentCheckable = checkable;

      if ('checkable' in _this.props) {
        currentCheckable = _this.props.checkable;
      }

      if (!currentCheckable) {
        return null;
      } // [Legacy] Custom element should be separate with `checkable` in future


      var $custom = typeof currentCheckable !== 'boolean' ? currentCheckable : null;

      if (typeof currentCheckable === 'boolean' && currentCheckable === true) {
        $custom = /*#__PURE__*/_react["default"].createElement(_checkbox["default"], {
          checked: checked,
          indeterminate: !checked && halfChecked,
          disabled: disabled || disableCheckbox,
          size: size
        });
      }

      return /*#__PURE__*/_react["default"].createElement("span", {
        className: (0, _commonTools.classnames)(prefixCls + "-checkbox", checked && prefixCls + "-checkbox-checked", !checked && halfChecked && prefixCls + "-checkbox-indeterminate", (disabled || disableCheckbox) && prefixCls + "-checkbox-disabled"),
        onClick: _this.onCheck
      }, $custom);
    };

    _this.renderSelector = function (filterTitle, isLeafNode) {
      var _this$props6 = _this.props,
          title = _this$props6.title,
          selected = _this$props6.selected,
          icon = _this$props6.icon;
      var _this$context$rcTree5 = _this.context.rcTree,
          prefixCls = _this$context$rcTree5.prefixCls,
          showIcon = _this$context$rcTree5.showIcon,
          treeIcon = _this$context$rcTree5.icon;

      var disabled = _this.isDisabled();

      var wrapClass = prefixCls + "-node-content-wrapper"; // Icon - Still show loading icon when loading without showIcon

      var $icon;

      if (showIcon) {
        var currentIcon = icon || treeIcon;
        $icon = currentIcon ? /*#__PURE__*/_react["default"].createElement("span", {
          className: (0, _commonTools.classnames)(prefixCls + "-iconEle", prefixCls + "-icon__customize")
        }, typeof currentIcon === 'function' ? /*#__PURE__*/_react["default"].createElement(currentIcon, (0, _extends2["default"])({}, _this.props)) : currentIcon) : null;
      }

      var currentTitle = title;

      if (typeof title === 'string' && typeof filterTitle === 'string' && !!filterTitle) {
        currentTitle = /*#__PURE__*/_react["default"].createElement(_searchText["default"], {
          prefixCls: prefixCls,
          text: title,
          searchValue: filterTitle,
          showSearch: true
        });
      } // Title


      var $title = /*#__PURE__*/_react["default"].createElement("span", {
        className: prefixCls + "-title"
      }, currentTitle);

      return /*#__PURE__*/_react["default"].createElement("span", {
        ref: _this.setSelectHandle,
        title: typeof title === 'string' ? title : '',
        className: (0, _commonTools.classnames)("" + wrapClass, wrapClass + "-" + (_this.getNodeState(isLeafNode) || 'normal'), !disabled && selected && prefixCls + "-node-selected"),
        onMouseEnter: _this.onMouseEnter,
        onMouseLeave: _this.onMouseLeave,
        onContextMenu: _this.onContextMenu,
        onClick: _this.onSelectorClick,
        onDoubleClick: _this.onSelectorDoubleClick
      }, $icon, $title);
    };

    _this.renderChildren = function (nodeList) {
      var _this$props7 = _this.props,
          expanded = _this$props7.expanded,
          pos = _this$props7.pos;
      var _this$context$rcTree6 = _this.context.rcTree,
          prefixCls = _this$context$rcTree6.prefixCls,
          renderTreeNode = _this$context$rcTree6.renderTreeNode;
      var $children;

      if (expanded) {
        $children = /*#__PURE__*/_react["default"].createElement("ul", {
          className: (0, _commonTools.classnames)(prefixCls + "-child-tree", expanded && prefixCls + "-child-tree-open"),
          "data-expanded": expanded,
          role: "group"
        }, (0, _treeTools.mapChildren)(nodeList, function (node, index) {
          return renderTreeNode(node, index, pos);
        }));
      }

      return $children;
    };

    return _this;
  }

  var _proto = TreeNode.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.syncLoadData(this.props);
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    this.syncLoadData(this.props);
  };

  _proto.isSelectable = function isSelectable() {
    var selectable = this.props.selectable;
    var treeSelectable = this.context.rcTree.selectable; // Ignore when selectable is undefined or null

    if (typeof selectable === 'boolean') {
      return selectable;
    }

    return treeSelectable;
  } // Load data to avoid default expanded tree without data
  ;

  _proto.render = function render() {
    var _classnames;

    var _this$props8 = this.props,
        loading = _this$props8.loading,
        className = _this$props8.className,
        style = _this$props8.style,
        isLeaf = _this$props8.isLeaf,
        expanded = _this$props8.expanded,
        selected = _this$props8.selected,
        checked = _this$props8.checked,
        halfChecked = _this$props8.halfChecked,
        otherProps = (0, _objectWithoutPropertiesLoose2["default"])(_this$props8, ["loading", "className", "style", "isLeaf", "expanded", "selected", "checked", "halfChecked"]);
    var _this$context$rcTree7 = this.context.rcTree,
        prefixCls = _this$context$rcTree7.prefixCls,
        filterTreeNode = _this$context$rcTree7.filterTreeNode;
    var disabled = this.isDisabled();
    var dataOrAriaAttributeProps = (0, _treeTools.getDataAndAria)(otherProps);
    var filterTitle = filterTreeNode && filterTreeNode(this) || false;
    var nodeList = this.getNodeChildren();
    var hasChildren = nodeList.length > 0;
    var isLeafNode = this.isLeaf(hasChildren);
    return /*#__PURE__*/_react["default"].createElement("li", (0, _extends2["default"])({
      className: (0, _commonTools.classnames)(className, (_classnames = {}, _classnames[prefixCls + "-treenode-disabled"] = disabled, _classnames[prefixCls + "-treenode-switcher-" + (expanded ? 'open' : 'close')] = !isLeaf, _classnames[prefixCls + "-treenode-checkbox-checked"] = checked, _classnames[prefixCls + "-treenode-checkbox-indeterminate"] = halfChecked, _classnames[prefixCls + "-treenode-selected"] = selected, _classnames[prefixCls + "-treenode-loading"] = loading, _classnames['current-filter-node'] = filterTitle, _classnames)),
      style: style,
      role: "treeitem"
    }, dataOrAriaAttributeProps), /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-treenode-container"
    }, this.renderSwitcher(isLeafNode), /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-treenode-container-title"
    }, this.renderCheckbox(), this.renderSelector(filterTitle, isLeafNode))), hasChildren && this.renderChildren(nodeList));
  };

  return TreeNode;
}(_react.PureComponent); // @ts-ignore


TreeNode.contextType = _context["default"];
TreeNode.defaultProps = {
  title: defaultTitle,
  size: 'medium',
  isFileNode: false,
  disabledAllWhenNodeDisabled: false
};
TreeNode.isTreeNode = 1;
var _default = TreeNode;
exports["default"] = _default;
module.exports = exports.default;
