// @baidu/one-ui@4.41.1/lib/components/tree/common/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.convertDataToTree = convertDataToTree;
exports.getPosition = getPosition;
exports.getKey = getKey;
exports.flattenTreeData = flattenTreeData;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _treeNode = _interopRequireDefault(require("./treeNode"));

var _childrenTools = _interopRequireDefault(require("../../../core/childrenTools"));

var internalProcessProps = function internalProcessProps(props) {
  return props;
};

function convertDataToTree(treeData, processer) {
  if (!treeData) {
    return [];
  }

  var _ref = processer || {},
      _ref$processProps = _ref.processProps,
      processProps = _ref$processProps === void 0 ? internalProcessProps : _ref$processProps;

  var list = Array.isArray(treeData) ? treeData : [treeData];
  return list.map(function (_ref2, index) {
    var children = _ref2.children,
        props = (0, _objectWithoutPropertiesLoose2["default"])(_ref2, ["children"]);
    var childrenNodes = convertDataToTree(children, processer);
    return /*#__PURE__*/_react["default"].createElement(_treeNode["default"], (0, _extends2["default"])({
      key: index
    }, processProps(props)), childrenNodes);
  });
}

function getPosition(level, index) {
  return level + "-" + index;
}

function getKey(key, pos) {
  if (key !== null && key !== undefined) {
    return key;
  }

  return pos;
}

function flattenTreeData(treeNodeList, expandedKeys) {
  var expandedKeySet = new Set(expandedKeys === true ? [] : expandedKeys);
  var flattenList = [];

  function dig(list, parent) {
    return list.map(function (treeNode, index) {
      var pos = getPosition(parent ? parent.pos : '0', index);
      var mergedKey = getKey(treeNode.key, pos); // Add FlattenDataNode into list

      var flattenNode = (0, _extends2["default"])({}, treeNode, {
        parent: parent,
        pos: pos,
        children: null,
        data: treeNode,
        level: parent ? parent.level + 1 : 0
      });
      var disabledCheckbox = treeNode.props && treeNode.props.disableCheckbox;
      var disabled = treeNode.props && treeNode.props.disabled;

      if (parent && parent.props && parent.props.disableCheckbox) {
        flattenNode.props = (0, _extends2["default"])({}, flattenNode.props, {
          disableCheckbox: disabledCheckbox || parent.props.disableCheckbox
        });
      }

      if (parent && parent.props && parent.props.disabled) {
        flattenNode.props = (0, _extends2["default"])({}, flattenNode.props, {
          disabled: disabled || parent.props.disabled
        });
      }

      flattenList.push(flattenNode); // Loop treeNode children

      if (expandedKeys === true || expandedKeySet.has(mergedKey)) {
        var children = (0, _childrenTools["default"])(treeNode.props.children) || [];
        flattenNode.children = children.length > 0 ? dig(children, flattenNode) : children;
      } else {
        flattenNode.children = [];
      }

      return flattenNode;
    });
  }

  dig(treeNodeList);
  return flattenList;
}
