// @baidu/one-ui@4.32.1/lib/components/embedded/horizon.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _dlsIconsReact = require("dls-icons-react");

/**
 * @deprecated
 */
var Horizon = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(Horizon, _PureComponent);

  function Horizon(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.onClose = function () {
      _this.setState({
        visible: false
      });

      _this.props.onClose();
    };

    _this.state = {
      visible: true
    };
    return _this;
  }

  var _proto = Horizon.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        prefixCls = _this$props.prefixCls,
        className = _this$props.className,
        closable = _this$props.closable,
        children = _this$props.children;
    var visible = this.state.visible;

    if ('visible' in this.props) {
      visible = this.props.visible;
    }

    if (!visible) {
      return null;
    }

    return /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + " " + className
    }, closable ? /*#__PURE__*/_react["default"].createElement("span", {
      className: prefixCls + "-close",
      onClick: this.onClose
    }, /*#__PURE__*/_react["default"].createElement(_dlsIconsReact.IconTimes, {
      className: prefixCls + "-close-icon"
    })) : null, /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-body"
    }, children));
  };

  return Horizon;
}(_react.PureComponent);

exports["default"] = Horizon;
Horizon.propTypes = {
  onClose: _propTypes["default"].func,
  closable: _propTypes["default"].bool,
  prefixCls: _propTypes["default"].string,
  className: _propTypes["default"].string,
  children: _propTypes["default"].node,
  visible: _propTypes["default"].bool
};
Horizon.defaultProps = {
  closable: true,
  prefixCls: 'one-embedded-horizon',
  className: ''
};
module.exports = exports.default;
