// @baidu/one-ui@4.32.1/lib/components/table/columnTitle.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _popover = _interopRequireDefault(require("../popover"));

/**
 * @file column title
 * @author tangjinhua
 * @date 2021-09-17
 */
;

var ColumnTitle = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(ColumnTitle, _PureComponent);

  function ColumnTitle() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.state = {
      showDesc: true
    };

    _this.hide = function () {
      _this.setState({
        showDesc: false
      });
    };

    _this.show = function () {
      _this.setState({
        showDesc: true
      });
    };

    return _this;
  }

  var _proto = ColumnTitle.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        desc = _this$props.desc,
        _this$props$title = _this$props.title,
        title = _this$props$title === void 0 ? null : _this$props$title;

    if (desc) {
      var content = desc;

      if ( /*#__PURE__*/_react["default"].isValidElement(desc)) {
        content = /*#__PURE__*/_react["default"].cloneElement(desc, {
          hide: this.hide,
          show: this.show
        });
      }

      return /*#__PURE__*/_react["default"].createElement(_popover["default"], {
        content: content,
        size: "small",
        overlayStyle: {
          display: this.state.showDesc ? '' : 'none'
        }
      }, title);
    }

    return title;
  };

  return ColumnTitle;
}(_react.PureComponent);

exports["default"] = ColumnTitle;
module.exports = exports.default;
