// @baidu/one-ui@4.41.1/es/components/dialog/dialog.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/extends";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import Button from '../button';
import DialogWrapper from './common/dialogWrapper';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
var widthSizeMap = {
  small: 400,
  medium: 600,
  large: 800
};
;
var Dialog = (_dec = withConfigConsumer('dialog'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Dialog, _PureComponent);

  /**
   * @deprecated
   */
  function Dialog(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.getWidthBySize = function () {
      var width = _this.props.width;

      if (widthSizeMap[width]) {
        return widthSizeMap[width];
      }

      return width;
    };

    _this.handleCancel = function (e) {
      var _this$props = _this.props,
          onCancel = _this$props.onCancel,
          visibleControlledWhenClose = _this$props.visibleControlledWhenClose;

      if (!visibleControlledWhenClose) {
        _this.setState({
          visible: false
        });
      }

      if (onCancel) {
        onCancel(e);
      }
    };

    _this.handleOk = function (e) {
      var onOk = _this.props.onOk;

      if (onOk) {
        onOk(e);
      }
    };

    _this.renderFooter = function () {
      var _this$props2 = _this.props,
          okText = _this$props2.okText,
          okType = _this$props2.okType,
          cancelText = _this$props2.cancelText,
          confirmLoading = _this$props2.confirmLoading,
          okProps = _this$props2.okProps,
          cancelProps = _this$props2.cancelProps,
          okOrder = _this$props2.okOrder,
          cancelOrder = _this$props2.cancelOrder,
          size = _this$props2.size,
          buttonPosition = _this$props2.buttonPosition;
      var buttonSize = _this.props.buttonSize || size;
      var okStyle = okOrder ? _extends({}, okProps.style || {}, {
        order: okOrder
      }) : _extends({}, okProps.style || {});
      var cancelStyle = cancelOrder ? _extends({}, cancelProps.style || {}, {
        order: cancelOrder
      }) : _extends({}, cancelProps.style || {});
      var confirmButton = /*#__PURE__*/React.createElement(Button, _extends({
        key: "confirm",
        type: okType,
        size: buttonSize,
        loading: confirmLoading,
        onClick: _this.handleOk
      }, okProps, {
        style: okStyle
      }), okText || '确定');
      var cancelButton = /*#__PURE__*/React.createElement(Button, _extends({
        key: "cancel",
        size: buttonSize,
        type: "normal",
        onClick: _this.handleCancel
      }, cancelProps, {
        style: cancelStyle
      }), cancelText || '取消');
      var defaultFooter = buttonPosition === 'right' ? [cancelButton, confirmButton] : [confirmButton, cancelButton];
      return defaultFooter;
    };

    _this.state = {
      visible: props.visible
    };
    return _this;
  }

  var _proto = Dialog.prototype;

  _proto.render = function render() {
    var _classnames;

    var _this$props3 = this.props,
        footer = _this$props3.footer,
        wrapClassName = _this$props3.wrapClassName,
        centered = _this$props3.centered,
        className = _this$props3.className,
        prefixCls = _this$props3.prefixCls,
        buttonPosition = _this$props3.buttonPosition,
        destroyOnClose = _this$props3.destroyOnClose,
        size = _this$props3.size,
        fullScreen = _this$props3.fullScreen,
        okOrder = _this$props3.okOrder,
        cancelOrder = _this$props3.cancelOrder,
        lightPrefix = _this$props3.lightPrefix,
        restProps = _objectWithoutPropertiesLoose(_this$props3, ["footer", "wrapClassName", "centered", "className", "prefixCls", "buttonPosition", "destroyOnClose", "size", "fullScreen", "okOrder", "cancelOrder", "lightPrefix"]);

    var visible = this.state.visible;
    var defaultFooter = this.renderFooter();

    if (!visible && destroyOnClose) {
      // 关闭的时候销毁节点
      return null;
    }

    var wrapClass = classnames(prefixCls + "-" + buttonPosition, prefixCls + "-" + size, (_classnames = {}, _classnames[prefixCls + "-centered"] = !!centered, _classnames[prefixCls + "-full-screen"] = fullScreen, _classnames[prefixCls + "-order-reverse"] = okOrder && cancelOrder && cancelOrder < okOrder, _classnames[prefixCls + "-order-normal"] = okOrder && cancelOrder && cancelOrder > okOrder, _classnames), wrapClassName, className);
    return /*#__PURE__*/React.createElement(DialogWrapper, _extends({}, restProps, {
      prefixCls: prefixCls,
      wrapClassName: wrapClass,
      width: this.getWidthBySize(),
      footer: footer === undefined ? defaultFooter : footer,
      visible: visible,
      onClose: this.handleCancel
    }));
  };

  return Dialog;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-dialog',
  size: 'medium',
  confirmLoading: false,
  visible: false,
  okType: 'primary',
  centered: true,
  destroyOnClose: false,
  mask: true,
  maskClosable: false,
  buttonPosition: 'left',
  footer: undefined,
  needCloseIcon: true,
  okProps: {},
  cancelProps: {},
  width: 'medium',
  fullScreen: false,
  visibleControlledWhenClose: false,
  contentClassName: '',
  maskClassName: ''
}, _class2.confirm = void 0, _class2.alert = void 0, _class2.toast = void 0, _class2.getDerivedStateFromProps = function (nextProps, prevState) {
  if ('visible' in nextProps && nextProps.visible !== prevState.visible) {
    return {
      visible: nextProps.visible
    };
  }

  return null;
}, _temp)) || _class);
export default Dialog;
