// @baidu/one-ui@4.32.1/src/components/switch/common/switch.tsx
import React, {PureComponent, ReactNode} from 'react';
import IconLoading from '../../loading';
import {classnames} from '../../../core/commonTools';
import {SwitchProps} from '../interface';

interface SwitchState {
    checked?: boolean;
};

type InnerSwitchProps = SwitchProps & {loadingIcon?: ReactNode, onClick?}

class InnerSwitch extends PureComponent<InnerSwitchProps, SwitchState> {

    static defaultProps = {
        className: '',
        defaultChecked: false
    }

    constructor(props) {
        super(props);
        let checked = false;
        if ('checked' in props) {
            checked = !!props.checked;
        }
        else {
            checked = !!props.defaultChecked;
        }
        this.state = {
            checked
        };
    }

    static getDerivedStateFromProps = nextProps => {
        const newState: SwitchState = {};
        if ('checked' in nextProps) {
            newState.checked = !!nextProps.checked;
        }
        return newState;
    }

    switchRef;

    setChecked(checked, e) {
        const {disabled, onChange, loading} = this.props;
        if (disabled || loading) {
            return;
        }
        if (!('checked' in this.props)) {
            this.setState({
                checked
            });
        }
        if (onChange) {
            onChange(checked, e);
        }
    }

    onHandleClick = e => {
        const checked = this.state.checked;
        const onClick = this.props.onClick;
        const newChecked = !checked;
        this.setChecked(newChecked, e);
        if (onClick) {
            onClick(newChecked, e);
        }
    };

    onHandleKeyDown = e => {
        if (e.keyCode === 37) {
            // Left
            this.setChecked(false, e);
        }
        else if (e.keyCode === 39) {
            // Right
            this.setChecked(true, e);
        }
    };

    saveRef = ref => {
        this.switchRef = ref;
    }

    focus() {
        if (this.switchRef) {
            this.switchRef.focus();
        }
    }

    blur() {
        if (this.switchRef) {
            this.switchRef.blur();
        }
    }

    render() {
        const {
            className,
            prefixCls,
            disabled,
            loadingIcon,
            loading,
            onChange,
            ...restProps
        } = this.props;
        const checked = this.state.checked;
        const switchClassName = classnames(prefixCls, {
            [`${prefixCls}-checked`]: checked,
            [`${prefixCls}-disabled`]: disabled || loading
        }, className);
        return (
            <button
                {...restProps}
                type="button"
                role="switch"
                aria-checked={checked}
                disabled={disabled}
                className={switchClassName}
                onKeyDown={this.onHandleKeyDown}
                onClick={this.onHandleClick}
                ref={this.saveRef}
            >
                <span className={`${prefixCls}-thumb`}>
                    {
                        loading ? (
                            loadingIcon || <span className={`${prefixCls}-loading-icon`}><IconLoading /></span>
                        ) : null
                    }
                </span>
            </button>
        );
    }
}

export default InnerSwitch;
