// @baidu/one-ui@4.32.1/lib/components/stack/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {
  Stack: true
};
exports["default"] = exports.Stack = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _interface = require("./interface");

Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  exports[key] = _interface[key];
});

/**
 * @visibleName Stack
 */
var Stack = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var _classNames, _classNames2, _ref2;

  var prefixCls = _ref.prefixCls,
      children = _ref.children,
      className = _ref.className,
      direction = _ref.direction,
      wrap = _ref.wrap,
      inline = _ref.inline,
      align = _ref.align,
      justify = _ref.justify,
      gap = _ref.gap,
      restProps = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["prefixCls", "children", "className", "direction", "wrap", "inline", "align", "justify", "gap"]);
  var rootClass = (0, _classnames["default"])(prefixCls, className, (_classNames = {}, _classNames[prefixCls + "-wrap"] = wrap, _classNames[prefixCls + "-inline"] = inline, _classNames[prefixCls + "-" + direction] = direction, _classNames[prefixCls + "-gap"] = gap, _classNames));
  var realAlign = align || (direction === 'row' ? 'center' : 'stretch');
  var innerClass = (0, _classnames["default"])(prefixCls + "-inner", (_classNames2 = {}, _classNames2[prefixCls + "-inner-gap-" + gap] = typeof gap === 'string', _classNames2[prefixCls + "-inner-align-" + realAlign] = realAlign, _classNames2[prefixCls + "-inner-justify-" + justify] = justify, _classNames2));
  var innerStyle = typeof gap === 'number' && gap > 0 ? (_ref2 = {}, _ref2["--" + (prefixCls + "-gap")] = gap + "px", _ref2) : null;
  return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
    className: rootClass
  }, restProps, {
    ref: ref
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: innerClass,
    style: innerStyle
  }, children));
});
exports.Stack = Stack;
Stack.defaultProps = {
  prefixCls: 'one-stack',
  direction: 'row',
  wrap: false,
  inline: false
};
Stack.displayName = 'Stack';
var _default = Stack;
exports["default"] = _default;
