// @baidu/one-ui@4.41.1/es/components/badge/badge.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";

var _dec, _class, _class2, _temp;

import React, { PureComponent } from 'react';
import omit from 'omit.js';
import { classnames } from '../../core/commonTools';
import { withConfigConsumer } from '../providerConfig/context';
;
var Badge = (_dec = withConfigConsumer('badge'), _dec(_class = (_temp = _class2 = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(Badge, _PureComponent);

  function Badge(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _this.render = function () {
      var _classnames, _classnames2, _classnames3;

      var _this$props = _this.props,
          showZero = _this$props.showZero,
          prefixCls = _this$props.prefixCls,
          overflowCount = _this$props.overflowCount,
          className = _this$props.className,
          style = _this$props.style,
          children = _this$props.children,
          isDot = _this$props.isDot,
          type = _this$props.type,
          text = _this$props.text,
          offset = _this$props.offset,
          color = _this$props.color,
          textContent = _this$props.textContent,
          lightPrefix = _this$props.lightPrefix,
          restProps = _objectWithoutPropertiesLoose(_this$props, ["showZero", "prefixCls", "overflowCount", "className", "style", "children", "isDot", "type", "text", "offset", "color", "textContent", "lightPrefix"]);

      var _this$state = _this.state,
          visible = _this$state.visible,
          count = _this$state.count;
      var displayCount = +count > +overflowCount ? overflowCount + "+" : count;
      var isZero = displayCount === 0;
      var isDotMode = isDot && !isZero || type; // 小圆点状态下，不展示数字

      if (isDotMode) {
        displayCount = '';
      }

      var isEmpty = (displayCount === null || displayCount === undefined || displayCount === '') && !textContent;
      var hidden = (isEmpty || isZero && !showZero) && !isDotMode || !visible;
      var typeCls = classnames((_classnames = {}, _classnames[prefixCls + "-type-dot"] = !!type, _classnames[prefixCls + "-type-" + type] = !!type, _classnames));
      var numberBoxCls = classnames((_classnames2 = {}, _classnames2[prefixCls + "-dot"] = isDotMode, _classnames2[prefixCls + "-count"] = !isDotMode, _classnames2[prefixCls + "-type-" + type] = !!type, _classnames2[prefixCls + "-hidden"] = hidden, _classnames2));
      var badgeCls = classnames(className, prefixCls, (_classnames3 = {}, _classnames3[prefixCls + "-type"] = !!type, _classnames3[prefixCls + "-not-a-wrapper"] = !children, _classnames3));
      var styleWithOffset = offset ? _extends({
        right: -offset[0],
        marginTop: offset[1]
      }, style) : style;

      if (color) {
        styleWithOffset = _extends({}, styleWithOffset, {
          backgroundColor: color
        });
      }

      var divProps = omit(restProps, ['visible']);

      if (!children && type) {
        var colorStyle = color ? {
          backgroundColor: color
        } : undefined;
        return /*#__PURE__*/React.createElement("span", _extends({}, divProps, {
          className: badgeCls,
          style: style
        }), /*#__PURE__*/React.createElement("span", {
          className: typeCls,
          style: colorStyle
        }), /*#__PURE__*/React.createElement("span", {
          className: prefixCls + "-type-text"
        }, text));
      }

      var showContent = displayCount; // 如果有文字形式，数字形式默认失效

      if (textContent) {
        showContent = textContent;
      }

      return /*#__PURE__*/React.createElement("span", _extends({}, divProps, {
        className: badgeCls
      }), children, /*#__PURE__*/React.createElement("span", {
        className: numberBoxCls,
        style: styleWithOffset
      }, showContent));
    };

    _this.state = {
      count: props.count,
      visible: props.visible
    };
    return _this;
  }

  return Badge;
}(PureComponent), _class2.defaultProps = {
  prefixCls: 'one-badge',
  count: null,
  showZero: false,
  isDot: false,
  overflowCount: 99,
  visible: true,
  textContent: ''
}, _class2.getDerivedStateFromProps = function (nextProps, prevState) {
  var newState = {};

  if ('visible' in nextProps && nextProps.visible !== prevState.visible) {
    newState.visible = nextProps.visible;
  }

  if ('count' in nextProps && nextProps.count !== prevState.count) {
    newState.count = nextProps.count;
  }

  return newState;
}, _temp)) || _class);
export default Badge;
