// @baidu/one-ui@4.41.1/lib/components/table/common/columnManager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _utils = require("./utils");

var ColumnManager = /*#__PURE__*/function () {
  function ColumnManager(columns) {
    this._cached = {};
    this.columns = void 0;
    this.columns = columns;
  }

  var _proto = ColumnManager.prototype;

  _proto.isAnyColumnsFixed = function isAnyColumnsFixed() {
    var _this = this;

    return this._cache('isAnyColumnsFixed', function () {
      return _this.columns.some(function (column) {
        return !!column.fixed;
      });
    });
  };

  _proto.isAnyColumnsLeftFixed = function isAnyColumnsLeftFixed() {
    var _this2 = this;

    return this._cache('isAnyColumnsLeftFixed', function () {
      return _this2.columns.some(function (column) {
        return column.fixed === 'left' || column.fixed === true;
      });
    });
  };

  _proto.isAnyColumnsRightFixed = function isAnyColumnsRightFixed() {
    var _this3 = this;

    return this._cache('isAnyColumnsRightFixed', function () {
      return _this3.columns.some(function (column) {
        return column.fixed === 'right';
      });
    });
  };

  _proto.leftColumns = function leftColumns() {
    var _this4 = this;

    return this._cache('leftColumns', function () {
      return _this4.groupedColumns().filter(function (column) {
        return column.fixed === 'left' || column.fixed === true;
      });
    });
  };

  _proto.rightColumns = function rightColumns() {
    var _this5 = this;

    return this._cache('rightColumns', function () {
      return _this5.groupedColumns().filter(function (column) {
        return column.fixed === 'right';
      });
    });
  };

  _proto.leafColumns = function leafColumns() {
    var _this6 = this;

    return this._cache('leafColumns', function () {
      return _this6._leafColumns(_this6.columns);
    });
  };

  _proto.leftLeafColumns = function leftLeafColumns() {
    var _this7 = this;

    return this._cache('leftLeafColumns', function () {
      return _this7._leafColumns(_this7.leftColumns());
    });
  };

  _proto.rightLeafColumns = function rightLeafColumns() {
    var _this8 = this;

    return this._cache('rightLeafColumns', function () {
      return _this8._leafColumns(_this8.rightColumns());
    });
  };

  _proto.allColumnsWidth = function allColumnsWidth(tableWidth, refresh) {
    if (refresh === void 0) {
      refresh = false;
    }

    var leafColumns = this.leafColumns();
    var fixedWidthColumns = leafColumns.filter(function (col) {
      return col.fixedWidth;
    });
    var fixedWidth = fixedWidthColumns.reduce(function (width, col) {
      return col.width + width;
    }, 0);
    var tableCalcWidth = tableWidth - fixedWidth;
    var normalWidth = (0, _utils.formatNormalWidth)(tableWidth, leafColumns);
    var curWidth = 0;
    return this._cache('columnsWidth', function () {
      return leafColumns.map(function (column) {
        var key = column.key || column.dataIndex;
        var width = column.fixedWidth ? column.width : (0, _utils.formatCellWidth)(column.width || normalWidth, tableCalcWidth);
        var minWidth = column.fixedWidth ? column.width : (0, _utils.formatCellWidth)(column.minWidth || column.width || normalWidth, tableCalcWidth);
        var maxWidth = column.fixedWidth ? column.width : (0, _utils.formatCellWidth)(column.maxWidth, tableCalcWidth) || null;
        curWidth = (0, _utils.formatCellLeft)(width, curWidth, tableCalcWidth);
        return {
          key: key,
          width: width,
          minWidth: minWidth,
          maxWidth: maxWidth,
          left: curWidth,
          fixed: column.fixed,
          fixedWidth: column.fixedWidth
        };
      });
    }, refresh);
  } // add appropriate rowspan and colspan to column
  ;

  _proto.groupedColumns = function groupedColumns() {
    var _this9 = this;

    return this._cache('groupedColumns', function () {
      var getGroupColumns = function getGroupColumns(columns, currentRow, parentColumn, rows) {
        if (currentRow === void 0) {
          currentRow = 0;
        }

        if (parentColumn === void 0) {
          parentColumn = {
            colSpan: 0
          };
        }

        if (rows === void 0) {
          rows = [];
        }

        // track how many rows we got
        rows[currentRow] = rows[currentRow] || [];
        var grouped = [];

        var setRowSpan = function setRowSpan(column) {
          var rowSpan = rows.length - currentRow;

          if (column && !column.children && rowSpan > 1 && (!column.rowSpan || column.rowSpan < rowSpan)) {
            column.rowSpan = rowSpan;
          }
        };

        columns.forEach(function (column, index) {
          var newColumn = (0, _extends2["default"])({}, column);
          rows[currentRow].push(newColumn);
          parentColumn.colSpan = parentColumn.colSpan || 0;

          if (newColumn.children && newColumn.children.length > 0) {
            newColumn.children = getGroupColumns(newColumn.children, currentRow + 1, newColumn, rows);
            parentColumn.colSpan += newColumn.colSpan;
          } else {
            parentColumn.colSpan++;
          } // update rowspan to all same row columns


          for (var i = 0; i < rows[currentRow].length - 1; ++i) {
            setRowSpan(rows[currentRow][i]);
          } // last column, update rowspan immediately


          if (index + 1 === columns.length) {
            setRowSpan(newColumn);
          }

          grouped.push(newColumn);
        });
        return grouped;
      };

      return getGroupColumns(_this9.columns);
    });
  };

  _proto.reset = function reset(columns) {
    this.columns = columns;
    this._cached = {};
  };

  _proto._cache = function _cache(name, fn, refresh) {
    if (name in this._cached && !refresh) {
      return this._cached[name];
    }

    this._cached[name] = fn();
    return this._cached[name];
  };

  _proto._leafColumns = function _leafColumns(columns) {
    var _this10 = this;

    var leafColumns = [];
    columns.forEach(function (column) {
      if (!column.children) {
        leafColumns.push(column);
      } else {
        leafColumns.push.apply(leafColumns, _this10._leafColumns(column.children));
      }
    });
    return leafColumns;
  };

  return ColumnManager;
}();

exports["default"] = ColumnManager;
module.exports = exports.default;
