// @baidu/one-ui@4.41.1/lib/components/table/common/selectionBox.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _checkbox = _interopRequireDefault(require("../../checkbox"));

var _radio = _interopRequireDefault(require("../../radio"));

;

var SelectionBox = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(SelectionBox, _PureComponent);

  function SelectionBox(_props) {
    var _this;

    _this = _PureComponent.call(this, _props) || this;
    _this.unsubscribe = void 0;

    _this.getCheckState = function (props) {
      var store = props.store,
          defaultSelection = props.defaultSelection,
          rowIndex = props.rowIndex;
      var checked = false;

      if (store.getState().selectionDirty) {
        checked = store.getState().selectedRowKeys.indexOf(rowIndex) >= 0;
      } else {
        checked = store.getState().selectedRowKeys.indexOf(rowIndex) >= 0 || defaultSelection.indexOf(rowIndex) >= 0;
      }

      return checked;
    };

    _this.state = {
      checked: _this.getCheckState(_props)
    };
    return _this;
  }

  var _proto = SelectionBox.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.subscribe();
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    if (this.unsubscribe) {
      this.unsubscribe();
    }
  };

  _proto.subscribe = function subscribe() {
    var _this2 = this;

    var store = this.props.store;
    this.unsubscribe = store.subscribe(function () {
      var checked = _this2.getCheckState(_this2.props);

      _this2.setState({
        checked: checked
      });
    });
  };

  _proto.render = function render() {
    var _this$props = this.props,
        type = _this$props.type,
        rowIndex = _this$props.rowIndex,
        disabled = _this$props.disabled,
        indeterminate = _this$props.indeterminate,
        onChange = _this$props.onChange;
    var checked = this.state.checked;

    if (type === 'radio') {
      return /*#__PURE__*/_react["default"].createElement(_radio["default"], {
        disabled: disabled,
        onChange: onChange,
        value: rowIndex,
        checked: checked
      });
    }

    return /*#__PURE__*/_react["default"].createElement(_checkbox["default"], {
      indeterminate: indeterminate,
      checked: checked,
      disabled: disabled,
      onChange: onChange
    });
  };

  return SelectionBox;
}(_react.PureComponent);

exports["default"] = SelectionBox;
module.exports = exports.default;
