// @baidu/one-ui@4.41.1/lib/components/table/common/shadow.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _miniStore = require("mini-store");

var ShadowArea = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2["default"])(ShadowArea, _PureComponent);

  function ShadowArea() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = ShadowArea.prototype;

  _proto.calcWidth = function calcWidth() {
    var _this$props = this.props,
        allColumnsWidth = _this$props.allColumnsWidth,
        placement = _this$props.placement;
    var filters = allColumnsWidth.filter(function (col) {
      return col.fixed === placement;
    });

    if (!filters.length) {
      return;
    }

    var totalWidth = 0;
    filters.forEach(function (col) {
      totalWidth += col.width;
    });
    return totalWidth;
  };

  _proto.render = function render() {
    var _this$props2 = this.props,
        prefixCls = _this$props2.prefixCls,
        placement = _this$props2.placement,
        allColumnsWidth = _this$props2.allColumnsWidth;

    if (!allColumnsWidth.length) {
      return null;
    }

    var width = this.calcWidth();

    if (!width) {
      return null;
    }

    return /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls + "-" + placement + "-shadow",
      style: {
        width: width
      }
    });
  };

  return ShadowArea;
}(_react.PureComponent);

ShadowArea.propTypes = {
  allColumnsWidth: _propTypes["default"].array,
  placement: _propTypes["default"].oneOf(['left', 'right']),
  prefixCls: _propTypes["default"].string
};

var _default = (0, _miniStore.connect)(function (state) {
  return {
    allColumnsWidth: state.allColumnsWidth
  };
})(ShadowArea);

exports["default"] = _default;
module.exports = exports.default;
