// @baidu/one-ui@4.41.1/es/components/table/common/dragLine.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
import { classnames } from '../../../core/commonTools';

var DragLine = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(DragLine, _PureComponent);

  function DragLine() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(args)) || this;
    _this.state = {
      dragLeft: 0,
      showLine: false
    };

    _this.setShowLine = function (showLine) {
      _this.setState({
        showLine: showLine
      });
    };

    _this.setDragLeft = function (dragLeft) {
      _this.setState({
        dragLeft: dragLeft
      });
    };

    return _this;
  }

  var _proto = DragLine.prototype;

  _proto.render = function render() {
    var _classnames;

    var prefixCls = this.props.prefixCls;
    var _this$state = this.state,
        dragLeft = _this$state.dragLeft,
        showLine = _this$state.showLine;
    return /*#__PURE__*/React.createElement("div", {
      key: "drag-line",
      style: {
        left: dragLeft
      },
      className: classnames(prefixCls + "-drag-line", (_classnames = {}, _classnames[prefixCls + "-drag-line-show"] = showLine, _classnames))
    });
  };

  return DragLine;
}(PureComponent);

export default DragLine;
