// @baidu/one-ui@4.41.1/es/components/table/common/headTable.js
import _extends from "@babel/runtime/helpers/extends";
import _omit from "lodash/omit";
import React, { useContext } from 'react';
import BaseTable from './baseTable';
import Context from './context';
import Loading from '../../loading';
import classNames from 'classnames';
export default function HeadTable(props) {
  var _classNames;

  var _useContext = useContext(Context),
      table = _useContext.table;

  var _table$props = table.props,
      prefixCls = _table$props.prefixCls,
      scroll = _table$props.scroll,
      showHeader = _table$props.showHeader,
      headerFixTop = _table$props.headerFixTop,
      loadingOption = _table$props.loadingOption,
      headerBottom = _table$props.headerBottom,
      loading = _table$props.loading;
  var columns = props.columns,
      tableClassName = props.tableClassName,
      expander = props.expander,
      onDragStart = props.onDragStart,
      onDraging = props.onDraging,
      onDragEnd = props.onDragEnd,
      onHeaderCellMouseEnter = props.onHeaderCellMouseEnter,
      onHeaderCellMouseLeave = props.onHeaderCellMouseLeave;
  var saveRef = table.saveRef,
      containerWidth = table.containerWidth;
  var useFixedHeader = table.props.useFixedHeader;
  var isFixedHeader = typeof headerFixTop === 'number' && headerFixTop != null;
  var refName = 'headTable';
  var headStyle = isFixedHeader ? {
    top: headerFixTop
  } : {};
  var showLoadBar = !loadingOption || loadingOption.type !== 'spinner';

  if (scroll.y || table.props.headerFixTop !== null || headerBottom || showLoadBar) {
    useFixedHeader = true;
  }

  var headContent = /*#__PURE__*/React.createElement("div", {
    className: classNames(prefixCls + "-header-container", (_classNames = {}, _classNames[prefixCls + "-sticky-header"] = isFixedHeader, _classNames)),
    style: headStyle
  }, !useFixedHeader || !showHeader ? null : /*#__PURE__*/React.createElement("div", {
    key: refName,
    ref: saveRef(refName),
    className: prefixCls + "-header"
  }, /*#__PURE__*/React.createElement(BaseTable, {
    tableClassName: tableClassName,
    hasHead: true,
    hasBody: false,
    fixedHeader: isFixedHeader || useFixedHeader,
    columns: columns,
    expander: expander,
    onDragStart: onDragStart,
    onDraging: onDraging,
    onDragEnd: onDragEnd,
    onHeaderCellMouseEnter: onHeaderCellMouseEnter,
    onHeaderCellMouseLeave: onHeaderCellMouseLeave
  })), showLoadBar ? /*#__PURE__*/React.createElement("div", {
    className: prefixCls + "-loading-bar"
  }, /*#__PURE__*/React.createElement(Loading.Bar, _extends({
    size: "xsmall",
    loading: loading
  }, _omit(loadingOption, ['type'])))) : null, headerBottom ? /*#__PURE__*/React.createElement("div", {
    className: prefixCls + "-header-bottom"
  }, headerBottom) : null);
  return headContent;
}
HeadTable.defaultProps = {
  refName: '',
  fixedStyle: {},
  onDragStart: function onDragStart() {},
  onDraging: function onDraging() {},
  onDragEnd: function onDragEnd() {}
};
