// @baidu/one-ui@4.41.1/es/components/table/common/selectionBox.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React, { PureComponent } from 'react';
import Checkbox from '../../checkbox';
import Radio from '../../radio';
;

var SelectionBox = /*#__PURE__*/function (_PureComponent) {
  _inheritsLoose(SelectionBox, _PureComponent);

  function SelectionBox(_props) {
    var _this;

    _this = _PureComponent.call(this, _props) || this;
    _this.unsubscribe = void 0;

    _this.getCheckState = function (props) {
      var store = props.store,
          defaultSelection = props.defaultSelection,
          rowIndex = props.rowIndex;
      var checked = false;

      if (store.getState().selectionDirty) {
        checked = store.getState().selectedRowKeys.indexOf(rowIndex) >= 0;
      } else {
        checked = store.getState().selectedRowKeys.indexOf(rowIndex) >= 0 || defaultSelection.indexOf(rowIndex) >= 0;
      }

      return checked;
    };

    _this.state = {
      checked: _this.getCheckState(_props)
    };
    return _this;
  }

  var _proto = SelectionBox.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.subscribe();
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    if (this.unsubscribe) {
      this.unsubscribe();
    }
  };

  _proto.subscribe = function subscribe() {
    var _this2 = this;

    var store = this.props.store;
    this.unsubscribe = store.subscribe(function () {
      var checked = _this2.getCheckState(_this2.props);

      _this2.setState({
        checked: checked
      });
    });
  };

  _proto.render = function render() {
    var _this$props = this.props,
        type = _this$props.type,
        rowIndex = _this$props.rowIndex,
        disabled = _this$props.disabled,
        indeterminate = _this$props.indeterminate,
        onChange = _this$props.onChange;
    var checked = this.state.checked;

    if (type === 'radio') {
      return /*#__PURE__*/React.createElement(Radio, {
        disabled: disabled,
        onChange: onChange,
        value: rowIndex,
        checked: checked
      });
    }

    return /*#__PURE__*/React.createElement(Checkbox, {
      indeterminate: indeterminate,
      checked: checked,
      disabled: disabled,
      onChange: onChange
    });
  };

  return SelectionBox;
}(PureComponent);

export { SelectionBox as default };
